

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Changes &mdash; MGKit: Metagenomic framework 0.5.3 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/graphviz.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="mgkit" href="api/modules.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> MGKit: Metagenomic framework
          

          
          </a>

          
            
            
              <div class="version">
                0.5.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1"><a class="reference internal" href="library.html">Library Reference</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Changes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#id1">0.5.3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id2">0.5.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id3">0.5.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id4">0.5.0</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#added">Added</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id5">0.4.4</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id6">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#changed">Changed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id7">0.4.3</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fixes">Fixes</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id8">0.4.2</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id9">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id10">Changed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id11">0.4.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id12">0.4.0</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id13">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id14">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#deprecated">Deprecated</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tests">Tests</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id15">0.3.4</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#scripts">Scripts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#python3">Python3</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id16">0.3.3</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id17">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id18">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id19">Deprecated</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id20">0.3.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id21">0.3.1</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id22">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id23">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fixed">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id24">Deprecated</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id25">0.3.0</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id26">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id27">Changed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id28">0.2.5</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id29">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id30">Added</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id31">0.2.4</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id32">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id33">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id34">0.2.3</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id35">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id36">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id37">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id38">0.2.2</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id39">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id40">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#removed">Removed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id41">0.2.1</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id42">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id43">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id44">Deprecated</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id45">Removed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id46">0.2.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id47">0.1.16</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id48">0.1.15</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id49">0.1.14</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id50">0.1.13</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id51">0.1.12</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id52">0.1.11</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MGKit: Metagenomic framework</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Changes</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/changes.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="changes">
<h1>Changes<a class="headerlink" href="#changes" title="Permalink to this headline">¶</a></h1>
<div class="section" id="id1">
<h2>0.5.3<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Added parameters to support the partial pN/pS calculations to <cite>mgkit.snps.funcs.combine_sample_snps</cite></p></li>
<li><p>Added options to pnps-gen script to output only pS or pN to script <cite>pnps-gen</cite></p></li>
</ul>
</div>
<div class="section" id="id2">
<h2>0.5.2<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Fixed a bug when using <em>–rank None</em> in <cite>pnps-gen</cite></p></li>
</ul>
</div>
<div class="section" id="id3">
<h2>0.5.1<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>get_gene_taxon_dataframe: changed: Changed in version 0.5.1: gene_map can be None, use_uid can be passed to the underline function</p></li>
<li><p>added option to include the lineage as a string in <em>pnps-gen</em></p></li>
<li><p>added option to use the <em>uids</em> from the GFF instead of <em>gene_id</em>, this does not require the GFF file, they are embedded into the <em>.pickle</em> file</p></li>
<li><p>by default <em>pnps-gen</em> returns the taxon included in the GFF and not a ranked taxon</p></li>
<li><p>added option to make a different type of table in <em>pnps-gen rank</em></p></li>
</ul>
</div>
<div class="section" id="id4">
<h2>0.5.0<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h2>
<div class="section" id="added">
<h3>Added<a class="headerlink" href="#added" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><cite>taxon-utils get</cite> command to query a taxonomy file</p></li>
<li><p><cite>pnps-gen</cite> to generate a table of pN/pS values</p></li>
</ul>
</div>
</div>
<div class="section" id="id5">
<h2>0.4.4<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id6">
<h3>Added<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.dict_to_text" title="mgkit.utils.dictionary.dict_to_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.dictionary.dict_to_text()</span></code></a> <a class="reference internal" href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.text_to_dict" title="mgkit.utils.dictionary.text_to_dict"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.dictionary.text_to_dict()</span></code></a> to read/write simple dictionary files (tables)</p></li>
<li><p><em>filter-gff overlap</em> command, added option to not use the strand information in filtering the overlaps and also to make multiple iterations (max 10) to better remove overlaps</p></li>
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.has_attr" title="mgkit.io.gff.Annotation.has_attr"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.has_attr()</span></code></a> and <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.del_attr" title="mgkit.io.gff.Annotation.del_attr"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.del_attr()</span></code></a></p></li>
<li><p>a new script, <em>edit-gff</em> to view a GFF as table and perform general edits on it</p></li>
</ul>
</div>
<div class="section" id="changed">
<h3>Changed<a class="headerlink" href="#changed" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><cite>download-ncbi-taxa.sh</cite> and <cite>download-uniprot-taxa.sh</cite> (<a class="reference internal" href="scripts/download-taxonomy.html#download-taxonomy"><span class="std std-ref">Download Taxonomy</span></a>): if a <cite>PROGBAR</cite> enviroment variable is set, the progress bar (default in <cite>wget</cite>) is used</p></li>
<li><p>changed <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.set_attr" title="mgkit.io.gff.Annotation.set_attr"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.set_attr()</span></code></a> to allow changing standard attributes</p></li>
<li><p>added some checks for unexpected lengths in <em>add-gff-info exp_syn</em>, check the log for this cases</p></li>
<li><p><a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.get_seq_expected_syn_count" title="mgkit.utils.sequence.get_seq_expected_syn_count"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.get_seq_expected_syn_count()</span></code></a> silently skips codons containing <em>N</em> or which is not of length 3</p></li>
</ul>
</div>
</div>
<div class="section" id="id7">
<h2>0.4.3<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h2>
<div class="section" id="fixes">
<h3>Fixes<a class="headerlink" href="#fixes" title="Permalink to this headline">¶</a></h3>
<p><a class="reference internal" href="api/mgkit.align.html#mgkit.align.SamtoolsDepth" title="mgkit.align.SamtoolsDepth"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.align.SamtoolsDepth</span></code></a> in version 0.4.2 was using a <a class="reference external" href="https://docs.python.org/3/library/weakref.html#weakref.WeakValueDictionary" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">weakref.WeakValueDictionary</span></code></a> to speed up the recovery of memory from the internal dictionary. In the tests on MacOS the memory was mostly kept, but on Linux when submitted as a job it seems to be freed instantly. This also impact the <cite>add-gff-info cov_samtools</cite> command, since it uses this class - it will run, but reports that the number of sequences not found in the samtools depth file is the same as the number of sequences in the GFF file.</p>
</div>
</div>
<div class="section" id="id8">
<h2>0.4.2<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Fixed reading of Expasy files. The reading was not changed to adopt Python3 conventions like the rest of the routines. Included a test that dowload the expasy file and parses it</p></li>
<li><p>Optimisations of <cite>add-gff-info cov_samtools</cite> and the <a class="reference internal" href="api/mgkit.align.html#module-mgkit.align" title="mgkit.align"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.align</span></code></a> routines used</p></li>
</ul>
<div class="section" id="id9">
<h3>Added<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.parse_expasy_dat" title="mgkit.mappings.enzyme.parse_expasy_dat"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.mappings.enzyme.parse_expasy_dat()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.advance_file" title="mgkit.align.SamtoolsDepth.advance_file"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.align.SamtoolsDepth.advance_file()</span></code></a></p></li>
<li><p>option <cite>-m</cite> to calculate average coverage in <cite>add-gff-info cov_samtools</cite></p></li>
</ul>
</div>
<div class="section" id="id10">
<h3>Changed<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>fix for detection of compressed files <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.parse_gff" title="mgkit.io.gff.parse_gff"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.parse_gff()</span></code></a></p></li>
<li><p>Fix for compressed files already opened in <a class="reference internal" href="api/mgkit.io.utils.html#mgkit.io.utils.open_file" title="mgkit.io.utils.open_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.utils.open_file()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.align.html#mgkit.align.SamtoolsDepth" title="mgkit.align.SamtoolsDepth"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.align.SamtoolsDepth</span></code></a>: several optimisations and changes to support a scanning approach, instead of lookup table. No exception is raised when a sequence is not found in the file, instead assuming that the coverage is 0</p></li>
<li><p><a class="reference internal" href="api/mgkit.align.html#mgkit.align.read_samtools_depth" title="mgkit.align.read_samtools_depth"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.align.read_samtools_depth()</span></code></a> was changed, and now it returns lists instead of numpy arrays - this increases the speed of reading to about 3-4x in some tests</p></li>
<li><p><a class="reference internal" href="api/mgkit.align.html#mgkit.align.read_samtools_depth" title="mgkit.align.read_samtools_depth"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.align.read_samtools_depth()</span></code></a> also assumes that lines read have a ‘n’ at the end and avoid using <cite>strip</cite> this should be a safe assumptions under Pyuthon3</p></li>
<li><p><a class="reference internal" href="api/mgkit.align.html#mgkit.align.SamtoolsDepth" title="mgkit.align.SamtoolsDepth"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.align.SamtoolsDepth</span></code></a> now uses a weakref.WeakValueDictionary for <code class="xref py py-class docutils literal notranslate"><span class="pre">SamtoolsDepth.data</span></code> to improve release of memory</p></li>
</ul>
</div>
</div>
<div class="section" id="id11">
<h2>0.4.1<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h2>
<p>Sanity checks for several mistakes, including never changed the Programming language version in the setup.py from 2.7. Tested installation under Python 3.6, with tox. Also removed the last bit of code using progressbar2.</p>
</div>
<div class="section" id="id12">
<h2>0.4.0<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h2>
<p>This version was tested under Python 3.5, but the tests (with tox) were run also under Python 2.7. However, from the next release Python 2.7 will be removed gradually (as Python 2.7 won’t be supported/patched anymore from next year).</p>
<div class="section" id="id13">
<h3>Added<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<p>Added <cite>–progress</cite> option to several scripts</p>
<p><a class="reference internal" href="api/mgkit.counts.glm.html#module-mgkit.counts.glm" title="mgkit.counts.glm"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.counts.glm</span></code></a>:</p>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.counts.glm.html#mgkit.counts.glm.optimise_alpha_scipy" title="mgkit.counts.glm.optimise_alpha_scipy"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.glm.optimise_alpha_scipy()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.counts.glm.html#mgkit.counts.glm.optimise_alpha_scipy_function" title="mgkit.counts.glm.optimise_alpha_scipy_function"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.glm.optimise_alpha_scipy_function()</span></code></a></p></li>
</ul>
<p><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.graphs</span></code></p>
<ul class="simple">
<li><p><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.graphs.Reaction</span></code></p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.graphs.merge_kgmls()</span></code></p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.graphs.parse_kgml_reactions()</span></code></p></li>
</ul>
<p><a class="reference internal" href="api/mgkit.taxon.html#module-mgkit.taxon" title="mgkit.taxon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.taxon</span></code></a>:</p>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.is_ranked_below" title="mgkit.taxon.Taxonomy.is_ranked_below"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.Taxonomy.is_ranked_below()</span></code></a></p></li>
</ul>
</div>
<div class="section" id="id14">
<h3>Changed<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<p>Requires <cite>pandas</cite> version &gt;=0.24 because now a pandas.SparseArray is used for <cite>add-gff-info cov_samtools</cite>. Before, when reading the depth files from <strong>samtools</strong> the array for each sequence was kept in memory, while now only the ones in the GFF file are used.</p>
<p><a class="reference internal" href="api/mgkit.align.html#module-mgkit.align" title="mgkit.align"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.align</span></code></a>:</p>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.align.html#mgkit.align.SamtoolsDepth" title="mgkit.align.SamtoolsDepth"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.align.SamtoolsDepth</span></code></a>: uses pandas.SparseArray now. It should use less memory, but needs pandas version &gt; 0.24</p></li>
<li><p><a class="reference internal" href="api/mgkit.align.html#mgkit.align.read_samtools_depth" title="mgkit.align.read_samtools_depth"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.align.read_samtools_depth()</span></code></a>: now returns 3 array, instead of 2. Also added seq_ids to skip lines</p></li>
</ul>
<p><a class="reference internal" href="api/mgkit.io.gff.html#module-mgkit.io.gff" title="mgkit.io.gff"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.gff</span></code></a></p>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.from_gff" title="mgkit.io.gff.from_gff"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.gff.from_gff</span></code></a>: added encoding parameter</p></li>
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.parse_gff" title="mgkit.io.gff.parse_gff"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.gff.parse_gff</span></code></a>: In some cases ASCII decoding is not enough, so it is parametrised now</p></li>
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.split_gff_file" title="mgkit.io.gff.split_gff_file"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.gff.split_gff_file</span></code></a>: added encoding parameter</p></li>
</ul>
<p><a class="reference internal" href="api/mgkit.mappings.eggnog.html#module-mgkit.mappings.eggnog" title="mgkit.mappings.eggnog"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.mappings.eggnog</span></code></a>:</p>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo" title="mgkit.mappings.eggnog.NOGInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.mappings.eggnog.NOGInfo</span></code></a>: made file reading compatible with Python 3</p></li>
</ul>
<p><a class="reference internal" href="api/mgkit.snps.funcs.html#module-mgkit.snps.funcs" title="mgkit.snps.funcs"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.snps.funcs</span></code></a>:</p>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.funcs.combine_sample_snps()</span></code></a>: added store_uids</p></li>
</ul>
</div>
<div class="section" id="deprecated">
<h3>Deprecated<a class="headerlink" href="#deprecated" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.io.blast.html#mgkit.io.blast.add_blast_result_to_annotation" title="mgkit.io.blast.add_blast_result_to_annotation"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.blast.add_blast_result_to_annotation()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.read_taxonomy" title="mgkit.taxon.Taxonomy.read_taxonomy"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.Taxonomy.read_taxonomy()</span></code></a>: use Taxonomy.read_from_ncbi_dump()</p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.Taxonomy.parse_uniprot_taxon()</span></code></p></li>
</ul>
</div>
<div class="section" id="tests">
<h3>Tests<a class="headerlink" href="#tests" title="Permalink to this headline">¶</a></h3>
<p>Removed the last portions that used <cite>nosetets</cite> and better integrated pytest with setup.py. Now uses <a class="reference external" href="https://ci.appveyor.com/project/setsuna80/mgkit">AppVeyor</a> for testing the build and running tests under Python 3.</p>
<p>In cases where the testing environment has no or limited internet connection, tests that require an internet connection can be skipped by setting the following environment variable before running the tests:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ export MGKIT_TESTS_CONN_SKIP=T
</pre></div>
</div>
</div>
</div>
<div class="section" id="id15">
<h2>0.3.4<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h2>
<p>General cleanup and testing release. Major changes:</p>
<ul class="simple">
<li><p>general moving to Python2 (2.7) and Python3 (3.5+) support, using the future package and when convenient checks for the version of python installed</p></li>
<li><p>setup includes now all the optional dependencies, since this makes it easier to deal with conda environments</p></li>
<li><p>move to pytest from nose, since it allows some functionality that interests me, along with the reorganisation of the test modules and skips of tests that cannot be executed (like mongodb)</p></li>
<li><p>move from urlib to using <cite>requests</cite>, which also helps with python3 support</p></li>
<li><p>more careful with some dependencies, like the lzma module and msgpack</p></li>
<li><p>addition of more tests, to help the porting to python3, along with a tox configuration</p></li>
<li><p><code class="xref py py-mod docutils literal notranslate"><span class="pre">matplotlib.pyplot</span></code> is still in the <a class="reference internal" href="api/mgkit.plots.unused.html#module-mgkit.plots.unused" title="mgkit.plots.unused"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.plots.unused</span></code></a>, but it is not imported when the parent package is, now. It is still needed in the <a class="reference internal" href="api/mgkit.plots.utils.html#module-mgkit.plots.utils" title="mgkit.plots.utils"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.plots.utils</span></code></a> functions, so the import has been moved inside the function. This should help with virtual environments and test suites</p></li>
<li><p>renamed <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonomy" title="mgkit.taxon.UniprotTaxonomy"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy</span></code></a> to <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy" title="mgkit.taxon.Taxonomy"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.taxon.Taxonomy</span></code></a>, since it’s really NCBI taxonomy and it’s preferred to download the data from there. Same for <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonTuple" title="mgkit.taxon.UniprotTaxonTuple"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonTuple</span></code></a> to <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.TaxonTuple" title="mgkit.taxon.TaxonTuple"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.taxon.TaxonTuple</span></code></a>, with an alias for old name there, but will be removed in a later version</p></li>
<li><p><cite>download_data</cite> was removed. Taxonomy should be downloaded using <cite>download-taxonomy.sh</cite>, and the <a class="reference internal" href="api/mgkit.mappings.html#module-mgkit.mappings" title="mgkit.mappings"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.mappings</span></code></a> is in need of refactoring to remove old and now ununsed functionality</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.get_ranked_id" title="mgkit.taxon.Taxonomy.get_ranked_id"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.Taxonomy.get_ranked_id()</span></code></a></p></li>
<li><p>using a sphinx plugin to render the jupyter notebooks instead of old solution</p></li>
<li><p>rerun most of the tutorial and updated commands for newest available software (samtools/bcftools) and preferred the SNP calling from bcftools</p></li>
</ul>
<div class="section" id="scripts">
<h3>Scripts<a class="headerlink" href="#scripts" title="Permalink to this headline">¶</a></h3>
<p>This is a summary of notable changes, it is advised to check the changes in the command line interface for several scripts</p>
<ul class="simple">
<li><p>changed several scripts command line interface, to adapt to the use of <em>click</em></p></li>
<li><p><cite>taxon-utils lca</cite> has one options only to specify the output format, also adding the option to output a format that can be used by <cite>add-gff-info addtaxa</cite></p></li>
<li><p><cite>taxon-utils filter</cite> support the filtering of table files, when they are in a 2-columns format, such as those that are downloaded by <cite>download-ncbi-taxa.sh</cite></p></li>
<li><p>removed the <em>eggnog</em> and <em>taxonomy</em> commands from <cite>add-gff-info</cite>, the former since it’s not that useful, the latter because it’s possible to achieve the same results using taxon-utils with the new output option</p></li>
<li><p>removed the <em>rand</em> command of <cite>fastq-utils</cite> since it was only for testing and the FastQ parser is the one from <a class="reference internal" href="api/mgkit.io.fastq.html#module-mgkit.io.fastq" title="mgkit.io.fastq"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.fastq</span></code></a></p></li>
<li><p>substantial changes where made to commands <em>values</em> and <em>sequence</em> of the <cite>filter-gff</cite> script</p></li>
<li><p><cite>sampling-utils rand_seq</cite> now can save the model used and reload it</p></li>
<li><p>removed <cite>download_data</cite> and <cite>download_profiles</cite>, since they are not going to be used in the next tutorial and it is preferred now to use BLAST and then find the LCA with <cite>taxon-utils</cite></p></li>
</ul>
</div>
<div class="section" id="python3">
<h3>Python3<a class="headerlink" href="#python3" title="Permalink to this headline">¶</a></h3>
<p>At the time of writing all tests pass on Python 3.5, but more tests are needed, along with some new ones for the blast parser and the scripts. Some important changes:</p>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation" title="mgkit.io.gff.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation</span></code></a> uses its <em>uid</em> to hash the instance. This allows the use in sets (mainly for filtering). However, hashing is not supported in <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange" title="mgkit.io.gff.GenomicRange"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.GenomicRange</span></code></a>.</p></li>
<li><p><a class="reference internal" href="api/mgkit.io.utils.html#mgkit.io.utils.open_file" title="mgkit.io.utils.open_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.utils.open_file()</span></code></a> now <em>always</em> opens and writes files in binary mode. This is one of the suggestions to keep compatibility between 2.x and 3.x. So if used directly the output must be decoded from <em>ascii</em>, which is the format used in text files (at least bioinformatics). However, this is not required for the parsers, like <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.parse_gff" title="mgkit.io.gff.parse_gff"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.parse_gff()</span></code></a>, <a class="reference internal" href="api/mgkit.io.fasta.html#mgkit.io.fasta.load_fasta" title="mgkit.io.fasta.load_fasta"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.fasta.load_fasta()</span></code></a> along with others (and the correspective <em>write_</em> functions): they return unicode strings when parsing and decode into <em>ascii</em> when writing.</p></li>
</ul>
<p>In general new projects will be worked on using Python 3.5 and the next releases will prioritise that (0.4.0 and later). If bugfixes are needed and Python 3 cannot be used, this version branch (0.3.x) will be used to fix bugs for users.</p>
</div>
</div>
<div class="section" id="id16">
<h2>0.3.3<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id17">
<h3>Added<a class="headerlink" href="#id17" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>module <a class="reference internal" href="api/mgkit.counts.glm.html#module-mgkit.counts.glm" title="mgkit.counts.glm"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.counts.glm</span></code></a>, with functions used to help the fit of Generalised Linear Models (GLM)</p></li>
<li><p><a class="reference internal" href="api/mgkit.io.fastq.html#mgkit.io.fastq.load_fastq_rename" title="mgkit.io.fastq.load_fastq_rename"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.fastq.load_fastq_rename()</span></code></a></p></li>
<li><p>added <cite>sync</cite>, <cite>sample_stream</cite> and <cite>rand_seq</cite> commands to <cite>sampling-utils</cite> script</p></li>
<li><p><a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.extrapolate_model" title="mgkit.utils.sequence.extrapolate_model"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.extrapolate_model()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.qualities_model_constant" title="mgkit.utils.sequence.qualities_model_constant"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.qualities_model_constant()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.qualities_model_decrease" title="mgkit.utils.sequence.qualities_model_decrease"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.qualities_model_decrease()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.random_qualities" title="mgkit.utils.sequence.random_qualities"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.random_qualities()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.random_sequences" title="mgkit.utils.sequence.random_sequences"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.random_sequences()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.random_sequences_codon" title="mgkit.utils.sequence.random_sequences_codon"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.random_sequences_codon()</span></code></a></p></li>
<li><p><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.get_lineage_string()</span></code></p></li>
<li><p><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.parse_gtdb_lineage()</span></code></p></li>
<li><p><a class="reference internal" href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.get_gene_info_iter" title="mgkit.net.uniprot.get_gene_info_iter"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.uniprot.get_gene_info_iter()</span></code></a></p></li>
</ul>
</div>
<div class="section" id="id18">
<h3>Changed<a class="headerlink" href="#id18" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.io.fastq.html#mgkit.io.fastq.write_fastq_sequence" title="mgkit.io.fastq.write_fastq_sequence"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.fastq.write_fastq_sequence()</span></code></a></p></li>
<li><p>added <cite>seq_id</cite> as a special attribute to <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_attr" title="mgkit.io.gff.Annotation.get_attr"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.get_attr()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.from_prodigal_frag" title="mgkit.io.gff.from_prodigal_frag"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.from_prodigal_frag()</span></code></a> is tested and fixed</p></li>
<li><p>added cache in <a class="reference internal" href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.HDFDict" title="mgkit.utils.dictionary.HDFDict"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.utils.dictionary.HDFDict</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.sequence_gc_content" title="mgkit.utils.sequence.sequence_gc_content"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.sequence_gc_content()</span></code></a> now returns 0.5 when denominator is 0</p></li>
<li><p><cite>add-gff-info addtaxa -a</cite> now accept <cite>seq_id</cite> as lookup, to use output from <cite>taxon-utils lca</cite> (after cutting output)</p></li>
</ul>
</div>
<div class="section" id="id19">
<h3>Deprecated<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.io.fastq.html#mgkit.io.fastq.convert_seqid_to_old" title="mgkit.io.fastq.convert_seqid_to_old"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.fastq.convert_seqid_to_old()</span></code></a></p></li>
</ul>
</div>
</div>
<div class="section" id="id20">
<h2>0.3.2<a class="headerlink" href="#id20" title="Permalink to this headline">¶</a></h2>
<p>Removed deprecated code</p>
</div>
<div class="section" id="id21">
<h2>0.3.1<a class="headerlink" href="#id21" title="Permalink to this headline">¶</a></h2>
<p>This release adds several scripts and commands. Successive releases 0.3.x releases will be used to fix bugs and refine the APIs and CLI. Most importantly, since the publishing of the first paper using the framework, the releases will go torward the removal of as much deprecated code as possible. At the same time, a general review of the code to be able to run on Python3 (probably via the <em>six</em> package) will start. The general idea is to reach as a full removal of legacy code in 0.4.0, while full Python3 compatibility is the aim of 0.5.0, which also means dropping dependencies that are not compatible with Python3.</p>
<div class="section" id="id22">
<h3>Added<a class="headerlink" href="#id22" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.graphs.from_kgml()</span></code> to make a graph from a KGML file (allows for directionality)</p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.graphs.add_module_compounds()</span></code>: updates a graph with compounds information as needed</p></li>
<li><p><a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.parse_reaction" title="mgkit.kegg.parse_reaction"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.kegg.parse_reaction()</span></code></a>: parses a reaction equation from Kegg</p></li>
<li><p>added <cite>–no-frame</cite> option to <a class="reference internal" href="scripts/hmmer2gff.html#hmmer2gff"><span class="std std-ref">hmmer2gff - Convert HMMER output to GFF</span></a>, to use non translated protein sequences. Also changed the <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.from_hmmer" title="mgkit.io.gff.from_hmmer"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.from_hmmer()</span></code></a> function to enable this behaviour</p></li>
<li><p>added options <cite>–num-gt</cite> and <cite>–num-lt</cite> to the <em>values</em> command of <a class="reference internal" href="scripts/filter-gff.html#filter-gff"><span class="std std-ref">filter-gff - Filter GFF annotations</span></a> to filter based on <cite>&gt;</cite> and <cite>&lt;</cite> inequality, in addition to <cite>&gt;=</cite> and <cite>&lt;=</cite></p></li>
<li><p>added <em>uid</em> as command in <a class="reference internal" href="scripts/fasta-utils.html#fasta-utils"><span class="std std-ref">fasta-utils - Fasta Utilities</span></a> to make unique fasta headers</p></li>
<li><p>methods to make <a class="reference internal" href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB" title="mgkit.db.mongo.GFFDB"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.db.mongo.GFFDB</span></code></a> to behave like a dictionary (an annotation <em>uid</em> can be used as a key to retrieve it, instead of a query), this includes the possibility to iterate over it, but what is yielded are the values, not the keys (i.e. <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation" title="mgkit.io.gff.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation</span></code></a> instances, not <em>uid</em>)</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.counts.func.html#mgkit.counts.func.from_gff" title="mgkit.counts.func.from_gff"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.func.from_gff()</span></code></a> to load count data stored inside a GFF, as is the case when the <em>counts</em> command of <a class="reference internal" href="scripts/add-gff-info.html#add-gff-info"><span class="std std-ref">add-gff-info - Add informations to GFF annotations</span></a> is used’</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.conv" title="mgkit.kegg.KeggClientRest.conv"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.kegg.KeggClientRest.conv()</span></code></a> and <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.find" title="mgkit.kegg.KeggClientRest.find"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.kegg.KeggClientRest.find()</span></code></a> operations to <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest" title="mgkit.kegg.KeggClientRest"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.kegg.KeggClientRest</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest" title="mgkit.kegg.KeggClientRest"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.kegg.KeggClientRest</span></code></a> now caches calls to several methods. The cache can be written to disk using <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.write_cache" title="mgkit.kegg.KeggClientRest.write_cache"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.kegg.KeggClientRest.write_cache()</span></code></a> or emptied via <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.empty_cache" title="mgkit.kegg.KeggClientRest.empty_cache"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.kegg.KeggClientRest.empty_cache()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.merge_dictionaries" title="mgkit.utils.dictionary.merge_dictionaries"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.dictionary.merge_dictionaries()</span></code></a> to merge multiple dictionaries where the keys contain different values</p></li>
<li><p>added a Docker file to make a preconfigured mgkit/jupyter build</p></li>
<li><p>added C functions (using <a class="reference external" href="www.cython.org">Cython</a>) for tetramer/kmer counting. The C functions are the default, with the pure python implementation having a <em>_</em> appended to their names. This is because the Cython functions cannot have docstrings</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.annotation_coverage_sorted" title="mgkit.io.gff.annotation_coverage_sorted"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.annotation_coverage_sorted()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.to_dict" title="mgkit.io.gff.Annotation.to_dict"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.to_dict()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.plots.utils.html#mgkit.plots.utils.legend_patches" title="mgkit.plots.utils.legend_patches"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.utils.legend_patches()</span></code></a> to create matplotlib patches, to be in legends</p></li>
<li><p>added scripts download IDs to taxa tables from NCBI/Uniprot</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.utils.html#mgkit.io.utils.group_tuples_by_key" title="mgkit.io.utils.group_tuples_by_key"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.utils.group_tuples_by_key()</span></code></a></p></li>
<li><p>added <em>cov</em> command to <a class="reference internal" href="scripts/get-gff-info.html#get-gff-info"><span class="std std-ref">get-gff-info - Extract informations to GFF annotations</span></a> and <a class="reference internal" href="scripts/filter-gff.html#filter-gff"><span class="std std-ref">filter-gff - Filter GFF annotations</span></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.fasta.html#mgkit.io.fasta.load_fasta_prodigal" title="mgkit.io.fasta.load_fasta_prodigal"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.fasta.load_fasta_prodigal()</span></code></a>, to load the fasta file from prodigal for called genes (tested on aminoacids)</p></li>
<li><p>added option to output a JSON file to the <em>lca</em> command in ref:<cite>taxon-utils</cite> and <em>cov</em> command in <a class="reference internal" href="scripts/get-gff-info.html#get-gff-info"><span class="std std-ref">get-gff-info - Extract informations to GFF annotations</span></a></p></li>
<li><p>added a bash script, <em>sort-gff.sh</em> to help sort a GFF</p></li>
<li><p>added <code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.get_lineage()</span></code> which simplifies the use of <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.get_lineage" title="mgkit.taxon.get_lineage"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.get_lineage()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.fastq.html#mgkit.io.fastq.load_fastq" title="mgkit.io.fastq.load_fastq"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.fastq.load_fastq()</span></code></a> as a simple parser for fastq files</p></li>
<li><p>added a new script, <a class="reference internal" href="scripts/sampling-utils.html#sampling-utils"><span class="std std-ref">sampling-utils - Resampling Utilities</span></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.utils.common.html#mgkit.utils.common.union_ranges" title="mgkit.utils.common.union_ranges"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.common.union_ranges()</span></code></a> and <a class="reference internal" href="api/mgkit.utils.common.html#mgkit.utils.common.complement_ranges" title="mgkit.utils.common.complement_ranges"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.common.complement_ranges()</span></code></a></p></li>
<li><p>added <em>to_hdf</em> command to <a class="reference internal" href="scripts/taxon_utils.html#taxon-utils"><span class="std std-ref">taxon-utils - Taxonomy Utilities</span></a> to create a HDF5 file to lookup taxa tables from NCBI/Uniprot</p></li>
<li><p>added <cite>–hdf-table</cite> option to <em>addtaxa</em> command in <a class="reference internal" href="scripts/add-gff-info.html#add-gff-info"><span class="std std-ref">add-gff-info - Add informations to GFF annotations</span></a></p></li>
<li><p><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.add_taxon()</span></code>, <code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.add_lineage()</span></code> and <code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.drop_taxon()</span></code></p></li>
</ul>
</div>
<div class="section" id="id23">
<h3>Changed<a class="headerlink" href="#id23" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>changed <em>domain</em> to <em>superkingdom</em> as for NCBI taxonomy in <code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.read_from_gtdb_taxonomy()</span></code></p></li>
<li><p>updated scripts documentation to include installed but non advertised scripts (like <span class="xref std std-ref">translate_seq</span>)</p></li>
<li><p><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.kegg.KeggReaction</span></code> was reworked to only store the equation information</p></li>
<li><p>some commands in <a class="reference internal" href="scripts/fastq-utils.html#fastq-utils"><span class="std std-ref">fastq-utils - Fastq Utilities</span></a> did not support standard in/out, also added the script usage to the script details</p></li>
<li><p><span class="xref std std-ref">translate_seq</span> now supports standard in/out</p></li>
<li><p>added <em>haplotypes</em> parameter to <a class="reference internal" href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.funcs.combine_sample_snps()</span></code></a></p></li>
<li><p>an annotation from <a class="reference internal" href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB" title="mgkit.db.mongo.GFFDB"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.db.mongo.GFFDB</span></code></a> now doesn’t include the lineage, because it conflicts with the string used in a GFF file</p></li>
<li><p>an <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.coverage" title="mgkit.io.gff.Annotation.coverage"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.coverage()</span></code></a> now returns a <cite>float</cite> instead od a <cite>int</cite></p></li>
<li><p>moved code from package <a class="reference internal" href="api/mgkit.io.html#module-mgkit.io" title="mgkit.io"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io</span></code></a> to <a class="reference internal" href="api/mgkit.io.utils.html#module-mgkit.io.utils" title="mgkit.io.utils"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.utils</span></code></a></p></li>
<li><p>changed behaviour of <a class="reference internal" href="api/mgkit.utils.common.html#mgkit.utils.common.union_range" title="mgkit.utils.common.union_range"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.common.union_range()</span></code></a></p></li>
<li><p>removed <code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utills.common.range_substract_()</span></code></p></li>
<li><p>added <em>progressbar2</em> as installation requirement</p></li>
<li><p>changed how <code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.find_by_name()</span></code></p></li>
</ul>
</div>
<div class="section" id="fixed">
<h3>Fixed<a class="headerlink" href="#fixed" title="Permalink to this headline">¶</a></h3>
<p>Besides smaller fixes:</p>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.plots.abund.html#mgkit.plots.abund.draw_circles" title="mgkit.plots.abund.draw_circles"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.abund.draw_circles()</span></code></a> behaviour when <cite>sizescale</cite> doesn’t have the same shape as <cite>order</cite></p></li>
<li><p>parser is now correct for <a class="reference internal" href="scripts/taxon_utils.html#taxon-utils"><span class="std std-ref">taxon-utils - Taxonomy Utilities</span></a>, to include the <a class="reference external" href="https://github.com/marbl/Krona/wiki">Krona</a> options</p></li>
<li><p>ondition when a blast output is empty, hence <em>lineno</em> is not initialised when a message is logged</p></li>
</ul>
</div>
<div class="section" id="id24">
<h3>Deprecated<a class="headerlink" href="#id24" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><span class="xref std std-ref">translate_seq</span> will be removed in version 0.4.0, instead use the similar command in <a class="reference internal" href="scripts/fasta-utils.html#fasta-utils"><span class="std std-ref">fasta-utils - Fasta Utilities</span></a></p></li>
</ul>
</div>
</div>
<div class="section" id="id25">
<h2>0.3.0<a class="headerlink" href="#id25" title="Permalink to this headline">¶</a></h2>
<p>A lot of bugs were fixed in this release, especially for reading NCBI taxonomy and using the <em>msgpack</em> format to save a UniprotTaxonomy instance. Also added a tutorial for profiling a microbial community using MGKit and BLAST (<a class="reference internal" href="pipeline/blast2lca.html#blast2lca"><span class="std std-ref">Profile a Community with BLAST</span></a>)</p>
<div class="section" id="id26">
<h3>Added<a class="headerlink" href="#id26" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.align.html#mgkit.align.read_samtools_depth" title="mgkit.align.read_samtools_depth"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.align.read_samtools_depth()</span></code></a> to read the samtools depth format iteratively (returns a generator)</p></li>
<li><p><a class="reference internal" href="api/mgkit.align.html#mgkit.align.SamtoolsDepth" title="mgkit.align.SamtoolsDepth"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.align.SamtoolsDepth</span></code></a>, used to cache the samtools depth format, while requesting region coverage</p></li>
<li><p><a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.find_submodules" title="mgkit.kegg.KeggModule.find_submodules"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.kegg.KeggModule.find_submodules()</span></code></a>, <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.parse_entry2" title="mgkit.kegg.KeggModule.parse_entry2"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.kegg.KeggModule.parse_entry2()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.get_mapping_level" title="mgkit.mappings.enzyme.get_mapping_level"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.mappings.enzyme.get_mapping_level()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.cache_dict_file" title="mgkit.utils.dictionary.cache_dict_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.dictionary.cache_dict_file()</span></code></a> to cache a large dictionary file (tab separated file with 2 columns), an example of its usage is in the documentation</p></li>
<li><p><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.read_from_gtdb_taxonomy()</span></code> to read a custom taxonomy from a tab separated file. The taxon_id are not guaranteed to be stable between runs</p></li>
<li><p>added <em>cov_samtools</em> to <em>add-gff-info</em> script</p></li>
<li><p>added <a class="reference internal" href="scripts/fasta-utils.html#module-mgkit.workflow.fasta_utils" title="mgkit.workflow.fasta_utils"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.workflow.fasta_utils</span></code></a> and correspondent script <em>fasta-utils</em></p></li>
<li><p>added options <em>-k</em> and <em>-kt</em> to <em>taxon_utils</em>, which outputs a file that can be used with Krona <em>ktImportText</em> (needs to use <em>-q</em> with this script)</p></li>
</ul>
</div>
<div class="section" id="id27">
<h3>Changed<a class="headerlink" href="#id27" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>added <em>no_zero</em> parameter to <a class="reference internal" href="api/mgkit.io.blast.html#mgkit.io.blast.parse_accession_taxa_table" title="mgkit.io.blast.parse_accession_taxa_table"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.blast.parse_accession_taxa_table()</span></code></a></p></li>
<li><p>changed behaviour of <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggModule" title="mgkit.kegg.KeggModule"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.kegg.KeggModule</span></code></a> and some of its methods.</p></li>
<li><p>added <em>with_last</em> parameter to <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.get_lineage" title="mgkit.taxon.get_lineage"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.get_lineage()</span></code></a></p></li>
<li><p>added <em>–split</em> option to <em>add-gff-info exp_syn</em> and <em>get-gff-info sequence</em> scripts, to emulate BLAST behaviour in parsing sequence headers</p></li>
<li><p>added <em>-c</em> option to <em>add-gff-info addtaxa</em></p></li>
</ul>
</div>
</div>
<div class="section" id="id28">
<h2>0.2.5<a class="headerlink" href="#id28" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id29">
<h3>Changed<a class="headerlink" href="#id29" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>added the <em>only_ranked</em> argument to <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.get_lineage" title="mgkit.taxon.get_lineage"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.get_lineage()</span></code></a></p></li>
<li><p><em>add-gff-info addtaxa</em> (<a class="reference internal" href="scripts/add-gff-info.html#add-gff-info"><span class="std std-ref">add-gff-info - Add informations to GFF annotations</span></a>) doesn’t preload the GFF file if a dictionary is used instead of the taxa table</p></li>
<li><p><em>blast2gff blastdb</em> ((<a class="reference internal" href="scripts/blast2gff.html#blast2gff"><span class="std std-ref">blast2gff - Convert BLAST output to GFF</span></a>) offers more options to control the format of the header in the DB used</p></li>
<li><p>added the <em>sequence</em> command to <em>filter-gff</em> (<a class="reference internal" href="scripts/filter-gff.html#filter-gff"><span class="std std-ref">filter-gff - Filter GFF annotations</span></a>), to filter all annotations on a per-sequence base, based on mean bitscore or other comparisons</p></li>
</ul>
</div>
<div class="section" id="id30">
<h3>Added<a class="headerlink" href="#id30" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>added <a class="reference internal" href="api/mgkit.counts.func.html#mgkit.counts.func.load_counts_from_gff" title="mgkit.counts.func.load_counts_from_gff"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.func.load_counts_from_gff()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.blast.html#mgkit.io.blast.parse_accession_taxa_table" title="mgkit.io.blast.parse_accession_taxa_table"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.blast.parse_accession_taxa_table()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.plots.abund.html#mgkit.plots.abund.draw_axis_internal_triangle" title="mgkit.plots.abund.draw_axis_internal_triangle"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.abund.draw_axis_internal_triangle()</span></code></a></p></li>
<li><p>added representation of <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonomy" title="mgkit.taxon.UniprotTaxonomy"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy</span></code></a>, it show the number of taxa in the instance</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.last_common_ancestor_multiple" title="mgkit.taxon.last_common_ancestor_multiple"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.last_common_ancestor_multiple()</span></code></a></p></li>
<li><p>added <em>taxon_utils</em> (<a class="reference internal" href="scripts/taxon_utils.html#taxon-utils"><span class="std std-ref">taxon-utils - Taxonomy Utilities</span></a>) to filter GFF based on their taxonomy and find the last common ancestor for a reference sequence based on either GFF annotations or a list of taxon_ids (in a text file)</p></li>
</ul>
</div>
</div>
<div class="section" id="id31">
<h2>0.2.4<a class="headerlink" href="#id31" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id32">
<h3>Changed<a class="headerlink" href="#id32" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.get_contigs_info" title="mgkit.utils.sequence.get_contigs_info"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.get_contigs_info()</span></code></a> now accepts a dictionary name-&gt;seq or a list of sequences, besides a file name (r536)</p></li>
<li><p><em>add-gff-info</em> <strong>counts</strong> command now removes trailing commas from the samples list</p></li>
<li><p>the axes are turned off after the dendrogram is plo</p></li>
</ul>
</div>
<div class="section" id="id33">
<h3>Fixed<a class="headerlink" href="#id33" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>the <strong>snp_parser</strong> script requirements were set wrong in <em>setup.py</em> (r540)</p></li>
<li><p>uncommented lines to download sample data to build documentation (r533)</p></li>
<li><p><em>add-gff-info</em> <strong>uniprot</strong> command now writes the <em>lineage</em> attribute correctly (r538)</p></li>
</ul>
</div>
</div>
<div class="section" id="id34">
<h2>0.2.3<a class="headerlink" href="#id34" title="Permalink to this headline">¶</a></h2>
<p>The installation dependencies are more flexible, with only <em>numpy</em> as being <strong>required</strong>. To install every needed packages, you can use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip install mgkit[full]
</pre></div>
</div>
<div class="section" id="id35">
<h3>Added<a class="headerlink" href="#id35" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>new option to pass the <em>query sequences</em> to <strong>blast2gff</strong>, this allows to add the correct frame of the annotation in the GFF</p></li>
<li><p>added the attributes <em>evalue</em>, <em>subject_start</em> and <em>subject_end</em> to the output of <em>blast2gff</em>. The subject start and end position allow to understand on which frame of the <em>subject sequence</em> the match was found</p></li>
<li><p>added the options to annotate the heatmap with the numbers. Also updated the relative example notebook</p></li>
<li><p>Added the option to reads the taxonomy from NCBI dump files, using <code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.read_from_ncbi_dump()</span></code>. This make it faster to get the taxonomy file</p></li>
<li><p>added argument to return information from <a class="reference internal" href="api/mgkit.net.embl.html#mgkit.net.embl.datawarehouse_search" title="mgkit.net.embl.datawarehouse_search"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.embl.datawarehouse_search()</span></code></a>, in the form of tab separated data. The argument <em>fields</em> can be used when <em>display</em> is set to <strong>report</strong>. An example on how to use it is in the function documentation</p></li>
<li><p>added a bash script <em>download-taxonomy.sh</em> that download the taxonomy</p></li>
<li><p>added script <em>venv-docs.sh</em> to build the documentation in HTML under a virtual environment. matplotlib on MacOS X raises a RuntimeError, because of a bug in <a class="reference external" href="https://github.com/pypa/virtualenv/issues/54">virtualenv</a>, the documentation can be first build with this, after the script <em>create-apidoc.sh</em> is create the API documentation. The rest of the documentation (e.g. the PDF) can be created with <em>make</em> as usual, afterwards</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.net.pfam.html#module-mgkit.net.pfam" title="mgkit.net.pfam"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.net.pfam</span></code></a>, with only one function at the moment, that returns the descriptions of the families.</p></li>
<li><p>added <em>pfam</em> command to <em>add-gff-info</em>, using the mentioned function, it adds the description of the Pfam families in the GFF file</p></li>
<li><p>added a new exception, used internally when an additional dependency is needed</p></li>
</ul>
</div>
<div class="section" id="id36">
<h3>Changed<a class="headerlink" href="#id36" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>using the NCBI taxonomy dump has two side effects:</p>
<blockquote>
<div><ul class="simple">
<li><p>the scientific/common names are kept as is, not lower cased as was before</p></li>
<li><p>a <em>merged</em> file is provided for <em>taxon_id</em> that changed. While the old taxon_id is kept in the taxonomy, this point to the new taxon, to keep backward compatibility</p></li>
</ul>
</div></blockquote>
</li>
<li><p>renamed the <em>add-gff-info</em> <em>gitaxa</em> command to <em>addtaxa</em>. It now accepts more data sources (dictionaries) and is more general</p></li>
<li><p>changed <a class="reference internal" href="api/mgkit.net.embl.html#mgkit.net.embl.datawarehouse_search" title="mgkit.net.embl.datawarehouse_search"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.embl.datawarehouse_search()</span></code></a> to automatically set the limit at 100,000 records</p></li>
<li><p>the taxonomy can now be saved using <a class="reference external" href="https://github.com/msgpack/msgpack-python">msgpack</a>, making it faster to read/write it. It’s also more compact and better compression ratio</p></li>
<li><p>the <a class="reference internal" href="api/mgkit.plots.heatmap.html#mgkit.plots.heatmap.grouped_spine" title="mgkit.plots.heatmap.grouped_spine"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.heatmap.grouped_spine()</span></code></a> now accept the rotation of the labels as option</p></li>
<li><p>added option to use another attribute for the <em>gene_id</em> in the <em>get-gff-info</em> script <em>gtf</em> command</p></li>
<li><p>added a function to compare the version of MGKit used, throwing a warning, when it’s different (<a class="reference internal" href="api/mgkit.html#mgkit.check_version" title="mgkit.check_version"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.check_version()</span></code></a>)</p></li>
<li><p>removed test for old SNPs structures and added the same tests for the new one</p></li>
<li><p><a class="reference internal" href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP" title="mgkit.snps.classes.GeneSNP"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.snps.classes.GeneSNP</span></code></a> now caches the number of synonymous and non-synonymous SNPs for better speed</p></li>
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.__contains__" title="mgkit.io.gff.GenomicRange.__contains__"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.GenomicRange.__contains__()</span></code></a> now also accepts a tuple (start, end) or another GenomicRange instance</p></li>
</ul>
</div>
<div class="section" id="id37">
<h3>Fixed<a class="headerlink" href="#id37" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>a bug in the <em>gitaxa</em> (now <em>addtaxa</em>) command: when a taxon_id was not found in the table, the wrong <em>taxon_name</em> and <em>lineage</em> was inserted</p></li>
<li><p>bug in <a class="reference internal" href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP" title="mgkit.snps.classes.GeneSNP"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.snps.classes.GeneSNP</span></code></a> that prevented the correct addition of values</p></li>
<li><p>fixed bug in <a class="reference internal" href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.flat_sample_snps" title="mgkit.snps.funcs.flat_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.funcs.flat_sample_snps()</span></code></a> with the new class</p></li>
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.parse_gff" title="mgkit.io.gff.parse_gff"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.parse_gff()</span></code></a> now correctly handles comment lines and stops parsing if the fasta file at the end of a GFF is found</p></li>
</ul>
</div>
</div>
<div class="section" id="id38">
<h2>0.2.2<a class="headerlink" href="#id38" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id39">
<h3>Added<a class="headerlink" href="#id39" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>new commands for the <strong>add-gff-info</strong> script (<a class="reference internal" href="scripts/add-gff-info.html#add-gff-info"><span class="std std-ref">add-gff-info - Add informations to GFF annotations</span></a>):</p>
<blockquote>
<div><ul class="simple">
<li><p><em>eggnog</em> to add information from eggNOG HMMs (at the moment the 4.5 Viral)</p></li>
<li><p><em>counts</em> and <em>fpkms</em> to add count data (correctly exported to mongodb)</p></li>
<li><p><em>gitaxa</em> to add taxonomy information directly from GI identifiers from NCBI</p></li>
</ul>
</div></blockquote>
</li>
<li><p>added <em>blastdb</em> command to <strong>blast2gff</strong> script (<a class="reference internal" href="scripts/blast2gff.html#blast2gff"><span class="std std-ref">blast2gff - Convert BLAST output to GFF</span></a>)</p></li>
<li><p>updated <a class="reference internal" href="gff.html#gff-specs"><span class="std std-ref">MGKit GFF Specifications</span></a></p></li>
<li><p>added <em>gtf</em> command to <strong>get-gff-info</strong> script (<a class="reference internal" href="scripts/get-gff-info.html#get-gff-info"><span class="std std-ref">get-gff-info - Extract informations to GFF annotations</span></a>) to convert a GFF to GTF, that is accepted by <a class="reference external" href="http://bioinf.wehi.edu.au/featureCounts/">featureCounts</a>, in conjunction with the <em>counts</em> command of <strong>add-gff-info</strong></p></li>
<li><p>added method to <a class="reference internal" href="api/mgkit.snps.classes.html#mgkit.snps.classes.RatioMixIn.calc_ratio_flag" title="mgkit.snps.classes.RatioMixIn.calc_ratio_flag"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.snps.classes.RatioMixIn.calc_ratio_flag</span></code></a> to calculate special cases of pN/pS</p></li>
</ul>
</div>
<div class="section" id="id40">
<h3>Changed<a class="headerlink" href="#id40" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>added argument in functions of the <a class="reference internal" href="api/mgkit.snps.conv_func.html#module-mgkit.snps.conv_func" title="mgkit.snps.conv_func"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.snps.conv_func</span></code></a> to bypass the default filters</p></li>
<li><p>added <em>use_uid</em> argument to <a class="reference internal" href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.funcs.combine_sample_snps()</span></code></a> to use the <em>uid</em> instead of the <em>gene_id</em> when calculating pN/pS</p></li>
<li><p>added <em>flag_values</em> argument to <a class="reference internal" href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.funcs.combine_sample_snps()</span></code></a> to use <a class="reference internal" href="api/mgkit.snps.classes.html#mgkit.snps.classes.RatioMixIn.calc_ratio_flag" title="mgkit.snps.classes.RatioMixIn.calc_ratio_flag"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.snps.classes.RatioMixIn.calc_ratio_flag</span></code></a> instead of <a class="reference internal" href="api/mgkit.snps.classes.html#mgkit.snps.classes.RatioMixIn.calc_ratio" title="mgkit.snps.classes.RatioMixIn.calc_ratio"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.snps.classes.RatioMixIn.calc_ratio</span></code></a></p></li>
</ul>
</div>
<div class="section" id="removed">
<h3>Removed<a class="headerlink" href="#removed" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>deprecated code from the <strong>snps</strong> package</p></li>
</ul>
</div>
</div>
<div class="section" id="id41">
<h2>0.2.1<a class="headerlink" href="#id41" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id42">
<h3>Added<a class="headerlink" href="#id42" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>added <a class="reference internal" href="api/mgkit.db.mongo.html#module-mgkit.db.mongo" title="mgkit.db.mongo"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.db.mongo</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.db.dbm.html#module-mgkit.db.dbm" title="mgkit.db.dbm"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.db.dbm</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_mappings" title="mgkit.io.gff.Annotation.get_mappings"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.get_mappings()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.to_json" title="mgkit.io.gff.Annotation.to_json"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.to_json()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.to_mongodb" title="mgkit.io.gff.Annotation.to_mongodb"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.to_mongodb()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.from_json" title="mgkit.io.gff.from_json"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.from_json()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.from_mongodb" title="mgkit.io.gff.from_mongodb"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.from_mongodb()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.get_lineage" title="mgkit.taxon.get_lineage"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.get_lineage()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.get_contigs_info" title="mgkit.utils.sequence.get_contigs_info"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.get_contigs_info()</span></code></a></p></li>
<li><p>added <cite>mongodb</cite> and <cite>dbm</cite> commands to script <cite>get-gff-info</cite></p></li>
<li><p>added <cite>kegg</cite> command to <cite>add-gff-info</cite> script, caching results and <cite>-d</cite> option to <cite>uniprot</cite> command</p></li>
<li><p>added <cite>-ft</cite> option to <cite>blast2gff</cite> script</p></li>
<li><p>added <cite>-ko</cite> option to <cite>download_profiles</cite></p></li>
<li><p>added new HMMER tutorial</p></li>
<li><p>added another notebook to the plot examples, for misc. tips</p></li>
<li><p>added a script that downloads from figshare the tutorial data]</p></li>
<li><p>added function to get an enzyme full name (<a class="reference internal" href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.get_enzyme_full_name" title="mgkit.mappings.enzyme.get_enzyme_full_name"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.mappings.enzyme.get_enzyme_full_name()</span></code></a>)</p></li>
<li><p>added example notebook for using GFF annotations and the <a class="reference internal" href="api/mgkit.db.dbm.html#module-mgkit.db.dbm" title="mgkit.db.dbm"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.db.dbm</span></code></a>, <a class="reference internal" href="api/mgkit.db.mongo.html#module-mgkit.db.mongo" title="mgkit.db.mongo"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.db.mongo</span></code></a> modules</p></li>
</ul>
</div>
<div class="section" id="id43">
<h3>Changed<a class="headerlink" href="#id43" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="api/mgkit.io.blast.html#mgkit.io.blast.parse_uniprot_blast" title="mgkit.io.blast.parse_uniprot_blast"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.blast.parse_uniprot_blast()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation" title="mgkit.io.gff.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange" title="mgkit.io.gff.GenomicRange"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.GenomicRange</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.from_hmmer" title="mgkit.io.gff.from_hmmer"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.from_hmmer()</span></code></a></p></li>
<li><p><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.read_taxonomy()</span></code></p></li>
<li><p><a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.parse_uniprot_taxon" title="mgkit.taxon.parse_uniprot_taxon"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.parse_uniprot_taxon()</span></code></a></p></li>
<li><p>changed behaviour of <cite>hmmer2gff</cite> script</p></li>
<li><p>changed tutorial notebook to specify the directory where the data is</p></li>
</ul>
</div>
<div class="section" id="id44">
<h3>Deprecated<a class="headerlink" href="#id44" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.filter.taxon.filter_taxonomy_by_lineage()</span></code></p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.filter.taxon.filter_taxonomy_by_rank()</span></code></p></li>
</ul>
</div>
<div class="section" id="id45">
<h3>Removed<a class="headerlink" href="#id45" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>removed old <cite>filter_gff</cite> script</p></li>
</ul>
</div>
</div>
<div class="section" id="id46">
<h2>0.2.0<a class="headerlink" href="#id46" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>added creation of wheel distribution</p></li>
<li><p>changes to ensure compatibility with alter pandas versions</p></li>
<li><p><a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_ec" title="mgkit.io.gff.Annotation.get_ec"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.get_ec()</span></code></a> now returns a set, reflected changes in tests</p></li>
<li><p>added a <cite>–cite</cite> option to scripts</p></li>
<li><p>fixes to tutorial</p></li>
<li><p>updated documentation for sphinx 1.3</p></li>
<li><p>changes to diagrams</p></li>
<li><p>added decoration to raise warnings for deprecated functions</p></li>
<li><p>added possibility for <a class="reference internal" href="api/mgkit.counts.func.html#mgkit.counts.func.load_sample_counts" title="mgkit.counts.func.load_sample_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.func.load_sample_counts()</span></code></a> info_dict to be a function instead of a dictionary</p></li>
<li><p>consolidation of some eggNOG structures</p></li>
<li><p>added more spine options in <a class="reference internal" href="api/mgkit.plots.heatmap.html#mgkit.plots.heatmap.grouped_spine" title="mgkit.plots.heatmap.grouped_spine"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.heatmap.grouped_spine()</span></code></a></p></li>
<li><p>added a <cite>length</cite> property to <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation" title="mgkit.io.gff.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation</span></code></a></p></li>
<li><p>changed <cite>filter-gff</cite> script to customise the filtering function, from the default one, also updated the relative documentation</p></li>
<li><p>fixed a few plot functions</p></li>
</ul>
</div>
<div class="section" id="id47">
<h2>0.1.16<a class="headerlink" href="#id47" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>changed default parameter for <a class="reference internal" href="api/mgkit.plots.boxplot.html#mgkit.plots.boxplot.add_values_to_boxplot" title="mgkit.plots.boxplot.add_values_to_boxplot"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.boxplot.add_values_to_boxplot()</span></code></a></p></li>
<li><p>Added <em>include_only</em> filter option to the default snp filters <a class="reference internal" href="api/mgkit.consts.html#mgkit.consts.DEFAULT_SNP_FILTER" title="mgkit.consts.DEFAULT_SNP_FILTER"><code class="xref py py-data docutils literal notranslate"><span class="pre">mgkit.consts.DEFAULT_SNP_FILTER</span></code></a></p></li>
<li><p>the default filter for SNPs now use an include only option, by default including only protozoa, archaea, fungi and bacteria in the matrix</p></li>
<li><p>added <em>widths</em> parameter to def <a class="reference internal" href="api/mgkit.plots.boxplot.html#mgkit.plots.boxplot.boxplot_dataframe" title="mgkit.plots.boxplot.boxplot_dataframe"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.boxplot.boxplot_dataframe()</span></code></a> function, added function <a class="reference internal" href="api/mgkit.plots.boxplot.html#mgkit.plots.boxplot.add_significance_to_boxplot" title="mgkit.plots.boxplot.add_significance_to_boxplot"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.boxplot.add_significance_to_boxplot()</span></code></a> and updated example boxplot notebook for new function example</p></li>
<li><p><em>use_dist</em> and <em>dist_func</em> parameters to the <a class="reference internal" href="api/mgkit.plots.heatmap.html#mgkit.plots.heatmap.dendrogram" title="mgkit.plots.heatmap.dendrogram"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.heatmap.dendrogram()</span></code></a> function</p></li>
<li><p>added a few constants and functions to calculate the distance matrices of taxa: <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.taxa_distance_matrix" title="mgkit.taxon.taxa_distance_matrix"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.taxa_distance_matrix()</span></code></a>, <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.distance_taxa_ancestor" title="mgkit.taxon.distance_taxa_ancestor"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.distance_taxa_ancestor()</span></code></a> and <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.distance_two_taxa" title="mgkit.taxon.distance_two_taxa"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.distance_two_taxa()</span></code></a></p></li>
<li><p><a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.link_ids" title="mgkit.kegg.KeggClientRest.link_ids"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.kegg.KeggClientRest.link_ids()</span></code></a> now accept a dictionary as list of ids</p></li>
<li><p>if the conversion of an Annotation attribute (first 8 columns) raises a ValueError in <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.from_gff" title="mgkit.io.gff.from_gff"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.from_gff()</span></code></a>, by default the parser keeps the string version (cases for phase, where is ‘.’ instead of a number)</p></li>
<li><p>treat cases where an attribute is set with no value in <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.from_gff" title="mgkit.io.gff.from_gff"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.from_gff()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.plots.colors.html#mgkit.plots.colors.palette_float_to_hex" title="mgkit.plots.colors.palette_float_to_hex"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.colors.palette_float_to_hex()</span></code></a> to convert floating value palettes to string</p></li>
<li><p>forces vertical alignment of tick labels in heatmaps</p></li>
<li><p>added parameter to get a consensus sequence for an AA alignment, by adding the <em>nucl</em> parameter to <a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.Alignment.get_consensus" title="mgkit.utils.sequence.Alignment.get_consensus"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.utils.sequence.Alignment.get_consensus()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.get_variant_sequence" title="mgkit.utils.sequence.get_variant_sequence"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.utils.sequence.get_variant_sequence()</span></code></a> to get variants of a sequence, essentially changing the sequence according to the SNPs passed</p></li>
<li><p>added method to get an aminoacid sequence from Annotation in <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_aa_seq" title="mgkit.io.gff.Annotation.get_aa_seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.get_aa_seq()</span></code></a> and added the possibility to pass a SNP to get the variant sequence of an Annotation in <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_nuc_seq" title="mgkit.io.gff.Annotation.get_nuc_seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.get_nuc_seq()</span></code></a>.</p></li>
<li><p>added <em>exp_syn</em> command to <cite>add-gff-info</cite> script</p></li>
<li><p>changed GTF file conversion</p></li>
<li><p>changed behaviour of <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.is_ancestor" title="mgkit.taxon.is_ancestor"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.is_ancestor()</span></code></a>: if a <em>taxon_id</em> raises a KeyError, False is now returned. In other words, if the taxon_id is not found in the taxonomy, it’s not an ancestor</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.__contains__" title="mgkit.io.gff.GenomicRange.__contains__"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.GenomicRange.__contains__()</span></code></a>. It tests if a position is inside the range</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.get_relative_pos" title="mgkit.io.gff.GenomicRange.get_relative_pos"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.GenomicRange.get_relative_pos()</span></code></a>. It returns a position relative to the GenomicRange start</p></li>
<li><p>fixed documentation and bugs (Annotation.get_nuc_seq)</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.is_syn" title="mgkit.io.gff.Annotation.is_syn"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.is_syn()</span></code></a>. It returns True if a SNP is synonymous and False if non-synonymous</p></li>
<li><p>added <em>to_nuc</em> parameter to <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.from_nuc_blast" title="mgkit.io.gff.from_nuc_blast"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.from_nuc_blast()</span></code></a> function. It to_nuc is False, it is assumed that the hit was against an amino acidic DB, in which case the phase should always set to 0</p></li>
<li><p>reworked internal of <cite>snp_parser</cite> script. It doesn’t use SNPDat anymore</p></li>
<li><p>updated tutorial</p></li>
<li><p>added ipython notebook as an example to explore data from the tutorial</p></li>
<li><p>cleaned deprecated code, fixed imports, added tests and documentation</p></li>
</ul>
</div>
<div class="section" id="id48">
<h2>0.1.15<a class="headerlink" href="#id48" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>changed name of <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.lowest_common_ancestor" title="mgkit.taxon.lowest_common_ancestor"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.lowest_common_ancestor()</span></code></a> to <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.last_common_ancestor" title="mgkit.taxon.last_common_ancestor"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.last_common_ancestor()</span></code></a>, the old function name points to the new one</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.counts.func.html#mgkit.counts.func.map_counts_to_category" title="mgkit.counts.func.map_counts_to_category"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.func.map_counts_to_category()</span></code></a> to remap counts from one ID to another</p></li>
<li><p>added <cite>get-gff-info</cite> script to extract information from GFF files</p></li>
<li><p>script <cite>download_data</cite> can now download only taxonomy data</p></li>
<li><p>added more script documentation</p></li>
<li><p>added examples on gene prediction</p></li>
<li><p>added function <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.from_hmmer" title="mgkit.io.gff.from_hmmer"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.from_hmmer()</span></code></a> to parse HMMER results and return <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation" title="mgkit.io.gff.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation</span></code></a> instances</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.to_gtf" title="mgkit.io.gff.Annotation.to_gtf"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.to_gtf()</span></code></a> to return a GTF line, <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.add_gc_content" title="mgkit.io.gff.Annotation.add_gc_content"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.add_gc_content()</span></code></a> and <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.add_gc_ratio" title="mgkit.io.gff.Annotation.add_gc_ratio"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.add_gc_ratio()</span></code></a> to calculate GC content and ratio respectively</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.parse_gff_files" title="mgkit.io.gff.parse_gff_files"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.parse_gff_files()</span></code></a> to parse multiple GFF files</p></li>
<li><p>added <em>uid_used</em> parameter to several functions in <a class="reference internal" href="api/mgkit.counts.func.html#module-mgkit.counts.func" title="mgkit.counts.func"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.counts.func</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.plots.abund.html#module-mgkit.plots.abund" title="mgkit.plots.abund"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.plots.abund</span></code></a> to plot abundance plots</p></li>
<li><p>added example notebooks for plots</p></li>
<li><p>HTSeq is now required only by the scripts that uses it, <em>snp_parser</em> and <em>fastq_utils</em></p></li>
<li><p>added function to convert numbers when reading from htseq count files</p></li>
<li><p>changed behavior of <em>-b</em> option in <cite>add-gff-info</cite> <em>taxonomy</em> command</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.get_annotation_map" title="mgkit.io.gff.get_annotation_map"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.get_annotation_map()</span></code></a></p></li>
</ul>
</div>
<div class="section" id="id49">
<h2>0.1.14<a class="headerlink" href="#id49" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>added ipthon notebooks to the documentation. As of this version the included ones (in <cite>docs/source/examples</cite>) are for two plot modules. Also added a bash script to convert them into rst files to be included with the documentation. The <em>.rst</em> are not versioned, and they must be rebuild, meaning that one of the requirements for building the docs is to have <a class="reference external" href="http://ipython.org">IPython</a> installed with the notebook extension</p></li>
<li><p>now importing some packages automatically import the subpackages as well</p></li>
<li><p>refactored <a class="reference internal" href="api/mgkit.plots.html#module-mgkit.plots" title="mgkit.plots"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.plots</span></code></a> into a package, with most of the original functions imported into it, for backward compatibility</p></li>
<li><p>added <code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.graphs.build_weighted_graph()</span></code></p></li>
<li><p>added <em>box_vert</em> parameter in <a class="reference internal" href="api/mgkit.plots.boxplot.html#mgkit.plots.boxplot.add_values_to_boxplot" title="mgkit.plots.boxplot.add_values_to_boxplot"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.plots.boxplot.add_values_to_boxplot()</span></code></a>, the default will be changed in a later version (kept for compatibility with older scripts/notebooks)</p></li>
<li><p>added an heatmap module to the plots package. Examples are in the notebook</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.align.html#mgkit.align.covered_annotation_bp" title="mgkit.align.covered_annotation_bp"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.align.covered_annotation_bp()</span></code></a> to find the number of bp covered by reads in annotations (as opposed to using the annotation length)</p></li>
<li><p>added documentation to <a class="reference internal" href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo" title="mgkit.mappings.eggnog.NOGInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.mappings.eggnog.NOGInfo</span></code></a> and an additional method</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.get_uniprot_ec_mappings" title="mgkit.net.uniprot.get_uniprot_ec_mappings"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.uniprot.get_uniprot_ec_mappings()</span></code></a> as it was used in a few scripts already</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.change_mapping_level" title="mgkit.mappings.enzyme.change_mapping_level"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.mappings.enzyme.change_mapping_level()</span></code></a> and other to deal with EC numbers. Also improved documentation with some examples</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.counts.func.html#mgkit.counts.func.load_sample_counts_to_genes" title="mgkit.counts.func.load_sample_counts_to_genes"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.func.load_sample_counts_to_genes()</span></code></a> and <a class="reference internal" href="api/mgkit.counts.func.html#mgkit.counts.func.load_sample_counts_to_taxon" title="mgkit.counts.func.load_sample_counts_to_taxon"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.func.load_sample_counts_to_taxon()</span></code></a>, for mapping counts to only genes or taxa. Also added <em>index</em> parameter in <a class="reference internal" href="api/mgkit.counts.func.html#mgkit.counts.func.map_counts" title="mgkit.counts.func.map_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.func.map_counts()</span></code></a> to accomodate the changes</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.get_ko_to_eggnog_mappings" title="mgkit.net.uniprot.get_ko_to_eggnog_mappings"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.uniprot.get_ko_to_eggnog_mappings()</span></code></a> to get mappings of KO identifiers to eggNOG</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.split_gff_file" title="mgkit.io.gff.split_gff_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.split_gff_file()</span></code></a> to split a gff into several ones, assuring that all annotations for a sequence is in the same file; useful to split massive GFF files before filtering</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.counts.func.html#mgkit.counts.func.load_deseq2_results" title="mgkit.counts.func.load_deseq2_results"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.func.load_deseq2_results()</span></code></a> to load DESeq2 results in <em>CSV</em> format</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.counts.scaling.html#mgkit.counts.scaling.scale_rpkm" title="mgkit.counts.scaling.scale_rpkm"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.scaling.scale_rpkm()</span></code></a> for scale with rpkm a count table</p></li>
<li><p>added caching options to <a class="reference internal" href="api/mgkit.counts.func.html#mgkit.counts.func.load_sample_counts" title="mgkit.counts.func.load_sample_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.counts.func.load_sample_counts()</span></code></a> and others</p></li>
<li><p>fixes and improvements to documentation</p></li>
</ul>
</div>
<div class="section" id="id50">
<h2>0.1.13<a class="headerlink" href="#id50" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>added counts package, including functions to load HTSeq-counts results and scaling</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.filter.taxon.html#mgkit.filter.taxon.filter_by_ancestor" title="mgkit.filter.taxon.filter_by_ancestor"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.filter.taxon.filter_by_ancestor()</span></code></a>, as a convenience function</p></li>
<li><p>deprecated functions in <a class="reference internal" href="api/mgkit.io.blast.html#module-mgkit.io.blast" title="mgkit.io.blast"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.blast</span></code></a> module, added more to parse blast outputs (some specific)</p></li>
<li><p><a class="reference internal" href="api/mgkit.io.fasta.html#mgkit.io.fasta.load_fasta" title="mgkit.io.fasta.load_fasta"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.fasta.load_fasta()</span></code></a> returns uppercase sequences, added a function (<a class="reference internal" href="api/mgkit.io.fasta.html#mgkit.io.fasta.split_fasta_file" title="mgkit.io.fasta.split_fasta_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.fasta.split_fasta_file()</span></code></a>) to split fasta files</p></li>
<li><p>added more methods to <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation" title="mgkit.io.gff.Annotation"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation</span></code></a> to complete API from old annotations</p></li>
<li><p>fixed <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.dbq" title="mgkit.io.gff.Annotation.dbq"><code class="xref py py-attr docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.dbq</span></code></a> property to return an <strong>int</strong> (bug in filtering with filter-gff)</p></li>
<li><p>added function to extract the sequences covered by annotations, using the <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_nuc_seq" title="mgkit.io.gff.Annotation.get_nuc_seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.get_nuc_seq()</span></code></a> method</p></li>
<li><p>added <code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.correct_old_annotations()</span></code> to update old annotated GFF to new conventions</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.group_annotations_by_ancestor" title="mgkit.io.gff.group_annotations_by_ancestor"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.group_annotations_by_ancestor()</span></code></a> and <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.group_annotations_sorted" title="mgkit.io.gff.group_annotations_sorted"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.group_annotations_sorted()</span></code></a></p></li>
<li><p>moved deprecated GFF classes/modules in <code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.gff_old</span></code></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.uniprot.html#module-mgkit.io.uniprot" title="mgkit.io.uniprot"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.uniprot</span></code></a> module to read/write Uniprot files</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.get_ids_names" title="mgkit.kegg.KeggClientRest.get_ids_names"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.kegg.KeggClientRest.get_ids_names()</span></code></a> to remove old methods to get specific class names used to retrieve (they are deprecated at the moment)</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggModule" title="mgkit.kegg.KeggModule"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.kegg.KeggModule</span></code></a> to parse a Kegg module entry</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.net.embl.html#mgkit.net.embl.datawarehouse_search" title="mgkit.net.embl.datawarehouse_search"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.embl.datawarehouse_search()</span></code></a> to search EMBL resources</p></li>
<li><p>made <a class="reference internal" href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.query_uniprot" title="mgkit.net.uniprot.query_uniprot"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.uniprot.query_uniprot()</span></code></a> more flexible</p></li>
<li><p>added/changed plot function in <a class="reference internal" href="api/mgkit.plots.html#module-mgkit.plots" title="mgkit.plots"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.plots</span></code></a></p></li>
<li><p>added enum34 as a dependency for Python versions below 3.4</p></li>
<li><p>changed classes to hold SNPs data: deprecated <code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.snps.classes.GeneSyn</span></code>, replaced by <a class="reference internal" href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP" title="mgkit.snps.classes.GeneSNP"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.snps.classes.GeneSNP</span></code></a> which the enum module for <a class="reference internal" href="api/mgkit.snps.classes.html#mgkit.snps.classes.SNPType" title="mgkit.snps.classes.SNPType"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.snps.classes.SNPType</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.NoLcaFound" title="mgkit.taxon.NoLcaFound"><code class="xref py py-exc docutils literal notranslate"><span class="pre">mgkit.taxon.NoLcaFound</span></code></a></p></li>
<li><p>fixed behaviour of <code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.get_ranked_taxon()</span></code> for newer taxonomies</p></li>
<li><p>change behaviour of <code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.is_ancestor()</span></code> to use module <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.is_ancestor" title="mgkit.taxon.is_ancestor"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.is_ancestor()</span></code></a> and accept multiple taxon IDs to test</p></li>
<li><p><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.taxon.UniprotTaxonomy.load_data()</span></code> now accept compressed data and file handles</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.lowest_common_ancestor" title="mgkit.taxon.lowest_common_ancestor"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.lowest_common_ancestor()</span></code></a> to find the lowest common ancestor of two taxon IDs</p></li>
<li><p>changed behaviour of <a class="reference internal" href="api/mgkit.taxon.html#mgkit.taxon.parse_uniprot_taxon" title="mgkit.taxon.parse_uniprot_taxon"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.taxon.parse_uniprot_taxon()</span></code></a></p></li>
<li><p>added functions to get GC content, ratio of a sequence and it composition to <a class="reference internal" href="api/mgkit.utils.sequence.html#module-mgkit.utils.sequence" title="mgkit.utils.sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.utils.sequence</span></code></a></p></li>
<li><p>added more options to <strong>blast2gff</strong> script</p></li>
<li><p>added <em>coverage</em>, <em>taxonomy</em> and <em>unipfile</em> to <strong>add-gff-info</strong></p></li>
<li><p>refactored <strong>snp_parser</strong> to use new classes</p></li>
<li><p>added possibility to use sorted GFF files as input for <strong>filter-gff</strong> to use less memory (the examples show how to use <em>sort</em> in Unix)</p></li>
</ul>
</div>
<div class="section" id="id51">
<h2>0.1.12<a class="headerlink" href="#id51" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>added functions to elongate annotations, measure the coverage of them and diff GFF files in <a class="reference internal" href="api/mgkit.io.gff.html#module-mgkit.io.gff" title="mgkit.io.gff"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.io.gff</span></code></a></p></li>
<li><p>added ranges_length and union_ranges to <a class="reference internal" href="api/mgkit.utils.common.html#module-mgkit.utils.common" title="mgkit.utils.common"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.utils.common</span></code></a></p></li>
<li><p>added script filter-gff, filter_gff will be deprecated</p></li>
<li><p>added script blast2gff to convert blast output to a GFF</p></li>
<li><p>removed unneeded dependencies to build docs</p></li>
<li><p>added script add-gff-info to add more annotations to GFF files</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.blast.html#mgkit.io.blast.parse_blast_tab" title="mgkit.io.blast.parse_blast_tab"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.blast.parse_blast_tab()</span></code></a> to parse BLAST tabular format</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.blast.html#mgkit.io.blast.parse_uniprot_blast" title="mgkit.io.blast.parse_uniprot_blast"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.blast.parse_uniprot_blast()</span></code></a> to return annotations from a BLAST tabular file</p></li>
<li><p>added <code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.graph</span></code> module</p></li>
<li><p>added classes <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation" title="mgkit.io.gff.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation</span></code></a> and <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange" title="mgkit.io.gff.GenomicRange"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.GenomicRange</span></code></a> and deprecated old classes to handle GFF annotations (API not stable)</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.io.gff.html#mgkit.io.gff.DuplicateKeyError" title="mgkit.io.gff.DuplicateKeyError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">mgkit.io.gff.DuplicateKeyError</span></code></a> raised in parsing GFF files</p></li>
<li><p>added functions used to return annotations from several sources</p></li>
<li><p>added option <cite>gff_type</cite> in <code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.gff.load_gff()</span></code></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.net.embl.html#mgkit.net.embl.dbfetch" title="mgkit.net.embl.dbfetch"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.embl.dbfetch()</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.get_gene_info" title="mgkit.net.uniprot.get_gene_info"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.uniprot.get_gene_info()</span></code></a> and <a class="reference internal" href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.query_uniprot" title="mgkit.net.uniprot.query_uniprot"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.uniprot.query_uniprot()</span></code></a> <a class="reference internal" href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.parse_uniprot_response" title="mgkit.net.uniprot.parse_uniprot_response"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.net.uniprot.parse_uniprot_response()</span></code></a></p></li>
<li><p>added apply_func_to_values to <a class="reference internal" href="api/mgkit.utils.dictionary.html#module-mgkit.utils.dictionary" title="mgkit.utils.dictionary"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.utils.dictionary</span></code></a></p></li>
<li><p>added <a class="reference internal" href="api/mgkit.snps.conv_func.html#mgkit.snps.conv_func.get_full_dataframe" title="mgkit.snps.conv_func.get_full_dataframe"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.conv_func.get_full_dataframe()</span></code></a>, <a class="reference internal" href="api/mgkit.snps.conv_func.html#mgkit.snps.conv_func.get_gene_taxon_dataframe" title="mgkit.snps.conv_func.get_gene_taxon_dataframe"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.conv_func.get_gene_taxon_dataframe()</span></code></a></p></li>
<li><p>added more tests</p></li>
</ul>
</div>
<div class="section" id="id52">
<h2>0.1.11<a class="headerlink" href="#id52" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>removed <cite>rst2pdf</cite> for generating a PDF for documentation. Latex is preferred</p></li>
<li><p>corrections to documentation and example script</p></li>
<li><p>removed need for joblib library in <cite>translate_seq</cite> script: used only if available (for using multiple processors)</p></li>
<li><p>deprecated <code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.funcs.combine_snps_in_dataframe()</span></code> and <code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.funcs.combine_snps_in_dataframe()</span></code>: <a class="reference internal" href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.funcs.combine_sample_snps()</span></code></a> should be used</p></li>
<li><p>refactored some tests and added more</p></li>
<li><p>added <cite>docs_req.txt</cite> to help build the documentation ont readthedocs.org</p></li>
<li><p>renamed <code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.snps.classes.GeneSyn</span></code> gid and taxon attributes to gene_id and taxon_id. The old names are still available for use (via properties), but the will be taken out in later versions. Old pickle data should be loaded and saved again before in this release</p></li>
<li><p>added a few convenience functions to ease the use of <a class="reference internal" href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">combine_sample_snps()</span></code></a></p></li>
<li><p>added function <a class="reference internal" href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.significance_test" title="mgkit.snps.funcs.significance_test"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.funcs.significance_test()</span></code></a> to test the distributions of genes share between two taxa.</p></li>
<li><p>fixed an issue with deinterleaving sequence data from khmer</p></li>
<li><p>added <a class="reference internal" href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.flat_sample_snps" title="mgkit.snps.funcs.flat_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.funcs.flat_sample_snps()</span></code></a></p></li>
<li><p>Added method to <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest" title="mgkit.kegg.KeggClientRest"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.kegg.KeggClientRest</span></code></a> to get names for all ids of a certain type (more generic than the various <cite>get_*_names</cite>)</p></li>
<li><p>added first implementation of <a class="reference internal" href="api/mgkit.kegg.html#mgkit.kegg.KeggModule" title="mgkit.kegg.KeggModule"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.kegg.KeggModule</span></code></a> class to parse a Kegg module entry</p></li>
<li><p><a class="reference internal" href="api/mgkit.snps.conv_func.html#mgkit.snps.conv_func.get_rank_dataframe" title="mgkit.snps.conv_func.get_rank_dataframe"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.conv_func.get_rank_dataframe()</span></code></a>, <a class="reference internal" href="api/mgkit.snps.conv_func.html#mgkit.snps.conv_func.get_gene_map_dataframe" title="mgkit.snps.conv_func.get_gene_map_dataframe"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.snps.conv_func.get_gene_map_dataframe()</span></code></a></p></li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="api/modules.html" class="btn btn-neutral float-left" title="mgkit" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2013-2020, Francesco Rubino

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>