
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Index &#8212; MGKit: Metagenomic framework 0.5.5 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/graphviz.css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="#" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            

<h1 id="index">Index</h1>

<div class="genindex-jumpbox">
 <a href="#Symbols"><strong>Symbols</strong></a>
 | <a href="#_"><strong>_</strong></a>
 | <a href="#A"><strong>A</strong></a>
 | <a href="#B"><strong>B</strong></a>
 | <a href="#C"><strong>C</strong></a>
 | <a href="#D"><strong>D</strong></a>
 | <a href="#E"><strong>E</strong></a>
 | <a href="#F"><strong>F</strong></a>
 | <a href="#G"><strong>G</strong></a>
 | <a href="#H"><strong>H</strong></a>
 | <a href="#I"><strong>I</strong></a>
 | <a href="#J"><strong>J</strong></a>
 | <a href="#K"><strong>K</strong></a>
 | <a href="#L"><strong>L</strong></a>
 | <a href="#M"><strong>M</strong></a>
 | <a href="#N"><strong>N</strong></a>
 | <a href="#O"><strong>O</strong></a>
 | <a href="#P"><strong>P</strong></a>
 | <a href="#Q"><strong>Q</strong></a>
 | <a href="#R"><strong>R</strong></a>
 | <a href="#S"><strong>S</strong></a>
 | <a href="#T"><strong>T</strong></a>
 | <a href="#U"><strong>U</strong></a>
 | <a href="#V"><strong>V</strong></a>
 | <a href="#W"><strong>W</strong></a>
 | <a href="#Y"><strong>Y</strong></a>
 
</div>
<h2 id="Symbols">Symbols</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    --attr-index &lt;attr_index&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-ai">edit-gff-table command line option</a>
</li>
      </ul></li>
      <li>
    --attr-value &lt;attr_value&gt;

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-k">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-k">blast2gff-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --attribute &lt;attribute&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-a">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-a">filter-gff-sequence command line option</a>
</li>
      </ul></li>
      <li>
    --attributes &lt;attributes&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-a">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-a">edit-gff-remove command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-a">edit-gff-view command line option</a>
</li>
      </ul></li>
      <li>
    --average

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-m">add-gff-info-cov_samtools command line option</a>
</li>
      </ul></li>
      <li>
    --bitscore &lt;bitscore&gt;

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-b">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-b">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-b">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    --buffer &lt;buffer&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-buffer">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --cache-table

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-c">add-gff-info-addtaxa command line option</a>
</li>
      </ul></li>
      <li>
    --choose-func &lt;choose_func&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-c">filter-gff-overlap command line option</a>
</li>
      </ul></li>
      <li>
    --chunk-size &lt;chunk_size&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-c">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    --cite

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cite">add-gff-info command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-cite">blast2gff command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-cite">edit-gff command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-cite">fasta-utils command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-cite">fastq-utils command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cite">filter-gff command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cite">get-gff-info command line option</a>
</li>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-cite">json2gff command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-cite">pnps-gen command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-cite">sampling-utils command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-cite">taxon-utils command line option</a>
</li>
      </ul></li>
      <li>
    --coding-prop &lt;coding_prop&gt;

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-r">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    --comment &lt;comment&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-c">edit-gff-table command line option</a>
</li>
      </ul></li>
      <li>
    --comparison &lt;comparison&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-c">filter-gff-sequence command line option</a>
</li>
      </ul></li>
      <li>
    --const-model

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-d">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    --count-files &lt;count_files&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-c">add-gff-info-counts command line option</a>
</li>
      </ul></li>
      <li>
    --db-quality &lt;db_quality&gt;

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-dbq">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-dbq">blast2gff-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --db-used &lt;db_used&gt;

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-db">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-db">blast2gff-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --depths &lt;depths&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-d">add-gff-info-cov_samtools command line option</a>
</li>
      </ul></li>
      <li>
    --description

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-d">add-gff-info-kegg command line option</a>
</li>
      </ul></li>
      <li>
    --dictionary &lt;dictionary&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-d">add-gff-info-addtaxa command line option</a>
</li>
      </ul></li>
      <li>
    --dist-loc &lt;dist_loc&gt;

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-x">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    --eggnog

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-e">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --email &lt;email&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-c">add-gff-info-kegg command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-c">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --enzymes

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-ec">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --exclude-taxon-id &lt;exclude_taxon_id&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-e">taxon-utils-filter command line option</a>
</li>
      </ul></li>
      <li>
    --exclude-taxon-name &lt;exclude_taxon_name&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-en">taxon-utils-filter command line option</a>
</li>
      </ul></li>
      <li>
    --fasta-file &lt;fasta_file&gt;

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-a">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-a">blast2gff-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --fastq

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-q">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-q">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-q">sampling-utils-sample_stream command line option</a>
</li>
      </ul></li>
      <li>
    --feat-type &lt;feat_type&gt;

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-ft">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-ft">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-ft">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    --featureCounts

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-e">add-gff-info-counts command line option</a>
</li>
      </ul></li>
      <li>
    --force-taxon-id

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-f">add-gff-info-unipfile command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-f">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --fpkms

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-f">add-gff-info-counts command line option</a>
</li>
      </ul></li>
      <li>
    --function &lt;function&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-f">filter-gff-sequence command line option</a>
</li>
      </ul></li>
      <li>
    --gc-content &lt;gc_content&gt;

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-gc">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    --gene-attr &lt;gene_attr&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-a">add-gff-info-addtaxa command line option</a>
</li>
      </ul></li>
      <li>
    --gene-id &lt;gene_id&gt;

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-g">get-gff-info-gtf command line option</a>
</li>
      </ul></li>
      <li>
    --gene-index &lt;gene_index&gt;

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-i">blast2gff-blastdb command line option</a>
</li>
      </ul></li>
      <li>
    --gene-map &lt;gene_map&gt;

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-g">pnps-gen-full command line option</a>
</li>
      </ul></li>
      <li>
    --gene-taxon-table &lt;gene_taxon_table&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-t">add-gff-info-addtaxa command line option</a>
</li>
      </ul></li>
      <li>
    --gzip

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-z">fasta-utils-split command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-z">get-gff-info-split command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-z">sampling-utils-sample command line option</a>
</li>
      </ul></li>
      <li>
    --hdf-table &lt;hdf_table&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-f">add-gff-info-addtaxa command line option</a>
</li>
      </ul></li>
      <li>
    --header

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-h">edit-gff-view command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-d">taxon-utils-get command line option</a>
</li>
      </ul></li>
      <li>
    --header-sep &lt;header_sep&gt;

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-s">blast2gff-blastdb command line option</a>
</li>
      </ul></li>
      <li>
    --id-attr &lt;id_attr&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-i">add-gff-info-pfam command line option</a>
</li>
      </ul></li>
      <li>
    --include-children

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-c">taxon-utils-get command line option</a>
</li>
      </ul></li>
      <li>
    --include-taxon-id &lt;include_taxon_id&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-i">taxon-utils-filter command line option</a>
</li>
      </ul></li>
      <li>
    --include-taxon-name &lt;include_taxon_name&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-in">taxon-utils-filter command line option</a>
</li>
      </ul></li>
      <li>
    --indent &lt;indent&gt;

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-i">get-gff-info-mongodb command line option</a>
</li>
      </ul></li>
      <li>
    --index-size &lt;index_size&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-s">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    --infer-params &lt;infer_params&gt;

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-i">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    --iterations &lt;iterations&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-n">filter-gff-overlap command line option</a>
</li>
      </ul></li>
      <li>
    --json-out

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-j">get-gff-info-cov command line option</a>
</li>
      </ul></li>
      <li>
    --keep-empty

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-k">edit-gff-view command line option</a>
</li>
      </ul></li>
      <li>
    --kegg-id &lt;kegg_id&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-m">add-gff-info-kegg command line option</a>
</li>
      </ul></li>
      <li>
    --kegg_orthologs

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-ko">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --key &lt;key&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-k">edit-gff-table command line option</a>
</li>
      </ul></li>
      <li>
    --key-index &lt;key_index&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-ki">edit-gff-table command line option</a>
</li>
      </ul></li>
      <li>
    --krona-total &lt;krona_total&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-kt">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    --length &lt;length&gt;

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-l">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    --lineage

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-l">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-l">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-l">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    --mapping &lt;mapping&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-m">add-gff-info-unipfile command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-m">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --mapping-file &lt;mapping_file&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-i">add-gff-info-unipfile command line option</a>
</li>
      </ul></li>
      <li>
    --master-file &lt;master_file&gt;

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-m">sampling-utils-sync command line option</a>
</li>
      </ul></li>
      <li>
    --max-seq &lt;max_seq&gt;

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-x">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-x">sampling-utils-sample_stream command line option</a>
</li>
      </ul></li>
      <li>
    --min-cov &lt;min_cov&gt;

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-c">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-c">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    --min-coverage &lt;min_coverage&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-c">filter-gff-cov command line option</a>
</li>
      </ul></li>
      <li>
    --min-num &lt;min_num&gt;

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-m">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-m">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    --no-cache

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-c">get-gff-info-mongodb command line option</a>
</li>
      </ul></li>
      <li>
    --no-lca &lt;no_lca&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-n">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-n">taxon-utils-lca_line command line option</a>
</li>
      </ul></li>
      <li>
    --no-split

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-n">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-n">blast2gff-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --no-strand

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-d">filter-gff-overlap command line option</a>
</li>
      </ul></li>
      <li>
    --no-wrap

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-w">fasta-utils-translate command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-w">get-gff-info-sequence command line option</a>
</li>
      </ul></li>
      <li>
    --num-ann &lt;num_ann&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-n">edit-gff-fields command line option</a>
</li>
      </ul></li>
      <li>
    --num-eq &lt;num_eq&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-num-eq">filter-gff-values command line option</a>
</li>
      </ul></li>
      <li>
    --num-ge &lt;num_ge&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-num-ge">filter-gff-values command line option</a>
</li>
      </ul></li>
      <li>
    --num-gt &lt;num_gt&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-num-gt">filter-gff-values command line option</a>
</li>
      </ul></li>
      <li>
    --num-le &lt;num_le&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-num-le">filter-gff-values command line option</a>
</li>
      </ul></li>
      <li>
    --num-lt &lt;num_lt&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-num-lt">filter-gff-values command line option</a>
</li>
      </ul></li>
      <li>
    --num-seqs &lt;num_seqs&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-n">add-gff-info-cov_samtools command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-n">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    --number &lt;number&gt;

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-n">fasta-utils-split command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-n">get-gff-info-split command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-n">sampling-utils-sample command line option</a>
</li>
      </ul></li>
      <li>
    --one-seq

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-1">fasta-utils-translate command line option</a>
</li>
      </ul></li>
      <li>
    --only-edited

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-o">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-o">edit-gff-table command line option</a>
</li>
      </ul></li>
      <li>
    --only-names &lt;only_names&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-o">taxon-utils-get command line option</a>
</li>
      </ul></li>
      <li>
    --only-pn

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-pn">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-pn">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    --only-ps

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-ps">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-ps">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    --only-ranked

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-a">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-a">taxon-utils-lca_line command line option</a>
</li>
      </ul></li>
      <li>
    --out-format &lt;out_format&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-f">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    --output-dir &lt;output_dir&gt;

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-dbm-d">get-gff-info-dbm command line option</a>
</li>
      </ul></li>
      <li>
    --overwrite

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-w">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-w">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    --partial

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-p">taxon-utils-get command line option</a>
</li>
      </ul></li>
      <li>
    --pathways

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-p">add-gff-info-kegg command line option</a>
</li>
      </ul></li>
      <li>
    --prefix &lt;prefix&gt;

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-p">fasta-utils-split command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-p">get-gff-info-split command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-p">sampling-utils-sample command line option</a>
</li>
      </ul></li>
      <li>
    --prob &lt;prob&gt;

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-r">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-r">sampling-utils-sample_stream command line option</a>
</li>
      </ul></li>
      <li>
    --progress

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-progress">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-progress">add-gff-info-counts command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-progress">add-gff-info-cov_samtools command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-progress">add-gff-info-exp_syn command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-progress">add-gff-info-unipfile command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-progress">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-progress">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-progress">fasta-utils-translate command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-progress">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-progress">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-progress">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-progress">filter-gff-values command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-progress">get-gff-info-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-progress">get-gff-info-mongodb command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-progress">get-gff-info-sequence command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-progress">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-progress">taxon-utils-filter command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-progress">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-progress">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    --protein-names

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-d">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --rank &lt;rank&gt;

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-r">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-r">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    --read-model &lt;read_model&gt;

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-a">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    --reference &lt;reference&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-r">add-gff-info-exp_syn command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-f">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-f">get-gff-info-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-f">get-gff-info-sequence command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-r">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    --remove-version

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-r">blast2gff-blastdb command line option</a>
</li>
      </ul></li>
      <li>
    --rename

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-r">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-r">get-gff-info-cov command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-m">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    --reverse

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-r">get-gff-info-sequence command line option</a>
</li>
      </ul></li>
      <li>
    --sample-alignment &lt;sample_alignment&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-a">add-gff-info-coverage command line option</a>
</li>
      </ul></li>
      <li>
    --samples &lt;samples&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-s">add-gff-info-counts command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-s">add-gff-info-cov_samtools command line option</a>
</li>
      </ul></li>
      <li>
    --save-model &lt;save_model&gt;

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-m">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    --separator &lt;separator&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-s">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-s">edit-gff-view command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-p">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-s">taxon-utils-get command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-s">taxon-utils-lca_line command line option</a>
</li>
      </ul></li>
      <li>
    --simple-table

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-p">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    --size &lt;size&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-s">filter-gff-overlap command line option</a>
</li>
      </ul></li>
      <li>
    --skip-no-taxon

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-e">add-gff-info-addtaxa command line option</a>
</li>
      </ul></li>
      <li>
    --skip-rows &lt;skip_rows&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-r">edit-gff-table command line option</a>
</li>
      </ul></li>
      <li>
    --snp-data &lt;snp_data&gt;

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-s">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-s">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    --sort-attr &lt;sort_attr&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-a">filter-gff-overlap command line option</a>
</li>
      </ul></li>
      <li>
    --sorted

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-t">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-t">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-t">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-s">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    --split

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-s">add-gff-info-exp_syn command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-s">get-gff-info-sequence command line option</a>
</li>
      </ul></li>
      <li>
    --str-eq &lt;str_eq&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-str-eq">filter-gff-values command line option</a>
</li>
      </ul></li>
      <li>
    --str-in &lt;str_in&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-str-in">filter-gff-values command line option</a>
</li>
      </ul></li>
      <li>
    --strand-specific

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-s">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-s">get-gff-info-cov command line option</a>
</li>
      </ul></li>
      <li>
    --strip

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-s">fastq-utils-di command line option</a>
</li>
      </ul></li>
      <li>
    --table

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-p">taxon-utils-filter command line option</a>
</li>
      </ul></li>
      <li>
    --table &lt;table&gt;

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-t">fasta-utils-uid command line option</a>
</li>
      </ul></li>
      <li>
    --table-file &lt;table_file&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-t">edit-gff-table command line option</a>
</li>
      </ul></li>
      <li>
    --table-name &lt;table_name&gt;

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-n">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    --taxon-db &lt;taxon_db&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-db">add-gff-info-addtaxa command line option</a>
</li>
      </ul></li>
      <li>
    --taxon-id

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-t">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    --taxon_ids &lt;taxon_ids&gt;

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-i">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-i">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    --taxonomy &lt;taxonomy&gt;

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-x">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-t">get-gff-info-mongodb command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-t">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-t">pnps-gen-rank command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-t">taxon-utils-filter command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-t">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-t">taxon-utils-lca_line command line option</a>
</li>
      </ul></li>
      <li>
    --trans-table &lt;trans_table&gt;

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-t">fasta-utils-translate command line option</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    --two-columns

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-2">pnps-gen-full command line option</a>
</li>
      </ul></li>
      <li>
    --uids &lt;uids&gt;

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-f">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-f">edit-gff-remove command line option</a>
</li>
      </ul></li>
      <li>
    --unstack

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-u">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    --use-accession

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-a">add-gff-info-pfam command line option</a>
</li>
      </ul></li>
      <li>
    --use-uid

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-u">pnps-gen-full command line option</a>
</li>
      </ul></li>
      <li>
    --value &lt;value&gt;

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-l">filter-gff-sequence command line option</a>
</li>
      </ul></li>
      <li>
    --verbose

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-v">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-v">add-gff-info-counts command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-v">add-gff-info-cov_samtools command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-v">add-gff-info-coverage command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-v">add-gff-info-exp_syn command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-v">add-gff-info-kegg command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-v">add-gff-info-pfam command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-v">add-gff-info-unipfile command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-v">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-v">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-v">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-v">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-v">edit-gff-fields command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-v">edit-gff-remove command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-v">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-v">edit-gff-view command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-v">fasta-utils-split command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-v">fasta-utils-translate command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-v">fasta-utils-uid command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-convert-v">fastq-utils-convert command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-v">fastq-utils-di command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-il-v">fastq-utils-il command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-v">fastq-utils-sort command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-v">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-v">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-v">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-v">filter-gff-values command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-v">get-gff-info-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-dbm-v">get-gff-info-dbm command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-v">get-gff-info-gtf command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-v">get-gff-info-mongodb command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-v">get-gff-info-sequence command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-v">get-gff-info-split command line option</a>
</li>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-mongodb-v">json2gff-mongodb command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-v">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-v">pnps-gen-rank command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-v">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-v">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-v">sampling-utils-sample_stream command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-v">sampling-utils-sync command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-v">taxon-utils-filter command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-v">taxon-utils-get command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-v">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-v">taxon-utils-lca_line command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-v">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    --version

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-version">add-gff-info command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-version">blast2gff command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-version">edit-gff command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-version">fasta-utils command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-version">fastq-utils command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-version">filter-gff command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-version">get-gff-info command line option</a>
</li>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-version">json2gff command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-version">pnps-gen command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-version">sampling-utils command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-version">taxon-utils command line option</a>
</li>
      </ul></li>
      <li>
    -1

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-1">fasta-utils-translate command line option</a>
</li>
      </ul></li>
      <li>
    -2

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-2">pnps-gen-full command line option</a>
</li>
      </ul></li>
      <li>
    -a

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-a">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-a">add-gff-info-coverage command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-a">add-gff-info-pfam command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-a">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-a">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-a">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-a">edit-gff-remove command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-a">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-a">edit-gff-view command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-a">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-a">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-a">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-a">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-a">taxon-utils-lca_line command line option</a>
</li>
      </ul></li>
      <li>
    -ai

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-ai">edit-gff-table command line option</a>
</li>
      </ul></li>
      <li>
    -b

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-b">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-b">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-b">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    -c

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-c">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-c">add-gff-info-counts command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-c">add-gff-info-kegg command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-c">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-c">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-c">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-c">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-c">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-c">get-gff-info-mongodb command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-c">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-c">pnps-gen-rank command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-c">taxon-utils-get command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-c">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    -d

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-d">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-d">add-gff-info-cov_samtools command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-d">add-gff-info-kegg command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-d">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-d">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-dbm-d">get-gff-info-dbm command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-d">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-d">taxon-utils-get command line option</a>
</li>
      </ul></li>
      <li>
    -db

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-db">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-db">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-db">blast2gff-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    -dbq

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-dbq">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-dbq">blast2gff-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    -e

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-e">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-e">add-gff-info-counts command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-e">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-e">taxon-utils-filter command line option</a>
</li>
      </ul></li>
      <li>
    -ec

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-ec">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    -en

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-en">taxon-utils-filter command line option</a>
</li>
      </ul></li>
      <li>
    -f

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-f">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-f">add-gff-info-counts command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-f">add-gff-info-unipfile command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-f">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-f">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-f">edit-gff-remove command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-f">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-f">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-f">get-gff-info-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-f">get-gff-info-sequence command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-f">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    -ft

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-ft">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-ft">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-ft">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    -g

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-g">get-gff-info-gtf command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-g">pnps-gen-full command line option</a>
</li>
      </ul></li>
      <li>
    -gc

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-gc">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    -h

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-h">edit-gff-view command line option</a>
</li>
      </ul></li>
      <li>
    -i

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-i">add-gff-info-pfam command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-i">add-gff-info-unipfile command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-i">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-i">get-gff-info-mongodb command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-i">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-i">pnps-gen-rank command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-i">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-i">taxon-utils-filter command line option</a>
</li>
      </ul></li>
      <li>
    -in

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-in">taxon-utils-filter command line option</a>
</li>
      </ul></li>
      <li>
    -j

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-j">get-gff-info-cov command line option</a>
</li>
      </ul></li>
      <li>
    -k

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-k">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-k">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-k">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-k">edit-gff-view command line option</a>
</li>
      </ul></li>
      <li>
    -ki

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-ki">edit-gff-table command line option</a>
</li>
      </ul></li>
      <li>
    -ko

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-ko">add-gff-info-uniprot command line option</a>
</li>
      </ul></li>
      <li>
    -kt

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-kt">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    -l

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-l">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-l">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-l">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-l">pnps-gen-rank command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-l">sampling-utils-rand_seq command line option</a>
</li>
      </ul></li>
      <li>
    -m

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-m">add-gff-info-cov_samtools command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-m">add-gff-info-kegg command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-m">add-gff-info-unipfile command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-m">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-m">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-m">pnps-gen-rank command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-m">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-m">sampling-utils-sync command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-m">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    -n

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-n">add-gff-info-cov_samtools command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-n">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-n">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-n">edit-gff-fields command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-n">fasta-utils-split command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-n">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-n">get-gff-info-split command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-n">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-n">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-n">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-n">taxon-utils-lca_line command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-n">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    -o

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-o">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-o">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-o">taxon-utils-get command line option</a>
</li>
      </ul></li>
      <li>
    -p

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-p">add-gff-info-kegg command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-p">fasta-utils-split command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-p">get-gff-info-split command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-p">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-p">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-p">taxon-utils-filter command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-p">taxon-utils-get command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-p">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    -pn

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-pn">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-pn">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    -ps

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-ps">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-ps">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    -q

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-q">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-q">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-q">sampling-utils-sample_stream command line option</a>
</li>
      </ul></li>
      <li>
    -r

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-r">add-gff-info-exp_syn command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-r">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-r">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-r">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-r">get-gff-info-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-r">get-gff-info-sequence command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-r">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-r">pnps-gen-rank command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-r">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-r">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-r">sampling-utils-sample_stream command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-r">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li>
    -s

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-s">add-gff-info-counts command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-s">add-gff-info-cov_samtools command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-s">add-gff-info-exp_syn command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-s">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-s">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-s">edit-gff-view command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-s">fastq-utils-di command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-s">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-s">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-s">get-gff-info-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-s">get-gff-info-sequence command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-s">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-s">pnps-gen-rank command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-s">taxon-utils-get command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-s">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-s">taxon-utils-lca_line command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-s">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    -t

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-t">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-t">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-t">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-t">fasta-utils-translate command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-t">fasta-utils-uid command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-t">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-t">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-t">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-t">get-gff-info-mongodb command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-t">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-t">pnps-gen-rank command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-t">taxon-utils-filter command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-t">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-t">taxon-utils-lca_line command line option</a>
</li>
      </ul></li>
      <li>
    -u

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-u">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-u">pnps-gen-rank command line option</a>
</li>
      </ul></li>
      <li>
    -v

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-v">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-v">add-gff-info-counts command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-v">add-gff-info-cov_samtools command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-v">add-gff-info-coverage command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-v">add-gff-info-exp_syn command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-v">add-gff-info-kegg command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-v">add-gff-info-pfam command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-v">add-gff-info-unipfile command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-v">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-v">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-v">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-v">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-v">edit-gff-fields command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-v">edit-gff-remove command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-v">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-v">edit-gff-view command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-v">fasta-utils-split command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-v">fasta-utils-translate command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-v">fasta-utils-uid command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-convert-v">fastq-utils-convert command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-v">fastq-utils-di command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-il-v">fastq-utils-il command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-v">fastq-utils-sort command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-v">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-v">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-v">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-v">filter-gff-values command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-v">get-gff-info-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-dbm-v">get-gff-info-dbm command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-v">get-gff-info-gtf command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-v">get-gff-info-mongodb command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-v">get-gff-info-sequence command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-v">get-gff-info-split command line option</a>
</li>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-mongodb-v">json2gff-mongodb command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-v">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-v">pnps-gen-rank command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-v">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-v">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-v">sampling-utils-sample_stream command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-v">sampling-utils-sync command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-v">taxon-utils-filter command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-v">taxon-utils-get command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-v">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-v">taxon-utils-lca_line command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-v">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    -w

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-w">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-w">fasta-utils-translate command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-w">get-gff-info-sequence command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-w">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
      <li>
    -x

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-x">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-x">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-x">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-x">sampling-utils-sample_stream command line option</a>
</li>
      </ul></li>
      <li>
    -z

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-z">fasta-utils-split command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-z">get-gff-info-split command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-z">sampling-utils-sample command line option</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="_">_</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.__contains__">__contains__() (mgkit.io.gff.GenomicRange method)</a>

      <ul>
        <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.__contains__">(mgkit.taxon.Taxonomy method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.__getitem__">__getitem__() (mgkit.db.mongo.GFFDB method)</a>

      <ul>
        <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.__getitem__">(mgkit.taxon.Taxonomy method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.__iter__">__iter__() (mgkit.db.mongo.GFFDB method)</a>

      <ul>
        <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.__iter__">(mgkit.taxon.Taxonomy method)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.__len__">__len__() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.__repr__">__repr__() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence._get_kmers">_get_kmers() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence._sequence_signature">_sequence_signature() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence._signatures_matrix">_signatures_matrix() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence._sliding_window">_sliding_window() (in module mgkit.utils.sequence)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="A">A</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.io.snpdat.html#id0">aa_change (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.aa_change">[1]</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.add">add() (mgkit.snps.classes.GeneSNP method)</a>
</li>
      <li>
    add-gff-info command line option

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cite">--cite</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-version">--version</a>
</li>
      </ul></li>
      <li>
    add-gff-info-addtaxa command line option

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-c">--cache-table</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-d">--dictionary &lt;dictionary&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-a">--gene-attr &lt;gene_attr&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-t">--gene-taxon-table &lt;gene_taxon_table&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-f">--hdf-table &lt;hdf_table&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-progress">--progress</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-e">--skip-no-taxon</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-db">--taxon-db &lt;taxon_db&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-x">--taxonomy &lt;taxonomy&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-v">--verbose</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-a">-a</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-c">-c</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-d">-d</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-db">-db</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-e">-e</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-f">-f</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-t">-t</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-v">-v</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-x">-x</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    add-gff-info-counts command line option

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-c">--count-files &lt;count_files&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-e">--featureCounts</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-f">--fpkms</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-progress">--progress</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-s">--samples &lt;samples&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-v">--verbose</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-c">-c</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-e">-e</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-f">-f</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-s">-s</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-v">-v</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    add-gff-info-cov_samtools command line option

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-m">--average</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-d">--depths &lt;depths&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-n">--num-seqs &lt;num_seqs&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-progress">--progress</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-s">--samples &lt;samples&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-v">--verbose</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-d">-d</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-m">-m</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-n">-n</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-s">-s</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-v">-v</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    add-gff-info-coverage command line option

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-a">--sample-alignment &lt;sample_alignment&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-v">--verbose</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-a">-a</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-v">-v</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    add-gff-info-exp_syn command line option

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-progress">--progress</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-r">--reference &lt;reference&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-s">--split</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-v">--verbose</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-r">-r</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-s">-s</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-v">-v</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    add-gff-info-kegg command line option

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-d">--description</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-c">--email &lt;email&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-m">--kegg-id &lt;kegg_id&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-p">--pathways</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-v">--verbose</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-c">-c</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-d">-d</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-m">-m</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-p">-p</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-v">-v</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    add-gff-info-pfam command line option

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-i">--id-attr &lt;id_attr&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-a">--use-accession</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-v">--verbose</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-a">-a</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-i">-i</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-v">-v</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    add-gff-info-unipfile command line option

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-f">--force-taxon-id</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-m">--mapping &lt;mapping&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-i">--mapping-file &lt;mapping_file&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-progress">--progress</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-v">--verbose</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-f">-f</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-i">-i</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-m">-m</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-v">-v</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    add-gff-info-uniprot command line option

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-buffer">--buffer &lt;buffer&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-e">--eggnog</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-c">--email &lt;email&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-ec">--enzymes</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-f">--force-taxon-id</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-ko">--kegg_orthologs</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-l">--lineage</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-m">--mapping &lt;mapping&gt;</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-d">--protein-names</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-t">--taxon-id</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-v">--verbose</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-c">-c</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-d">-d</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-e">-e</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-ec">-ec</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-f">-f</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-ko">-ko</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-l">-l</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-m">-m</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-t">-t</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-v">-v</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.workflow.utils.html#mgkit.workflow.utils.add_basic_options">add_basic_options() (in module mgkit.workflow.utils)</a>
</li>
      <li><a href="api/mgkit.io.blast.html#mgkit.io.blast.add_blast_result_to_annotation">add_blast_result_to_annotation() (in module mgkit.io.blast)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.add_coverage_info">add_coverage_info() (in module mgkit.align)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.add_exp_syn_count">add_exp_syn_count() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.add_gc_content">add_gc_content() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.add_gc_ratio">add_gc_ratio() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.add_lineage">add_lineage() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.Alignment.add_seq">add_seq() (mgkit.utils.sequence.Alignment method)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.Alignment.add_seqs">add_seqs() (mgkit.utils.sequence.Alignment method)</a>
</li>
      <li><a href="api/mgkit.plots.boxplot.html#mgkit.plots.boxplot.add_significance_to_boxplot">add_significance_to_boxplot() (in module mgkit.plots.boxplot)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.add_snp">add_snp() (mgkit.snps.classes.GeneSNP method)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.add_taxon">add_taxon() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.workflow.add_gff_info.html#mgkit.workflow.add_gff_info.add_uniprot_info">add_uniprot_info() (in module mgkit.workflow.add_gff_info)</a>
</li>
      <li><a href="api/mgkit.plots.boxplot.html#mgkit.plots.boxplot.add_values_to_boxplot">add_values_to_boxplot() (in module mgkit.plots.boxplot)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.advance_file">advance_file() (mgkit.align.SamtoolsDepth method)</a>
</li>
      <li><a href="api/mgkit.filter.lists.html#mgkit.filter.lists.aggr_filtered_list">aggr_filtered_list() (in module mgkit.filter.lists)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.Alignment">Alignment (class in mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id1">ann_frame (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.ann_frame">[1]</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.annotate_sequence">annotate_sequence() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation">Annotation (class in mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.annotation_coverage">annotation_coverage() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.annotation_coverage_sorted">annotation_coverage_sorted() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.annotation_elongation">annotation_elongation() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.api_url">api_url (mgkit.kegg.KeggClientRest attribute)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.apply_func_to_values">apply_func_to_values() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.utils.common.html#mgkit.utils.common.apply_func_window">apply_func_window() (in module mgkit.utils.common)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.attr">attr (mgkit.io.gff.Annotation attribute)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.AttributeNotFound">AttributeNotFound</a>
</li>
      <li><a href="api/mgkit.utils.common.html#mgkit.utils.common.average_length">average_length() (in module mgkit.utils.common)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="B">B</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.utils.trans_tables.html#mgkit.utils.trans_tables.BAC_PLT">BAC_PLT (in module mgkit.utils.trans_tables)</a>
</li>
      <li><a href="api/mgkit.plots.unused.html#mgkit.plots.unused.barchart_categories">barchart_categories() (in module mgkit.plots.unused)</a>
</li>
      <li><a href="api/mgkit.plots.heatmap.html#mgkit.plots.heatmap.baseheatmap">baseheatmap() (in module mgkit.plots.heatmap)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.batch_load_htseq_counts">batch_load_htseq_counts() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.utils.common.html#mgkit.utils.common.between">between() (in module mgkit.utils.common)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.bitscore">bitscore() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.consts.html#mgkit.consts.BLACK_LIST">BLACK_LIST (in module mgkit.consts)</a>
</li>
      <li>
    blast2gff command line option

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-cite">--cite</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-version">--version</a>
</li>
      </ul></li>
      <li>
    blast2gff-blastdb command line option

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-k">--attr-value &lt;attr_value&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-b">--bitscore &lt;bitscore&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-dbq">--db-quality &lt;db_quality&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-db">--db-used &lt;db_used&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-a">--fasta-file &lt;fasta_file&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-ft">--feat-type &lt;feat_type&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-i">--gene-index &lt;gene_index&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-s">--header-sep &lt;header_sep&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-n">--no-split</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-progress">--progress</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-r">--remove-version</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-v">--verbose</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-a">-a</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-b">-b</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-db">-db</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-dbq">-dbq</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-ft">-ft</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-i">-i</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-k">-k</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-n">-n</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-r">-r</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-s">-s</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-v">-v</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-arg-BLAST_FILE">BLAST_FILE</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-arg-GFF_FILE">GFF_FILE</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    blast2gff-uniprot command line option

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-k">--attr-value &lt;attr_value&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-b">--bitscore &lt;bitscore&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-dbq">--db-quality &lt;db_quality&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-db">--db-used &lt;db_used&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-a">--fasta-file &lt;fasta_file&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-ft">--feat-type &lt;feat_type&gt;</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-n">--no-split</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-progress">--progress</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-v">--verbose</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-a">-a</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-b">-b</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-db">-db</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-dbq">-dbq</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-ft">-ft</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-k">-k</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-n">-n</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-v">-v</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-arg-BLAST_FILE">BLAST_FILE</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-arg-GFF_FILE">GFF_FILE</a>
</li>
      </ul></li>
      <li>
    BLAST_FILE

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-arg-BLAST_FILE">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-arg-BLAST_FILE">blast2gff-uniprot command line option</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.plots.boxplot.html#mgkit.plots.boxplot.boxplot_dataframe">boxplot_dataframe() (in module mgkit.plots.boxplot)</a>
</li>
      <li><a href="api/mgkit.plots.boxplot.html#mgkit.plots.boxplot.boxplot_dataframe_multindex">boxplot_dataframe_multindex() (in module mgkit.plots.boxplot)</a>
</li>
      <li><a href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.build_rank_matrix">build_rank_matrix() (in module mgkit.snps.funcs)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="C">C</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonTuple.c_name">c_name (mgkit.taxon.UniprotTaxonTuple attribute)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.cache">cache (mgkit.kegg.KeggClientRest attribute)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.cache_dict_file">cache_dict_file (class in mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.mappings.pandas_map.html#mgkit.mappings.pandas_map.calc_coefficient_of_variation">calc_coefficient_of_variation() (in module mgkit.mappings.pandas_map)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.calc_n50">calc_n50() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.RatioMixIn.calc_pn">calc_pn() (mgkit.snps.classes.RatioMixIn method)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.RatioMixIn.calc_ps">calc_ps() (mgkit.snps.classes.RatioMixIn method)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.RatioMixIn.calc_ratio">calc_ratio() (mgkit.snps.classes.RatioMixIn method)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.RatioMixIn.calc_ratio_flag">calc_ratio_flag() (mgkit.snps.classes.RatioMixIn method)</a>
</li>
      <li><a href="api/mgkit.io.fastq.html#mgkit.io.fastq.CASAVA_HEADER_NEW">CASAVA_HEADER_NEW (in module mgkit.io.fastq)</a>
</li>
      <li><a href="api/mgkit.io.fastq.html#mgkit.io.fastq.CASAVA_HEADER_OLD">CASAVA_HEADER_OLD (in module mgkit.io.fastq)</a>
</li>
      <li><a href="api/mgkit.mappings.cazy.html#mgkit.mappings.cazy.CAZY_FAMILIES">CAZY_FAMILIES (in module mgkit.mappings.cazy)</a>
</li>
      <li><a href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.change_mapping_level">change_mapping_level() (in module mgkit.mappings.enzyme)</a>
</li>
      <li><a href="api/mgkit.io.fastq.html#mgkit.io.fastq.check_fastq_type">check_fastq_type() (in module mgkit.io.fastq)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.check_snp_in_seq">check_snp_in_seq() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.check_snp_in_set">check_snp_in_set() (in module mgkit.workflow.snp_parser)</a>
</li>
      <li><a href="api/mgkit.html#mgkit.check_version">check_version() (in module mgkit)</a>
</li>
      <li><a href="api/mgkit.filter.gff.html#mgkit.filter.gff.choose_annotation">choose_annotation() (in module mgkit.filter.gff)</a>
</li>
      <li><a href="api/mgkit.io.fastq.html#mgkit.io.fastq.choose_header_type">choose_header_type() (in module mgkit.io.fastq)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id2">chr_name (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.chr_name">[1]</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id3">chr_pos (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.chr_pos">[1]</a>
</li>
      <li><a href="api/mgkit.html#mgkit.cite">cite() (in module mgkit)</a>
</li>
      <li><a href="api/mgkit.workflow.utils.html#mgkit.workflow.utils.cite_callback">cite_callback() (in module mgkit.workflow.utils)</a>
</li>
      <li><a href="api/mgkit.workflow.utils.html#mgkit.workflow.utils.CiteAction">CiteAction (class in mgkit.workflow.utils)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.classes">classes (mgkit.kegg.KeggModule attribute)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.closed">closed (mgkit.align.SamtoolsDepth attribute)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id4">codon (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.codon">[1]</a>
</li>
      <li><a href="api/mgkit.plots.abund.html#mgkit.plots.abund.col_func_firstel">col_func_firstel() (in module mgkit.plots.abund)</a>
</li>
      <li><a href="api/mgkit.plots.abund.html#mgkit.plots.abund.col_func_name">col_func_name() (in module mgkit.plots.abund)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.plots.abund.html#mgkit.plots.abund.col_func_taxon">col_func_taxon() (in module mgkit.plots.abund)</a>
</li>
      <li><a href="api/mgkit.logger.html#mgkit.logger.ColorFormatter">ColorFormatter (class in mgkit.logger)</a>
</li>
      <li><a href="api/mgkit.logger.html#mgkit.logger.ColorFormatter.colors">colors (mgkit.logger.ColorFormatter attribute)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.combine_dict">combine_dict() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.combine_dict_one_value">combine_dict_one_value() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps">combine_sample_snps() (in module mgkit.snps.funcs)</a>
</li>
      <li><a href="api/mgkit.workflow.sampling_utils.html#mgkit.workflow.sampling_utils.compare_header">compare_header() (in module mgkit.workflow.sampling_utils)</a>
</li>
      <li><a href="api/mgkit.utils.common.html#mgkit.utils.common.complement_ranges">complement_ranges() (in module mgkit.utils.common)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.compounds">compounds (mgkit.kegg.KeggModule attribute)</a>
</li>
      <li><a href="api/mgkit.io.utils.html#mgkit.io.utils.compressed_handle">compressed_handle() (in module mgkit.io.utils)</a>
</li>
      <li><a href="api/mgkit.mappings.pandas_map.html#mgkit.mappings.pandas_map.concatenate_and_rename_tables">concatenate_and_rename_tables() (in module mgkit.mappings.pandas_map)</a>
</li>
      <li><a href="api/mgkit.logger.html#mgkit.logger.config_log">config_log() (in module mgkit.logger)</a>
</li>
      <li><a href="api/mgkit.logger.html#mgkit.logger.config_log_to_file">config_log_to_file() (in module mgkit.logger)</a>
</li>
      <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.conn">conn (mgkit.db.mongo.GFFDB attribute)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.contact">contact (mgkit.kegg.KeggClientRest attribute)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.conv">conv() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.convert_aa_to_nuc_coord">convert_aa_to_nuc_coord() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.convert_gff_to_gtf">convert_gff_to_gtf() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.convert_record">convert_record() (mgkit.db.mongo.GFFDB method)</a>
</li>
      <li><a href="api/mgkit.io.fastq.html#mgkit.io.fastq.convert_seqid_to_new">convert_seqid_to_new() (in module mgkit.io.fastq)</a>
</li>
      <li><a href="api/mgkit.io.fastq.html#mgkit.io.fastq.convert_seqid_to_old">convert_seqid_to_old() (in module mgkit.io.fastq)</a>
</li>
      <li><a href="api/mgkit.mappings.utils.html#mgkit.mappings.utils.count_genes_in_mapping">count_genes_in_mapping() (in module mgkit.mappings.utils)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.counts">counts() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#id0">coverage (mgkit.snps.classes.GeneSNP attribute)</a>, <a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.coverage">[1]</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.coverage">coverage() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.covered_annotation_bp">covered_annotation_bp() (in module mgkit.align)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.cpd_desc_re">cpd_desc_re (mgkit.kegg.KeggClientRest attribute)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.cpd_re">cpd_re (mgkit.kegg.KeggClientRest attribute)</a>
</li>
      <li><a href="api/mgkit.db.dbm.html#mgkit.db.dbm.create_gff_dbm">create_gff_dbm() (in module mgkit.db.dbm)</a>
</li>
      <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.cursor">cursor() (mgkit.db.mongo.GFFDB method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="D">D</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.data">data (mgkit.align.SamtoolsDepth attribute)</a>
</li>
      <li><a href="api/mgkit.net.embl.html#mgkit.net.embl.datawarehouse_search">datawarehouse_search() (in module mgkit.net.embl)</a>
</li>
      <li><a href="api/mgkit.db.dbm.html#mgkit.db.dbm.GFFDB.db">db (mgkit.db.dbm.GFFDB attribute)</a>

      <ul>
        <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.db">(mgkit.db.mongo.GFFDB attribute)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.db">db() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.net.embl.html#mgkit.net.embl.dbfetch">dbfetch() (in module mgkit.net.embl)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.dbq">dbq() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.html#mgkit.DEBUG">DEBUG (in module mgkit)</a>
</li>
      <li><a href="api/mgkit.consts.html#mgkit.consts.DEFAULT_SNP_FILTER">DEFAULT_SNP_FILTER (in module mgkit.consts)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.del_attr">del_attr() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.plots.heatmap.html#mgkit.plots.heatmap.dendrogram">dendrogram() (in module mgkit.plots.heatmap)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.density">density (mgkit.align.SamtoolsDepth attribute)</a>
</li>
      <li><a href="api/mgkit.html#mgkit.DependencyError">DependencyError</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.utils.common.html#mgkit.utils.common.deprecated">deprecated() (in module mgkit.utils.common)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.dict_to_text">dict_to_text() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.diff_gff">diff_gff() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.distance_taxa_ancestor">distance_taxa_ancestor() (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.distance_two_taxa">distance_two_taxa() (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.plots.abund.html#mgkit.plots.abund.draw_1d_grid">draw_1d_grid() (in module mgkit.plots.abund)</a>
</li>
      <li><a href="api/mgkit.plots.abund.html#mgkit.plots.abund.draw_axis_internal_triangle">draw_axis_internal_triangle() (in module mgkit.plots.abund)</a>
</li>
      <li><a href="api/mgkit.plots.abund.html#mgkit.plots.abund.draw_circles">draw_circles() (in module mgkit.plots.abund)</a>
</li>
      <li><a href="api/mgkit.plots.abund.html#mgkit.plots.abund.draw_triangle_grid">draw_triangle_grid() (in module mgkit.plots.abund)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.drop_sequence">drop_sequence() (mgkit.align.SamtoolsDepth method)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.drop_taxon">drop_taxon() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.utils.trans_tables.html#mgkit.utils.trans_tables.DRS_MIT">DRS_MIT (in module mgkit.utils.trans_tables)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.DuplicateKeyError">DuplicateKeyError</a>
</li>
  </ul></td>
</tr></table>

<h2 id="E">E</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    edit-gff command line option

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-cite">--cite</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-version">--version</a>
</li>
      </ul></li>
      <li>
    edit-gff-add command line option

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-a">--attributes &lt;attributes&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-o">--only-edited</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-w">--overwrite</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-f">--uids &lt;uids&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-v">--verbose</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-a">-a</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-f">-f</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-o">-o</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-v">-v</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-w">-w</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    edit-gff-fields command line option

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-n">--num-ann &lt;num_ann&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-v">--verbose</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-n">-n</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-v">-v</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-arg-GFF_FILE">GFF_FILE</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-arg-TXT_FILE">TXT_FILE</a>
</li>
      </ul></li>
      <li>
    edit-gff-remove command line option

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-a">--attributes &lt;attributes&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-f">--uids &lt;uids&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-v">--verbose</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-a">-a</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-f">-f</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-v">-v</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    edit-gff-table command line option

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-ai">--attr-index &lt;attr_index&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-a">--attribute &lt;attribute&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-c">--comment &lt;comment&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-k">--key &lt;key&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-ki">--key-index &lt;key_index&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-o">--only-edited</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-s">--separator &lt;separator&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-r">--skip-rows &lt;skip_rows&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-t">--table-file &lt;table_file&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-v">--verbose</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-a">-a</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-ai">-ai</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-c">-c</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-k">-k</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-ki">-ki</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-o">-o</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-r">-r</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-s">-s</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-t">-t</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-v">-v</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    edit-gff-view command line option

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-a">--attributes &lt;attributes&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-h">--header</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-k">--keep-empty</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-s">--separator &lt;separator&gt;</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-v">--verbose</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-a">-a</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-h">-h</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-k">-k</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-s">-s</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-v">-v</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.EGGNOG_CAT">EGGNOG_CAT (in module mgkit.mappings.eggnog)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.EGGNOG_CAT_KEYS">EGGNOG_CAT_KEYS (in module mgkit.mappings.eggnog)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.EGGNOG_CAT_MAP">EGGNOG_CAT_MAP (in module mgkit.mappings.eggnog)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.EGGNOG_CAT_NAMES">EGGNOG_CAT_NAMES (in module mgkit.mappings.eggnog)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.elongate_annotations">elongate_annotations() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.net.embl.html#mgkit.net.embl.EMBL_DBID">EMBL_DBID (in module mgkit.net.embl)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.empty_cache">empty_cache() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.end">end (mgkit.io.gff.GenomicRange attribute)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.entry">entry (mgkit.kegg.KeggModule attribute)</a>
</li>
      <li><a href="api/mgkit.net.embl.html#mgkit.net.embl.EntryNotFound">EntryNotFound</a>
</li>
      <li><a href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.ENZCLASS_REGEX">ENZCLASS_REGEX (in module mgkit.mappings.enzyme)</a>
</li>
      <li><a href="api/mgkit.workflow.utils.html#mgkit.workflow.utils.exit_script">exit_script() (in module mgkit.workflow.utils)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id5">exon (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.exon">[1]</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#id1">exp_nonsyn (mgkit.snps.classes.GeneSNP attribute)</a>, <a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.exp_nonsyn">[1]</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.exp_nonsyn">exp_nonsyn() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#id2">exp_syn (mgkit.snps.classes.GeneSNP attribute)</a>, <a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.exp_syn">[1]</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.exp_syn">exp_syn() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.expand_from_list">expand_from_list() (mgkit.io.gff.GenomicRange method)</a>
</li>
      <li><a href="api/mgkit.filter.reads.html#mgkit.filter.reads.expected_error_rate">expected_error_rate() (in module mgkit.filter.reads)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.extract_nuc_seqs">extract_nuc_seqs() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.extrapolate_model">extrapolate_model() (in module mgkit.utils.sequence)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="F">F</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    fasta-utils command line option

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-cite">--cite</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-version">--version</a>
</li>
      </ul></li>
      <li>
    fasta-utils-split command line option

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-z">--gzip</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-n">--number &lt;number&gt;</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-p">--prefix &lt;prefix&gt;</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-v">--verbose</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-n">-n</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-p">-p</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-v">-v</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-z">-z</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-arg-FASTA_FILE">FASTA_FILE</a>
</li>
      </ul></li>
      <li>
    fasta-utils-translate command line option

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-w">--no-wrap</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-1">--one-seq</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-progress">--progress</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-t">--trans-table &lt;trans_table&gt;</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-v">--verbose</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-1">-1</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-t">-t</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-v">-v</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-w">-w</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-arg-FASTA_FILE">FASTA_FILE</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    fasta-utils-uid command line option

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-t">--table &lt;table&gt;</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-v">--verbose</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-t">-t</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-v">-v</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-arg-FASTA_FILE">FASTA_FILE</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    FASTA_FILE

      <ul>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-split-arg-FASTA_FILE">fasta-utils-split command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-arg-FASTA_FILE">fasta-utils-translate command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-arg-FASTA_FILE">fasta-utils-uid command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-convert-arg-FASTA_FILE">fastq-utils-convert command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-arg-FASTA_FILE">get-gff-info-sequence command line option</a>
</li>
      </ul></li>
      <li>
    fastq-utils command line option

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-cite">--cite</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-version">--version</a>
</li>
      </ul></li>
      <li>
    fastq-utils-convert command line option

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-convert-v">--verbose</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-convert-v">-v</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-convert-arg-FASTA_FILE">FASTA_FILE</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-convert-arg-FASTQ_FILE">FASTQ_FILE</a>
</li>
      </ul></li>
      <li>
    fastq-utils-di command line option

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-s">--strip</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-v">--verbose</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-s">-s</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-v">-v</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-arg-FASTQ_FILE">FASTQ_FILE</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-arg-MATE1_FILE">MATE1_FILE</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-arg-MATE2_FILE">MATE2_FILE</a>
</li>
      </ul></li>
      <li>
    fastq-utils-il command line option

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-il-v">--verbose</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-il-v">-v</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-il-arg-FASTQ_FILE">FASTQ_FILE</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-il-arg-MATE1_FILE">MATE1_FILE</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-il-arg-MATE2_FILE">MATE2_FILE</a>
</li>
      </ul></li>
      <li>
    fastq-utils-sort command line option

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-v">--verbose</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-v">-v</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-arg-MATE1_INPUT">MATE1_INPUT</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-arg-MATE1_OUTPUT">MATE1_OUTPUT</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-arg-MATE2_INPUT">MATE2_INPUT</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-arg-MATE2_OUTPUT">MATE2_OUTPUT</a>
</li>
      </ul></li>
      <li>
    FASTQ_FILE

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-convert-arg-FASTQ_FILE">fastq-utils-convert command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-arg-FASTQ_FILE">fastq-utils-di command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-il-arg-FASTQ_FILE">fastq-utils-il command line option</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.io.snpdat.html#id6">feat_dist (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.feat_dist">[1]</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id7">feat_end (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.feat_end">[1]</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id8">feat_num (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.feat_num">[1]</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id9">feat_start (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.feat_start">[1]</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.feat_type">feat_type (mgkit.io.gff.Annotation attribute)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id10">feature (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.feature">[1]</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.file_handle">file_handle (mgkit.align.SamtoolsDepth attribute)</a>
</li>
      <li>
    filter-gff command line option

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cite">--cite</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-version">--version</a>
</li>
      </ul></li>
      <li>
    filter-gff-cov command line option

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-c">--min-coverage &lt;min_coverage&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-progress">--progress</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-f">--reference &lt;reference&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-r">--rename</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-t">--sorted</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-s">--strand-specific</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-v">--verbose</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-c">-c</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-f">-f</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-r">-r</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-s">-s</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-t">-t</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-v">-v</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    filter-gff-overlap command line option

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-c">--choose-func &lt;choose_func&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-n">--iterations &lt;iterations&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-d">--no-strand</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-progress">--progress</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-s">--size &lt;size&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-a">--sort-attr &lt;sort_attr&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-t">--sorted</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-v">--verbose</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-a">-a</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-c">-c</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-d">-d</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-n">-n</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-s">-s</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-t">-t</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-v">-v</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    filter-gff-sequence command line option

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-a">--attribute &lt;attribute&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-c">--comparison &lt;comparison&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-f">--function &lt;function&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-progress">--progress</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-t">--sorted</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-l">--value &lt;value&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-v">--verbose</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-a">-a</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-c">-c</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-f">-f</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-l">-l</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-t">-t</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-v">-v</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    filter-gff-values command line option

      <ul>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-num-eq">--num-eq &lt;num_eq&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-num-ge">--num-ge &lt;num_ge&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-num-gt">--num-gt &lt;num_gt&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-num-le">--num-le &lt;num_le&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-num-lt">--num-lt &lt;num_lt&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-progress">--progress</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-str-eq">--str-eq &lt;str_eq&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-str-in">--str-in &lt;str_in&gt;</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-v">--verbose</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-v">-v</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.filter.gff.html#mgkit.filter.gff.filter_annotations">filter_annotations() (in module mgkit.filter.gff)</a>
</li>
      <li><a href="api/mgkit.filter.gff.html#mgkit.filter.gff.filter_attr_num">filter_attr_num() (in module mgkit.filter.gff)</a>
</li>
      <li><a href="api/mgkit.filter.gff.html#mgkit.filter.gff.filter_attr_num_s">filter_attr_num_s() (in module mgkit.filter.gff)</a>
</li>
      <li><a href="api/mgkit.filter.gff.html#mgkit.filter.gff.filter_attr_str">filter_attr_str() (in module mgkit.filter.gff)</a>
</li>
      <li><a href="api/mgkit.filter.gff.html#mgkit.filter.gff.filter_base">filter_base() (in module mgkit.filter.gff)</a>
</li>
      <li><a href="api/mgkit.filter.gff.html#mgkit.filter.gff.filter_base_num">filter_base_num() (in module mgkit.filter.gff)</a>
</li>
      <li><a href="api/mgkit.filter.taxon.html#mgkit.filter.taxon.filter_by_ancestor">filter_by_ancestor() (in module mgkit.filter.taxon)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.filter_counts">filter_counts() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.workflow.filter_gff.html#mgkit.workflow.filter_gff.filter_eq">filter_eq() (in module mgkit.workflow.filter_gff)</a>
</li>
      <li><a href="api/mgkit.snps.filter.html#mgkit.snps.filter.filter_genesyn_by_coverage">filter_genesyn_by_coverage() (in module mgkit.snps.filter)</a>
</li>
      <li><a href="api/mgkit.snps.filter.html#mgkit.snps.filter.filter_genesyn_by_gene_id">filter_genesyn_by_gene_id() (in module mgkit.snps.filter)</a>
</li>
      <li><a href="api/mgkit.snps.filter.html#mgkit.snps.filter.filter_genesyn_by_taxon_id">filter_genesyn_by_taxon_id() (in module mgkit.snps.filter)</a>
</li>
      <li><a href="api/mgkit.workflow.filter_gff.html#mgkit.workflow.filter_gff.filter_gt">filter_gt() (in module mgkit.workflow.filter_gff)</a>
</li>
      <li><a href="api/mgkit.workflow.filter_gff.html#mgkit.workflow.filter_gff.filter_in">filter_in() (in module mgkit.workflow.filter_gff)</a>
</li>
      <li><a href="api/mgkit.filter.gff.html#mgkit.filter.gff.filter_len">filter_len() (in module mgkit.filter.gff)</a>
</li>
      <li><a href="api/mgkit.workflow.filter_gff.html#mgkit.workflow.filter_gff.filter_lt">filter_lt() (in module mgkit.workflow.filter_gff)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.filter_nan">filter_nan() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.filter_ratios_by_numbers">filter_ratios_by_numbers() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.filter.taxon.html#mgkit.filter.taxon.filter_taxon_by_id_list">filter_taxon_by_id_list() (in module mgkit.filter.taxon)</a>
</li>
      <li><a href="api/mgkit.filter.common.html#mgkit.filter.common.FilterFails">FilterFails</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.find">find() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.find_annotation">find_annotation() (mgkit.db.mongo.GFFDB method)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.find_by_name">find_by_name() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.workflow.filter_gff.html#mgkit.workflow.filter_gff.find_comparison">find_comparison() (in module mgkit.workflow.filter_gff)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.find_id_in_dict">find_id_in_dict() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.find_submodules">find_submodules() (mgkit.kegg.KeggModule method)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.first_cp">first_cp() (mgkit.kegg.KeggModule property)</a>
</li>
      <li><a href="api/mgkit.counts.glm.html#mgkit.counts.glm.fit_lowess_interpolate">fit_lowess_interpolate() (in module mgkit.counts.glm)</a>
</li>
      <li><a href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.flat_sample_snps">flat_sample_snps() (in module mgkit.snps.funcs)</a>
</li>
      <li><a href="api/mgkit.plots.colors.html#mgkit.plots.colors.float_to_hex_color">float_to_hex_color() (in module mgkit.plots.colors)</a>
</li>
      <li><a href="api/mgkit.logger.html#mgkit.logger.ColorFormatter.format">format() (mgkit.logger.ColorFormatter method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.fpkms">fpkms() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id11">frame (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.frame">[1]</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.from_aa_blast_frag">from_aa_blast_frag() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.from_gff">from_gff() (in module mgkit.counts.func)</a>

      <ul>
        <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.from_gff">(in module mgkit.io.gff)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.from_glimmer3">from_glimmer3() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.from_hmmer">from_hmmer() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.from_json">from_json() (in module mgkit.io.gff)</a>

      <ul>
        <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.from_json">(mgkit.snps.classes.GeneSNP method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.from_mongodb">from_mongodb() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.from_nuc_blast">from_nuc_blast() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.from_nuc_blast_frag">from_nuc_blast_frag() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.from_prodigal_frag">from_prodigal_frag() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.from_sequence">from_sequence() (in module mgkit.io.gff)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="G">G</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.gen_name_map">gen_name_map() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id12">gene_id (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.gene_id">[1]</a>

      <ul>
        <li><a href="api/mgkit.snps.classes.html#id3">(mgkit.snps.classes.GeneSNP attribute)</a>, <a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.gene_id">[1]</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.gene_id">gene_id() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id13">gene_name (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.gene_name">[1]</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP">GeneSNP (class in mgkit.snps.classes)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange">GenomicRange (class in mgkit.io.gff)</a>
</li>
      <li>
    get-gff-info command line option

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cite">--cite</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-version">--version</a>
</li>
      </ul></li>
      <li>
    get-gff-info-cov command line option

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-j">--json-out</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-progress">--progress</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-f">--reference &lt;reference&gt;</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-r">--rename</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-s">--strand-specific</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-v">--verbose</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-f">-f</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-j">-j</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-r">-r</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-s">-s</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-v">-v</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-arg-GFF_FILE">GFF_FILE</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    get-gff-info-dbm command line option

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-dbm-d">--output-dir &lt;output_dir&gt;</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-dbm-v">--verbose</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-dbm-d">-d</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-dbm-v">-v</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-dbm-arg-GFF_FILE">GFF_FILE</a>
</li>
      </ul></li>
      <li>
    get-gff-info-gtf command line option

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-g">--gene-id &lt;gene_id&gt;</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-v">--verbose</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-g">-g</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-v">-v</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-arg-GFF_FILE">GFF_FILE</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-arg-GTF_FILE">GTF_FILE</a>
</li>
      </ul></li>
      <li>
    get-gff-info-mongodb command line option

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-i">--indent &lt;indent&gt;</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-c">--no-cache</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-progress">--progress</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-t">--taxonomy &lt;taxonomy&gt;</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-v">--verbose</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-c">-c</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-i">-i</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-t">-t</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-v">-v</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-arg-GFF_FILE">GFF_FILE</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    get-gff-info-sequence command line option

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-w">--no-wrap</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-progress">--progress</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-f">--reference &lt;reference&gt;</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-r">--reverse</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-s">--split</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-v">--verbose</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-f">-f</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-r">-r</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-s">-s</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-v">-v</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-w">-w</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-arg-FASTA_FILE">FASTA_FILE</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-arg-GFF_FILE">GFF_FILE</a>
</li>
      </ul></li>
      <li>
    get-gff-info-split command line option

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-z">--gzip</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-n">--number &lt;number&gt;</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-p">--prefix &lt;prefix&gt;</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-v">--verbose</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-n">-n</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-p">-p</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-v">-v</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-z">-z</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-arg-GFF_FILE">GFF_FILE</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.workflow.hmmer2gff.html#mgkit.workflow.hmmer2gff.get_aa_data">get_aa_data() (in module mgkit.workflow.hmmer2gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_aa_seq">get_aa_seq() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.get_ancestor_map">get_ancestor_map() (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.get_annotation_map">get_annotation_map() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_attr">get_attr() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.Alignment.get_consensus">get_consensus() (mgkit.utils.sequence.Alignment method)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.get_contigs_info">get_contigs_info() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.snps.filter.html#mgkit.snps.filter.get_default_filters">get_default_filters() (in module mgkit.snps.filter)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_ec">get_ec() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.get_entry">get_entry() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.get_enzyme_full_name">get_enzyme_full_name() (in module mgkit.mappings.enzyme)</a>
</li>
      <li><a href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.get_enzyme_level">get_enzyme_level() (in module mgkit.mappings.enzyme)</a>
</li>
      <li><a href="api/mgkit.snps.conv_func.html#mgkit.snps.conv_func.get_full_dataframe">get_full_dataframe() (in module mgkit.snps.conv_func)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo.get_gene_funccat">get_gene_funccat() (mgkit.mappings.eggnog.NOGInfo method)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.get_gene_info">get_gene_info() (in module mgkit.net.uniprot)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.get_gene_info_iter">get_gene_info_iter() (in module mgkit.net.uniprot)</a>
</li>
      <li><a href="api/mgkit.snps.conv_func.html#mgkit.snps.conv_func.get_gene_map_dataframe">get_gene_map_dataframe() (in module mgkit.snps.conv_func)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo.get_gene_nog">get_gene_nog() (mgkit.mappings.eggnog.NOGInfo method)</a>
</li>
      <li><a href="api/mgkit.snps.conv_func.html#mgkit.snps.conv_func.get_gene_taxon_dataframe">get_gene_taxon_dataframe() (in module mgkit.snps.conv_func)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.get_general_eggnog_cat">get_general_eggnog_cat() (in module mgkit.mappings.eggnog)</a>
</li>
      <li><a href="api/mgkit.plots.utils.html#mgkit.plots.utils.get_grid_figure">get_grid_figure() (in module mgkit.plots.utils)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.get_ids_names">get_ids_names() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.get_ko_to_eggnog_mappings">get_ko_to_eggnog_mappings() (in module mgkit.net.uniprot)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.get_lineage">get_lineage() (in module mgkit.taxon)</a>

      <ul>
        <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.get_lineage">(mgkit.taxon.Taxonomy method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.get_lineage_string">get_lineage_string() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_mapping">get_mapping() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.get_mapping_level">get_mapping_level() (in module mgkit.mappings.enzyme)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.get_mappings">get_mappings() (in module mgkit.net.uniprot)</a>

      <ul>
        <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_mappings">(mgkit.io.gff.Annotation method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.get_name_map">get_name_map() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo.get_nog_funccat">get_nog_funccat() (mgkit.mappings.eggnog.NOGInfo method)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo.get_nog_gencat">get_nog_gencat() (mgkit.mappings.eggnog.NOGInfo method)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo.get_nogs_funccat">get_nogs_funccat() (mgkit.mappings.eggnog.NOGInfo method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_nuc_seq">get_nuc_seq() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.get_number_of_samples">get_number_of_samples() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.get_ortholog_pathways">get_ortholog_pathways() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.get_pathway_links">get_pathway_links() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.net.pfam.html#mgkit.net.pfam.get_pfam_families">get_pfam_families() (in module mgkit.net.pfam)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.Alignment.get_position">get_position() (mgkit.utils.sequence.Alignment method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.get_range">get_range() (mgkit.io.gff.GenomicRange method)</a>
</li>
      <li><a href="api/mgkit.snps.conv_func.html#mgkit.snps.conv_func.get_rank_dataframe">get_rank_dataframe() (in module mgkit.snps.conv_func)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.get_ranked_id">get_ranked_id() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.get_ranked_taxon">get_ranked_taxon() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.get_reaction_equations">get_reaction_equations() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.get_region_coverage">get_region_coverage() (in module mgkit.align)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.get_relative_pos">get_relative_pos() (mgkit.io.gff.GenomicRange method)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.get_seq_expected_syn_count">get_seq_expected_syn_count() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.Alignment.get_seq_len">get_seq_len() (mgkit.utils.sequence.Alignment method)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.get_seq_number_of_syn">get_seq_number_of_syn() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.net.embl.html#mgkit.net.embl.get_sequences_by_ids">get_sequences_by_ids() (in module mgkit.net.embl)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.get_sequences_by_ko">get_sequences_by_ko() (in module mgkit.net.uniprot)</a>
</li>
      <li><a href="api/mgkit.plots.utils.html#mgkit.plots.utils.get_single_figure">get_single_figure() (in module mgkit.plots.utils)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.Alignment.get_snps">get_snps() (mgkit.utils.sequence.Alignment method)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.get_syn_matrix">get_syn_matrix() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.get_syn_matrix_all">get_syn_matrix_all() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.plots.unused.html#mgkit.plots.unused.get_taxon_colors_new">get_taxon_colors_new() (in module mgkit.plots.unused)</a>
</li>
      <li><a href="api/mgkit.workflow.taxon_utils.html#mgkit.workflow.taxon_utils.get_taxon_info">get_taxon_info() (in module mgkit.workflow.taxon_utils)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.get_uid_info">get_uid_info() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.get_uid_info_ann">get_uid_info_ann() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.get_uniprot_ec_mappings">get_uniprot_ec_mappings() (in module mgkit.net.uniprot)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.get_variant_sequence">get_variant_sequence() (in module mgkit.utils.sequence)</a>
</li>
      <li>
    GFF_FILE

      <ul>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-blastdb-arg-GFF_FILE">blast2gff-blastdb command line option</a>
</li>
        <li><a href="scripts/blast2gff.html#cmdoption-blast2gff-uniprot-arg-GFF_FILE">blast2gff-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-arg-GFF_FILE">edit-gff-fields command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-arg-GFF_FILE">get-gff-info-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-dbm-arg-GFF_FILE">get-gff-info-dbm command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-arg-GFF_FILE">get-gff-info-gtf command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-arg-GFF_FILE">get-gff-info-mongodb command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-sequence-arg-GFF_FILE">get-gff-info-sequence command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-split-arg-GFF_FILE">get-gff-info-split command line option</a>
</li>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-mongodb-arg-GFF_FILE">json2gff-mongodb command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-arg-GFF_FILE">taxon-utils-lca command line option</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.db.dbm.html#mgkit.db.dbm.GFFDB">GFFDB (class in mgkit.db.dbm)</a>

      <ul>
        <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB">(class in mgkit.db.mongo)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.mappings.utils.html#mgkit.mappings.utils.group_annotation_by_mapping">group_annotation_by_mapping() (in module mgkit.mappings.utils)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.group_annotations">group_annotations() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.group_annotations_by_ancestor">group_annotations_by_ancestor() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.group_annotations_sorted">group_annotations_sorted() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.mappings.pandas_map.html#mgkit.mappings.pandas_map.group_dataframe_by_mapping">group_dataframe_by_mapping() (in module mgkit.mappings.pandas_map)</a>
</li>
      <li><a href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.group_rank_matrix">group_rank_matrix() (in module mgkit.snps.funcs)</a>
</li>
      <li><a href="api/mgkit.io.utils.html#mgkit.io.utils.group_tuples_by_key">group_tuples_by_key() (in module mgkit.io.utils)</a>
</li>
      <li><a href="api/mgkit.plots.heatmap.html#mgkit.plots.heatmap.grouped_spine">grouped_spine() (in module mgkit.plots.heatmap)</a>
</li>
      <li>
    GTF_FILE

      <ul>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-gtf-arg-GTF_FILE">get-gff-info-gtf command line option</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="H">H</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.has_attr">has_attr() (mgkit.io.gff.Annotation method)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.HDFDict">HDFDict (class in mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.plots.heatmap.html#mgkit.plots.heatmap.heatmap_clustered">heatmap_clustered() (in module mgkit.plots.heatmap)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="I">I</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.id_prefix">id_prefix (mgkit.kegg.KeggClientRest attribute)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id14">in_feat (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.in_feat">[1]</a>
</li>
      <li><a href="api/mgkit.workflow.sampling_utils.html#mgkit.workflow.sampling_utils.infer_parameters">infer_parameters() (in module mgkit.workflow.sampling_utils)</a>
</li>
      <li><a href="api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.init_count_set">init_count_set() (in module mgkit.workflow.snp_parser)</a>
</li>
      <li>
    INPUT_FILE

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-arg-INPUT_FILE">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-arg-INPUT_FILE">add-gff-info-counts command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-arg-INPUT_FILE">add-gff-info-cov_samtools command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-arg-INPUT_FILE">add-gff-info-coverage command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-arg-INPUT_FILE">add-gff-info-exp_syn command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-arg-INPUT_FILE">add-gff-info-kegg command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-arg-INPUT_FILE">add-gff-info-pfam command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-arg-INPUT_FILE">add-gff-info-unipfile command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-arg-INPUT_FILE">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-arg-INPUT_FILE">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-arg-INPUT_FILE">edit-gff-remove command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-arg-INPUT_FILE">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-arg-INPUT_FILE">edit-gff-view command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-arg-INPUT_FILE">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-arg-INPUT_FILE">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-arg-INPUT_FILE">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-arg-INPUT_FILE">filter-gff-values command line option</a>
</li>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-mongodb-arg-INPUT_FILE">json2gff-mongodb command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-arg-INPUT_FILE">sampling-utils-sample command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-arg-INPUT_FILE">sampling-utils-sample_stream command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-arg-INPUT_FILE">sampling-utils-sync command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-arg-INPUT_FILE">taxon-utils-filter command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-arg-INPUT_FILE">taxon-utils-lca_line command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-arg-INPUT_FILE">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.insert_many">insert_many() (mgkit.db.mongo.GFFDB method)</a>
</li>
      <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.insert_one">insert_one() (mgkit.db.mongo.GFFDB method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.intersect">intersect() (mgkit.io.gff.GenomicRange method)</a>
</li>
      <li><a href="api/mgkit.utils.trans_tables.html#mgkit.utils.trans_tables.INV_MIT">INV_MIT (in module mgkit.utils.trans_tables)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.is_ancestor">is_ancestor() (in module mgkit.taxon)</a>

      <ul>
        <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.is_ancestor">(mgkit.taxon.Taxonomy method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.is_ranked_below">is_ranked_below() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.is_syn">is_syn() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.db.dbm.html#mgkit.db.dbm.GFFDB.items">items() (mgkit.db.dbm.GFFDB method)</a>

      <ul>
        <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.items">(mgkit.db.mongo.GFFDB method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.iter_ids">iter_ids() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.db.dbm.html#mgkit.db.dbm.GFFDB.iteritems">iteritems() (mgkit.db.dbm.GFFDB method)</a>

      <ul>
        <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.iteritems">(mgkit.db.mongo.GFFDB method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.db.dbm.html#mgkit.db.dbm.GFFDB.itervalues">itervalues() (mgkit.db.dbm.GFFDB method)</a>

      <ul>
        <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.itervalues">(mgkit.db.mongo.GFFDB method)</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="J">J</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    json2gff command line option

      <ul>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-cite">--cite</a>
</li>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-version">--version</a>
</li>
      </ul></li>
      <li>
    json2gff-mongodb command line option

      <ul>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-mongodb-v">--verbose</a>
</li>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-mongodb-v">-v</a>
</li>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-mongodb-arg-GFF_FILE">GFF_FILE</a>
</li>
        <li><a href="scripts/json2gff.html#cmdoption-json2gff-mongodb-arg-INPUT_FILE">INPUT_FILE</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="K">K</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest">KeggClientRest (class in mgkit.kegg)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule">KeggModule (class in mgkit.kegg)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.keys">keys() (mgkit.db.mongo.GFFDB method)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.ko_desc_re">ko_desc_re (mgkit.kegg.KeggClientRest attribute)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.ko_to_mapping">ko_to_mapping() (in module mgkit.net.uniprot)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="L">L</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.last_common_ancestor">last_common_ancestor() (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.last_common_ancestor_multiple">last_common_ancestor_multiple() (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.last_cp">last_cp() (mgkit.kegg.KeggModule property)</a>
</li>
      <li><a href="api/mgkit.plots.utils.html#mgkit.plots.utils.legend_patches">legend_patches() (in module mgkit.plots.utils)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.length">length() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.LEVEL1_NAMES">LEVEL1_NAMES (in module mgkit.mappings.enzyme)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonTuple.lineage">lineage (mgkit.taxon.UniprotTaxonTuple attribute)</a>
</li>
      <li><a href="api/mgkit.plots.unused.html#mgkit.plots.unused.lineplot_values_on_second_axis">lineplot_values_on_second_axis() (in module mgkit.plots.unused)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.link">link() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.link_ids">link_ids() (in module mgkit.utils.dictionary)</a>

      <ul>
        <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.link_ids">(mgkit.kegg.KeggClientRest method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.list_ids">list_ids() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.load_cache">load_cache() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.load_counts_from_gff">load_counts_from_gff() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.load_data">load_data() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo.load_description">load_description() (mgkit.mappings.eggnog.NOGInfo method)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.load_deseq2_results">load_deseq2_results() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.io.fasta.html#mgkit.io.fasta.load_fasta">load_fasta() (in module mgkit.io.fasta)</a>
</li>
      <li><a href="api/mgkit.workflow.blast2gff.html#mgkit.workflow.blast2gff.load_fasta_file">load_fasta_file() (in module mgkit.workflow.blast2gff)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.io.fasta.html#mgkit.io.fasta.load_fasta_files">load_fasta_files() (in module mgkit.io.fasta)</a>
</li>
      <li><a href="api/mgkit.io.fasta.html#mgkit.io.fasta.load_fasta_prodigal">load_fasta_prodigal() (in module mgkit.io.fasta)</a>
</li>
      <li><a href="api/mgkit.io.fasta.html#mgkit.io.fasta.load_fasta_rename">load_fasta_rename() (in module mgkit.io.fasta)</a>
</li>
      <li><a href="api/mgkit.io.fastq.html#mgkit.io.fastq.load_fastq">load_fastq() (in module mgkit.io.fastq)</a>
</li>
      <li><a href="api/mgkit.io.fastq.html#mgkit.io.fastq.load_fastq_rename">load_fastq_rename() (in module mgkit.io.fastq)</a>
</li>
      <li><a href="api/mgkit.workflow.add_gff_info.html#mgkit.workflow.add_gff_info.load_featurecounts_files">load_featurecounts_files() (in module mgkit.workflow.add_gff_info)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo.load_funccat">load_funccat() (mgkit.mappings.eggnog.NOGInfo method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.load_gff_base_info">load_gff_base_info() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.load_gff_mappings">load_gff_mappings() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.workflow.add_gff_info.html#mgkit.workflow.add_gff_info.load_htseq_count_files">load_htseq_count_files() (in module mgkit.workflow.add_gff_info)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.load_htseq_counts">load_htseq_counts() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo.load_members">load_members() (mgkit.mappings.eggnog.NOGInfo method)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.load_sample_counts">load_sample_counts() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.load_sample_counts_to_genes">load_sample_counts_to_genes() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.load_sample_counts_to_taxon">load_sample_counts_to_taxon() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.workflow.fasta_utils.html#mgkit.workflow.fasta_utils.load_trans_table">load_trans_table() (in module mgkit.workflow.fasta_utils)</a>
</li>
      <li><a href="api/mgkit.net.embl.html#mgkit.net.embl.LOG">LOG (in module mgkit.net.embl)</a>
</li>
      <li><a href="api/mgkit.counts.glm.html#mgkit.counts.glm.lowess_ci_bootstrap">lowess_ci_bootstrap() (in module mgkit.counts.glm)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.lowest_common_ancestor">lowest_common_ancestor() (in module mgkit.taxon)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="M">M</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.workflow.hmmer2gff.html#mgkit.workflow.hmmer2gff.main">main() (in module mgkit.workflow.hmmer2gff)</a>

      <ul>
        <li><a href="api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.main">(in module mgkit.workflow.snp_parser)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.workflow.filter_gff.html#mgkit.workflow.filter_gff.make_choose_func">make_choose_func() (in module mgkit.workflow.filter_gff)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.make_reverse_table">make_reverse_table() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.mappings.pandas_map.html#mgkit.mappings.pandas_map.make_stat_table">make_stat_table() (in module mgkit.mappings.pandas_map)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.map_counts">map_counts() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.map_counts_to_category">map_counts_to_category() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.snps.mapper.html#mgkit.snps.mapper.map_gene_id">map_gene_id() (in module mgkit.snps.mapper)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.map_gene_id_to_map">map_gene_id_to_map() (in module mgkit.counts.func)</a>
</li>
      <li><a href="api/mgkit.mappings.taxon.html#mgkit.mappings.taxon.map_taxon_by_id_list">map_taxon_by_id_list() (in module mgkit.mappings.taxon)</a>
</li>
      <li><a href="api/mgkit.snps.mapper.html#mgkit.snps.mapper.map_taxon_id_to_ancestor">map_taxon_id_to_ancestor() (in module mgkit.snps.mapper)</a>
</li>
      <li><a href="api/mgkit.counts.func.html#mgkit.counts.func.map_taxon_id_to_rank">map_taxon_id_to_rank() (in module mgkit.counts.func)</a>

      <ul>
        <li><a href="api/mgkit.snps.mapper.html#mgkit.snps.mapper.map_taxon_id_to_rank">(in module mgkit.snps.mapper)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.plots.unused.html#mgkit.plots.unused.map_taxon_to_colours">map_taxon_to_colours() (in module mgkit.plots.unused)</a>
</li>
      <li><a href="api/mgkit.io.uniprot.html#mgkit.io.uniprot.MAPPINGS">MAPPINGS (in module mgkit.io.uniprot)</a>
</li>
      <li>
    MATE1_FILE

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-arg-MATE1_FILE">fastq-utils-di command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-il-arg-MATE1_FILE">fastq-utils-il command line option</a>
</li>
      </ul></li>
      <li>
    MATE1_INPUT

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-arg-MATE1_INPUT">fastq-utils-sort command line option</a>
</li>
      </ul></li>
      <li>
    MATE1_OUTPUT

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-arg-MATE1_OUTPUT">fastq-utils-sort command line option</a>
</li>
      </ul></li>
      <li>
    MATE2_FILE

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-di-arg-MATE2_FILE">fastq-utils-di command line option</a>
</li>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-il-arg-MATE2_FILE">fastq-utils-il command line option</a>
</li>
      </ul></li>
      <li>
    MATE2_INPUT

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-arg-MATE2_INPUT">fastq-utils-sort command line option</a>
</li>
      </ul></li>
      <li>
    MATE2_OUTPUT

      <ul>
        <li><a href="scripts/fastq-utils.html#cmdoption-fastq-utils-sort-arg-MATE2_OUTPUT">fastq-utils-sort command line option</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.max_size">max_size (mgkit.align.SamtoolsDepth attribute)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.max_size_dict">max_size_dict (mgkit.align.SamtoolsDepth attribute)</a>
</li>
      <li><a href="api/mgkit.simple_cache.html#mgkit.simple_cache.memoize">memoize (class in mgkit.simple_cache)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.merge_dictionaries">merge_dictionaries() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id15">messages (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.messages">[1]</a>
</li>
      <li>
    mgkit

      <ul>
        <li><a href="api/mgkit.html#module-mgkit">module</a>
</li>
      </ul></li>
      <li>
    mgkit.align

      <ul>
        <li><a href="api/mgkit.align.html#module-mgkit.align">module</a>
</li>
      </ul></li>
      <li>
    mgkit.consts

      <ul>
        <li><a href="api/mgkit.consts.html#module-mgkit.consts">module</a>
</li>
      </ul></li>
      <li>
    mgkit.counts

      <ul>
        <li><a href="api/mgkit.counts.html#module-mgkit.counts">module</a>
</li>
      </ul></li>
      <li>
    mgkit.counts.func

      <ul>
        <li><a href="api/mgkit.counts.func.html#module-mgkit.counts.func">module</a>
</li>
      </ul></li>
      <li>
    mgkit.counts.glm

      <ul>
        <li><a href="api/mgkit.counts.glm.html#module-mgkit.counts.glm">module</a>
</li>
      </ul></li>
      <li>
    mgkit.counts.scaling

      <ul>
        <li><a href="api/mgkit.counts.scaling.html#module-mgkit.counts.scaling">module</a>
</li>
      </ul></li>
      <li>
    mgkit.db

      <ul>
        <li><a href="api/mgkit.db.html#module-mgkit.db">module</a>
</li>
      </ul></li>
      <li>
    mgkit.db.dbm

      <ul>
        <li><a href="api/mgkit.db.dbm.html#module-mgkit.db.dbm">module</a>
</li>
      </ul></li>
      <li>
    mgkit.db.mongo

      <ul>
        <li><a href="api/mgkit.db.mongo.html#module-mgkit.db.mongo">module</a>
</li>
      </ul></li>
      <li>
    mgkit.filter

      <ul>
        <li><a href="api/mgkit.filter.html#module-mgkit.filter">module</a>
</li>
      </ul></li>
      <li>
    mgkit.filter.common

      <ul>
        <li><a href="api/mgkit.filter.common.html#module-mgkit.filter.common">module</a>
</li>
      </ul></li>
      <li>
    mgkit.filter.gff

      <ul>
        <li><a href="api/mgkit.filter.gff.html#module-mgkit.filter.gff">module</a>
</li>
      </ul></li>
      <li>
    mgkit.filter.lists

      <ul>
        <li><a href="api/mgkit.filter.lists.html#module-mgkit.filter.lists">module</a>
</li>
      </ul></li>
      <li>
    mgkit.filter.reads

      <ul>
        <li><a href="api/mgkit.filter.reads.html#module-mgkit.filter.reads">module</a>
</li>
      </ul></li>
      <li>
    mgkit.filter.taxon

      <ul>
        <li><a href="api/mgkit.filter.taxon.html#module-mgkit.filter.taxon">module</a>
</li>
      </ul></li>
      <li>
    mgkit.io

      <ul>
        <li><a href="api/mgkit.io.html#module-mgkit.io">module</a>
</li>
      </ul></li>
      <li>
    mgkit.io.blast

      <ul>
        <li><a href="api/mgkit.io.blast.html#module-mgkit.io.blast">module</a>
</li>
      </ul></li>
      <li>
    mgkit.io.fasta

      <ul>
        <li><a href="api/mgkit.io.fasta.html#module-mgkit.io.fasta">module</a>
</li>
      </ul></li>
      <li>
    mgkit.io.fastq

      <ul>
        <li><a href="api/mgkit.io.fastq.html#module-mgkit.io.fastq">module</a>
</li>
      </ul></li>
      <li>
    mgkit.io.gff

      <ul>
        <li><a href="api/mgkit.io.gff.html#module-mgkit.io.gff">module</a>
</li>
      </ul></li>
      <li>
    mgkit.io.glimmer

      <ul>
        <li><a href="api/mgkit.io.glimmer.html#module-mgkit.io.glimmer">module</a>
</li>
      </ul></li>
      <li>
    mgkit.io.snpdat

      <ul>
        <li><a href="api/mgkit.io.snpdat.html#module-mgkit.io.snpdat">module</a>
</li>
      </ul></li>
      <li>
    mgkit.io.uniprot

      <ul>
        <li><a href="api/mgkit.io.uniprot.html#module-mgkit.io.uniprot">module</a>
</li>
      </ul></li>
      <li>
    mgkit.io.utils

      <ul>
        <li><a href="api/mgkit.io.utils.html#module-mgkit.io.utils">module</a>
</li>
      </ul></li>
      <li>
    mgkit.kegg

      <ul>
        <li><a href="api/mgkit.kegg.html#module-mgkit.kegg">module</a>
</li>
      </ul></li>
      <li>
    mgkit.logger

      <ul>
        <li><a href="api/mgkit.logger.html#module-mgkit.logger">module</a>
</li>
      </ul></li>
      <li>
    mgkit.mappings

      <ul>
        <li><a href="api/mgkit.mappings.html#module-mgkit.mappings">module</a>
</li>
      </ul></li>
      <li>
    mgkit.mappings.cazy

      <ul>
        <li><a href="api/mgkit.mappings.cazy.html#module-mgkit.mappings.cazy">module</a>
</li>
      </ul></li>
      <li>
    mgkit.mappings.eggnog

      <ul>
        <li><a href="api/mgkit.mappings.eggnog.html#module-mgkit.mappings.eggnog">module</a>
</li>
      </ul></li>
      <li>
    mgkit.mappings.enzyme

      <ul>
        <li><a href="api/mgkit.mappings.enzyme.html#module-mgkit.mappings.enzyme">module</a>
</li>
      </ul></li>
      <li>
    mgkit.mappings.go

      <ul>
        <li><a href="api/mgkit.mappings.go.html#module-mgkit.mappings.go">module</a>
</li>
      </ul></li>
      <li>
    mgkit.mappings.pandas_map

      <ul>
        <li><a href="api/mgkit.mappings.pandas_map.html#module-mgkit.mappings.pandas_map">module</a>
</li>
      </ul></li>
      <li>
    mgkit.mappings.taxon

      <ul>
        <li><a href="api/mgkit.mappings.taxon.html#module-mgkit.mappings.taxon">module</a>
</li>
      </ul></li>
      <li>
    mgkit.mappings.utils

      <ul>
        <li><a href="api/mgkit.mappings.utils.html#module-mgkit.mappings.utils">module</a>
</li>
      </ul></li>
      <li>
    mgkit.net

      <ul>
        <li><a href="api/mgkit.net.html#module-mgkit.net">module</a>
</li>
      </ul></li>
      <li>
    mgkit.net.embl

      <ul>
        <li><a href="api/mgkit.net.embl.html#module-mgkit.net.embl">module</a>
</li>
      </ul></li>
      <li>
    mgkit.net.pfam

      <ul>
        <li><a href="api/mgkit.net.pfam.html#module-mgkit.net.pfam">module</a>
</li>
      </ul></li>
      <li>
    mgkit.net.uniprot

      <ul>
        <li><a href="api/mgkit.net.uniprot.html#module-mgkit.net.uniprot">module</a>
</li>
      </ul></li>
      <li>
    mgkit.net.utils

      <ul>
        <li><a href="api/mgkit.net.utils.html#module-mgkit.net.utils">module</a>
</li>
      </ul></li>
      <li>
    mgkit.plots

      <ul>
        <li><a href="api/mgkit.plots.html#module-mgkit.plots">module</a>
</li>
      </ul></li>
      <li>
    mgkit.plots.abund

      <ul>
        <li><a href="api/mgkit.plots.abund.html#module-mgkit.plots.abund">module</a>
</li>
      </ul></li>
      <li>
    mgkit.plots.boxplot

      <ul>
        <li><a href="api/mgkit.plots.boxplot.html#module-mgkit.plots.boxplot">module</a>
</li>
      </ul></li>
      <li>
    mgkit.plots.colors

      <ul>
        <li><a href="api/mgkit.plots.colors.html#module-mgkit.plots.colors">module</a>
</li>
      </ul></li>
      <li>
    mgkit.plots.heatmap

      <ul>
        <li><a href="api/mgkit.plots.heatmap.html#module-mgkit.plots.heatmap">module</a>
</li>
      </ul></li>
      <li>
    mgkit.plots.unused

      <ul>
        <li><a href="api/mgkit.plots.unused.html#module-mgkit.plots.unused">module</a>
</li>
      </ul></li>
      <li>
    mgkit.plots.utils

      <ul>
        <li><a href="api/mgkit.plots.utils.html#module-mgkit.plots.utils">module</a>
</li>
      </ul></li>
      <li>
    mgkit.simple_cache

      <ul>
        <li><a href="api/mgkit.simple_cache.html#module-mgkit.simple_cache">module</a>
</li>
      </ul></li>
      <li>
    mgkit.snps

      <ul>
        <li><a href="api/mgkit.snps.html#module-mgkit.snps">module</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    mgkit.snps.classes

      <ul>
        <li><a href="api/mgkit.snps.classes.html#module-mgkit.snps.classes">module</a>
</li>
      </ul></li>
      <li>
    mgkit.snps.conv_func

      <ul>
        <li><a href="api/mgkit.snps.conv_func.html#module-mgkit.snps.conv_func">module</a>
</li>
      </ul></li>
      <li>
    mgkit.snps.filter

      <ul>
        <li><a href="api/mgkit.snps.filter.html#module-mgkit.snps.filter">module</a>
</li>
      </ul></li>
      <li>
    mgkit.snps.funcs

      <ul>
        <li><a href="api/mgkit.snps.funcs.html#module-mgkit.snps.funcs">module</a>
</li>
      </ul></li>
      <li>
    mgkit.snps.mapper

      <ul>
        <li><a href="api/mgkit.snps.mapper.html#module-mgkit.snps.mapper">module</a>
</li>
      </ul></li>
      <li>
    mgkit.taxon

      <ul>
        <li><a href="api/mgkit.taxon.html#module-mgkit.taxon">module</a>
</li>
      </ul></li>
      <li>
    mgkit.utils

      <ul>
        <li><a href="api/mgkit.utils.html#module-mgkit.utils">module</a>
</li>
      </ul></li>
      <li>
    mgkit.utils.common

      <ul>
        <li><a href="api/mgkit.utils.common.html#module-mgkit.utils.common">module</a>
</li>
      </ul></li>
      <li>
    mgkit.utils.dictionary

      <ul>
        <li><a href="api/mgkit.utils.dictionary.html#module-mgkit.utils.dictionary">module</a>
</li>
      </ul></li>
      <li>
    mgkit.utils.sequence

      <ul>
        <li><a href="api/mgkit.utils.sequence.html#module-mgkit.utils.sequence">module</a>
</li>
      </ul></li>
      <li>
    mgkit.utils.trans_tables

      <ul>
        <li><a href="api/mgkit.utils.trans_tables.html#module-mgkit.utils.trans_tables">module</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow

      <ul>
        <li><a href="api/mgkit.workflow.html#module-mgkit.workflow">module</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.add_gff_info

      <ul>
        <li><a href="api/mgkit.workflow.add_gff_info.html#module-mgkit.workflow.add_gff_info">module</a>, <a href="scripts/add-gff-info.html#module-mgkit.workflow.add_gff_info">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.blast2gff

      <ul>
        <li><a href="api/mgkit.workflow.blast2gff.html#module-mgkit.workflow.blast2gff">module</a>, <a href="scripts/blast2gff.html#module-mgkit.workflow.blast2gff">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.edit_gff

      <ul>
        <li><a href="scripts/edit_gff.html#module-mgkit.workflow.edit_gff">module</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.extract_gff_info

      <ul>
        <li><a href="api/mgkit.workflow.extract_gff_info.html#module-mgkit.workflow.extract_gff_info">module</a>, <a href="scripts/get-gff-info.html#module-mgkit.workflow.extract_gff_info">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.fasta_utils

      <ul>
        <li><a href="api/mgkit.workflow.fasta_utils.html#module-mgkit.workflow.fasta_utils">module</a>, <a href="scripts/fasta-utils.html#module-mgkit.workflow.fasta_utils">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.fastq_utils

      <ul>
        <li><a href="api/mgkit.workflow.fastq_utils.html#module-mgkit.workflow.fastq_utils">module</a>, <a href="scripts/fastq-utils.html#module-mgkit.workflow.fastq_utils">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.filter_gff

      <ul>
        <li><a href="api/mgkit.workflow.filter_gff.html#module-mgkit.workflow.filter_gff">module</a>, <a href="scripts/filter-gff.html#module-mgkit.workflow.filter_gff">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.hmmer2gff

      <ul>
        <li><a href="api/mgkit.workflow.hmmer2gff.html#module-mgkit.workflow.hmmer2gff">module</a>, <a href="scripts/hmmer2gff.html#module-mgkit.workflow.hmmer2gff">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.json2gff

      <ul>
        <li><a href="api/mgkit.workflow.json2gff.html#module-mgkit.workflow.json2gff">module</a>, <a href="scripts/json2gff.html#module-mgkit.workflow.json2gff">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.pnps_gen

      <ul>
        <li><a href="scripts/pnps_gen.html#module-mgkit.workflow.pnps_gen">module</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.sampling_utils

      <ul>
        <li><a href="api/mgkit.workflow.sampling_utils.html#module-mgkit.workflow.sampling_utils">module</a>, <a href="scripts/sampling-utils.html#module-mgkit.workflow.sampling_utils">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.snp_parser

      <ul>
        <li><a href="api/mgkit.workflow.snp_parser.html#module-mgkit.workflow.snp_parser">module</a>, <a href="scripts/snp_analysis.html#module-mgkit.workflow.snp_parser">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.taxon_utils

      <ul>
        <li><a href="api/mgkit.workflow.taxon_utils.html#module-mgkit.workflow.taxon_utils">module</a>, <a href="scripts/taxon_utils.html#module-mgkit.workflow.taxon_utils">[1]</a>
</li>
      </ul></li>
      <li>
    mgkit.workflow.utils

      <ul>
        <li><a href="api/mgkit.workflow.utils.html#module-mgkit.workflow.utils">module</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.consts.html#mgkit.consts.MIN_COV">MIN_COV (in module mgkit.consts)</a>
</li>
      <li><a href="api/mgkit.consts.html#mgkit.consts.MIN_NUM">MIN_NUM (in module mgkit.consts)</a>
</li>
      <li>
    module

      <ul>
        <li><a href="api/mgkit.html#module-mgkit">mgkit</a>
</li>
        <li><a href="api/mgkit.align.html#module-mgkit.align">mgkit.align</a>
</li>
        <li><a href="api/mgkit.consts.html#module-mgkit.consts">mgkit.consts</a>
</li>
        <li><a href="api/mgkit.counts.html#module-mgkit.counts">mgkit.counts</a>
</li>
        <li><a href="api/mgkit.counts.func.html#module-mgkit.counts.func">mgkit.counts.func</a>
</li>
        <li><a href="api/mgkit.counts.glm.html#module-mgkit.counts.glm">mgkit.counts.glm</a>
</li>
        <li><a href="api/mgkit.counts.scaling.html#module-mgkit.counts.scaling">mgkit.counts.scaling</a>
</li>
        <li><a href="api/mgkit.db.html#module-mgkit.db">mgkit.db</a>
</li>
        <li><a href="api/mgkit.db.dbm.html#module-mgkit.db.dbm">mgkit.db.dbm</a>
</li>
        <li><a href="api/mgkit.db.mongo.html#module-mgkit.db.mongo">mgkit.db.mongo</a>
</li>
        <li><a href="api/mgkit.filter.html#module-mgkit.filter">mgkit.filter</a>
</li>
        <li><a href="api/mgkit.filter.common.html#module-mgkit.filter.common">mgkit.filter.common</a>
</li>
        <li><a href="api/mgkit.filter.gff.html#module-mgkit.filter.gff">mgkit.filter.gff</a>
</li>
        <li><a href="api/mgkit.filter.lists.html#module-mgkit.filter.lists">mgkit.filter.lists</a>
</li>
        <li><a href="api/mgkit.filter.reads.html#module-mgkit.filter.reads">mgkit.filter.reads</a>
</li>
        <li><a href="api/mgkit.filter.taxon.html#module-mgkit.filter.taxon">mgkit.filter.taxon</a>
</li>
        <li><a href="api/mgkit.io.html#module-mgkit.io">mgkit.io</a>
</li>
        <li><a href="api/mgkit.io.blast.html#module-mgkit.io.blast">mgkit.io.blast</a>
</li>
        <li><a href="api/mgkit.io.fasta.html#module-mgkit.io.fasta">mgkit.io.fasta</a>
</li>
        <li><a href="api/mgkit.io.fastq.html#module-mgkit.io.fastq">mgkit.io.fastq</a>
</li>
        <li><a href="api/mgkit.io.gff.html#module-mgkit.io.gff">mgkit.io.gff</a>
</li>
        <li><a href="api/mgkit.io.glimmer.html#module-mgkit.io.glimmer">mgkit.io.glimmer</a>
</li>
        <li><a href="api/mgkit.io.snpdat.html#module-mgkit.io.snpdat">mgkit.io.snpdat</a>
</li>
        <li><a href="api/mgkit.io.uniprot.html#module-mgkit.io.uniprot">mgkit.io.uniprot</a>
</li>
        <li><a href="api/mgkit.io.utils.html#module-mgkit.io.utils">mgkit.io.utils</a>
</li>
        <li><a href="api/mgkit.kegg.html#module-mgkit.kegg">mgkit.kegg</a>
</li>
        <li><a href="api/mgkit.logger.html#module-mgkit.logger">mgkit.logger</a>
</li>
        <li><a href="api/mgkit.mappings.html#module-mgkit.mappings">mgkit.mappings</a>
</li>
        <li><a href="api/mgkit.mappings.cazy.html#module-mgkit.mappings.cazy">mgkit.mappings.cazy</a>
</li>
        <li><a href="api/mgkit.mappings.eggnog.html#module-mgkit.mappings.eggnog">mgkit.mappings.eggnog</a>
</li>
        <li><a href="api/mgkit.mappings.enzyme.html#module-mgkit.mappings.enzyme">mgkit.mappings.enzyme</a>
</li>
        <li><a href="api/mgkit.mappings.go.html#module-mgkit.mappings.go">mgkit.mappings.go</a>
</li>
        <li><a href="api/mgkit.mappings.pandas_map.html#module-mgkit.mappings.pandas_map">mgkit.mappings.pandas_map</a>
</li>
        <li><a href="api/mgkit.mappings.taxon.html#module-mgkit.mappings.taxon">mgkit.mappings.taxon</a>
</li>
        <li><a href="api/mgkit.mappings.utils.html#module-mgkit.mappings.utils">mgkit.mappings.utils</a>
</li>
        <li><a href="api/mgkit.net.html#module-mgkit.net">mgkit.net</a>
</li>
        <li><a href="api/mgkit.net.embl.html#module-mgkit.net.embl">mgkit.net.embl</a>
</li>
        <li><a href="api/mgkit.net.pfam.html#module-mgkit.net.pfam">mgkit.net.pfam</a>
</li>
        <li><a href="api/mgkit.net.uniprot.html#module-mgkit.net.uniprot">mgkit.net.uniprot</a>
</li>
        <li><a href="api/mgkit.net.utils.html#module-mgkit.net.utils">mgkit.net.utils</a>
</li>
        <li><a href="api/mgkit.plots.html#module-mgkit.plots">mgkit.plots</a>
</li>
        <li><a href="api/mgkit.plots.abund.html#module-mgkit.plots.abund">mgkit.plots.abund</a>
</li>
        <li><a href="api/mgkit.plots.boxplot.html#module-mgkit.plots.boxplot">mgkit.plots.boxplot</a>
</li>
        <li><a href="api/mgkit.plots.colors.html#module-mgkit.plots.colors">mgkit.plots.colors</a>
</li>
        <li><a href="api/mgkit.plots.heatmap.html#module-mgkit.plots.heatmap">mgkit.plots.heatmap</a>
</li>
        <li><a href="api/mgkit.plots.unused.html#module-mgkit.plots.unused">mgkit.plots.unused</a>
</li>
        <li><a href="api/mgkit.plots.utils.html#module-mgkit.plots.utils">mgkit.plots.utils</a>
</li>
        <li><a href="api/mgkit.simple_cache.html#module-mgkit.simple_cache">mgkit.simple_cache</a>
</li>
        <li><a href="api/mgkit.snps.html#module-mgkit.snps">mgkit.snps</a>
</li>
        <li><a href="api/mgkit.snps.classes.html#module-mgkit.snps.classes">mgkit.snps.classes</a>
</li>
        <li><a href="api/mgkit.snps.conv_func.html#module-mgkit.snps.conv_func">mgkit.snps.conv_func</a>
</li>
        <li><a href="api/mgkit.snps.filter.html#module-mgkit.snps.filter">mgkit.snps.filter</a>
</li>
        <li><a href="api/mgkit.snps.funcs.html#module-mgkit.snps.funcs">mgkit.snps.funcs</a>
</li>
        <li><a href="api/mgkit.snps.mapper.html#module-mgkit.snps.mapper">mgkit.snps.mapper</a>
</li>
        <li><a href="api/mgkit.taxon.html#module-mgkit.taxon">mgkit.taxon</a>
</li>
        <li><a href="api/mgkit.utils.html#module-mgkit.utils">mgkit.utils</a>
</li>
        <li><a href="api/mgkit.utils.common.html#module-mgkit.utils.common">mgkit.utils.common</a>
</li>
        <li><a href="api/mgkit.utils.dictionary.html#module-mgkit.utils.dictionary">mgkit.utils.dictionary</a>
</li>
        <li><a href="api/mgkit.utils.sequence.html#module-mgkit.utils.sequence">mgkit.utils.sequence</a>
</li>
        <li><a href="api/mgkit.utils.trans_tables.html#module-mgkit.utils.trans_tables">mgkit.utils.trans_tables</a>
</li>
        <li><a href="api/mgkit.workflow.html#module-mgkit.workflow">mgkit.workflow</a>
</li>
        <li><a href="api/mgkit.workflow.add_gff_info.html#module-mgkit.workflow.add_gff_info">mgkit.workflow.add_gff_info</a>, <a href="scripts/add-gff-info.html#module-mgkit.workflow.add_gff_info">[1]</a>
</li>
        <li><a href="api/mgkit.workflow.blast2gff.html#module-mgkit.workflow.blast2gff">mgkit.workflow.blast2gff</a>, <a href="scripts/blast2gff.html#module-mgkit.workflow.blast2gff">[1]</a>
</li>
        <li><a href="scripts/edit_gff.html#module-mgkit.workflow.edit_gff">mgkit.workflow.edit_gff</a>
</li>
        <li><a href="api/mgkit.workflow.extract_gff_info.html#module-mgkit.workflow.extract_gff_info">mgkit.workflow.extract_gff_info</a>, <a href="scripts/get-gff-info.html#module-mgkit.workflow.extract_gff_info">[1]</a>
</li>
        <li><a href="api/mgkit.workflow.fasta_utils.html#module-mgkit.workflow.fasta_utils">mgkit.workflow.fasta_utils</a>, <a href="scripts/fasta-utils.html#module-mgkit.workflow.fasta_utils">[1]</a>
</li>
        <li><a href="api/mgkit.workflow.fastq_utils.html#module-mgkit.workflow.fastq_utils">mgkit.workflow.fastq_utils</a>, <a href="scripts/fastq-utils.html#module-mgkit.workflow.fastq_utils">[1]</a>
</li>
        <li><a href="api/mgkit.workflow.filter_gff.html#module-mgkit.workflow.filter_gff">mgkit.workflow.filter_gff</a>, <a href="scripts/filter-gff.html#module-mgkit.workflow.filter_gff">[1]</a>
</li>
        <li><a href="api/mgkit.workflow.hmmer2gff.html#module-mgkit.workflow.hmmer2gff">mgkit.workflow.hmmer2gff</a>, <a href="scripts/hmmer2gff.html#module-mgkit.workflow.hmmer2gff">[1]</a>
</li>
        <li><a href="api/mgkit.workflow.json2gff.html#module-mgkit.workflow.json2gff">mgkit.workflow.json2gff</a>, <a href="scripts/json2gff.html#module-mgkit.workflow.json2gff">[1]</a>
</li>
        <li><a href="scripts/pnps_gen.html#module-mgkit.workflow.pnps_gen">mgkit.workflow.pnps_gen</a>
</li>
        <li><a href="api/mgkit.workflow.sampling_utils.html#module-mgkit.workflow.sampling_utils">mgkit.workflow.sampling_utils</a>, <a href="scripts/sampling-utils.html#module-mgkit.workflow.sampling_utils">[1]</a>
</li>
        <li><a href="api/mgkit.workflow.snp_parser.html#module-mgkit.workflow.snp_parser">mgkit.workflow.snp_parser</a>, <a href="scripts/snp_analysis.html#module-mgkit.workflow.snp_parser">[1]</a>
</li>
        <li><a href="api/mgkit.workflow.taxon_utils.html#module-mgkit.workflow.taxon_utils">mgkit.workflow.taxon_utils</a>, <a href="scripts/taxon_utils.html#module-mgkit.workflow.taxon_utils">[1]</a>
</li>
        <li><a href="api/mgkit.workflow.utils.html#module-mgkit.workflow.utils">mgkit.workflow.utils</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="N">N</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.name">name (mgkit.kegg.KeggModule attribute)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.cache_dict_file.next">next() (mgkit.utils.dictionary.cache_dict_file method)</a>
</li>
      <li><a href="api/mgkit.net.embl.html#mgkit.net.embl.NoEntryFound">NoEntryFound</a>
</li>
      <li><a href="api/mgkit.mappings.eggnog.html#mgkit.mappings.eggnog.NOGInfo">NOGInfo (class in mgkit.mappings.eggnog)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.NoLcaFound">NoLcaFound</a>
</li>
      <li><a href="api/mgkit.net.embl.html#mgkit.net.embl.NONE_FOUND">NONE_FOUND (in module mgkit.net.embl)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.SNPType.nonsyn">nonsyn (mgkit.snps.classes.SNPType attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.nonsyn">nonsyn() (mgkit.snps.classes.GeneSNP property)</a>
</li>
      <li><a href="api/mgkit.net.embl.html#mgkit.net.embl.NOT_FOUND">NOT_FOUND (in module mgkit.net.embl)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.not_found">not_found (mgkit.align.SamtoolsDepth attribute)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id16">nuc_change (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.nuc_change">[1]</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id17">nuc_ref (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.nuc_ref">[1]</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id18">num_features (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.num_features">[1]</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id19">num_stops (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.num_stops">[1]</a>
</li>
  </ul></td>
</tr></table>

<h2 id="O">O</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.io.utils.html#mgkit.io.utils.open_file">open_file() (in module mgkit.io.utils)</a>
</li>
      <li><a href="api/mgkit.counts.glm.html#mgkit.counts.glm.optimise_alpha_scipy">optimise_alpha_scipy() (in module mgkit.counts.glm)</a>
</li>
      <li><a href="api/mgkit.counts.glm.html#mgkit.counts.glm.optimise_alpha_scipy_function">optimise_alpha_scipy_function() (in module mgkit.counts.glm)</a>
</li>
      <li><a href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.order_ratios">order_ratios() (in module mgkit.snps.funcs)</a>
</li>
      <li>
    OUTPUT_FILE

      <ul>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-addtaxa-arg-OUTPUT_FILE">add-gff-info-addtaxa command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-counts-arg-OUTPUT_FILE">add-gff-info-counts command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-cov_samtools-arg-OUTPUT_FILE">add-gff-info-cov_samtools command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-coverage-arg-OUTPUT_FILE">add-gff-info-coverage command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-exp_syn-arg-OUTPUT_FILE">add-gff-info-exp_syn command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-kegg-arg-OUTPUT_FILE">add-gff-info-kegg command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-pfam-arg-OUTPUT_FILE">add-gff-info-pfam command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-unipfile-arg-OUTPUT_FILE">add-gff-info-unipfile command line option</a>
</li>
        <li><a href="scripts/add-gff-info.html#cmdoption-add-gff-info-uniprot-arg-OUTPUT_FILE">add-gff-info-uniprot command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-add-arg-OUTPUT_FILE">edit-gff-add command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-remove-arg-OUTPUT_FILE">edit-gff-remove command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-table-arg-OUTPUT_FILE">edit-gff-table command line option</a>
</li>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-view-arg-OUTPUT_FILE">edit-gff-view command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-translate-arg-OUTPUT_FILE">fasta-utils-translate command line option</a>
</li>
        <li><a href="scripts/fasta-utils.html#cmdoption-fasta-utils-uid-arg-OUTPUT_FILE">fasta-utils-uid command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-cov-arg-OUTPUT_FILE">filter-gff-cov command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-overlap-arg-OUTPUT_FILE">filter-gff-overlap command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-sequence-arg-OUTPUT_FILE">filter-gff-sequence command line option</a>
</li>
        <li><a href="scripts/filter-gff.html#cmdoption-filter-gff-values-arg-OUTPUT_FILE">filter-gff-values command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-cov-arg-OUTPUT_FILE">get-gff-info-cov command line option</a>
</li>
        <li><a href="scripts/get-gff-info.html#cmdoption-get-gff-info-mongodb-arg-OUTPUT_FILE">get-gff-info-mongodb command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-arg-OUTPUT_FILE">sampling-utils-rand_seq command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-arg-OUTPUT_FILE">sampling-utils-sample_stream command line option</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-arg-OUTPUT_FILE">sampling-utils-sync command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-arg-OUTPUT_FILE">taxon-utils-filter command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-arg-OUTPUT_FILE">taxon-utils-get command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-arg-OUTPUT_FILE">taxon-utils-lca command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-arg-OUTPUT_FILE">taxon-utils-lca_line command line option</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-arg-OUTPUT_FILE">taxon-utils-to_hdf command line option</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.workflow.taxon_utils.html#mgkit.workflow.taxon_utils.output_taxon_line">output_taxon_line() (in module mgkit.workflow.taxon_utils)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="P">P</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.plots.colors.html#mgkit.plots.colors.palette_float_to_hex">palette_float_to_hex() (in module mgkit.plots.colors)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonTuple.parent_id">parent_id (mgkit.taxon.UniprotTaxonTuple attribute)</a>
</li>
      <li><a href="api/mgkit.io.blast.html#mgkit.io.blast.parse_accession_taxa_table">parse_accession_taxa_table() (in module mgkit.io.blast)</a>
</li>
      <li><a href="api/mgkit.io.blast.html#mgkit.io.blast.parse_blast_tab">parse_blast_tab() (in module mgkit.io.blast)</a>
</li>
      <li><a href="api/mgkit.workflow.hmmer2gff.html#mgkit.workflow.hmmer2gff.parse_domain_table_contigs">parse_domain_table_contigs() (in module mgkit.workflow.hmmer2gff)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.parse_entry">parse_entry() (mgkit.kegg.KeggModule method)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.parse_entry2">parse_entry2() (mgkit.kegg.KeggModule method)</a>
</li>
      <li><a href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.parse_expasy_dat">parse_expasy_dat() (in module mgkit.mappings.enzyme)</a>
</li>
      <li><a href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.parse_expasy_dat_section">parse_expasy_dat_section() (in module mgkit.mappings.enzyme)</a>
</li>
      <li><a href="api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.parse_expasy_file">parse_expasy_file() (in module mgkit.mappings.enzyme)</a>
</li>
      <li><a href="api/mgkit.io.blast.html#mgkit.io.blast.parse_fragment_blast">parse_fragment_blast() (in module mgkit.io.blast)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.parse_gff">parse_gff() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.parse_gff_files">parse_gff_files() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.io.glimmer.html#mgkit.io.glimmer.parse_glimmer3">parse_glimmer3() (in module mgkit.io.glimmer)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.parse_gtdb_lineage">parse_gtdb_lineage() (mgkit.taxon.Taxonomy static method)</a>
</li>
      <li><a href="api/mgkit.workflow.add_gff_info.html#mgkit.workflow.add_gff_info.parse_hdf5_arg">parse_hdf5_arg() (in module mgkit.workflow.add_gff_info)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.parse_ncbi_taxonomy_merged_file">parse_ncbi_taxonomy_merged_file() (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.parse_ncbi_taxonomy_names_file">parse_ncbi_taxonomy_names_file() (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.parse_ncbi_taxonomy_nodes_file">parse_ncbi_taxonomy_nodes_file() (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.parse_reaction">parse_reaction() (in module mgkit.kegg)</a>

      <ul>
        <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.parse_reaction">(mgkit.kegg.KeggModule static method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.io.blast.html#mgkit.io.blast.parse_uniprot_blast">parse_uniprot_blast() (in module mgkit.io.blast)</a>
</li>
      <li><a href="api/mgkit.io.uniprot.html#mgkit.io.uniprot.parse_uniprot_mappings">parse_uniprot_mappings() (in module mgkit.io.uniprot)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.parse_uniprot_response">parse_uniprot_response() (in module mgkit.net.uniprot)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.parse_uniprot_taxon">parse_uniprot_taxon() (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.parse_vcf">parse_vcf() (in module mgkit.workflow.snp_parser)</a>
</li>
      <li><a href="api/mgkit.workflow.filter_gff.html#mgkit.workflow.filter_gff.perseq_calc_threshold">perseq_calc_threshold() (in module mgkit.workflow.filter_gff)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.phase">phase (mgkit.io.gff.Annotation attribute)</a>
</li>
      <li><a href="api/mgkit.snps.filter.html#mgkit.snps.filter.pipe_filters">pipe_filters() (in module mgkit.snps.filter)</a>
</li>
      <li><a href="api/mgkit.plots.unused.html#mgkit.plots.unused.plot_contig_assignment_bar">plot_contig_assignment_bar() (in module mgkit.plots.unused)</a>
</li>
      <li><a href="api/mgkit.plots.unused.html#mgkit.plots.unused.plot_scatter_2d">plot_scatter_2d() (in module mgkit.plots.unused)</a>
</li>
      <li><a href="api/mgkit.plots.unused.html#mgkit.plots.unused.plot_scatter_3d">plot_scatter_3d() (in module mgkit.plots.unused)</a>
</li>
      <li>
    pnps-gen command line option

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-cite">--cite</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-version">--version</a>
</li>
      </ul></li>
      <li>
    pnps-gen-full command line option

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-g">--gene-map &lt;gene_map&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-l">--lineage</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-c">--min-cov &lt;min_cov&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-m">--min-num &lt;min_num&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-pn">--only-pn</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-ps">--only-ps</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-r">--rank &lt;rank&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-p">--separator &lt;separator&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-s">--snp-data &lt;snp_data&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-i">--taxon_ids &lt;taxon_ids&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-t">--taxonomy &lt;taxonomy&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-2">--two-columns</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-u">--use-uid</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-v">--verbose</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-2">-2</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-c">-c</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-g">-g</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-i">-i</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-l">-l</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-m">-m</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-p">-p</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-pn">-pn</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-ps">-ps</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-r">-r</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-s">-s</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-t">-t</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-u">-u</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-v">-v</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-arg-TXT_FILE">TXT_FILE</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    pnps-gen-rank command line option

      <ul>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-l">--lineage</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-c">--min-cov &lt;min_cov&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-m">--min-num &lt;min_num&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-pn">--only-pn</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-ps">--only-ps</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-r">--rank &lt;rank&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-s">--snp-data &lt;snp_data&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-i">--taxon_ids &lt;taxon_ids&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-t">--taxonomy &lt;taxonomy&gt;</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-u">--unstack</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-v">--verbose</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-c">-c</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-i">-i</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-l">-l</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-m">-m</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-pn">-pn</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-ps">-ps</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-r">-r</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-s">-s</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-t">-t</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-u">-u</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-v">-v</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-arg-TXT_FILE">TXT_FILE</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.workflow.utils.html#mgkit.workflow.utils.PrintManAction">PrintManAction (class in mgkit.workflow.utils)</a>
</li>
      <li><a href="api/mgkit.plots.abund.html#mgkit.plots.abund.project_point">project_point() (in module mgkit.plots.abund)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id20">protein_id (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.protein_id">[1]</a>
</li>
      <li><a href="api/mgkit.utils.trans_tables.html#mgkit.utils.trans_tables.PRT_MIT">PRT_MIT (in module mgkit.utils.trans_tables)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.put_gaps_in_nuc_seq">put_gaps_in_nuc_seq() (in module mgkit.utils.sequence)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="Q">Q</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.qualities_model_constant">qualities_model_constant() (in module mgkit.utils.sequence)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.qualities_model_decrease">qualities_model_decrease() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.query_uniprot">query_uniprot() (in module mgkit.net.uniprot)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="R">R</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.random_qualities">random_qualities() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.random_sequences">random_sequences() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.random_sequences_codon">random_sequences_codon() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.common.html#mgkit.utils.common.range_intersect">range_intersect() (in module mgkit.utils.common)</a>
</li>
      <li><a href="api/mgkit.utils.common.html#mgkit.utils.common.range_substract">range_substract() (in module mgkit.utils.common)</a>
</li>
      <li><a href="api/mgkit.utils.common.html#mgkit.utils.common.ranges_length">ranges_length() (in module mgkit.utils.common)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonTuple.rank">rank (mgkit.taxon.UniprotTaxonTuple attribute)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.RatioMixIn">RatioMixIn (class in mgkit.snps.classes)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.reactions">reactions (mgkit.kegg.KeggModule attribute)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.read_from_gtdb_taxonomy">read_from_gtdb_taxonomy() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.read_from_ncbi_dump">read_from_ncbi_dump() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.read_samtools_depth">read_samtools_depth() (in module mgkit.align)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.read_taxonomy">read_taxonomy() (mgkit.taxon.Taxonomy method)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id21">region (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.region">[1]</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.region">region() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth.region_coverage">region_coverage() (mgkit.align.SamtoolsDepth method)</a>
</li>
      <li><a href="api/mgkit.workflow.fastq_utils.html#mgkit.workflow.fastq_utils.report_counts">report_counts() (in module mgkit.workflow.fastq_utils)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.REV_COMP">REV_COMP (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.reverse_aa_coord">reverse_aa_coord() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.reverse_complement">reverse_complement() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.reverse_complement_old">reverse_complement_old() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.reverse_mapping">reverse_mapping() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.rn_eq_re">rn_eq_re (mgkit.kegg.KeggClientRest attribute)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.rn_name_re">rn_name_re (mgkit.kegg.KeggClientRest attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="S">S</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonTuple.s_name">s_name (mgkit.taxon.UniprotTaxonTuple attribute)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.sample_coverage">sample_coverage() (mgkit.io.gff.Annotation property)</a>
</li>
      <li>
    sampling-utils command line option

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-cite">--cite</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-version">--version</a>
</li>
      </ul></li>
      <li>
    sampling-utils-rand_seq command line option

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-r">--coding-prop &lt;coding_prop&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-d">--const-model</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-x">--dist-loc &lt;dist_loc&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-q">--fastq</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-gc">--gc-content &lt;gc_content&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-i">--infer-params &lt;infer_params&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-l">--length &lt;length&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-n">--num-seqs &lt;num_seqs&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-progress">--progress</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-a">--read-model &lt;read_model&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-m">--save-model &lt;save_model&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-v">--verbose</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-a">-a</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-d">-d</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-gc">-gc</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-i">-i</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-l">-l</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-m">-m</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-n">-n</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-q">-q</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-r">-r</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-v">-v</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-x">-x</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-rand_seq-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    sampling-utils-sample command line option

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-q">--fastq</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-z">--gzip</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-x">--max-seq &lt;max_seq&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-n">--number &lt;number&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-p">--prefix &lt;prefix&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-r">--prob &lt;prob&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-v">--verbose</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-n">-n</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-p">-p</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-q">-q</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-r">-r</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-v">-v</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-x">-x</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-z">-z</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample-arg-INPUT_FILE">INPUT_FILE</a>
</li>
      </ul></li>
      <li>
    sampling-utils-sample_stream command line option

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-q">--fastq</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-x">--max-seq &lt;max_seq&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-r">--prob &lt;prob&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-v">--verbose</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-q">-q</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-r">-r</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-v">-v</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-x">-x</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sample_stream-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    sampling-utils-sync command line option

      <ul>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-m">--master-file &lt;master_file&gt;</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-v">--verbose</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-m">-m</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-v">-v</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/sampling-utils.html#cmdoption-sampling-utils-sync-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.align.html#mgkit.align.SamtoolsDepth">SamtoolsDepth (class in mgkit.align)</a>
</li>
      <li><a href="api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.save_data">save_data() (in module mgkit.workflow.snp_parser)</a>

      <ul>
        <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy.save_data">(mgkit.taxon.Taxonomy method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.counts.scaling.html#mgkit.counts.scaling.scale_deseq">scale_deseq() (in module mgkit.counts.scaling)</a>
</li>
      <li><a href="api/mgkit.counts.scaling.html#mgkit.counts.scaling.scale_factor_deseq">scale_factor_deseq() (in module mgkit.counts.scaling)</a>
</li>
      <li><a href="api/mgkit.counts.scaling.html#mgkit.counts.scaling.scale_rpkm">scale_rpkm() (in module mgkit.counts.scaling)</a>
</li>
      <li><a href="api/mgkit.plots.unused.html#mgkit.plots.unused.scatter_gene_values">scatter_gene_values() (in module mgkit.plots.unused)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.score">score (mgkit.io.gff.Annotation attribute)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.seq_id">seq_id (mgkit.io.gff.GenomicRange attribute)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.sequence_composition">sequence_composition() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.sequence_gc_content">sequence_gc_content() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.sequence_gc_ratio">sequence_gc_ratio() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.set_attr">set_attr() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.set_mapping">set_mapping() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.workflow.hmmer2gff.html#mgkit.workflow.hmmer2gff.set_parser">set_parser() (in module mgkit.workflow.hmmer2gff)</a>

      <ul>
        <li><a href="api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.set_parser">(in module mgkit.workflow.snp_parser)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.workflow.filter_gff.html#mgkit.workflow.filter_gff.setup_filters">setup_filters() (in module mgkit.workflow.filter_gff)</a>
</li>
      <li><a href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.significance_test">significance_test() (in module mgkit.snps.funcs)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.snpdat_reader">snpdat_reader() (in module mgkit.io.snpdat)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow">SNPDatRow (class in mgkit.io.snpdat)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#id4">snps (mgkit.snps.classes.GeneSNP attribute)</a>, <a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.snps">[1]</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.SNPType">SNPType (class in mgkit.snps.classes)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.source">source (mgkit.io.gff.Annotation attribute)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.split_dictionary_by_value">split_dictionary_by_value() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.io.fasta.html#mgkit.io.fasta.split_fasta_file">split_fasta_file() (in module mgkit.io.fasta)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.split_gff_file">split_gff_file() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.workflow.add_gff_info.html#mgkit.workflow.add_gff_info.split_sample_alg">split_sample_alg() (in module mgkit.workflow.add_gff_info)</a>
</li>
      <li><a href="api/mgkit.io.utils.html#mgkit.io.utils.split_write">split_write() (in module mgkit.io.utils)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.start">start (mgkit.io.gff.GenomicRange attribute)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.strand">strand (mgkit.io.gff.GenomicRange attribute)</a>

      <ul>
        <li><a href="api/mgkit.io.snpdat.html#id22">(mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.strand">[1]</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.SNPType.syn">syn (mgkit.snps.classes.SNPType attribute)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.syn">syn() (mgkit.snps.classes.GeneSNP property)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id23">synonymous (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.synonymous">[1]</a>
</li>
  </ul></td>
</tr></table>

<h2 id="T">T</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.taxa_distance_matrix">taxa_distance_matrix() (in module mgkit.taxon)</a>
</li>
      <li>
    taxon-utils command line option

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-cite">--cite</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-version">--version</a>
</li>
      </ul></li>
      <li>
    taxon-utils-filter command line option

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-e">--exclude-taxon-id &lt;exclude_taxon_id&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-en">--exclude-taxon-name &lt;exclude_taxon_name&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-i">--include-taxon-id &lt;include_taxon_id&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-in">--include-taxon-name &lt;include_taxon_name&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-progress">--progress</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-p">--table</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-t">--taxonomy &lt;taxonomy&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-v">--verbose</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-e">-e</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-en">-en</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-i">-i</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-in">-in</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-p">-p</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-t">-t</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-v">-v</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-filter-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    taxon-utils-get command line option

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-d">--header</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-c">--include-children</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-o">--only-names &lt;only_names&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-p">--partial</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-s">--separator &lt;separator&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-v">--verbose</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-c">-c</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-d">-d</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-o">-o</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-p">-p</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-s">-s</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-v">-v</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-arg-TAXONOMY_FILE">TAXONOMY_FILE</a>
</li>
      </ul></li>
      <li>
    taxon-utils-lca command line option

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-b">--bitscore &lt;bitscore&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-ft">--feat-type &lt;feat_type&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-kt">--krona-total &lt;krona_total&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-n">--no-lca &lt;no_lca&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-a">--only-ranked</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-f">--out-format &lt;out_format&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-progress">--progress</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-r">--reference &lt;reference&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-m">--rename</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-p">--simple-table</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-s">--sorted</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-t">--taxonomy &lt;taxonomy&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-v">--verbose</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-a">-a</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-b">-b</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-f">-f</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-ft">-ft</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-kt">-kt</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-m">-m</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-n">-n</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-p">-p</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-r">-r</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-s">-s</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-t">-t</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-v">-v</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-arg-GFF_FILE">GFF_FILE</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    taxon-utils-lca_line command line option

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-n">--no-lca &lt;no_lca&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-a">--only-ranked</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-s">--separator &lt;separator&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-t">--taxonomy &lt;taxonomy&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-v">--verbose</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-a">-a</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-n">-n</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-s">-s</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-t">-t</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-v">-v</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-lca_line-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li>
    taxon-utils-to_hdf command line option

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-c">--chunk-size &lt;chunk_size&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-s">--index-size &lt;index_size&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-w">--overwrite</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-progress">--progress</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-n">--table-name &lt;table_name&gt;</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-v">--verbose</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-c">-c</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-n">-n</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-s">-s</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-v">-v</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-w">-w</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-arg-INPUT_FILE">INPUT_FILE</a>
</li>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-to_hdf-arg-OUTPUT_FILE">OUTPUT_FILE</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.plots.unused.html#mgkit.plots.unused.TAXON_COLOURS">TAXON_COLOURS (in module mgkit.plots.unused)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.taxon_db">taxon_db() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#id5">taxon_id (mgkit.snps.classes.GeneSNP attribute)</a>, <a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.taxon_id">[1]</a>

      <ul>
        <li><a href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonTuple.taxon_id">(mgkit.taxon.UniprotTaxonTuple attribute)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.taxon_id">taxon_id() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.TAXON_RANKS">TAXON_RANKS (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.TAXON_ROOTS">TAXON_ROOTS (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.Taxonomy">Taxonomy (class in mgkit.taxon)</a>
</li>
      <li>
    TAXONOMY_FILE

      <ul>
        <li><a href="scripts/taxon_utils.html#cmdoption-taxon-utils-get-arg-TAXONOMY_FILE">taxon-utils-get command line option</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.TaxonTuple">TaxonTuple (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.utils.dictionary.html#mgkit.utils.dictionary.text_to_dict">text_to_dict() (in module mgkit.utils.dictionary)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.to_dict">to_dict() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggModule.to_edges">to_edges() (mgkit.kegg.KeggModule method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.to_file">to_file() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.to_gff">to_gff() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.to_gtf">to_gtf() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.to_json">to_json() (mgkit.io.gff.Annotation method)</a>

      <ul>
        <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.to_json">(mgkit.snps.classes.GeneSNP method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.to_mongodb">to_mongodb() (mgkit.io.gff.Annotation method)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.TRANS_TABLE">TRANS_TABLE (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id24">transcript_id (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.transcript_id">[1]</a>
</li>
      <li><a href="api/mgkit.io.snpdat.html#id25">transcript_name (mgkit.io.snpdat.SNPDatRow attribute)</a>, <a href="api/mgkit.io.snpdat.html#mgkit.io.snpdat.SNPDatRow.transcript_name">[1]</a>
</li>
      <li><a href="api/mgkit.workflow.fasta_utils.html#mgkit.workflow.fasta_utils.translate_seq">translate_seq() (in module mgkit.workflow.fasta_utils)</a>
</li>
      <li><a href="api/mgkit.utils.sequence.html#mgkit.utils.sequence.translate_sequence">translate_sequence() (in module mgkit.utils.sequence)</a>
</li>
      <li><a href="api/mgkit.filter.reads.html#mgkit.filter.reads.trim_by_ee">trim_by_ee() (in module mgkit.filter.reads)</a>
</li>
      <li>
    TXT_FILE

      <ul>
        <li><a href="scripts/edit_gff.html#cmdoption-edit-gff-fields-arg-TXT_FILE">edit-gff-fields command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-full-arg-TXT_FILE">pnps-gen-full command line option</a>
</li>
        <li><a href="scripts/pnps_gen.html#cmdoption-pnps-gen-rank-arg-TXT_FILE">pnps-gen-rank command line option</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="U">U</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.snps.classes.html#id6">uid (mgkit.snps.classes.GeneSNP attribute)</a>, <a href="api/mgkit.snps.classes.html#mgkit.snps.classes.GeneSNP.uid">[1]</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.Annotation.uid">uid() (mgkit.io.gff.Annotation property)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.GenomicRange.union">union() (mgkit.io.gff.GenomicRange method)</a>
</li>
      <li><a href="api/mgkit.utils.common.html#mgkit.utils.common.union_range">union_range() (in module mgkit.utils.common)</a>
</li>
      <li><a href="api/mgkit.utils.common.html#mgkit.utils.common.union_ranges">union_ranges() (in module mgkit.utils.common)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.UNIPROT_GET">UNIPROT_GET (in module mgkit.net.uniprot)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.UNIPROT_MAP">UNIPROT_MAP (in module mgkit.net.uniprot)</a>
</li>
      <li><a href="api/mgkit.io.uniprot.html#mgkit.io.uniprot.uniprot_mappings_to_dict">uniprot_mappings_to_dict() (in module mgkit.io.uniprot)</a>
</li>
      <li><a href="api/mgkit.net.uniprot.html#mgkit.net.uniprot.UNIPROT_TAXONOMY">UNIPROT_TAXONOMY (in module mgkit.net.uniprot)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonomy">UniprotTaxonomy (in module mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.taxon.html#mgkit.taxon.UniprotTaxonTuple">UniprotTaxonTuple (class in mgkit.taxon)</a>
</li>
      <li><a href="api/mgkit.utils.trans_tables.html#mgkit.utils.trans_tables.UNIVERSAL">UNIVERSAL (in module mgkit.utils.trans_tables)</a>
</li>
      <li><a href="api/mgkit.snps.classes.html#mgkit.snps.classes.SNPType.unknown">unknown (mgkit.snps.classes.SNPType attribute)</a>
</li>
      <li><a href="api/mgkit.io.utils.html#mgkit.io.utils.UnsupportedFormat">UnsupportedFormat</a>
</li>
      <li><a href="api/mgkit.workflow.add_gff_info.html#mgkit.workflow.add_gff_info.update_annotation_coverage">update_annotation_coverage() (in module mgkit.workflow.add_gff_info)</a>
</li>
      <li><a href="api/mgkit.net.utils.html#mgkit.net.utils.url_open">url_open() (in module mgkit.net.utils)</a>
</li>
      <li><a href="api/mgkit.net.utils.html#mgkit.net.utils.url_read">url_read() (in module mgkit.net.utils)</a>
</li>
      <li><a href="api/mgkit.net.embl.html#mgkit.net.embl.URL_REST">URL_REST (in module mgkit.net.embl)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="V">V</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.workflow.blast2gff.html#mgkit.workflow.blast2gff.validate_params">validate_params() (in module mgkit.workflow.blast2gff)</a>

      <ul>
        <li><a href="api/mgkit.workflow.filter_gff.html#mgkit.workflow.filter_gff.validate_params">(in module mgkit.workflow.filter_gff)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.workflow.taxon_utils.html#mgkit.workflow.taxon_utils.validate_taxon_ids">validate_taxon_ids() (in module mgkit.workflow.taxon_utils)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.workflow.taxon_utils.html#mgkit.workflow.taxon_utils.validate_taxon_names">validate_taxon_names() (in module mgkit.workflow.taxon_utils)</a>
</li>
      <li><a href="api/mgkit.db.dbm.html#mgkit.db.dbm.GFFDB.values">values() (mgkit.db.dbm.GFFDB method)</a>

      <ul>
        <li><a href="api/mgkit.db.mongo.html#mgkit.db.mongo.GFFDB.values">(mgkit.db.mongo.GFFDB method)</a>
</li>
      </ul></li>
      <li><a href="api/mgkit.counts.glm.html#mgkit.counts.glm.variance_to_alpha">variance_to_alpha() (in module mgkit.counts.glm)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="W">W</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.kegg.html#mgkit.kegg.KeggClientRest.write_cache">write_cache() (mgkit.kegg.KeggClientRest method)</a>
</li>
      <li><a href="api/mgkit.io.fasta.html#mgkit.io.fasta.write_fasta_sequence">write_fasta_sequence() (in module mgkit.io.fasta)</a>
</li>
      <li><a href="api/mgkit.io.fastq.html#mgkit.io.fastq.write_fastq_sequence">write_fastq_sequence() (in module mgkit.io.fastq)</a>
</li>
      <li><a href="api/mgkit.io.gff.html#mgkit.io.gff.write_gff">write_gff() (in module mgkit.io.gff)</a>
</li>
      <li><a href="api/mgkit.workflow.taxon_utils.html#mgkit.workflow.taxon_utils.write_json">write_json() (in module mgkit.workflow.taxon_utils)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.workflow.taxon_utils.html#mgkit.workflow.taxon_utils.write_krona">write_krona() (in module mgkit.workflow.taxon_utils)</a>
</li>
      <li><a href="api/mgkit.workflow.taxon_utils.html#mgkit.workflow.taxon_utils.write_lca_gff">write_lca_gff() (in module mgkit.workflow.taxon_utils)</a>
</li>
      <li><a href="api/mgkit.workflow.taxon_utils.html#mgkit.workflow.taxon_utils.write_lca_tab">write_lca_tab() (in module mgkit.workflow.taxon_utils)</a>
</li>
      <li><a href="api/mgkit.workflow.taxon_utils.html#mgkit.workflow.taxon_utils.write_lca_tab_simple">write_lca_tab_simple() (in module mgkit.workflow.taxon_utils)</a>
</li>
      <li><a href="api/mgkit.workflow.taxon_utils.html#mgkit.workflow.taxon_utils.write_no_lca">write_no_lca() (in module mgkit.workflow.taxon_utils)</a>
</li>
      <li><a href="api/mgkit.snps.funcs.html#mgkit.snps.funcs.write_sign_genes_table">write_sign_genes_table() (in module mgkit.snps.funcs)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="Y">Y</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="api/mgkit.utils.trans_tables.html#mgkit.utils.trans_tables.YST_ALT">YST_ALT (in module mgkit.utils.trans_tables)</a>
</li>
  </ul></td>
</tr></table>



          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">MGKit: Metagenomic framework</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1"><a class="reference internal" href="library.html">Library Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="changes.html">Changes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2013-2020, Francesco Rubino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>