
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Gene Prediction &#8212; MGKit: Metagenomic framework 0.5.5 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Gene Prediction with BLAST+" href="blast-pred.html" />
    <link rel="prev" title="Profile a Community with BLAST" href="blast2lca.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="gene-prediction">
<h1>Gene Prediction<a class="headerlink" href="#gene-prediction" title="Permalink to this headline">¶</a></h1>
<p>Gene prediction is an essential portion of a metagenomic pipeline, because there is no a priori knowledge of what genes are in the samples. Moreover, a gene must be taxonomically annotated to correlate its function to the taxonomic group it belongs to.</p>
<p>There different ways to predict genes, with some relying on general function domains like the ones from <a class="reference external" href="http://pfam.xfam.org/">PFam</a> or others. This type of collections is very useful in identifying proteins in an unknown sequence. The main drawback for the examined datasets is that it’s not possible to identify the organism but only the general function of a sequence.</p>
<p>A second approach is to use orthologs, genes derived from the same ancestral sequence with their separation originated from a speciation process. As functionality is preserved among them, they are a good choice when approaching samples where multiple organisms are present. Two collections, <a class="reference external" href="http://eggnog.embl.de">eggNOG</a> and <a class="reference external" href="http://www.kegg.jp/kegg/ko.html">Kegg Orthologs</a>, are highly curated. A single ortholog gene identifier maps to several genes from different organisms, so the characterisation of an ortholog gene propagate to all its associated genes. This includes links to pathways in the case of Kegg and to functional categories in the case of eggNOG.</p>
<p>These genes are shared between organisms, so a single ortholog gene corresponds to several genes in different organisms. In some cases this is a preferred approach, as it allows a good resolution in the function, especially because this collections are linked to pathways in the case of Kegg and to functional categories in the case of eggNOG. The downside is that the collection of gene is not extensive and it is not connected to a taxonomic identification.</p>
<p>Another approach is to use genes from general public databases, like <a class="reference external" href="http://www.uniprot.org">Uniprot</a>. While more general a collection, compared to Kegg Orthologs or eggNOG, it offers mappings to these two collections, as well as others. It does contain when available taxonomic information of its genes and it is divided into a manually curated portion (SwissProt) and an automated one (TrEMBL). This separation allows to have mixing annotations from both portions while preferentially use the ones from SwissProt.</p>
<p>In general the framework does not enforce one collection over another and in fact ortholog genes were used in one study, while Uniprot genes were used in others.</p>
<div class="section" id="prediction-software">
<h2>Prediction Software<a class="headerlink" href="#prediction-software" title="Permalink to this headline">¶</a></h2>
<p>The prediction of genes requires both a collection and specific softwares to find homologous sequneces. There are two classes of software that can be used for gene prediction: one is profile based and the second uses similarity search.</p>
<p>An example of software using profile search is <a class="reference external" href="http://hmmer.janelia.org/">HMMER</a>. This approach uses an alignment of similar sequences to create an hidden Markov model (HMM) profile that is used to identify sequences that are similar to said profile. Curated profiles can be created from the eggNOG collection or other collection, but is also possible to automate the process of creating custom profiles.</p>
<p>The similarity search approach, is used in <a class="reference external" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">BLAST+</a>, where a collection of sequences is first indexed and then all words in the index are searched against the query sequence and the most similar ones are investigated further to report a region of similarity.</p>
</div>
<div class="section" id="general-procedure">
<h2>General Procedure<a class="headerlink" href="#general-procedure" title="Permalink to this headline">¶</a></h2>
<div><svg height="640" viewBox="0 0 736 640" width="736" xmlns="http://www.w3.org/2000/svg" xmlns:inkspace="http://www.inkscape.org/namespaces/inkscape" xmlns:xlink="http://www.w3.org/1999/xlink">
  <defs id="defs_block">
    <filter height="1.504" id="filter_blur" inkspace:collect="always" width="1.1575" x="-0.07875" y="-0.252">
      <feGaussianBlur id="feGaussianBlur3780" inkspace:collect="always" stdDeviation="4.2" />
    </filter>
  </defs>
  <title>blockdiag</title>
  <desc />
  <polygon fill="rgb(0,0,0)" points="99,54 235,54 211,134 75,134 99,54" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" />
  <polygon fill="rgb(0,0,0)" points="95,50 231,50 207,130 71,130 95,50" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" />
  <polygon fill="rgb(0,0,0)" points="91,46 227,46 203,126 67,126 91,46" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" />
  <polygon fill="rgb(0,0,0)" points="315,46 451,46 427,126 291,126 315,46" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" />
  <rect fill="rgb(0,0,0)" height="80" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" width="160" x="67" y="166" />
  <rect fill="rgb(0,0,0)" height="80" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" width="160" x="67" y="286" />
  <rect fill="rgb(0,0,0)" height="80" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" width="160" x="291" y="406" />
  <rect fill="rgb(0,0,0)" height="80" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" width="160" x="515" y="406" />
  <rect fill="rgb(0,0,0)" height="80" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" width="160" x="291" y="166" />
  <rect fill="rgb(0,0,0)" height="80" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" width="160" x="291" y="286" />
  <rect fill="rgb(0,0,0)" height="80" stroke="rgb(0,0,0)" style="filter:url(#filter_blur);opacity:0.7;fill-opacity:1" width="160" x="515" y="526" />
  <polygon fill="rgb(55,126,184)" points="96,48 232,48 208,128 72,128 96,48" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(55,126,184)" points="92,44 228,44 204,124 68,124 92,44" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(55,126,184)" points="88,40 224,40 200,120 64,120 88,40" stroke="rgb(0,0,0)" />
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="87" x="144.5" y="79">Functional</text>
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="87" x="144.5" y="97">Prediction</text>
  <polygon fill="rgb(55,126,184)" points="312,40 448,40 424,120 288,120 312,40" stroke="rgb(0,0,0)" />
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="78" x="368.0" y="79">Taxonomic</text>
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="87" x="368.5" y="97">Prediction</text>
  <rect fill="rgb(228,26,28)" height="80" stroke="rgb(0,0,0)" width="160" x="64" y="160" />
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="87" x="144.5" y="199">Convert to</text>
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="26" x="144.0" y="217">GFF</text>
  <rect fill="rgb(228,26,28)" height="80" stroke="rgb(0,0,0)" width="160" x="64" y="280" />
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="52" x="144.0" y="319">Filter</text>
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="26" x="144.0" y="337">GFF</text>
  <rect fill="rgb(228,26,28)" height="80" stroke="rgb(0,0,0)" width="160" x="288" y="400" />
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="61" x="368.5" y="439">Extract</text>
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="78" x="368.0" y="457">Sequences</text>
  <rect fill="rgb(228,26,28)" height="80" stroke="rgb(0,0,0)" width="160" x="512" y="400" />
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="69" x="592.5" y="439">Complete</text>
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="96" x="592.0" y="457">Annotations</text>
  <rect fill="rgb(228,26,28)" height="80" stroke="rgb(0,0,0)" width="160" x="288" y="160" />
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="104" x="368.0" y="199">Add Taxonomy</text>
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="96" x="368.0" y="217">Information</text>
  <rect fill="rgb(152,78,163)" height="80" stroke="rgb(0,0,0)" width="160" x="288" y="280" />
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="78" x="368.0" y="328">Draft GFF</text>
  <rect fill="rgb(152,78,163)" height="80" stroke="rgb(0,0,0)" width="160" x="512" y="520" />
  <text fill="rgb(255,255,255)" font-family="sans-serif" font-size="16" font-style="normal" font-weight="normal" text-anchor="middle" textLength="78" x="592.0" y="568">Final GFF</text>
  <path d="M 144 128 L 144 152" fill="none" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(0,0,0)" points="144,159 140,152 148,152 144,159" stroke="rgb(0,0,0)" />
  <path d="M 368 120 L 368 152" fill="none" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(0,0,0)" points="368,159 364,152 372,152 368,159" stroke="rgb(0,0,0)" />
  <path d="M 144 240 L 144 272" fill="none" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(0,0,0)" points="144,279 140,272 148,272 144,279" stroke="rgb(0,0,0)" />
  <path d="M 224 320 L 280 320" fill="none" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(0,0,0)" points="287,320 280,316 280,324 287,320" stroke="rgb(0,0,0)" />
  <path d="M 368 480 L 368 500" fill="none" stroke="rgb(0,0,0)" />
  <path d="M 368 500 L 464 500" fill="none" stroke="rgb(0,0,0)" />
  <path d="M 464 500 L 464 25" fill="none" stroke="rgb(0,0,0)" />
  <path d="M 368 25 L 464 25" fill="none" stroke="rgb(0,0,0)" />
  <path d="M 368 25 L 368 32" fill="none" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(0,0,0)" points="368,39 364,32 372,32 368,39" stroke="rgb(0,0,0)" />
  <path d="M 368 240 L 368 272" fill="none" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(0,0,0)" points="368,279 364,272 372,272 368,279" stroke="rgb(0,0,0)" />
  <path d="M 592 480 L 592 512" fill="none" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(0,0,0)" points="592,519 588,512 596,512 592,519" stroke="rgb(0,0,0)" />
  <path d="M 368 360 L 368 392" fill="none" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(0,0,0)" points="368,399 364,392 372,392 368,399" stroke="rgb(0,0,0)" />
  <path d="M 368 360 L 368 380" fill="none" stroke="rgb(0,0,0)" />
  <path d="M 368 380 L 460.0 380" fill="none" stroke="rgb(0,0,0)" />
  <path d="M 460.0 380.0 A4.0,4.0 0 0 1 468.0 380.0" fill="none" stroke="rgb(0,0,0)" />
  <path d="M 468.0 380 L 592 380" fill="none" stroke="rgb(0,0,0)" />
  <path d="M 592 380 L 592 392" fill="none" stroke="rgb(0,0,0)" />
  <polygon fill="rgb(0,0,0)" points="592,399 588,392 596,392 592,399" stroke="rgb(0,0,0)" />
</svg>
</div>
<p>The end result of the full process is a <a class="reference external" href="http://www.sequenceontology.org/gff3.shtml">GFF file</a> including <em>gene_id</em> <em>taxon_id</em> and <em>uid</em> attributes. This attributes are need to identify univocally the annotation (<em>uid</em>), the gene that was functionally predicted (<em>gene_id</em>) and the organism it belongs to (<em>taxon_id</em>). A more detailed explanation of these attributes and others is in <a class="reference internal" href="../gff.html#gff-specs"><span class="std std-ref">MGKit GFF Specifications</span></a>.</p>
<p>The choice of the format is based on the fact that it is to manipulate without ad-hoc tools, as it is a text file, and it is accepted as input file by several bioinformatics tools.</p>
</div>
<div class="section" id="functional-prediction">
<h2>Functional Prediction<a class="headerlink" href="#functional-prediction" title="Permalink to this headline">¶</a></h2>
<p>The first step of the pipeline is to generate functional prediction information of the metagenomic sequences. This can be achieved using any tool of choice, with HMMER and BLAST+ preferred among others. While BLAST+ is being extensively tested, any program that outputs prediction data in the same tab separated format as BLAST+ can be used, including <a class="reference external" href="http://www.drive5.com/usearch/">USEARCH</a> and <a class="reference external" href="http://omics.informatics.indiana.edu/mg/RAPSearch2/">RAPSearch2</a>.</p>
<p>The framework provides two script, one for HMMER output <a class="reference internal" href="../scripts/hmmer2gff.html#hmmer2gff"><span class="std std-ref">hmmer2gff - Convert HMMER output to GFF</span></a> and one for BLAST+ tab separated format <a class="reference internal" href="../scripts/blast2gff.html#blast2gff"><span class="std std-ref">blast2gff - Convert BLAST output to GFF</span></a>, to convert predictions to GFF annotations.</p>
<p>Usually a filter on the quality of the prediction is used, with 40 bit as a minimum indication of homology and 60 being a better one. If more than one gene collection is used, for example both SwissProt and TrEMBL, it is advised to keep track of which collection the annotation comes from and giving the chosen collection a quality index (<em>dbq</em> attribute in the GFF), with higher values assigned to preferred collections.</p>
<div class="section" id="generate-profiles">
<h3>Generate Profiles<a class="headerlink" href="#generate-profiles" title="Permalink to this headline">¶</a></h3>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>More detailed information is in <span class="xref std std-ref">script-download-profiles</span></p>
</div>
<p>The framework provides, via a script and the following guidelines, a way to generate profiles for Kegg Orthologs genes, using Uniprot as repository of sequnece data.</p>
<p>The process of building the profiles to be used with HMMER is a step that involves several tasks:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>download of data</p></li>
<li><p>alignment of sequences</p></li>
<li><p>conversion in HMMER profiles.</p></li>
</ol>
</div></blockquote>
<p>The first step involves, for all ortholog genes, to download all sequences available for each taxon level of interest: this will produce a series of file which contain the amino-acid sequences for each tuple gene-taxon. The sequences downloaded are aligned using <a class="reference external" href="http://www.clustal.org/omega/">Clustal Omega</a> and for each alignment a profile is built.</p>
<p>Building profiles in this way, by going through all ortholog genes and choosing the taxon level desired, opens the possibility of incrementally refining the profiling of a metagenome without having to rerun all profiles again, as only the new ones need to be run. Filtering the all the results is a much faster operation.</p>
</div>
</div>
<div class="section" id="filter-annotations">
<h2>Filter Annotations<a class="headerlink" href="#filter-annotations" title="Permalink to this headline">¶</a></h2>
<p>The number of predictions generated by the chosen prediction software can be very high, with a lot of them having just a few base pairs difference. When this involves the same functional prediction, it is safe to use the one with the best score.</p>
<p>However, when multiple genes are predicted on roughly the same region of a sequence, the choice of the annotation to keep is more difficult. Overlapping annotations can be either a weaker prediction, or the result of a chimeric sequence, as it can happen in metagenomic assemblies.</p>
<p>To solve this problem a script (filter-gff) was written that filters annotations when an overlap occurs. The algorithm scans the list of all annotations in a single sequence, sorted by their bit score, trying to find annotations that overlap. The filter is triggered when two annotations overlap, for at least 100bp by default, and the annotation to keep is chosen using a function that maximise three parameters: db quality (dbq), bit score (bitscore) and annotation length, in order of priority. This greatly reduces the number of annotations remaining and keeps the best possible annotations.</p>
<p>The choice of the 100 bp, as default value for an overlap to trigger filtering between two annotations, is based on the comparison of 36 prokaryotic genomes retrieved from <a class="reference external" href="https://genome.ucsc.edu">UCSC</a> gene overlaps.</p>
<table class="docutils align-default" id="id1">
<caption><span class="caption-text">Archaeal Genomes</span><a class="headerlink" href="#id1" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Crenarchaea</p></th>
<th class="head"><p>Euryarchaea</p></th>
<th class="head"><p>Thaumarchaea</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Acidianus hospitalis</p></td>
<td><p>Archaeoglobus fulgidus</p></td>
<td><p>Cenarchaeum symbiosum</p></td>
</tr>
<tr class="row-odd"><td><p>Desulfurococcus kamchatkensis</p></td>
<td><p>Haloarcula marismortui</p></td>
<td><p>Nitrosopumilus maritimus</p></td>
</tr>
<tr class="row-even"><td><p>Hyperthermus butylicus</p></td>
<td><p>Methanobrevibacter ruminantium M1</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Pyrobaculum islandicum</p></td>
<td><p>Methanobrevibacter smithii</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Thermoproteus tenax Kra1</p></td>
<td><p>Thermococcus barophilus MP</p></td>
<td></td>
</tr>
<tr class="row-odd"><td></td>
<td><p>Thermococcus onnurineus</p></td>
<td></td>
</tr>
</tbody>
</table>
<table class="docutils align-default" id="id2">
<caption><span class="caption-text">Bacterial Genomes</span><a class="headerlink" href="#id2" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Actinobacteria</p></th>
<th class="head"><p>Aquificae</p></th>
<th class="head"><p>Bacteroidetes</p></th>
<th class="head"><p>Proteobacteria</p></th>
<th class="head"><p>Spirochaetes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Acidothermus cellulolyticus 11B</p></td>
<td><p>Aquifex aeolicus</p></td>
<td><p>Bacteroides thetaiotaomicron</p></td>
<td><p>Blochmannia floridanus</p></td>
<td><p>Borrelia burgdorferi</p></td>
</tr>
<tr class="row-odd"><td><p>Bifidobacterium longum</p></td>
<td><p>Hydrogenivirga sp. 128</p></td>
<td><p>Cytophaga hutchinsonii</p></td>
<td><p>Candidatus Carsonella ruddii</p></td>
<td><p>Leptospira interogans</p></td>
</tr>
<tr class="row-even"><td><p>Mycobacterium tuberculosis</p></td>
<td><p>Hydrogenobaculum 3684</p></td>
<td><p>Gramella forsetii</p></td>
<td><p>Photobacterium profundum</p></td>
<td><p>Treponema pallidum</p></td>
</tr>
<tr class="row-odd"><td><p>Nocardioides JS614</p></td>
<td><p>Persephonella marina</p></td>
<td><p>Salinibacter ruber</p></td>
<td><p>Salmonella typhi</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Rhodococcus RHA1</p></td>
<td><p>Sulfurihydrogenbium YO3AOP1</p></td>
<td></td>
<td><p>Shewanella oneidensis</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Tropheryma whipplei TW08 27</p></td>
<td><p>Sulfurihydrogenibium yellowstonense</p></td>
<td></td>
<td><p>Vibrio parahaemolyticus</p></td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="taxonomic-prediction">
<h2>Taxonomic Prediction<a class="headerlink" href="#taxonomic-prediction" title="Permalink to this headline">¶</a></h2>
<p>When using Uniprot to functionally predict genes in a sequence, the metadata available for the gene may contain taxonomic information. However, while a gene from one species may have been predicted in the data, this prediction may be incorrect. There are various reasons, closely related organisms, lack of specific genes for a class of organisms or annotations, among others.</p>
<p>In this cases the approach taken in the framework is to extract the predicted nucleotide sequences using the tool of choice, provieded that it names the sequences using the <em>uid</em> attribute of an annotation, or the provided script (<a class="reference internal" href="../scripts/get-gff-info.html#get-gff-info"><span class="std std-ref">get-gff-info - Extract informations to GFF annotations</span></a>). The sequences included in the file can be used with a similarity search program as BLAST to find the closest related sequences.</p>
<p>The collection used for this is the <em>nt</em> database from NCBI and a search against it can provide a better taxonomic assignment. The default behaviour is to take the taxonomic prediction with the highest score. It is recommended to use only predictions with a bit score of 60 or higher.</p>
<p>An included script <a class="reference internal" href="../scripts/add-gff-info.html#add-gff-info"><span class="std std-ref">add-gff-info - Add informations to GFF annotations</span></a> provides the functionality necessary to add the taxonomic assignments to the GFF file. It also includes a last common ancestor (LCA) algorithm to resolve ambiguous assignments.</p>
<div class="section" id="last-common-ancestor">
<h3>Last Common Ancestor<a class="headerlink" href="#last-common-ancestor" title="Permalink to this headline">¶</a></h3>
<p>While the default behaviour is to take a prediction with the highest score, this may not be correct if more predictions have similar score. For this reason a last common ancestor (LCA) algorithm can be enabled on the predictions that are a set number of bits from the highest one, with the default value used 10.</p>
<p>The algorithm works by collecting all taxonomic predictions for a sequence, that falls within the chosen threshold, and traversing the taxonomy to find the last common ancestor. If no common ancestor can be found, the taxonomic predictions are discarded.</p>
</div>
</div>
<div class="section" id="complete-annotations">
<h2>Complete Annotations<a class="headerlink" href="#complete-annotations" title="Permalink to this headline">¶</a></h2>
<p>When a GFF file is produced by the framework, it can be integrated with the taxonomic information from Uniprot, if that was the collection used to predict genes.</p>
<p>The process add mapping attributes to the GFF file, with eggNOG and Kegg Orthologs for example, while also completing the taxonomic assignment of annotations that were not assigned taxonomically. This can be done with an included script <a class="reference internal" href="../scripts/add-gff-info.html#add-gff-info"><span class="std std-ref">add-gff-info - Add informations to GFF annotations</span></a> and the completed GFF can be used for further analysis</p>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="blast-pred.html">Gene Prediction with BLAST+</a><ul>
<li class="toctree-l2"><a class="reference internal" href="blast-pred.html#functional-prediction">Functional Prediction</a></li>
<li class="toctree-l2"><a class="reference internal" href="blast-pred.html#filter-gff">Filter GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="blast-pred.html#taxonomic-prediction">Taxonomic Prediction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="blast-pred.html#taxonomic-refinement">Taxonomic Refinement</a></li>
<li class="toctree-l3"><a class="reference internal" href="blast-pred.html#complete-annotations">Complete Annotations</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">MGKit: Metagenomic framework</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Metagenomic Pipeline</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="Exploring-Metagenomic-Data.html">Tutorial - Exploring the Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="blast2lca.html">Profile a Community with BLAST</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Gene Prediction</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1"><a class="reference internal" href="../library.html">Library Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Metagenomic Pipeline</a><ul>
      <li>Previous: <a href="blast2lca.html" title="previous chapter">Profile a Community with BLAST</a></li>
      <li>Next: <a href="blast-pred.html" title="next chapter">Gene Prediction with BLAST+</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2013-2020, Francesco Rubino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/pipeline/gene_prediction.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>