
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>mgkit.counts.scaling &#8212; MGKit: Metagenomic framework 0.5.7 documentation</title>
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css" />
    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for mgkit.counts.scaling</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Scaling functions for counts</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">division</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">pandas</span>


<div class="viewcode-block" id="scale_factor_deseq"><a class="viewcode-back" href="../../../api/mgkit.counts.scaling.html#mgkit.counts.scaling.scale_factor_deseq">[docs]</a><span class="k">def</span> <span class="nf">scale_factor_deseq</span><span class="p">(</span><span class="n">dataframe</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionadded:: 0.1.13</span>

<span class="sd">    Returns the scale factor according to he deseq paper. The columns of the</span>
<span class="sd">    dataframe are the samples.</span>

<span class="sd">    size factor :math:`\\hat{s}_{j}` for sample *j* (from DESeq paper).</span>

<span class="sd">    .. math::</span>

<span class="sd">        \\hat{s}_{j} = median_{i} (</span>
<span class="sd">        \\frac</span>
<span class="sd">            {k_{ij}}</span>
<span class="sd">            {</span>
<span class="sd">                \\left (</span>
<span class="sd">                \\prod_{v=1}^{m}</span>
<span class="sd">                    k_{iv}</span>
<span class="sd">                \\right )^{1/m}</span>
<span class="sd">           }</span>
<span class="sd">        )</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># calc the genes geometric mean over all samples</span>
    <span class="n">gmean</span> <span class="o">=</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">stats</span><span class="o">.</span><span class="n">gmean</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># keeps only the genes whose geometric mean is &gt; 0</span>
    <span class="n">gmean</span> <span class="o">=</span> <span class="n">gmean</span><span class="p">[</span><span class="n">gmean</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>

    <span class="n">sample_factors</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># calc the scaling factor for each sample</span>
    <span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">genes</span> <span class="ow">in</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

        <span class="n">scale_factor</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">genes</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gmean</span><span class="o">.</span><span class="n">index</span><span class="p">]</span> <span class="o">/</span> <span class="n">gmean</span><span class="p">)</span>

        <span class="n">sample_factors</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">scale_factor</span>

    <span class="k">return</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">sample_factors</span><span class="p">)</span></div>


<div class="viewcode-block" id="scale_deseq"><a class="viewcode-back" href="../../../api/mgkit.counts.scaling.html#mgkit.counts.scaling.scale_deseq">[docs]</a><span class="k">def</span> <span class="nf">scale_deseq</span><span class="p">(</span><span class="n">dataframe</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionadded:: 0.1.13</span>

<span class="sd">    Scale a dataframe using the deseq scaling. Uses :func:`scale_factor_deseq`</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">scale_factors</span> <span class="o">=</span> <span class="n">scale_factor_deseq</span><span class="p">(</span><span class="n">dataframe</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">dataframe</span> <span class="o">/</span> <span class="n">scale_factors</span></div>


<div class="viewcode-block" id="scale_rpkm"><a class="viewcode-back" href="../../../api/mgkit.counts.scaling.html#mgkit.counts.scaling.scale_rpkm">[docs]</a><span class="k">def</span> <span class="nf">scale_rpkm</span><span class="p">(</span><span class="n">dataframe</span><span class="p">,</span> <span class="n">gene_len</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionadded:: 0.1.14</span>

<span class="sd">    Perform an RPKM scaling of the pandas dataframe/series supplied using the</span>
<span class="sd">    *gene_len* series containing the gene sizes for all elements of *dataframe*</span>

<span class="sd">    .. math::</span>

<span class="sd">        RPKM =\\frac {10^{9} \\cdot C} {N \\cdot L}</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">gene_len</span> <span class="o">=</span> <span class="n">gene_len</span><span class="p">[</span><span class="n">dataframe</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
    <span class="n">tot_reads</span> <span class="o">=</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

    <span class="k">return</span> <span class="p">(</span><span class="mi">10</span> <span class="o">**</span> <span class="mi">9</span><span class="p">)</span> <span class="o">*</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">gene_len</span> <span class="o">*</span> <span class="n">tot_reads</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">)</span></div>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">MGKit: Metagenomic framework</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../library.html">Library Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changes.html">Changes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  <li><a href="../../mgkit.html">mgkit</a><ul>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2013-2020, Francesco Rubino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>