
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>mgkit.mappings.enzyme &#8212; MGKit: Metagenomic framework 0.5.7 documentation</title>
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css" />
    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for mgkit.mappings.enzyme</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">.. versionadded:: 0.1.14</span>

<span class="sd">EC mappings</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">from</span> <span class="nn">future.utils</span> <span class="kn">import</span> <span class="n">viewitems</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">from</span> <span class="nn">..io</span> <span class="kn">import</span> <span class="n">open_file</span>

<span class="n">LEVEL1_NAMES</span> <span class="o">=</span> <span class="p">{</span>
    <span class="mi">1</span><span class="p">:</span> <span class="s1">&#39;oxidoreductases&#39;</span><span class="p">,</span>
    <span class="mi">2</span><span class="p">:</span> <span class="s1">&#39;transferases&#39;</span><span class="p">,</span>
    <span class="mi">3</span><span class="p">:</span> <span class="s1">&#39;hydrolases&#39;</span><span class="p">,</span>
    <span class="mi">4</span><span class="p">:</span> <span class="s1">&#39;lyases&#39;</span><span class="p">,</span>
    <span class="mi">5</span><span class="p">:</span> <span class="s1">&#39;isomerases&#39;</span><span class="p">,</span>
    <span class="mi">6</span><span class="p">:</span> <span class="s1">&#39;ligases&#39;</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Top level classification names</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">ENZCLASS_REGEX</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;^(\d)\. ?([\d-]+)\. ?([\d-]+)\. ?([\d-]+) +(.+)\.&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Used to get the description for the higher level enzyme classes from the file</span>
<span class="sd">*enzclass.txt* on `expasy &lt;http://expasy.org&gt;`_</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="parse_expasy_file"><a class="viewcode-back" href="../../../api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.parse_expasy_file">[docs]</a><span class="k">def</span> <span class="nf">parse_expasy_file</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionchanged:: 0.4.2</span>
<span class="sd">        changed to work on python 3.x</span>

<span class="sd">    Used to load enzyme descriptions from the file *enzclass.txt* on</span>
<span class="sd">    `expasy &lt;http://expasy.org&gt;`_.</span>

<span class="sd">    The FTP url for enzclass.txt is:</span>
<span class="sd">    `&lt;ftp://ftp.expasy.org/databases/enzyme/enzclass.txt&gt;`_</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">open_file</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;rb&#39;</span><span class="p">):</span>
        <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;ascii&#39;</span><span class="p">)</span>
        <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">ENZCLASS_REGEX</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">match</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">ec</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">value</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">match</span><span class="o">.</span><span class="n">groups</span><span class="p">()[:</span><span class="mi">4</span><span class="p">]</span> <span class="k">if</span> <span class="n">value</span> <span class="o">!=</span> <span class="s1">&#39;-&#39;</span><span class="p">)</span>

        <span class="n">labels</span><span class="p">[</span><span class="n">ec</span><span class="p">]</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">labels</span></div>


<div class="viewcode-block" id="get_enzyme_level"><a class="viewcode-back" href="../../../api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.get_enzyme_level">[docs]</a><span class="k">def</span> <span class="nf">get_enzyme_level</span><span class="p">(</span><span class="n">ec</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">4</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionadded:: 0.1.14</span>

<span class="sd">    Returns an enzyme class at a specific level , between 1 and 4 (by default</span>
<span class="sd">    the most specific, 4)</span>

<span class="sd">    Arguments:</span>
<span class="sd">        ec (str): a string representing an EC number (e.g. 1.2.4.10)</span>
<span class="sd">        level (int): from 1 to 4, to get a different level specificity of in</span>
<span class="sd">            the enzyme classification</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: the EC number at the requested specificity</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; from mgkit.mappings.enzyme import get_enzyme_level</span>
<span class="sd">        &gt;&gt;&gt; get_enzyme_level(&#39;1.1.3.4&#39;, 1)</span>
<span class="sd">        &#39;1&#39;</span>
<span class="sd">        &gt;&gt;&gt; get_enzyme_level(&#39;1.1.3.4&#39;, 2)</span>
<span class="sd">        &#39;1.1&#39;</span>
<span class="sd">        &gt;&gt;&gt; get_enzyme_level(&#39;1.1.3.4&#39;, 3)</span>
<span class="sd">        &#39;1.1.3&#39;</span>
<span class="sd">        &gt;&gt;&gt; get_enzyme_level(&#39;1.1.3.4&#39;, 4)</span>
<span class="sd">        &#39;1.1.3.4&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="s1">&#39;.&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ec</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[:</span><span class="n">level</span><span class="p">])</span></div>


<div class="viewcode-block" id="change_mapping_level"><a class="viewcode-back" href="../../../api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.change_mapping_level">[docs]</a><span class="k">def</span> <span class="nf">change_mapping_level</span><span class="p">(</span><span class="n">ec_map</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">3</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionadded:: 0.1.14</span>

<span class="sd">    Given a dictionary, whose values are dictionaries, in which a key is named</span>
<span class="sd">    *ec* and its value is an iterable of EC numbers, returns an iterator that</span>
<span class="sd">    can be used to build a dictionary with the same top level keys and the</span>
<span class="sd">    values are sets of the transformed EC numbers.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        ec_map (dict): dictionary generated by</span>
<span class="sd">            :func:`mgkit.net.uniprot.get_gene_info`</span>
<span class="sd">        level (int): number from 1 to 4, to specify the level of the mapping,</span>
<span class="sd">            passed to :func:`get_enzyme_level`</span>

<span class="sd">    Yields:</span>
<span class="sd">        tuple: a tuple (gene_id, set(ECs)), which can be passed to *dict* to</span>
<span class="sd">        make a dictionary</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; from mgkit.net.uniprot import get_gene_info</span>
<span class="sd">        &gt;&gt;&gt; from mgkit.mappings.enzyme import change_mapping_level</span>
<span class="sd">        &gt;&gt;&gt; ec_map = get_gene_info(&#39;Q9HFQ1&#39;, columns=&#39;ec&#39;)</span>
<span class="sd">        {&#39;Q9HFQ1&#39;: {&#39;ec&#39;: &#39;1.1.3.4&#39;}}</span>
<span class="sd">        &gt;&gt;&gt; dict(change_mapping_level(ec_map, level=2))</span>
<span class="sd">        {&#39;Q9HFQ1&#39;: {&#39;1.1&#39;}}</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">ecdict</span> <span class="ow">in</span> <span class="n">viewitems</span><span class="p">(</span><span class="n">ec_map</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">ec_list</span> <span class="o">=</span> <span class="n">ecdict</span><span class="p">[</span><span class="s1">&#39;ec&#39;</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ec_list</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">ec_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">ec_list</span><span class="p">]</span>

        <span class="k">yield</span> <span class="n">gene_id</span><span class="p">,</span> <span class="nb">set</span><span class="p">(</span>
            <span class="n">get_enzyme_level</span><span class="p">(</span><span class="n">ec_id</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">level</span><span class="p">)</span> <span class="k">for</span> <span class="n">ec_id</span> <span class="ow">in</span> <span class="n">ec_list</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="get_mapping_level"><a class="viewcode-back" href="../../../api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.get_mapping_level">[docs]</a><span class="k">def</span> <span class="nf">get_mapping_level</span><span class="p">(</span><span class="n">ec_map</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">3</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionadded:: 0.3.0</span>

<span class="sd">    Given a dictionary, whose values are iterable of EC numbers, returns an</span>
<span class="sd">    iterator that can be used to build a dictionary with the same top level</span>
<span class="sd">    keys and the values are sets of the transformed EC numbers.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        ec_map (dict): dictionary genes to EC</span>
<span class="sd">        level (int): number from 1 to 4, to specify the level of the mapping,</span>
<span class="sd">            passed to :func:`get_enzyme_level`</span>

<span class="sd">    Yields:</span>
<span class="sd">        tuple: a tuple (gene_id, set(ECs)), which can be passed to *dict* to</span>
<span class="sd">        make a dictionary</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">ec_list</span> <span class="ow">in</span> <span class="n">viewitems</span><span class="p">(</span><span class="n">ec_map</span><span class="p">):</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">ec_list</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ec_list</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">ec_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">ec_list</span><span class="p">]</span>

        <span class="k">yield</span> <span class="n">gene_id</span><span class="p">,</span> <span class="nb">set</span><span class="p">(</span>
            <span class="n">get_enzyme_level</span><span class="p">(</span><span class="n">ec_id</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">level</span><span class="p">)</span> <span class="k">for</span> <span class="n">ec_id</span> <span class="ow">in</span> <span class="n">ec_list</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="get_enzyme_full_name"><a class="viewcode-back" href="../../../api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.get_enzyme_full_name">[docs]</a><span class="k">def</span> <span class="nf">get_enzyme_full_name</span><span class="p">(</span><span class="n">ec_id</span><span class="p">,</span> <span class="n">ec_names</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;, &#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionadded:: 0.2.1</span>

<span class="sd">    From a EC identifiers and a dictionary of names builds a comma separated</span>
<span class="sd">    name (by default) that identifies the function of the enzyme.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        ec_id (str): EC identifier</span>
<span class="sd">        ec_names (dict): a dictionary of names that can be produced using</span>
<span class="sd">            :func:`parse_expasy_file`</span>
<span class="sd">        sep (str): string used to join the names</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: the enzyme classification name</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">name_list</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">name_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">ec_names</span><span class="p">[</span><span class="n">ec_id</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="n">ec_id</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ec_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">ec_id</span><span class="p">:</span>
            <span class="k">break</span>

    <span class="k">return</span> <span class="n">sep</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">reversed</span><span class="p">(</span><span class="n">name_list</span><span class="p">))</span></div>


<div class="viewcode-block" id="parse_expasy_dat_section"><a class="viewcode-back" href="../../../api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.parse_expasy_dat_section">[docs]</a><span class="k">def</span> <span class="nf">parse_expasy_dat_section</span><span class="p">(</span><span class="n">expasy_dat_section</span><span class="p">,</span> <span class="n">skip_comments</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">skip_codes</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionadded:: 0.4.2</span>

<span class="sd">    Parses an entry of the `enzyme.dat` file in expasy, used internally by</span>
<span class="sd">    :func:`mgkit.mappings.enzyme.parse_expasy_dat`, with the other arguments</span>
<span class="sd">    being passed over from it.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: dictionary with the entry, with keys being the codes of the entry</span>
<span class="sd">        and the values the lines</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parsed_lines</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">expasy_dat_section</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">skip_comments</span> <span class="ow">and</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;CC&#39;</span><span class="p">):</span>
            <span class="k">continue</span>
        <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">maxsplit</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">skip_codes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">key</span> <span class="ow">in</span> <span class="n">skip_codes</span><span class="p">):</span>
            <span class="k">continue</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">parsed_lines</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">parsed_lines</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">value</span><span class="p">]</span>

    <span class="k">if</span> <span class="s1">&#39;CA&#39;</span> <span class="ow">in</span> <span class="n">parsed_lines</span><span class="p">:</span>
        <span class="n">ca_lines</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">parsed_lines</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">])</span>
        <span class="n">parsed_lines</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ca_lines</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;. &#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">parsed_lines</span></div>

<div class="viewcode-block" id="parse_expasy_dat"><a class="viewcode-back" href="../../../api/mgkit.mappings.enzyme.html#mgkit.mappings.enzyme.parse_expasy_dat">[docs]</a><span class="k">def</span> <span class="nf">parse_expasy_dat</span><span class="p">(</span><span class="n">expasy_dat</span><span class="p">,</span> <span class="n">keep_empty</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">skip_comments</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">skip_codes</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionadded:: 0.4.2</span>

<span class="sd">    Parses the information in `enzyme.dat` file in expasy, a flat file</span>
<span class="sd">    containting the information about the enzyme classification.</span>

<span class="sd">    It can be downloaded at:</span>
<span class="sd">    `&lt;ftp://ftp.expasy.org/databases/enzyme/enzyme.dat&gt;`_</span>

<span class="sd">    Arguments:</span>
<span class="sd">        expasy_dat (str): file name or handle to an expasy.dat file</span>
<span class="sd">        keep_empty (bool): section that are empty are removed by default</span>
<span class="sd">        skip_comments (bool): used to avoid returning comments (lines starting)</span>
<span class="sd">            with `CC` in the file</span>
<span class="sd">        skip_codes (set, tuple): set or tuple or list to skip specific parts</span>
<span class="sd">            of the file, like `skip_comments`</span>
<span class="sd">    Yields:</span>
<span class="sd">        dict: dictionary with each entry in the file, where the keys are the</span>
<span class="sd">        codes and the values are the lines included in the file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">section</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">open_file</span><span class="p">(</span><span class="n">expasy_dat</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">):</span>
        <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;ascii&#39;</span><span class="p">)</span>

        <span class="c1"># section end</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;//&#39;</span><span class="p">):</span>
            <span class="n">parsed_section</span> <span class="o">=</span> <span class="n">parse_expasy_dat_section</span><span class="p">(</span><span class="n">section</span><span class="p">,</span> <span class="n">skip_comments</span><span class="o">=</span><span class="n">skip_comments</span><span class="p">,</span> <span class="n">skip_codes</span><span class="o">=</span><span class="n">skip_codes</span><span class="p">)</span>
            <span class="n">section</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="n">parsed_section</span> <span class="ow">or</span> <span class="n">keep_empty</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">parsed_section</span>
            <span class="k">continue</span>

        <span class="n">section</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span></div>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">MGKit: Metagenomic framework</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../library.html">Library Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changes.html">Changes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  <li><a href="../../mgkit.html">mgkit</a><ul>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2013-2020, Francesco Rubino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>