

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>get-gff-info - Extract informations to GFF annotations &mdash; MGKit: Metagenomic framework 0.5.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="hmmer2gff - Convert HMMER output to GFF" href="hmmer2gff.html" />
    <link rel="prev" title="add-gff-info - Add informations to GFF annotations" href="add-gff-info.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> MGKit: Metagenomic framework
          

          
          </a>

          
            
            
              <div class="version">
                0.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Scripts Details</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="blast2gff.html">blast2gff - Convert BLAST output to GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="filter-gff.html">filter-gff - Filter GFF annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="add-gff-info.html">add-gff-info - Add informations to GFF annotations</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">get-gff-info - Extract informations to GFF annotations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-mgkit.workflow.extract_gff_info">Overview</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sequence-command">sequence command</a></li>
<li class="toctree-l4"><a class="reference internal" href="#dbm-command">dbm command</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mongodb-command">mongodb command</a></li>
<li class="toctree-l4"><a class="reference internal" href="#gtf-command">gtf command</a></li>
<li class="toctree-l4"><a class="reference internal" href="#split-command">split command</a></li>
<li class="toctree-l4"><a class="reference internal" href="#cov-command">cov command</a></li>
<li class="toctree-l4"><a class="reference internal" href="#changes">Changes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#options">Options</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#get-gff-info">get-gff-info</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="hmmer2gff.html">hmmer2gff - Convert HMMER output to GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="snp_analysis.html">snp_parser - SNPs analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="download-taxonomy.html">Download Taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="download-taxonomy.html#download-accession-taxonid">Download Accession/TaxonID</a></li>
<li class="toctree-l2"><a class="reference internal" href="taxon_utils.html">taxon-utils - Taxonomy Utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="fasta-utils.html">fasta-utils - Fasta Utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="fastq-utils.html">fastq-utils - Fastq Utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="json2gff.html">json2gff - Convert JSON to GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="sampling-utils.html">sampling-utils - Resampling Utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="edit_gff.html">edit-gff - GFF Viewer and Editor</a></li>
<li class="toctree-l2"><a class="reference internal" href="pnps_gen.html">pnps-gen - pN/pS Table Generation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1"><a class="reference internal" href="../library.html">Library Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MGKit: Metagenomic framework</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">Scripts Details</a> &raquo;</li>
        
      <li>get-gff-info - Extract informations to GFF annotations</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/scripts/get-gff-info.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="get-gff-info-extract-informations-to-gff-annotations">
<span id="get-gff-info"></span><h1>get-gff-info - Extract informations to GFF annotations<a class="headerlink" href="#get-gff-info-extract-informations-to-gff-annotations" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-mgkit.workflow.extract_gff_info">
<span id="overview"></span><h2>Overview<a class="headerlink" href="#module-mgkit.workflow.extract_gff_info" title="Permalink to this headline">¶</a></h2>
<p>Extract information from GFF files</p>
<div class="section" id="sequence-command">
<h3>sequence command<a class="headerlink" href="#sequence-command" title="Permalink to this headline">¶</a></h3>
<p>Used to extract the nucleotidic sequences from GFF annotations. It requires the
<em>fasta</em> file containing the sequences referenced in the GFF <cite>seq_id</cite> attribute
(first column of the raw GFF).</p>
<p>The sequnces extract have as identifier the <cite>uid</cite> stored in the GFF file and by
default the sequnece is not reverse complemented if the annotation is on the
<strong>-</strong> strand, but this can be changed by using the <cite>-r</cite> option.</p>
<p>The sequences are wrapped at 60 characters, as per FASTA specs, but this
behavior can be disabled by specifing the <cite>-w</cite> option.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The reference file is loaded in memory</p>
</div>
</div>
<div class="section" id="dbm-command">
<h3>dbm command<a class="headerlink" href="#dbm-command" title="Permalink to this headline">¶</a></h3>
<p>Creates a dbm DB using the <cite>semidbm</cite> package. The database can then be loaded
using <a class="reference internal" href="../api/mgkit.db.dbm.html#mgkit.db.dbm.GFFDB" title="mgkit.db.dbm.GFFDB"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.db.dbm.GFFDB</span></code></a></p>
</div>
<div class="section" id="mongodb-command">
<h3>mongodb command<a class="headerlink" href="#mongodb-command" title="Permalink to this headline">¶</a></h3>
<p>Outputs annotations in a format supported by MongoDB. More information about it
can be found in <a class="reference internal" href="../api/mgkit.db.mongo.html#module-mgkit.db.mongo" title="mgkit.db.mongo"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.db.mongo</span></code></a></p>
</div>
<div class="section" id="gtf-command">
<h3>gtf command<a class="headerlink" href="#gtf-command" title="Permalink to this headline">¶</a></h3>
<p>Outputs annotations in the GTF format</p>
</div>
<div class="section" id="split-command">
<h3>split command<a class="headerlink" href="#split-command" title="Permalink to this headline">¶</a></h3>
<p>Splits a GFF file into smaller chunks, ensuring that all of a sequence
annotations are in the same file.</p>
</div>
<div class="section" id="cov-command">
<h3>cov command<a class="headerlink" href="#cov-command" title="Permalink to this headline">¶</a></h3>
<p>Calculate annotation coverage for each contig in a GFF file. The command can be
run as strand specific (not by default) and requires the reference file to
which the annotation refer to. The output file is a tab separated one, with the
first column being the sequence name, the second is the strand (+, -, or NA if
not strand specific) and the third is the percentage of the sequence covered by
annotations.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The GFF file is assumed to be sorted, by sequence or sequence-strand if
wanted. The GFF file can be sorted using <cite>sort -s -k 1,1 -k 7,7</cite> for strand
specific, or <cite>sort -s -k 1,1</cite> if not strand specific.</p>
</div>
</div>
<div class="section" id="changes">
<h3>Changes<a class="headerlink" href="#changes" title="Permalink to this headline">¶</a></h3>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.4: </span>using <em>click</em> instead of <em>argparse</em>, renamed <em>split</em> command <em>–json</em> to
<em>–json-out</em></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.1: </span>added <em>cov</em> command</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.0: </span>added <em>–split</em> option to <em>sequence</em> command</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.6: </span>added <em>split</em> command, <em>–indent</em> option to <em>mongodb</em></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.3: </span>added <em>–gene-id</em> option to <em>gtf</em> command</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.2.2: </span>added <em>gtf</em> command</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.2.1: </span><em>dbm</em> and <em>mongodb</em> commands</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.15.</span></p>
</div>
</div>
</div>
<div class="section" id="options">
<h2>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<div class="section" id="get-gff-info">
<h3>get-gff-info<a class="headerlink" href="#get-gff-info" title="Permalink to this headline">¶</a></h3>
<p>Main function</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>get-gff-info <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-version">
<code class="sig-name descname">--version</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-version" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the version and exit.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-cite">
<code class="sig-name descname">--cite</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-cite" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<div class="section" id="get-gff-info-cov">
<h4>cov<a class="headerlink" href="#get-gff-info-cov" title="Permalink to this headline">¶</a></h4>
<p>Report on how much a sequence length is covered
by annotations in [gff-file]</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>get-gff-info cov <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>GFF_FILE<span class="o">]</span> <span class="o">[</span>OUTPUT_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-cov-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-cov-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-cov-f">
<code class="sig-name descname">-f</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--reference</code><code class="sig-prename descclassname"> &lt;reference&gt;</code><a class="headerlink" href="#cmdoption-get-gff-info-cov-f" title="Permalink to this definition">¶</a></dt>
<dd><p>Reference FASTA file for the GFF  [required]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-cov-j">
<code class="sig-name descname">-j</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--json-out</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-cov-j" title="Permalink to this definition">¶</a></dt>
<dd><p>The output will be a JSON dictionary</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-cov-s">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--strand-specific</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-cov-s" title="Permalink to this definition">¶</a></dt>
<dd><p>If the coverage must be calculated on each strand</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-cov-r">
<code class="sig-name descname">-r</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--rename</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-cov-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Emulate BLAST output (use only the header part before the first space)</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-cov-progress">
<code class="sig-name descname">--progress</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-cov-progress" title="Permalink to this definition">¶</a></dt>
<dd><p>Shows Progress Bar</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-cov-arg-gff-file">
<code class="sig-name descname">GFF_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-cov-arg-gff-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-cov-arg-output-file">
<code class="sig-name descname">OUTPUT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-cov-arg-output-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
<div class="section" id="get-gff-info-dbm">
<h4>dbm<a class="headerlink" href="#get-gff-info-dbm" title="Permalink to this headline">¶</a></h4>
<p>Creates a dbm database with annotations from file
[gff-file] into db [output-dir]</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>get-gff-info dbm <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>GFF_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-dbm-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-dbm-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-dbm-d">
<code class="sig-name descname">-d</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output-dir</code><code class="sig-prename descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-get-gff-info-dbm-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory for the database  [default: gff-dbm]</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-dbm-arg-gff-file">
<code class="sig-name descname">GFF_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-dbm-arg-gff-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
<div class="section" id="get-gff-info-gtf">
<h4>gtf<a class="headerlink" href="#get-gff-info-gtf" title="Permalink to this headline">¶</a></h4>
<p>Extract annotations from a GFF file [gff-file] to
a GTF file [gtf-file]</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>get-gff-info gtf <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>GFF_FILE<span class="o">]</span> <span class="o">[</span>GTF_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-gtf-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-gtf-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-gtf-g">
<code class="sig-name descname">-g</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--gene-id</code><code class="sig-prename descclassname"> &lt;gene_id&gt;</code><a class="headerlink" href="#cmdoption-get-gff-info-gtf-g" title="Permalink to this definition">¶</a></dt>
<dd><p>GFF attribute to use for the GTF <em>gene_id</em> attribute  [default: gene_id]</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-gtf-arg-gff-file">
<code class="sig-name descname">GFF_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-gtf-arg-gff-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-gtf-arg-gtf-file">
<code class="sig-name descname">GTF_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-gtf-arg-gtf-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
<div class="section" id="get-gff-info-mongodb">
<h4>mongodb<a class="headerlink" href="#get-gff-info-mongodb" title="Permalink to this headline">¶</a></h4>
<p>Extract annotations from a GFF [gff-file] file
and makes output for MongoDB [output-file]</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>get-gff-info mongodb <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>GFF_FILE<span class="o">]</span> <span class="o">[</span>OUTPUT_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-mongodb-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-mongodb-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-mongodb-t">
<code class="sig-name descname">-t</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--taxonomy</code><code class="sig-prename descclassname"> &lt;taxonomy&gt;</code><a class="headerlink" href="#cmdoption-get-gff-info-mongodb-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Taxonomy used to populate the lineage</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-mongodb-c">
<code class="sig-name descname">-c</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--no-cache</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-mongodb-c" title="Permalink to this definition">¶</a></dt>
<dd><p>No cache for the lineage function</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-mongodb-i">
<code class="sig-name descname">-i</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--indent</code><code class="sig-prename descclassname"> &lt;indent&gt;</code><a class="headerlink" href="#cmdoption-get-gff-info-mongodb-i" title="Permalink to this definition">¶</a></dt>
<dd><p>If used, the json will be written in a human readble form</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-mongodb-progress">
<code class="sig-name descname">--progress</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-mongodb-progress" title="Permalink to this definition">¶</a></dt>
<dd><p>Shows Progress Bar</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-mongodb-arg-gff-file">
<code class="sig-name descname">GFF_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-mongodb-arg-gff-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-mongodb-arg-output-file">
<code class="sig-name descname">OUTPUT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-mongodb-arg-output-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
<div class="section" id="get-gff-info-sequence">
<h4>sequence<a class="headerlink" href="#get-gff-info-sequence" title="Permalink to this headline">¶</a></h4>
<p>Extract the nucleotidic sequences of
annotations from [gff-file] to [fasta-file]</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>get-gff-info sequence <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>GFF_FILE<span class="o">]</span> <span class="o">[</span>FASTA_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-sequence-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-sequence-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-sequence-r">
<code class="sig-name descname">-r</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--reverse</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-sequence-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverse complement sequences on the - strand</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-sequence-w">
<code class="sig-name descname">-w</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--no-wrap</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-sequence-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Write the sequences on one line</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-sequence-s">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--split</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-sequence-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Split the sequence header of the reference at the first space, to emulate BLAST behaviour</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-sequence-f">
<code class="sig-name descname">-f</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--reference</code><code class="sig-prename descclassname"> &lt;reference&gt;</code><a class="headerlink" href="#cmdoption-get-gff-info-sequence-f" title="Permalink to this definition">¶</a></dt>
<dd><p>Fasta file containing the reference sequences of the GFF file</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-sequence-progress">
<code class="sig-name descname">--progress</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-sequence-progress" title="Permalink to this definition">¶</a></dt>
<dd><p>Shows Progress Bar</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-sequence-arg-gff-file">
<code class="sig-name descname">GFF_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-sequence-arg-gff-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-sequence-arg-fasta-file">
<code class="sig-name descname">FASTA_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-sequence-arg-fasta-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
<div class="section" id="get-gff-info-split">
<h4>split<a class="headerlink" href="#get-gff-info-split" title="Permalink to this headline">¶</a></h4>
<p>Split annotations from a GFF file [gff-file] to
several files starting with [prefix]</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>get-gff-info split <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>GFF_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-split-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-split-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-split-p">
<code class="sig-name descname">-p</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--prefix</code><code class="sig-prename descclassname"> &lt;prefix&gt;</code><a class="headerlink" href="#cmdoption-get-gff-info-split-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Prefix for the file name in output  [default: split]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-split-n">
<code class="sig-name descname">-n</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--number</code><code class="sig-prename descclassname"> &lt;number&gt;</code><a class="headerlink" href="#cmdoption-get-gff-info-split-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of chunks into which split the GFF file  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-get-gff-info-split-z">
<code class="sig-name descname">-z</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--gzip</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-split-z" title="Permalink to this definition">¶</a></dt>
<dd><p>gzip output files</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="option">
<dt id="cmdoption-get-gff-info-split-arg-gff-file">
<code class="sig-name descname">GFF_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-get-gff-info-split-arg-gff-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="hmmer2gff.html" class="btn btn-neutral float-right" title="hmmer2gff - Convert HMMER output to GFF" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="add-gff-info.html" class="btn btn-neutral float-left" title="add-gff-info - Add informations to GFF annotations" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2013-2015, Francesco Rubino

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>