
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>taxon-utils - Taxonomy Utilities &#8212; MGKit: Metagenomic framework 0.5.7 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="fasta-utils - Fasta Utilities" href="fasta-utils.html" />
    <link rel="prev" title="Download Taxonomy" href="download-taxonomy.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="taxon-utils-taxonomy-utilities">
<span id="taxon-utils"></span><h1>taxon-utils - Taxonomy Utilities<a class="headerlink" href="#taxon-utils-taxonomy-utilities" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-mgkit.workflow.taxon_utils">
<span id="overview"></span><h2>Overview<a class="headerlink" href="#module-mgkit.workflow.taxon_utils" title="Permalink to this headline">¶</a></h2>
<p>The script contains commands used to access functionality related to
taxonomy, without the need to write ad-hoc code for functionality that
can be part of a workflow. One example is access to the the last common
ancestor function contained in the <a class="reference internal" href="../api/mgkit.taxon.html#module-mgkit.taxon" title="mgkit.taxon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.taxon</span></code></a>.</p>
<div class="section" id="last-common-ancestor-lca-and-lca-line">
<h3>Last Common Ancestor (lca and lca_line)<a class="headerlink" href="#last-common-ancestor-lca-and-lca-line" title="Permalink to this headline">¶</a></h3>
<p>These commands expose the functionality of
<code class="xref py py-func docutils literal notranslate"><span class="pre">last_common_ancestor_multiple()</span></code>, making it accessible via the command
line. They differ in the input file format and the choice of output files.</p>
<p>the <em>lca</em> command can be used to define the last common ancestor of contigs
from the annotation in a GFF file. The command uses the <em>taxon_ids</em> from all
annotations belonging to a contig/sequence, if they have a <strong>bitscore</strong> higher
or equal to the one passed (50 by default). The default output of the command
is a tab separated file where the first column is the contig/sequence name,
the taxon_id of the last common ancestor, its scientific/common name and its
lineage.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">contig_21</span>   <span class="mi">172788</span>  <span class="n">uncultured</span> <span class="n">phototrophic</span> <span class="n">eukaryote</span>   <span class="n">cellular</span> <span class="n">organisms</span><span class="p">,</span><span class="n">environmental</span> <span class="n">samples</span>
</pre></div>
</div>
<p>If the <em>-r</em> is used, by passing the fasta file containing the nucleotide
sequences the output file is a GFF where for each an annotation for the full
contig length contains the same information of the tab separated file format.</p>
<p>The <strong>lca_line</strong> command accept as input a file where each line consist of a
list of taxon_ids. The separator for the list can be changed and it defaults to
TAB. The last common ancestor for all taxa on a line is searched. The ouput of
this command is the same as the tab separated file of the <strong>lca</strong> command, with
the difference that instead of the first column, which in this command becames
a list of all <em>taxon_ids</em> that were used to find the last common ancestor for
that line. The list of <em>taxon_ids</em> is separated by semicolon “;”.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Both also accept the <em>-n</em> option, to report the config/line and the
taxon_ids that had no common ancestors. These are treated as errors and do
not appear in the output file.</p>
</div>
<div class="section" id="krona-output">
<h4>Krona Output<a class="headerlink" href="#krona-output" title="Permalink to this headline">¶</a></h4>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
<p>The <em>lca</em> command supports the writing of a file compatible with Krona. The
output file can be used with the <em>ktImportText/ImportText.pl</em> script included
with <a class="reference external" href="https://github.com/marbl/Krona/wiki">KronaTools</a>. Specifically, the
output from <em>taxon_utils</em> will be a file with all the lineages found (tab
separated), that can be used with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ktImportText -q taxon_utils_ouput
</pre></div>
</div>
<p>Note the use of <em>-q</em> to make the script count the lineages. Sequences with no
LCA found will be marked as <em>No LCA</em> in the graph, the <em>-n</em> is not required.</p>
<p>If a fasta file is passed, the format is changed to add the number of base pairs
for each contig, showing the number of bases for each taxonomic assignment. The
option <cite>-kt</cite> is not needed in this case. In that case, to generate a Krona plot:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ktImportText taxon_utils_ouput
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Please note that the output won’t include any sequence that didn’t have a
hit with the software used. If that’s important, the <strong>-kt</strong> option can be
used to add a number of <em>Unknown</em> lines at the end, to read the total
supplied.</p>
</div>
</div>
</div>
<div class="section" id="filter-by-taxon">
<h3>Filter by Taxon<a class="headerlink" href="#filter-by-taxon" title="Permalink to this headline">¶</a></h3>
<p>The <strong>filter</strong> command of this script allows to filter a GFF file using the
<em>taxon_id</em> attribute to include only some annotations, or exclude some. The
filter is based on the <cite>mgkit.taxon.is_ancestor</cite> function, and the
<cite>mgkit.filter.taxon.filter_taxon_by_id_list</cite>. It can also filter a table (tab
separated values) when the first element is an ID and the second is a taxon_id.
An example of a table of this sort is the output of the <cite>download-ncbi-taxa.sh</cite>
and <cite>download-uniprot-taxa.sh</cite>, where each accession of a database is associated
to a taxon_id.</p>
<p>Multiple taxon_id can be passed, either for inclusion or exclusion. If both
exclusion and inclusion is used, the first check is on the inclusion and then on
the exclusion. In alternative to passing taxon_id, taxon_names can be passed,
with values such as ‘cellular organisms’ that needs to be quoted. Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ taxon-utils filter -i 2 -in archaea -en prevotella -t taxonomy.pickle in.gff out.gff
</pre></div>
</div>
<p>Which will keep only line that are from Bacteria (taxon_id=2) and exclude those
from the genus <em>Prevotella</em>. It will be also include Archaea.</p>
<p>Multiple inclusion and exclusion flags can be put:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ taxon-utils filter -i 2 -i 2172 -t taxonomy in.gff out.gff
</pre></div>
</div>
<p>In particular, the inclusion flag is tested first and then the exclusion is
tested. So a line like this one:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">printf</span> <span class="s2">&quot;TEST\t838\nTEST\t1485&quot;</span> <span class="p">|</span> taxon-utils filter -p -t taxonomy.pickle -i <span class="m">2</span> -i <span class="m">1485</span> -e <span class="m">838</span>
</pre></div>
</div>
<p>Will produce <strong>TEST 1485</strong>, because both Prevotella (838) and Clostridium (1485)
are Bacteria (2) OR Prevotella, but Prevotella must be excluded according to
the exclusion option. This line also illustrate that a tab-separated file, where
the second column contains taxon IDs, can be filtered. In particular it can be
applied to files produced by <cite>download-ncbi-taxa.sh</cite> or
<cite>download-uniprot-taxa.sh</cite> (see <a class="reference internal" href="download-taxonomy.html#download-taxonomy"><span class="std std-ref">Download Taxonomy</span></a>).</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Annotations with no taxon_id are not included in the output of both filters</p>
</div>
</div>
<div class="section" id="convert-taxa-tables-to-hdf5">
<h3>Convert Taxa Tables to HDF5<a class="headerlink" href="#convert-taxa-tables-to-hdf5" title="Permalink to this headline">¶</a></h3>
<p>This command is used to convert the taxa tables download from Uniprot and NCBI,
using the scripts mentioned in <span class="xref std std-ref">download-data</span>,
<cite>download-uniprot-taxa.sh</cite> and <cite>download-ncbi-taxa</cite> into a HDF5 file that can
be used with the <em>addtaxa</em> command in <a class="reference internal" href="add-gff-info.html#add-gff-info"><span class="std std-ref">add-gff-info - Add informations to GFF annotations</span></a>.</p>
<p>The advantage is a faster lookup of the IDs. The other is a smaller memory
footprint when a great number of annotations are kept in memory.</p>
</div>
<div class="section" id="extract-taxonomy">
<h3>Extract taxonomy<a class="headerlink" href="#extract-taxonomy" title="Permalink to this headline">¶</a></h3>
<p>This command allows to print the taxonomy in a file created with MGKit. The
default behaviour is to print the scientific name, taxon_id, rank and lineage
(only ranked taxa) in a tab separated file (or standard output). IDs can also
be passed along with names:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>taxon-utils get -i <span class="m">2147</span> -o prevotella -i <span class="m">3688</span> -i <span class="m">1485</span> -o methanobrevibacter taxonomy.msgpack

INFO - mgkit.taxon: Loading taxonomy from file taxonomy.msgpack
Acholeplasma    <span class="m">2147</span>    genus   Bacteria<span class="p">;</span>Tenericutes<span class="p">;</span>Mollicutes<span class="p">;</span>Acholeplasmatales<span class="p">;</span>Acholeplasmataceae<span class="p">;</span>Acholeplasma
Prevotella      <span class="m">838</span>     genus   Bacteria<span class="p">;</span>Bacteroidetes<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Prevotellaceae<span class="p">;</span>Prevotella
Salicaceae      <span class="m">3688</span>    family  Eukaryota<span class="p">;</span>Viridiplantae<span class="p">;</span>Streptophyta<span class="p">;</span>Magnoliopsida<span class="p">;</span>Malpighiales<span class="p">;</span>Salicaceae
Methanobrevibacter      <span class="m">2172</span>    genus   Archaea<span class="p">;</span>Euryarchaeota<span class="p">;</span>Methanobacteria<span class="p">;</span>Methanobacteriales<span class="p">;</span>Methanobacteriaceae<span class="p">;</span>Methanobrevibacter
Clostridium     <span class="m">1485</span>    genus   Bacteria<span class="p">;</span>Firmicutes<span class="p">;</span>Clostridia<span class="p">;</span>Clostridiales<span class="p">;</span>Clostridiaceae<span class="p">;</span>Clostridium
</pre></div>
</div>
<p>The taxa separator can be changed from <strong>;</strong> with <cite>-x</cite> and common names instead of scientific names
can be used with the <cite>-a</cite> option.</p>
<div class="section" id="match-names">
<h4>Match names<a class="headerlink" href="#match-names" title="Permalink to this headline">¶</a></h4>
<p>Besides the option to change the column separator, it is possible to only
print specific taxa (case insensitive search, but need the correct name) and
print headers for the output table. A partial and fuzzy search are performed,
however these are only reported, unless option <cite>-p</cite> is passed.</p>
<p>The lineage is printed only for the taxa requested, but all children can be
included by using the <cite>-c</cite> option.</p>
</div>
</div>
<div class="section" id="import-taxonomy">
<h3>Import Taxonomy<a class="headerlink" href="#import-taxonomy" title="Permalink to this headline">¶</a></h3>
<p>The <cite>import</cite> command allows the use of taxonomies other than NCBI and write a file that can be
used with other scripts. At the moment only PhyloPhlan is supported.</p>
<div class="section" id="phylophlan">
<h4>PhyloPhlan<a class="headerlink" href="#phylophlan" title="Permalink to this headline">¶</a></h4>
<p>Tested with version 3 of the script, the taxonomy file name is similar to <cite>SGB.Sep20.txt.bz2</cite>.
The file contains also NCBI IDs for known taxa and those will be kept. The IDs created
for non-NCBI taxa will be <em>negative</em> integers, to distinguish them.</p>
</div>
</div>
<div class="section" id="changes">
<h3>Changes<a class="headerlink" href="#changes" title="Permalink to this headline">¶</a></h3>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.5.7: </span>added <em>–only-ids</em> <em>-x</em> and <em>-a</em> option to <em>get</em> command. Added <em>import</em> command</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.5.0: </span>added <em>get</em> command to taxon-utils to print the taxonomy or search in it</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.4: </span>changed interface and behaviour for <em>filter</em>, also now can filter tables;
<em>lca</em> has changed the interface and allows the output of a 2 column table</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.1: </span>added <em>to_hdf</em> command</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.1: </span>added <em>-j</em> option to <em>lca</em>, which outputs a JSON file with the LCA results</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.0: </span>added <em>-k</em> and <em>-kt</em> options for Krona output, lineage now includes the LCA
also added <em>-a</em> option to select between lineages with only ranked taxa.
Now it defaults to all components.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.6: </span>added <em>feat-type</em> option to <em>lca</em> command, added phylum output to nolca</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.2.5.</span></p>
</div>
</div>
</div>
<div class="section" id="options">
<h2>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<div class="section" id="taxon-utils">
<h3>taxon-utils<a class="headerlink" href="#taxon-utils" title="Permalink to this headline">¶</a></h3>
<p>Main function</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>taxon-utils <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-version">
<code class="sig-name descname">--version</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-version" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the version and exit.</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-cite">
<code class="sig-name descname">--cite</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-cite" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<div class="section" id="taxon-utils-filter">
<h4>filter<a class="headerlink" href="#taxon-utils-filter" title="Permalink to this headline">¶</a></h4>
<p>Filter a GFF file or a table based on taxonomy</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>taxon-utils filter <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>INPUT_FILE<span class="o">]</span> <span class="o">[</span>OUTPUT_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-filter-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-filter-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-filter-p">
<code class="sig-name descname">-p</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--table</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-filter-p" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-filter-t">
<code class="sig-name descname">-t</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--taxonomy</code><code class="sig-prename descclassname"> &lt;taxonomy&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-filter-t" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Required</strong> Taxonomy file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-filter-i">
<code class="sig-name descname">-i</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--include-taxon-id</code><code class="sig-prename descclassname"> &lt;include_taxon_id&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-filter-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Include only taxon_ids</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-filter-in">
<code class="sig-name descname">-in</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--include-taxon-name</code><code class="sig-prename descclassname"> &lt;include_taxon_name&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-filter-in" title="Permalink to this definition">¶</a></dt>
<dd><p>Include only taxon_names</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-filter-e">
<code class="sig-name descname">-e</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--exclude-taxon-id</code><code class="sig-prename descclassname"> &lt;exclude_taxon_id&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-filter-e" title="Permalink to this definition">¶</a></dt>
<dd><p>Exclude taxon_ids</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-filter-en">
<code class="sig-name descname">-en</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--exclude-taxon-name</code><code class="sig-prename descclassname"> &lt;exclude_taxon_name&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-filter-en" title="Permalink to this definition">¶</a></dt>
<dd><p>Exclude taxon_names</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-filter-progress">
<code class="sig-name descname">--progress</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-filter-progress" title="Permalink to this definition">¶</a></dt>
<dd><p>Shows Progress Bar</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-filter-arg-INPUT_FILE">
<code class="sig-name descname">INPUT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-filter-arg-INPUT_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-filter-arg-OUTPUT_FILE">
<code class="sig-name descname">OUTPUT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-filter-arg-OUTPUT_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
<div class="section" id="taxon-utils-get">
<h4>get<a class="headerlink" href="#taxon-utils-get" title="Permalink to this headline">¶</a></h4>
<p>Extract the taxonomy as a CSV file</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>taxon-utils get <span class="o">[</span>OPTIONS<span class="o">]</span> TAXONOMY_FILE <span class="o">[</span>OUTPUT_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-get-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-get-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-d">
<code class="sig-name descname">-d</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--header</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-get-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Include header in the output</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-a">
<code class="sig-name descname">-a</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--use-cname</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-get-a" title="Permalink to this definition">¶</a></dt>
<dd><p>Use the common name if present</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-s">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--separator</code><code class="sig-prename descclassname"> &lt;separator&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-get-s" title="Permalink to this definition">¶</a></dt>
<dd><p>column separator</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-x">
<code class="sig-name descname">-x</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--tax-sep</code><code class="sig-prename descclassname"> &lt;tax_sep&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-get-x" title="Permalink to this definition">¶</a></dt>
<dd><p>taxa separator</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>;</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-o">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--only-names</code><code class="sig-prename descclassname"> &lt;only_names&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-get-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Only get matched taxon names</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-i">
<code class="sig-name descname">-i</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--only-ids</code><code class="sig-prename descclassname"> &lt;only_ids&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-get-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Only get matched taxon IDs</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-name-file">
<code class="sig-name descname">--name-file</code><code class="sig-prename descclassname"> &lt;name_file&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-get-name-file" title="Permalink to this definition">¶</a></dt>
<dd><p>File with names to search</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-id-file">
<code class="sig-name descname">--id-file</code><code class="sig-prename descclassname"> &lt;id_file&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-get-id-file" title="Permalink to this definition">¶</a></dt>
<dd><p>File with IDs to search</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-p">
<code class="sig-name descname">-p</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--partial</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-get-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Use partial matches if any found (implies -o)</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-z">
<code class="sig-name descname">-z</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--no-fuzzy</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-get-z" title="Permalink to this definition">¶</a></dt>
<dd><p>Avoid fuzzy name search</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-c">
<code class="sig-name descname">-c</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--include-children</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-get-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Include taxa that are children of the requested (implies -o)</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-get-arg-TAXONOMY_FILE">
<code class="sig-name descname">TAXONOMY_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-get-arg-TAXONOMY_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-get-arg-OUTPUT_FILE">
<code class="sig-name descname">OUTPUT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-get-arg-OUTPUT_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
<div class="section" id="taxon-utils-import">
<h4>import<a class="headerlink" href="#taxon-utils-import" title="Permalink to this headline">¶</a></h4>
<p>Create a MGKit taxonomy from an alternative taxonomy</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>taxon-utils import <span class="o">[</span>OPTIONS<span class="o">]</span> IMPORT_FILE TAXONOMY_FILE
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-import-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-import-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-import-t">
<code class="sig-name descname">-t</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--tax-type</code><code class="sig-prename descclassname"> &lt;tax_type&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-import-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Type of taxonomy to import</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>phylophlan</p>
</dd>
<dt class="field-even">Options</dt>
<dd class="field-even"><p>phylophlan</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-import-arg-IMPORT_FILE">
<code class="sig-name descname">IMPORT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-import-arg-IMPORT_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-import-arg-TAXONOMY_FILE">
<code class="sig-name descname">TAXONOMY_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-import-arg-TAXONOMY_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

</div>
<div class="section" id="taxon-utils-lca">
<h4>lca<a class="headerlink" href="#taxon-utils-lca" title="Permalink to this headline">¶</a></h4>
<p>Finds the last common ancestor for each sequence
in a GFF file</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>taxon-utils lca <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>GFF_FILE<span class="o">]</span> <span class="o">[</span>OUTPUT_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-t">
<code class="sig-name descname">-t</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--taxonomy</code><code class="sig-prename descclassname"> &lt;taxonomy&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca-t" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Required</strong> Taxonomy file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-n">
<code class="sig-name descname">-n</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--no-lca</code><code class="sig-prename descclassname"> &lt;no_lca&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca-n" title="Permalink to this definition">¶</a></dt>
<dd><p>File to which write records with no LCA</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-a">
<code class="sig-name descname">-a</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--only-ranked</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca-a" title="Permalink to this definition">¶</a></dt>
<dd><p>If set, only taxa that have a rank will be used in the lineageself. This is not advised for lineages such as Viruses, where the top levels have no rank</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-b">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--bitscore</code><code class="sig-prename descclassname"> &lt;bitscore&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum bitscore accepted</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>0</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-m">
<code class="sig-name descname">-m</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--rename</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Emulates BLAST behaviour for headers (keep left of first space)</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-s">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--sorted</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca-s" title="Permalink to this definition">¶</a></dt>
<dd><p>If the GFF file is sorted (all of a sequence annotations are contiguos) can use less memory, <cite>sort -s -k 1,1</cite> can be used</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-ft">
<code class="sig-name descname">-ft</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--feat-type</code><code class="sig-prename descclassname"> &lt;feat_type&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca-ft" title="Permalink to this definition">¶</a></dt>
<dd><p>Feature type used if the output is a GFF (default is <em>LCA</em>)</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>LCA</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-g">
<code class="sig-name descname">-g</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--group-by-attr</code><code class="sig-prename descclassname"> &lt;group_by_attr&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute to get the LCA for - default to sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>seq_id</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-r">
<code class="sig-name descname">-r</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--reference</code><code class="sig-prename descclassname"> &lt;reference&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Required reference file for the GFF, if krona is the format, contig lengths are added</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-p">
<code class="sig-name descname">-p</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--simple-table</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses a 2 column table format (seq_id taxon_id) TAB separated</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-kt">
<code class="sig-name descname">-kt</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--krona-total</code><code class="sig-prename descclassname"> &lt;krona_total&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca-kt" title="Permalink to this definition">¶</a></dt>
<dd><p>Total number of raw sequences (used to output correct percentages in Krona</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-f">
<code class="sig-name descname">-f</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--out-format</code><code class="sig-prename descclassname"> &lt;out_format&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca-f" title="Permalink to this definition">¶</a></dt>
<dd><p>Format of output file</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>tab</p>
</dd>
<dt class="field-even">Options</dt>
<dd class="field-even"><p>krona|json|tab|gff</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-progress">
<code class="sig-name descname">--progress</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca-progress" title="Permalink to this definition">¶</a></dt>
<dd><p>Shows Progress Bar</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-arg-GFF_FILE">
<code class="sig-name descname">GFF_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca-arg-GFF_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca-arg-OUTPUT_FILE">
<code class="sig-name descname">OUTPUT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca-arg-OUTPUT_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
<div class="section" id="taxon-utils-lca-line">
<h4>lca_line<a class="headerlink" href="#taxon-utils-lca-line" title="Permalink to this headline">¶</a></h4>
<p>Finds the last common ancestor for all IDs in
a text file line</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>taxon-utils lca_line <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>INPUT_FILE<span class="o">]</span> <span class="o">[</span>OUTPUT_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-lca_line-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca_line-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca_line-t">
<code class="sig-name descname">-t</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--taxonomy</code><code class="sig-prename descclassname"> &lt;taxonomy&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca_line-t" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Required</strong> Taxonomy file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca_line-n">
<code class="sig-name descname">-n</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--no-lca</code><code class="sig-prename descclassname"> &lt;no_lca&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca_line-n" title="Permalink to this definition">¶</a></dt>
<dd><p>File to which write records with no LCA</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca_line-a">
<code class="sig-name descname">-a</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--only-ranked</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca_line-a" title="Permalink to this definition">¶</a></dt>
<dd><p>If set, only taxa that have a rank will be used in the lineageself. This is not advised for lineages such as Viruses, where the top levels have no rank</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca_line-s">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--separator</code><code class="sig-prename descclassname"> &lt;separator&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-lca_line-s" title="Permalink to this definition">¶</a></dt>
<dd><p>separator for taxon_ids (defaults to TAB)</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-lca_line-arg-INPUT_FILE">
<code class="sig-name descname">INPUT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca_line-arg-INPUT_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-lca_line-arg-OUTPUT_FILE">
<code class="sig-name descname">OUTPUT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-lca_line-arg-OUTPUT_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
<div class="section" id="taxon-utils-to-hdf">
<h4>to_hdf<a class="headerlink" href="#taxon-utils-to-hdf" title="Permalink to this headline">¶</a></h4>
<p>Convert a taxa table to HDF5, with the input as
tabular format, defaults to stdin. Output file, defaults to
(taxa-table.hf5)</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>taxon-utils to_hdf <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>INPUT_FILE<span class="o">]</span> <span class="o">[</span>OUTPUT_FILE<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-to_hdf-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--verbose</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-to_hdf-v" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-to_hdf-n">
<code class="sig-name descname">-n</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--table-name</code><code class="sig-prename descclassname"> &lt;table_name&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-to_hdf-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of the table/storage to use</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>taxa</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-to_hdf-w">
<code class="sig-name descname">-w</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--overwrite</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-to_hdf-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Overwrite the file, instead of appending to it</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-to_hdf-s">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--index-size</code><code class="sig-prename descclassname"> &lt;index_size&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-to_hdf-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum number of characters for the gene_id</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>12</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-to_hdf-c">
<code class="sig-name descname">-c</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--chunk-size</code><code class="sig-prename descclassname"> &lt;chunk_size&gt;</code><a class="headerlink" href="#cmdoption-taxon-utils-to_hdf-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Chunk size to use when reading the input file</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>5000000</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-to_hdf-progress">
<code class="sig-name descname">--progress</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-to_hdf-progress" title="Permalink to this definition">¶</a></dt>
<dd><p>Shows Progress Bar</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-taxon-utils-to_hdf-arg-INPUT_FILE">
<code class="sig-name descname">INPUT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-to_hdf-arg-INPUT_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-taxon-utils-to_hdf-arg-OUTPUT_FILE">
<code class="sig-name descname">OUTPUT_FILE</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-taxon-utils-to_hdf-arg-OUTPUT_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument</p>
</dd></dl>

</div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">MGKit: Metagenomic framework</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Scripts Details</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="blast2gff.html">blast2gff - Convert BLAST output to GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="filter-gff.html">filter-gff - Filter GFF annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="add-gff-info.html">add-gff-info - Add informations to GFF annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="get-gff-info.html">get-gff-info - Extract informations to GFF annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="hmmer2gff.html">hmmer2gff - Convert HMMER output to GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="snp_analysis.html">snp_parser - SNPs analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="download-taxonomy.html">Download Taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="download-taxonomy.html#download-accession-taxonid">Download Accession/TaxonID</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">taxon-utils - Taxonomy Utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="fasta-utils.html">fasta-utils - Fasta Utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="fastq-utils.html">fastq-utils - Fastq Utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="json2gff.html">json2gff - Convert JSON to GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="sampling-utils.html">sampling-utils - Resampling Utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="edit_gff.html">edit-gff - GFF Viewer and Editor</a></li>
<li class="toctree-l2"><a class="reference internal" href="pnps_gen.html">pnps-gen - pN/pS Table Generation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1"><a class="reference internal" href="../library.html">Library Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Scripts Details</a><ul>
      <li>Previous: <a href="download-taxonomy.html" title="previous chapter">Download Taxonomy</a></li>
      <li>Next: <a href="fasta-utils.html" title="next chapter">fasta-utils - Fasta Utilities</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2013-2020, Francesco Rubino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/scripts/taxon_utils.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>