

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>mgkit.io.uniprot &mdash; MGKit: Metagenomic framework 0.5.3 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/graphviz.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home" alt="Documentation Home"> MGKit: Metagenomic framework
          

          
          </a>

          
            
            
              <div class="version">
                0.5.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../library.html">Library Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changes.html">Changes</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">MGKit: Metagenomic framework</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
          <li><a href="../../mgkit.html">mgkit</a> &raquo;</li>
        
      <li>mgkit.io.uniprot</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for mgkit.io.uniprot</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">.. versionadded:: 0.1.13</span>

<span class="sd">Uniprot file formats</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">.</span> <span class="kn">import</span> <span class="n">open_file</span>

<span class="n">LOG</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<span class="n">NUM_LINES</span> <span class="o">=</span> <span class="mi">10</span> <span class="o">**</span> <span class="mi">7</span>


<span class="n">MAPPINGS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;taxonomy&#39;</span><span class="p">:</span> <span class="s1">&#39;NCBI_TaxID&#39;</span><span class="p">,</span>
    <span class="s1">&#39;eggnog&#39;</span><span class="p">:</span> <span class="s1">&#39;eggNOG&#39;</span><span class="p">,</span>
    <span class="s1">&#39;ko&#39;</span><span class="p">:</span> <span class="s1">&#39;KO&#39;</span><span class="p">,</span>
    <span class="s1">&#39;kegg&#39;</span><span class="p">:</span> <span class="s1">&#39;KEGG&#39;</span><span class="p">,</span>
    <span class="s1">&#39;biocyc&#39;</span><span class="p">:</span> <span class="s1">&#39;BioCyc&#39;</span><span class="p">,</span>
    <span class="s1">&#39;unipathway&#39;</span><span class="p">:</span> <span class="s1">&#39;UniPathway&#39;</span><span class="p">,</span>
    <span class="s1">&#39;embl&#39;</span><span class="p">:</span> <span class="s1">&#39;EMBL&#39;</span><span class="p">,</span>
    <span class="s1">&#39;embl_cds&#39;</span><span class="p">:</span> <span class="s1">&#39;EMBL-CDS&#39;</span><span class="p">,</span>
    <span class="s1">&#39;gi&#39;</span><span class="p">:</span> <span class="s1">&#39;GI&#39;</span><span class="p">,</span>
    <span class="s1">&#39;string&#39;</span><span class="p">:</span> <span class="s1">&#39;STRING&#39;</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Some of the mappings contained in the idmapping.dat.gz</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="parse_uniprot_mappings"><a class="viewcode-back" href="../../../api/mgkit.io.uniprot.html#mgkit.io.uniprot.parse_uniprot_mappings">[docs]</a><span class="k">def</span> <span class="nf">parse_uniprot_mappings</span><span class="p">(</span><span class="n">file_handle</span><span class="p">,</span> <span class="n">gene_ids</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mappings</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                           <span class="n">num_lines</span><span class="o">=</span><span class="n">NUM_LINES</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parses a Uniprot mapping `file &lt;ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/idmapping.dat.gz&gt;`_,</span>
<span class="sd">    returning a generator with the mappings.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        file_handle (str, file): file name or open file handle</span>
<span class="sd">        gene_ids (None, set): if not None, the returned mappings are for the</span>
<span class="sd">            gene IDs specified</span>
<span class="sd">        mappings (None, set): mappings to be returned</span>
<span class="sd">        num_lines (None, int): number of which a message is logged. If None,</span>
<span class="sd">            no message is logged</span>

<span class="sd">    Yields:</span>
<span class="sd">        tuple: the first element is the gene ID, the second is the mapping type</span>
<span class="sd">        and third element is the mapped ID</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">file_handle</span> <span class="o">=</span> <span class="n">open_file</span><span class="p">(</span><span class="n">file_handle</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>

    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Loading Uniprot Mappings from file (</span><span class="si">%s</span><span class="s2">)&quot;</span><span class="p">,</span>
        <span class="nb">getattr</span><span class="p">(</span><span class="n">file_handle</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="nb">repr</span><span class="p">(</span><span class="n">file_handle</span><span class="p">))</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">gene_ids</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">gene_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">)</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Mappings for </span><span class="si">%d</span><span class="s2"> gene_ids will be returned&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">mappings</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mappings</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">mappings</span><span class="p">)</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Mappings to &#39;</span><span class="si">%s</span><span class="s2">&#39; will be returned&quot;</span><span class="p">,</span>
            <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mappings</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">file_handle</span><span class="p">):</span>
        <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;ascii&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span><span class="n">num_lines</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">((</span><span class="n">idx</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="n">num_lines</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
            <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parsed </span><span class="si">%d</span><span class="s2"> lines&quot;</span><span class="p">,</span> <span class="n">idx</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

        <span class="n">gene_id</span><span class="p">,</span> <span class="n">mapping</span><span class="p">,</span> <span class="n">map_id</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span><span class="n">gene_ids</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">gene_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">):</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="p">(</span><span class="n">mappings</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">mapping</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mappings</span><span class="p">):</span>
            <span class="k">continue</span>

        <span class="k">yield</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">mapping</span><span class="p">,</span> <span class="n">map_id</span>

    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Read </span><span class="si">%d</span><span class="s2"> lines&quot;</span><span class="p">,</span> <span class="n">idx</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span></div>


<div class="viewcode-block" id="uniprot_mappings_to_dict"><a class="viewcode-back" href="../../../api/mgkit.io.uniprot.html#mgkit.io.uniprot.uniprot_mappings_to_dict">[docs]</a><span class="k">def</span> <span class="nf">uniprot_mappings_to_dict</span><span class="p">(</span><span class="n">file_handle</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">mappings</span><span class="p">,</span> <span class="n">num_lines</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionchanged:: 0.3.4</span>
<span class="sd">        added *num_lines*</span>

<span class="sd">    Parses a Uniprot mapping `file &lt;ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/idmapping.dat.gz&gt;`_,</span>
<span class="sd">    returning a generator of dictionaries with the mappings requested.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        file_handle (str, file): file name or open file handle</span>
<span class="sd">        gene_ids (None, set): if not None, the returned mappings are for the</span>
<span class="sd">            gene IDs specified</span>
<span class="sd">        mappings (None, set): mappings to be returned</span>
<span class="sd">        num_lines (int, None): passed to :func:`parse_uniprot_mappings`</span>

<span class="sd">    Yields:</span>
<span class="sd">        tuple: the first element is the gene ID, the second is a dictionary</span>
<span class="sd">        with all the mappings found, the key is the mapping type and the value</span>
<span class="sd">        is a list of all mapped IDs</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">iterator</span> <span class="o">=</span> <span class="n">parse_uniprot_mappings</span><span class="p">(</span>
        <span class="n">file_handle</span><span class="p">,</span>
        <span class="n">gene_ids</span><span class="o">=</span><span class="n">gene_ids</span><span class="p">,</span>
        <span class="n">mappings</span><span class="o">=</span><span class="n">mappings</span><span class="p">,</span>
        <span class="n">num_lines</span><span class="o">=</span><span class="n">num_lines</span>
    <span class="p">)</span>

    <span class="n">curr_gene</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="n">curr_maps</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">mapping</span><span class="p">,</span> <span class="n">map_id</span> <span class="ow">in</span> <span class="n">iterator</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">curr_gene</span> <span class="o">==</span> <span class="n">gene_id</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">curr_maps</span><span class="p">[</span><span class="n">mapping</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">map_id</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">curr_maps</span><span class="p">[</span><span class="n">mapping</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">map_id</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">curr_gene</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">curr_gene</span><span class="p">,</span> <span class="n">curr_maps</span>
                <span class="n">curr_maps</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">curr_gene</span> <span class="o">=</span> <span class="n">gene_id</span>
            <span class="n">curr_maps</span><span class="p">[</span><span class="n">mapping</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">map_id</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">yield</span> <span class="n">curr_gene</span><span class="p">,</span> <span class="n">curr_maps</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2013-2020, Francesco Rubino

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>