
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>mgkit.workflow.pnps_gen &#8212; MGKit: Metagenomic framework 0.5.7 documentation</title>
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css" />
    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for mgkit.workflow.pnps_gen</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Calculates pN/pS values</span>
<span class="sd">***********************</span>

<span class="sd">This script calculates pN/pS using the data produced by the script `vcf`</span>
<span class="sd">command. The result table is a CSV file.</span>

<span class="sd">Parse VCF Files</span>
<span class="sd">***************</span>

<span class="sd">The `vcf` command will parse a VCF file to produce the pickle file that is used</span>
<span class="sd">to calculate the pN/pS.</span>

<span class="sd">Calculate Rank pN/pS</span>
<span class="sd">********************</span>

<span class="sd">The `rank` command of the script reads SNPs information and calculate for each</span>
<span class="sd">element of a specific taxonomic rank (species, genus, family, etc.) its pN/pS.</span>
<span class="sd">Another option is the *None* rank, which makes the script use the taxonomic ID</span>
<span class="sd">found in the annotations.</span>

<span class="sd">For example, choosing the rank *genus* a table will be produced, similar to::</span>

<span class="sd">    Prevotella,0.0001,1,1.1,0.4</span>
<span class="sd">    Methanobrevibacter,1,0.5,0.6,0.8</span>

<span class="sd">A pN/pS value for each genus and sample (4 in this case) will be calculated.</span>

<span class="sd">It is important to specify the taxonomic IDs to include in tha calculations. By</span>
<span class="sd">default only bacteria are included. To get those values, the taxonomy can be</span>
<span class="sd">queried using `taxon-utils get`.</span>

<span class="sd">Calculate Gene/Rank pN/pS</span>
<span class="sd">*************************</span>

<span class="sd">The `full` command create a gene/taxon table of pN/pS, internally is a pandas</span>
<span class="sd">MultiIndex DataFrame, written in CSV format after script execution. The</span>
<span class="sd">difference with the `rank` is the pN/pS calculation is for each gene/taxon and</span>
<span class="sd">by default the `gene_id` from the original GFF file is used (which is stored</span>
<span class="sd">in the file generated by `snp_parser`). If other gene IDs needs to be used, a</span>
<span class="sd">table file can be provided, which can be passed in two column formats.</span>

<span class="sd">The default in MGKit is to use Uniprot gene IDs for the functions, but we may</span>
<span class="sd">want to examine the Kegg Orthologs instead. A table can be passed where the</span>
<span class="sd">first column in the gene_id stored in the GFF file and the second is the KO::</span>

<span class="sd">    Q7N6F9  K05685</span>
<span class="sd">    Q7N6F9  K01242</span>
<span class="sd">    G7E4F2  K05625</span>

<span class="sd">The `Q7N6F9` gene_id is repeated because it has multiple correspondences to KOs</span>
<span class="sd">and this format needs to be selected using the `-2` option of the command.</span>

<span class="sd">The default type of table expected by the command is a table with a gene ID as</span>
<span class="sd">first column one or more tab separated columns with mappings. The previous</span>
<span class="sd">table would look like this::</span>

<span class="sd">    Q7N6F9  K05685  K01242</span>
<span class="sd">    G7E4F2  K05625</span>

<span class="sd">These tables can be created from the original GFF file, assuming that mappings</span>
<span class="sd">to KO, EC Numbers are included, with a command line like this::</span>

<span class="sd">     edit-gff view -a gene_id -a map_KO final.contigs-a3.gff.gz | tr &#39;,&#39; &#39;\t&#39;</span>

<span class="sd">Extracting the KOs (which are comma separated in a MGKit GFF file) and changing</span>
<span class="sd">any comma to *tab*. This table can be passed to the script and will make it</span>
<span class="sd">possible to calculate the pN/pS for the KOs associated to the genes. Only gene</span>
<span class="sd">IDs present in this file have a calculated pN/pS.</span>

<span class="sd">Normally you combine the all isoforms with the same the *gene_id* to produce a</span>
<span class="sd">single pN/pS, but if it&#39;s needed, the `-u` option can be used to calculate a</span>
<span class="sd">pN/pS for each line in the GFF file.</span>

<span class="sd">Changes</span>
<span class="sd">*******</span>

<span class="sd">.. versionchanged:: 0.5.7</span>
<span class="sd">    added *vcf* command to parse VCF files and generate data for the script</span>

<span class="sd">.. versionchanged:: 0.5.1</span>
<span class="sd">    bug fix</span>

<span class="sd">.. versionchanged:: 0.5.1</span>

<span class="sd">    * added option to include the lineage as a string</span>
<span class="sd">    * added option to use the *uids* from the GFF instead of *gene_id*, this</span>
<span class="sd">      does not require the GFF file, they are embedded into the *.pickle* file</span>

<span class="sd">.. versionadded:: 0.5.0</span>
<span class="sd">&quot;&quot;&quot;</span>


<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">OrderedDict</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">click</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">functools</span>
<span class="kn">import</span> <span class="nn">vcf</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">from</span> <span class="nn">mgkit</span> <span class="kn">import</span> <span class="n">taxon</span>
<span class="kn">from</span> <span class="nn">mgkit.snps</span> <span class="kn">import</span> <span class="n">conv_func</span>
<span class="kn">import</span> <span class="nn">mgkit.snps.mapper</span>
<span class="kn">import</span> <span class="nn">mgkit.snps.funcs</span>
<span class="kn">import</span> <span class="nn">mgkit.snps.filter</span> <span class="k">as</span> <span class="nn">snp_filter</span>
<span class="kn">from</span> <span class="nn">..snps.classes</span> <span class="kn">import</span> <span class="n">GeneSNP</span><span class="p">,</span> <span class="n">SNPType</span>
<span class="kn">from</span> <span class="nn">..io</span> <span class="kn">import</span> <span class="n">gff</span><span class="p">,</span> <span class="n">fasta</span>
<span class="kn">from</span> <span class="nn">.</span> <span class="kn">import</span> <span class="n">utils</span>
<span class="kn">from</span> <span class="nn">..utils</span> <span class="kn">import</span> <span class="n">dictionary</span>
<span class="kn">from</span> <span class="nn">..</span> <span class="kn">import</span> <span class="n">logger</span>

<span class="n">LOG</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<span class="c1"># changing the value for the script to remove subclass as rank</span>
<span class="n">taxon</span><span class="o">.</span><span class="n">TAXON_RANKS</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">rank</span> <span class="k">for</span> <span class="n">rank</span> <span class="ow">in</span> <span class="n">taxon</span><span class="o">.</span><span class="n">TAXON_RANKS</span>
                          <span class="k">if</span> <span class="ow">not</span> <span class="n">rank</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;sub&#39;</span><span class="p">))</span>


<div class="viewcode-block" id="get_lineage"><a class="viewcode-back" href="../../../api/mgkit.workflow.pnps_gen.html#mgkit.workflow.pnps_gen.get_lineage">[docs]</a><span class="k">def</span> <span class="nf">get_lineage</span><span class="p">(</span><span class="n">taxonomy</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">taxon_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxonomy</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">taxon_id</span>
    <span class="k">return</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">get_lineage_string</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">only_ranked</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                       <span class="n">with_last</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">add_rank</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                       <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;;&#39;</span><span class="p">)</span></div>


<span class="nd">@click</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">version_option</span><span class="p">()</span>
<span class="nd">@utils</span><span class="o">.</span><span class="n">cite_option</span>
<span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="s2">&quot;Main function&quot;</span>
    <span class="k">pass</span>


<span class="nd">@main</span><span class="o">.</span><span class="n">command</span><span class="p">(</span><span class="s1">&#39;rank&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;&quot;&quot;Calculates pN/pS for a taxonomic rank&quot;&quot;&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-v&#39;</span><span class="p">,</span> <span class="s1">&#39;--verbose&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-t&#39;</span><span class="p">,</span> <span class="s1">&#39;--taxonomy&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;rb&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxonomy file&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="s1">&#39;--snp-data&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;rb&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;SNP data, output of `snp_parser`&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="s1">&#39;--rank&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;order&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Taxonomic rank&#39;</span><span class="p">,</span>
              <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">Choice</span><span class="p">(</span><span class="n">taxon</span><span class="o">.</span><span class="n">TAXON_RANKS</span> <span class="o">+</span> <span class="p">(</span><span class="s1">&#39;None&#39;</span><span class="p">,),</span>
                                <span class="n">case_sensitive</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-m&#39;</span><span class="p">,</span> <span class="s1">&#39;--min-num&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">IntRange</span><span class="p">(</span><span class="nb">min</span><span class="o">=</span><span class="mi">2</span><span class="p">),</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Minimum number of samples with a pN/pS to accept&#39;</span><span class="p">,</span>
              <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-c&#39;</span><span class="p">,</span> <span class="s1">&#39;--min-cov&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">IntRange</span><span class="p">(</span><span class="nb">min</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Minimum coverage for SNPs to be accepted&#39;</span><span class="p">,</span>
              <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="s1">&#39;--taxon_ids&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">INT</span><span class="p">,</span> <span class="n">multiple</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Taxon IDs to include&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,),</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-u&#39;</span><span class="p">,</span> <span class="s1">&#39;--unstack&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Samples are not in columns but as an array&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-l&#39;</span><span class="p">,</span> <span class="s1">&#39;--lineage&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Use lineage string instead of taxon_id&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-ps&#39;</span><span class="p">,</span> <span class="s1">&#39;--only-ps&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Only calculate pS values&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-pn&#39;</span><span class="p">,</span> <span class="s1">&#39;--only-pn&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Only calculate pN values&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">argument</span><span class="p">(</span><span class="s1">&#39;txt_file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">gen_rank</span><span class="p">(</span><span class="n">verbose</span><span class="p">,</span> <span class="n">taxonomy</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">,</span> <span class="n">rank</span><span class="p">,</span> <span class="n">min_num</span><span class="p">,</span> <span class="n">min_cov</span><span class="p">,</span>
             <span class="n">taxon_ids</span><span class="p">,</span> <span class="n">unstack</span><span class="p">,</span> <span class="n">lineage</span><span class="p">,</span> <span class="n">only_ps</span><span class="p">,</span> <span class="n">only_pn</span><span class="p">,</span> <span class="n">txt_file</span><span class="p">):</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">config_log</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span> <span class="k">if</span> <span class="n">verbose</span> <span class="k">else</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

    <span class="n">taxonomy</span> <span class="o">=</span> <span class="n">taxon</span><span class="o">.</span><span class="n">Taxonomy</span><span class="p">(</span><span class="n">taxonomy</span><span class="p">)</span>

    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Only taxa below </span><span class="si">%s</span><span class="s1"> will be included&#39;</span><span class="p">,</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">taxonomy</span><span class="p">[</span><span class="n">taxon_id</span><span class="p">]</span><span class="o">.</span><span class="n">s_name</span> <span class="k">for</span> <span class="n">taxon_id</span> <span class="ow">in</span> <span class="n">taxon_ids</span><span class="p">))</span>
    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Rank &quot;</span><span class="si">%s</span><span class="s1">&quot; and above will be included&#39;</span><span class="p">,</span> <span class="n">rank</span><span class="p">)</span>

    <span class="n">snp_data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">snp_data</span><span class="p">)</span>

    <span class="n">filters</span> <span class="o">=</span> <span class="n">snp_filter</span><span class="o">.</span><span class="n">get_default_filters</span><span class="p">(</span><span class="n">taxonomy</span><span class="p">,</span> <span class="n">min_cov</span><span class="o">=</span><span class="n">min_cov</span><span class="p">,</span>
                                             <span class="n">include_only</span><span class="o">=</span><span class="n">taxon_ids</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">rank</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxon</span><span class="o">.</span><span class="n">TAXON_RANKS</span><span class="p">:</span>
        <span class="n">rank</span> <span class="o">=</span> <span class="kc">None</span>
    
    <span class="k">if</span> <span class="n">rank</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">taxon_func</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">taxon_func</span> <span class="o">=</span> <span class="n">functools</span><span class="o">.</span><span class="n">partial</span><span class="p">(</span>
            <span class="n">mgkit</span><span class="o">.</span><span class="n">snps</span><span class="o">.</span><span class="n">mapper</span><span class="o">.</span><span class="n">map_taxon_id_to_rank</span><span class="p">,</span>
            <span class="n">taxonomy</span><span class="o">=</span><span class="n">taxonomy</span><span class="p">,</span>
            <span class="n">rank</span><span class="o">=</span><span class="n">rank</span>
        <span class="p">)</span>
    
    <span class="n">partial_calc</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">partial_syn</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">if</span> <span class="n">only_ps</span> <span class="ow">or</span> <span class="n">only_pn</span><span class="p">:</span>
        <span class="n">partial_calc</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="n">only_pn</span><span class="p">:</span>
            <span class="n">partial_syn</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Only &quot;</span><span class="si">%s</span><span class="s1">&quot; will be calculated &#39;</span><span class="p">,</span> <span class="s2">&quot;pS&quot;</span>  <span class="k">if</span> <span class="n">partial_syn</span> <span class="k">else</span> <span class="s2">&quot;pN&quot;</span><span class="p">)</span>

    <span class="n">pnps</span> <span class="o">=</span> <span class="n">mgkit</span><span class="o">.</span><span class="n">snps</span><span class="o">.</span><span class="n">funcs</span><span class="o">.</span><span class="n">combine_sample_snps</span><span class="p">(</span><span class="n">snp_data</span><span class="p">,</span> <span class="n">min_num</span><span class="p">,</span> <span class="n">filters</span><span class="p">,</span> <span class="n">index_type</span><span class="o">=</span><span class="s1">&#39;taxon&#39;</span><span class="p">,</span>
                                                <span class="n">taxon_func</span><span class="o">=</span><span class="n">taxon_func</span><span class="p">,</span> <span class="n">partial_calc</span><span class="o">=</span><span class="n">partial_calc</span><span class="p">,</span>
                                                <span class="n">partial_syn</span><span class="o">=</span><span class="n">partial_syn</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">lineage</span><span class="p">:</span>
        <span class="n">pnps</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">get_lineage</span><span class="p">(</span><span class="n">taxonomy</span><span class="p">,</span> <span class="n">x</span><span class="p">),</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">pnps</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;taxon&#39;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">unstack</span><span class="p">:</span>
        <span class="n">pnps</span> <span class="o">=</span> <span class="n">pnps</span><span class="o">.</span><span class="n">unstack</span><span class="p">()</span>

    <span class="n">pnps</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">txt_file</span><span class="p">)</span>


<div class="viewcode-block" id="read_gene_map_default"><a class="viewcode-back" href="../../../api/mgkit.workflow.pnps_gen.html#mgkit.workflow.pnps_gen.read_gene_map_default">[docs]</a><span class="k">def</span> <span class="nf">read_gene_map_default</span><span class="p">(</span><span class="n">file_handle</span><span class="p">,</span> <span class="n">separator</span><span class="p">):</span>

    <span class="n">gene_map</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file_handle</span><span class="p">:</span>
        <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">separator</span><span class="p">)</span>
        <span class="n">gene_map</span><span class="p">[</span><span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>

    <span class="k">return</span> <span class="n">gene_map</span></div>


<div class="viewcode-block" id="read_gene_map_two_columns"><a class="viewcode-back" href="../../../api/mgkit.workflow.pnps_gen.html#mgkit.workflow.pnps_gen.read_gene_map_two_columns">[docs]</a><span class="k">def</span> <span class="nf">read_gene_map_two_columns</span><span class="p">(</span><span class="n">file_handle</span><span class="p">,</span> <span class="n">separator</span><span class="p">):</span>

    <span class="n">gene_map</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file_handle</span><span class="p">:</span>
        <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">separator</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">gene_map</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">gene_map</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">value</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">gene_map</span></div>


<span class="nd">@main</span><span class="o">.</span><span class="n">command</span><span class="p">(</span><span class="s1">&#39;full&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;&quot;&quot;Calculates pN/pS&quot;&quot;&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-v&#39;</span><span class="p">,</span> <span class="s1">&#39;--verbose&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-t&#39;</span><span class="p">,</span> <span class="s1">&#39;--taxonomy&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;rb&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxonomy file&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="s1">&#39;--snp-data&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;rb&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;SNP data, output of `snp_parser`&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="s1">&#39;--rank&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Taxonomic rank&#39;</span><span class="p">,</span>
              <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">Choice</span><span class="p">(</span><span class="n">taxon</span><span class="o">.</span><span class="n">TAXON_RANKS</span> <span class="o">+</span> <span class="p">(</span><span class="s1">&#39;None&#39;</span><span class="p">,),</span>
                                <span class="n">case_sensitive</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-m&#39;</span><span class="p">,</span> <span class="s1">&#39;--min-num&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">IntRange</span><span class="p">(</span><span class="nb">min</span><span class="o">=</span><span class="mi">2</span><span class="p">),</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Minimum number of samples with a pN/pS to accept&#39;</span><span class="p">,</span>
              <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-c&#39;</span><span class="p">,</span> <span class="s1">&#39;--min-cov&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">IntRange</span><span class="p">(</span><span class="nb">min</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Minimum coverage for SNPs to be accepted&#39;</span><span class="p">,</span>
              <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="s1">&#39;--taxon-ids&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">INT</span><span class="p">,</span> <span class="n">multiple</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Taxon IDs to include&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-u&#39;</span><span class="p">,</span> <span class="s1">&#39;--use-uid&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Use uids from the GFF file instead of gene_id as genes&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="s1">&#39;--gene-map&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Dictionary to map *gene_id* to another ID&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-2&#39;</span><span class="p">,</span> <span class="s1">&#39;--two-columns&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;gene-map is a two columns table with repeated keys&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-p&#39;</span><span class="p">,</span> <span class="s1">&#39;--separator&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;column separator for gene-map file&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-l&#39;</span><span class="p">,</span> <span class="s1">&#39;--lineage&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Use lineage string instead of taxon_id&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-e&#39;</span><span class="p">,</span> <span class="s1">&#39;--parquet&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Output a Parquet file instead of CSV&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-ps&#39;</span><span class="p">,</span> <span class="s1">&#39;--only-ps&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Only calculate pS values&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-pn&#39;</span><span class="p">,</span> <span class="s1">&#39;--only-pn&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Only calculate pN values&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">argument</span><span class="p">(</span><span class="s1">&#39;output_file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">writable</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">gen_full</span><span class="p">(</span><span class="n">verbose</span><span class="p">,</span> <span class="n">taxonomy</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">,</span> <span class="n">rank</span><span class="p">,</span> <span class="n">min_num</span><span class="p">,</span> <span class="n">min_cov</span><span class="p">,</span>
             <span class="n">taxon_ids</span><span class="p">,</span> <span class="n">use_uid</span><span class="p">,</span> <span class="n">gene_map</span><span class="p">,</span> <span class="n">two_columns</span><span class="p">,</span> <span class="n">separator</span><span class="p">,</span> <span class="n">lineage</span><span class="p">,</span>
             <span class="n">parquet</span><span class="p">,</span> <span class="n">only_ps</span><span class="p">,</span> <span class="n">only_pn</span><span class="p">,</span> <span class="n">output_file</span><span class="p">):</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">config_log</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span> <span class="k">if</span> <span class="n">verbose</span> <span class="k">else</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">gene_map</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Reading gene-map&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">two_columns</span><span class="p">:</span>
            <span class="n">gene_map</span> <span class="o">=</span> <span class="n">read_gene_map_two_columns</span><span class="p">(</span><span class="n">gene_map</span><span class="p">,</span> <span class="n">separator</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">gene_map</span> <span class="o">=</span> <span class="n">read_gene_map_default</span><span class="p">(</span><span class="n">gene_map</span><span class="p">,</span> <span class="n">separator</span><span class="p">)</span>
        
        <span class="n">gene_map</span> <span class="o">=</span> <span class="n">functools</span><span class="o">.</span><span class="n">partial</span><span class="p">(</span>
            <span class="n">mgkit</span><span class="o">.</span><span class="n">snps</span><span class="o">.</span><span class="n">mapper</span><span class="o">.</span><span class="n">map_gene_id</span><span class="p">,</span>
            <span class="n">gene_map</span><span class="o">=</span><span class="n">gene_map</span>
        <span class="p">)</span>

    <span class="n">taxonomy</span> <span class="o">=</span> <span class="n">taxon</span><span class="o">.</span><span class="n">Taxonomy</span><span class="p">(</span><span class="n">taxonomy</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">rank</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Rank &quot;</span><span class="si">%s</span><span class="s1">&quot; and above will be included&#39;</span><span class="p">,</span> <span class="n">rank</span><span class="p">)</span>

    <span class="n">snp_data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">snp_data</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">rank</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxon</span><span class="o">.</span><span class="n">TAXON_RANKS</span><span class="p">:</span>
        <span class="n">rank</span> <span class="o">=</span> <span class="kc">None</span>
    
    <span class="k">if</span> <span class="n">rank</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">taxon_func</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">taxon_func</span> <span class="o">=</span> <span class="n">functools</span><span class="o">.</span><span class="n">partial</span><span class="p">(</span>
            <span class="n">mgkit</span><span class="o">.</span><span class="n">snps</span><span class="o">.</span><span class="n">mapper</span><span class="o">.</span><span class="n">map_taxon_id_to_rank</span><span class="p">,</span>
            <span class="n">taxonomy</span><span class="o">=</span><span class="n">taxonomy</span><span class="p">,</span>
            <span class="n">rank</span><span class="o">=</span><span class="n">rank</span>
        <span class="p">)</span>

    <span class="n">filters</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">functools</span><span class="o">.</span><span class="n">partial</span><span class="p">(</span>
            <span class="n">snp_filter</span><span class="o">.</span><span class="n">filter_genesyn_by_coverage</span><span class="p">,</span>
            <span class="n">min_cov</span><span class="o">=</span><span class="n">min_cov</span>
        <span class="p">)</span>
    <span class="p">]</span>

    <span class="k">if</span> <span class="n">taxon_ids</span><span class="p">:</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Only taxa below </span><span class="si">%s</span><span class="s1"> will be included&#39;</span><span class="p">,</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">taxonomy</span><span class="p">[</span><span class="n">taxon_id</span><span class="p">]</span><span class="o">.</span><span class="n">s_name</span> <span class="k">for</span> <span class="n">taxon_id</span> <span class="ow">in</span> <span class="n">taxon_ids</span><span class="p">))</span>
        <span class="n">filters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="n">functools</span><span class="o">.</span><span class="n">partial</span><span class="p">(</span>
                <span class="n">snp_filter</span><span class="o">.</span><span class="n">filter_genesyn_by_taxon_id</span><span class="p">,</span>
                <span class="n">taxonomy</span><span class="o">=</span><span class="n">taxonomy</span><span class="p">,</span>
                <span class="n">filter_list</span><span class="o">=</span><span class="n">taxon_ids</span><span class="p">,</span>
                <span class="n">exclude</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">func</span><span class="o">=</span><span class="n">taxon</span><span class="o">.</span><span class="n">is_ancestor</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="n">partial_calc</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">partial_syn</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">if</span> <span class="n">only_ps</span> <span class="ow">or</span> <span class="n">only_pn</span><span class="p">:</span>
        <span class="n">partial_calc</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="n">only_pn</span><span class="p">:</span>
            <span class="n">partial_syn</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Only &quot;</span><span class="si">%s</span><span class="s1">&quot; will be calculated &#39;</span><span class="p">,</span> <span class="s2">&quot;pS&quot;</span>  <span class="k">if</span> <span class="n">partial_syn</span> <span class="k">else</span> <span class="s2">&quot;pN&quot;</span><span class="p">)</span>

    <span class="n">pnps</span> <span class="o">=</span> <span class="n">mgkit</span><span class="o">.</span><span class="n">snps</span><span class="o">.</span><span class="n">funcs</span><span class="o">.</span><span class="n">combine_sample_snps</span><span class="p">(</span><span class="n">snp_data</span><span class="p">,</span> <span class="n">min_num</span><span class="p">,</span> <span class="n">filters</span><span class="p">,</span> <span class="n">gene_func</span><span class="o">=</span><span class="n">gene_map</span><span class="p">,</span>
                                                <span class="n">taxon_func</span><span class="o">=</span><span class="n">taxon_func</span><span class="p">,</span> <span class="n">use_uid</span><span class="o">=</span><span class="n">use_uid</span><span class="p">,</span>
                                                <span class="n">partial_calc</span><span class="o">=</span><span class="n">partial_calc</span><span class="p">,</span> <span class="n">partial_syn</span><span class="o">=</span><span class="n">partial_syn</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">lineage</span><span class="p">:</span>
        <span class="n">pnps</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>  <span class="n">get_lineage</span><span class="p">(</span><span class="n">taxonomy</span><span class="p">,</span> <span class="n">x</span><span class="p">),</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">pnps</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;taxon&#39;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">parquet</span><span class="p">:</span>
        <span class="n">pnps</span><span class="o">.</span><span class="n">to_parquet</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pnps</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>


<div class="viewcode-block" id="init_count_set"><a class="viewcode-back" href="../../../api/mgkit.workflow.pnps_gen.html#mgkit.workflow.pnps_gen.init_count_set">[docs]</a><span class="k">def</span> <span class="nf">init_count_set</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">seqs</span><span class="p">):</span>
    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Init data structures&quot;</span><span class="p">)</span>

    <span class="n">samples</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">annotations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sample_coverage</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

    <span class="n">snp_data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
        <span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="p">{})</span> <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">samples</span>
    <span class="p">)</span>

    <span class="k">for</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">annotations</span><span class="p">):</span>

        <span class="n">taxon_id</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">taxon_id</span>

        <span class="n">uid</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">uid</span>

        <span class="n">sample_coverage</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">sample_coverage</span>
        <span class="n">annotation</span><span class="o">.</span><span class="n">add_exp_syn_count</span><span class="p">(</span><span class="n">seqs</span><span class="p">[</span><span class="n">annotation</span><span class="o">.</span><span class="n">seq_id</span><span class="p">])</span>

        <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">sample_coverage</span><span class="p">:</span>
            <span class="n">snp_data</span><span class="p">[</span><span class="n">sample</span><span class="p">][</span><span class="n">uid</span><span class="p">]</span> <span class="o">=</span> <span class="n">GeneSNP</span><span class="p">(</span>
                <span class="n">uid</span><span class="o">=</span><span class="n">uid</span><span class="p">,</span>
                <span class="n">gene_id</span><span class="o">=</span><span class="n">annotation</span><span class="o">.</span><span class="n">gene_id</span><span class="p">,</span>
                <span class="n">taxon_id</span><span class="o">=</span><span class="n">taxon_id</span><span class="p">,</span>
                <span class="n">exp_syn</span><span class="o">=</span><span class="n">annotation</span><span class="o">.</span><span class="n">exp_syn</span><span class="p">,</span>
                <span class="n">exp_nonsyn</span><span class="o">=</span><span class="n">annotation</span><span class="o">.</span><span class="n">exp_nonsyn</span><span class="p">,</span>
                <span class="n">coverage</span><span class="o">=</span><span class="n">sample_coverage</span><span class="p">[</span><span class="n">sample</span><span class="p">],</span>
            <span class="p">)</span>

    <span class="k">return</span> <span class="n">snp_data</span></div>


<div class="viewcode-block" id="init_count_set_sample_files"><a class="viewcode-back" href="../../../api/mgkit.workflow.pnps_gen.html#mgkit.workflow.pnps_gen.init_count_set_sample_files">[docs]</a><span class="k">def</span> <span class="nf">init_count_set_sample_files</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">seqs</span><span class="p">,</span> <span class="n">cov_files</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    .. versionadded:: 0.5.7</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Init data structures&quot;</span><span class="p">)</span>

    <span class="n">snp_data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
        <span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="p">{})</span> <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">cov_files</span>
    <span class="p">)</span>

    <span class="n">annotations</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">(</span><span class="o">*</span><span class="n">annotations</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

    <span class="k">for</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Adding Expected Values&quot;</span><span class="p">):</span>
        <span class="n">annotation</span><span class="o">.</span><span class="n">add_exp_syn_count</span><span class="p">(</span><span class="n">seqs</span><span class="p">[</span><span class="n">annotation</span><span class="o">.</span><span class="n">seq_id</span><span class="p">])</span>

    <span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">cov_file</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">cov_files</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">desc</span><span class="o">=</span><span class="s1">&#39;Adding Coverage&#39;</span><span class="p">):</span>
        <span class="n">cov_info</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="n">dictionary</span><span class="o">.</span><span class="n">text_to_dict</span><span class="p">(</span><span class="n">cov_file</span><span class="p">,</span> <span class="n">value_func</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">skip_empty</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">skip_comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">:</span>

            <span class="n">taxon_id</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">taxon_id</span>
            <span class="n">uid</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">uid</span>

            <span class="n">snp_data</span><span class="p">[</span><span class="n">sample</span><span class="p">][</span><span class="n">uid</span><span class="p">]</span> <span class="o">=</span> <span class="n">GeneSNP</span><span class="p">(</span>
                <span class="n">uid</span><span class="o">=</span><span class="n">uid</span><span class="p">,</span>
                <span class="n">gene_id</span><span class="o">=</span><span class="n">annotation</span><span class="o">.</span><span class="n">gene_id</span><span class="p">,</span>
                <span class="n">taxon_id</span><span class="o">=</span><span class="n">taxon_id</span><span class="p">,</span>
                <span class="n">exp_syn</span><span class="o">=</span><span class="n">annotation</span><span class="o">.</span><span class="n">exp_syn</span><span class="p">,</span>
                <span class="n">exp_nonsyn</span><span class="o">=</span><span class="n">annotation</span><span class="o">.</span><span class="n">exp_nonsyn</span><span class="p">,</span>
                <span class="n">coverage</span><span class="o">=</span><span class="n">cov_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">uid</span><span class="p">,</span> <span class="mf">0.</span><span class="p">),</span>
            <span class="p">)</span>

    <span class="k">return</span> <span class="n">snp_data</span></div>

<div class="viewcode-block" id="check_snp_in_set"><a class="viewcode-back" href="../../../api/mgkit.workflow.pnps_gen.html#mgkit.workflow.pnps_gen.check_snp_in_set">[docs]</a><span class="k">def</span> <span class="nf">check_snp_in_set</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">change</span><span class="p">,</span> <span class="n">annotations</span><span class="p">,</span> <span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Used by :func:`parse_vcf` to check if a SNP</span>

<span class="sd">    :param iterable samples: list of samples that contain the SNP</span>
<span class="sd">    :param dict snp_data: dictionary from :func:`init_count_set` with per</span>
<span class="sd">        sample SNPs information</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">pos</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">annotation</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">is_syn</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">change</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
            <span class="n">snp_type</span> <span class="o">=</span> <span class="n">SNPType</span><span class="o">.</span><span class="n">syn</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">snp_type</span> <span class="o">=</span> <span class="n">SNPType</span><span class="o">.</span><span class="n">nonsyn</span>

        <span class="n">uid</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">uid</span>
        <span class="n">rel_pos</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get_relative_pos</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">:</span>
            <span class="n">snp_data</span><span class="p">[</span><span class="n">sample</span><span class="p">][</span><span class="n">uid</span><span class="p">]</span><span class="o">.</span><span class="n">add_snp</span><span class="p">(</span><span class="n">rel_pos</span><span class="p">,</span> <span class="n">change</span><span class="p">,</span> <span class="n">snp_type</span><span class="o">=</span><span class="n">snp_type</span><span class="p">)</span></div>


<div class="viewcode-block" id="parse_vcf"><a class="viewcode-back" href="../../../api/mgkit.workflow.pnps_gen.html#mgkit.workflow.pnps_gen.parse_vcf">[docs]</a><span class="k">def</span> <span class="nf">parse_vcf</span><span class="p">(</span><span class="n">vcf_handle</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">,</span> <span class="n">annotations</span><span class="p">,</span> <span class="n">seqs</span><span class="p">,</span> <span class="n">min_qual</span><span class="p">,</span> <span class="n">min_reads</span><span class="p">,</span> <span class="n">min_freq</span><span class="p">,</span> 
                <span class="n">sample_ids</span><span class="p">,</span> <span class="n">num_lines</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>

    <span class="c1"># total number of SNPs accepted</span>
    <span class="n">accepted_snps</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># number of SNPs skipped for low depth</span>
    <span class="n">skip_dp</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># number of SNPs skipped for low allele frequency</span>
    <span class="n">skip_af</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># number of SNPs skipped for low quality</span>
    <span class="n">skip_qual</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># indels</span>
    <span class="n">skip_indels</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">line_no</span><span class="p">,</span> <span class="n">vcf_record</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">vcf_handle</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">CHROM</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">is_indel</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">&#39;DP&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">min_reads</span><span class="p">:</span>
            <span class="c1"># not enough reads (depth) for the SNP</span>
            <span class="n">skip_dp</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">QUAL</span> <span class="o">&lt;</span> <span class="n">min_qual</span><span class="p">:</span>
            <span class="c1"># low quality SNP</span>
            <span class="n">skip_qual</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">continue</span>

        <span class="n">alleles_freq</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">ALT</span><span class="p">),</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">aaf</span><span class="p">))</span>
        <span class="c1"># Used to keep track of the presence of an allele on which sample</span>
        <span class="n">alleles_sample</span> <span class="o">=</span> <span class="p">{</span>
            <span class="nb">str</span><span class="p">(</span><span class="n">allele</span><span class="p">):</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">ALT</span>
        <span class="p">}</span>

        <span class="k">for</span> <span class="n">sample_info</span> <span class="ow">in</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">samples</span><span class="p">:</span>
            <span class="c1"># equivalent to GT=. no call was made</span>
            <span class="k">if</span> <span class="n">sample_info</span><span class="o">.</span><span class="n">gt_bases</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="c1"># Not considering phases/unphased</span>
            <span class="c1"># but we really are looking at haployd</span>
            <span class="c1"># stress in documentation and examples</span>
            <span class="n">sample_info_gt</span> <span class="o">=</span> <span class="n">sample_info</span><span class="o">.</span><span class="n">gt_bases</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;!&#39;</span><span class="p">,</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s1">&#39;/&#39;</span> <span class="ow">in</span> <span class="n">sample_info_gt</span><span class="p">:</span>
                <span class="n">sample_info_gt</span> <span class="o">=</span> <span class="n">sample_info_gt</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span>
            <span class="c1"># Haployd</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">sample_info_gt</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample_info_gt</span><span class="p">]</span>

            <span class="k">for</span> <span class="n">change</span> <span class="ow">in</span> <span class="n">sample_info_gt</span><span class="p">:</span>
                <span class="c1"># if it&#39;s the reference continue</span>
                <span class="k">if</span> <span class="n">change</span> <span class="o">==</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">REF</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">allele_freq</span> <span class="o">=</span> <span class="n">alleles_freq</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">change</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">allele_freq</span> <span class="o">&lt;</span> <span class="n">min_freq</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">alleles_sample</span><span class="p">[</span><span class="n">change</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">[</span><span class="n">sample_info</span><span class="o">.</span><span class="n">sample</span><span class="p">])</span>

        <span class="n">seq</span> <span class="o">=</span> <span class="n">seqs</span><span class="p">[</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">CHROM</span><span class="p">]</span>
        <span class="n">ann_seq</span> <span class="o">=</span> <span class="n">annotations</span><span class="p">[</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">CHROM</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">change</span><span class="p">,</span> <span class="n">samples</span> <span class="ow">in</span> <span class="n">alleles_sample</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">samples</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">check_snp_in_set</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">,</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">POS</span><span class="p">,</span>
                             <span class="n">change</span><span class="p">,</span> <span class="n">ann_seq</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>
            <span class="n">accepted_snps</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">verbose</span> <span class="ow">and</span> <span class="p">(</span><span class="n">line_no</span> <span class="o">%</span> <span class="p">(</span><span class="n">num_lines</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
            <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Line </span><span class="si">%d</span><span class="s2">, SNPs passed </span><span class="si">%d</span><span class="s2">; skipped for: qual </span><span class="si">%d</span><span class="s2">, &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;depth </span><span class="si">%d</span><span class="s2">, freq </span><span class="si">%d</span><span class="s2">, indels </span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">line_no</span><span class="p">,</span> <span class="n">accepted_snps</span><span class="p">,</span> <span class="n">skip_qual</span><span class="p">,</span> <span class="n">skip_dp</span><span class="p">,</span> <span class="n">skip_af</span><span class="p">,</span> <span class="n">skip_indels</span>
            <span class="p">)</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Finished parsing, SNPs passed </span><span class="si">%d</span><span class="s2">; skipped for: qual </span><span class="si">%d</span><span class="s2">, &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;depth </span><span class="si">%d</span><span class="s2">, freq </span><span class="si">%d</span><span class="s2">, indels </span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">accepted_snps</span><span class="p">,</span> <span class="n">skip_qual</span><span class="p">,</span> <span class="n">skip_dp</span><span class="p">,</span> <span class="n">skip_af</span><span class="p">,</span> <span class="n">skip_indels</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">accepted_snps</span><span class="p">,</span> <span class="n">skip_qual</span><span class="p">,</span> <span class="n">skip_dp</span><span class="p">,</span> <span class="n">skip_af</span><span class="p">,</span> <span class="n">skip_indels</span></div>


<div class="viewcode-block" id="save_data"><a class="viewcode-back" href="../../../api/mgkit.workflow.pnps_gen.html#mgkit.workflow.pnps_gen.save_data">[docs]</a><span class="k">def</span> <span class="nf">save_data</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Pickle data structures to the disk.</span>

<span class="sd">    :param str output_file: base name for pickle files</span>
<span class="sd">    :param dict snp_data: dictionary from :func:`init_count_set` with per</span>
<span class="sd">        sample SNPs information</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Saving sample SNPs to </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">snp_data</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span></div>


<span class="nd">@main</span><span class="o">.</span><span class="n">command</span><span class="p">(</span><span class="s1">&#39;vcf&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;&quot;&quot;parse a VCF file and a GFF file to produce the</span>
<span class="s2">                data used for `pnps-gen`</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-v&#39;</span><span class="p">,</span> <span class="s1">&#39;--verbose&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-ft&#39;</span><span class="p">,</span> <span class="s1">&#39;--feature&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;CDS&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">STRING</span><span class="p">,</span>
              <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Feature to use in the GFF file&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="s1">&#39;--gff-file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;rb&#39;</span><span class="p">),</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;GFF file to use&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-a&#39;</span><span class="p">,</span> <span class="s1">&#39;--fasta-file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;rb&#39;</span><span class="p">),</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Reference file (FASTA) for the GFF&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-q&#39;</span><span class="p">,</span> <span class="s1">&#39;--min-qual&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">INT</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Minimum quality for SNPs (Phred score)&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-f&#39;</span><span class="p">,</span> <span class="s1">&#39;--min-freq&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=.</span><span class="mi">01</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">FLOAT</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Minimum allele frequency&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="s1">&#39;--min-reads&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">INT</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Minimum number of reads to accept the SNP&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-m&#39;</span><span class="p">,</span> <span class="s1">&#39;--sample-ids&#39;</span><span class="p">,</span> <span class="n">multiple</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
              <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">STRING</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;&#39;&#39;the ids of the samples used in the analysis,</span>
<span class="s1">              must be the same as in the GFF file&#39;&#39;&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-n&#39;</span><span class="p">,</span> <span class="s1">&#39;--num-lines&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">10</span><span class="o">**</span><span class="mi">5</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">INT</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of VCF lines after which printing status&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">argument</span><span class="p">(</span><span class="s1">&#39;vcf-file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">),</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">argument</span><span class="p">(</span><span class="s1">&#39;output-file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;wb&#39;</span><span class="p">))</span>
<span class="k">def</span> <span class="nf">parse_command</span><span class="p">(</span><span class="n">verbose</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">gff_file</span><span class="p">,</span> <span class="n">fasta_file</span><span class="p">,</span> <span class="n">min_qual</span><span class="p">,</span> <span class="n">min_freq</span><span class="p">,</span>
                  <span class="n">min_reads</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span> <span class="n">num_lines</span><span class="p">,</span> <span class="n">vcf_file</span><span class="p">,</span> <span class="n">output_file</span><span class="p">):</span>

    <span class="n">mgkit</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">config_log</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span> <span class="k">if</span> <span class="n">verbose</span> <span class="k">else</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

    <span class="n">vcf_handle</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">Reader</span><span class="p">(</span><span class="n">fsock</span><span class="o">=</span><span class="n">vcf_file</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">vcf_handle</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">):</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">exit_script</span><span class="p">(</span><span class="s2">&quot;The number of sample names is wrong: VCF (</span><span class="si">{}</span><span class="s2">) -&gt; Passed (</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">vcf_handle</span><span class="o">.</span><span class="n">samples</span><span class="p">),</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)),</span> <span class="mi">1</span>
        <span class="p">)</span>

    <span class="c1"># Loads them as list because it&#39;s easier to init the data structure</span>
    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading annotations from GFF File, using feat_type: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">feature</span><span class="p">)</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">seq_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">test_names</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">for</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">gff</span><span class="o">.</span><span class="n">parse_gff</span><span class="p">(</span><span class="n">gff_file</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">feat_type</span> <span class="o">!=</span> <span class="n">feature</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">annotations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">annotation</span><span class="p">)</span>
        <span class="n">seq_ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">annotation</span><span class="o">.</span><span class="n">seq_id</span><span class="p">)</span>
        <span class="c1"># checks if sample names are wrong</span>
        <span class="k">if</span> <span class="n">test_names</span><span class="p">:</span>
            <span class="n">sample_names_gff</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">annotation</span><span class="o">.</span><span class="n">sample_coverage</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
            <span class="k">if</span> <span class="n">sample_names_gff</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">):</span>
                <span class="n">utils</span><span class="o">.</span><span class="n">exit_script</span><span class="p">(</span><span class="s2">&quot;Sample names are wrong: GFF (</span><span class="si">{}</span><span class="s2">) -&gt; Passed (</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_names_gff</span><span class="p">),</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)),</span> <span class="mi">2</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">test_names</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">annotations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sample_coverage</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">):</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">exit_script</span><span class="p">(</span><span class="s2">&quot;The number of sample names is wrong: VCF (</span><span class="si">{}</span><span class="s2">) -&gt; Passed (</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">vcf_handle</span><span class="o">.</span><span class="n">samples</span><span class="p">),</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)),</span> <span class="mi">1</span>
        <span class="p">)</span>

    <span class="n">sample_ids</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">vcf_handle</span><span class="o">.</span><span class="n">samples</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">))</span>
    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sample IDs from VCF to GFF: </span><span class="si">%r</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">)</span>

    <span class="n">seqs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">seq_id</span><span class="p">:</span> <span class="n">seq</span>
        <span class="k">for</span> <span class="n">seq_id</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">fasta</span><span class="o">.</span><span class="n">load_fasta</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">seq_id</span> <span class="ow">in</span> <span class="n">seq_ids</span>
    <span class="p">}</span>

    <span class="n">snp_data</span> <span class="o">=</span> <span class="n">init_count_set</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">seqs</span><span class="p">)</span>

    <span class="n">annotations</span> <span class="o">=</span> <span class="n">gff</span><span class="o">.</span><span class="n">group_annotations</span><span class="p">(</span>
        <span class="n">annotations</span><span class="p">,</span>
        <span class="n">key_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">seq_id</span>
    <span class="p">)</span>

    <span class="n">parse_vcf</span><span class="p">(</span><span class="n">vcf_handle</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">,</span> <span class="n">annotations</span><span class="p">,</span> <span class="n">seqs</span><span class="p">,</span>
              <span class="n">min_qual</span><span class="p">,</span> <span class="n">min_reads</span><span class="p">,</span> <span class="n">min_freq</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span> <span class="n">num_lines</span><span class="p">)</span>

    <span class="n">save_data</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">)</span>


<span class="nd">@main</span><span class="o">.</span><span class="n">command</span><span class="p">(</span><span class="s1">&#39;vcf_alt&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;&quot;&quot;parse a VCF file and a GFF file to produce the</span>
<span class="s2">                data used for `pnps-gen`, uses file a list for sample coverage</span>
<span class="s2">                instead of taking information from the GFF file</span>
<span class="s2">                &quot;&quot;&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-v&#39;</span><span class="p">,</span> <span class="s1">&#39;--verbose&#39;</span><span class="p">,</span> <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-ft&#39;</span><span class="p">,</span> <span class="s1">&#39;--feature&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;CDS&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">STRING</span><span class="p">,</span>
              <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Feature to use in the GFF file&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="s1">&#39;--gff-file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;rb&#39;</span><span class="p">),</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;GFF file to use&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-a&#39;</span><span class="p">,</span> <span class="s1">&#39;--fasta-file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;rb&#39;</span><span class="p">),</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Reference file (FASTA) for the GFF&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-q&#39;</span><span class="p">,</span> <span class="s1">&#39;--min-qual&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">INT</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Minimum quality for SNPs (Phred score)&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-f&#39;</span><span class="p">,</span> <span class="s1">&#39;--min-freq&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=.</span><span class="mi">01</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">FLOAT</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Minimum allele frequency&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="s1">&#39;--min-reads&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">INT</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Minimum number of reads to accept the SNP&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-n&#39;</span><span class="p">,</span> <span class="s1">&#39;--num-lines&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">10</span><span class="o">**</span><span class="mi">5</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">INT</span><span class="p">,</span> <span class="n">show_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of VCF lines after which printing status&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-l&#39;</span><span class="p">,</span> <span class="s1">&#39;--sample-file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">),</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;File with list of coverage files and sample names (TAB separated)&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="s1">&#39;--file-list&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">),</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;File with list of VCF files (one per line)&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s1">&#39;-u&#39;</span><span class="p">,</span> <span class="s1">&#39;--uid-map&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">),</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
              <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Only load annotations from a specific map file&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">argument</span><span class="p">(</span><span class="s1">&#39;vcf-file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">file_okay</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">readable</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">argument</span><span class="p">(</span><span class="s1">&#39;output-file&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s1">&#39;wb&#39;</span><span class="p">))</span>
<span class="k">def</span> <span class="nf">parse_alt_command</span><span class="p">(</span><span class="n">verbose</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">gff_file</span><span class="p">,</span> <span class="n">fasta_file</span><span class="p">,</span> <span class="n">min_qual</span><span class="p">,</span> <span class="n">min_freq</span><span class="p">,</span>
                  <span class="n">min_reads</span><span class="p">,</span> <span class="n">num_lines</span><span class="p">,</span> <span class="n">sample_file</span><span class="p">,</span> <span class="n">file_list</span><span class="p">,</span> <span class="n">uid_map</span><span class="p">,</span> <span class="n">vcf_file</span><span class="p">,</span> <span class="n">output_file</span><span class="p">):</span>

    <span class="n">mgkit</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">config_log</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span> <span class="k">if</span> <span class="n">verbose</span> <span class="k">else</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">file_list</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">vcf_handles</span> <span class="o">=</span> <span class="p">[</span><span class="n">vcf</span><span class="o">.</span><span class="n">Reader</span><span class="p">(</span><span class="n">fsock</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">))]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">vcf_handles</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">vcf</span><span class="o">.</span><span class="n">Reader</span><span class="p">(</span><span class="n">fsock</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span> <span class="s1">&#39;r&#39;</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file_list</span>
        <span class="p">]</span>

    <span class="n">cov_files</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">sample_file</span><span class="p">:</span>
        <span class="n">sample_id</span><span class="p">,</span> <span class="n">cov_file</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">cov_files</span><span class="p">[</span><span class="n">sample_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">cov_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">vcf_handles</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cov_files</span><span class="p">):</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">exit_script</span><span class="p">(</span><span class="s2">&quot;The number of sample names is wrong: VCF (</span><span class="si">{}</span><span class="s2">) -&gt; Passed (</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">vcf_handles</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">samples</span><span class="p">),</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cov_files</span><span class="p">)),</span> <span class="mi">1</span>
        <span class="p">)</span>
    
    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;VCF sample ID -&gt; Sample ID&quot;</span><span class="p">)</span>
    <span class="n">sample_ids</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">vcf_handles</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">samples</span><span class="p">,</span> <span class="n">cov_files</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">vcf_sample</span><span class="p">,</span> <span class="n">sample_id</span> <span class="ow">in</span> <span class="n">sample_ids</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> -&gt; </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">vcf_sample</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">uid_map</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Loading uid from file </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">uid_map</span><span class="p">)</span>
        <span class="n">uid_map</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
            <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">uid_map</span>
        <span class="p">)</span>

    <span class="c1"># Loads them as list because it&#39;s easier to init the data structure</span>
    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading annotations from GFF File, using feat_type: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">feature</span><span class="p">)</span>
    <span class="n">ann_iterator</span> <span class="o">=</span> <span class="n">gff</span><span class="o">.</span><span class="n">parse_gff</span><span class="p">(</span><span class="n">gff_file</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">uid_map</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ann_iterator</span> <span class="o">=</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">uid</span> <span class="ow">in</span> <span class="n">uid_map</span><span class="p">,</span> <span class="n">ann_iterator</span><span class="p">)</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="n">gff</span><span class="o">.</span><span class="n">group_annotations</span><span class="p">(</span>
        <span class="p">(</span>
            <span class="n">annotation</span>
            <span class="k">for</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">ann_iterator</span>
            <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">feat_type</span> <span class="o">==</span> <span class="n">feature</span>
        <span class="p">),</span>
        <span class="n">key_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">seq_id</span>
    <span class="p">)</span>

    <span class="n">seqs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">seq_id</span><span class="p">:</span> <span class="n">seq</span>
        <span class="k">for</span> <span class="n">seq_id</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">fasta</span><span class="o">.</span><span class="n">load_fasta</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">seq_id</span> <span class="ow">in</span> <span class="n">annotations</span>
    <span class="p">}</span>

    <span class="n">snp_data</span> <span class="o">=</span> <span class="n">init_count_set_sample_files</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">seqs</span><span class="p">,</span> <span class="n">cov_files</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">vcf_handles</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">vcf_handles</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">vcf_handles</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Parsing VCF Files&quot;</span><span class="p">)</span>
    
    <span class="n">accepted_snps</span><span class="p">,</span> <span class="n">skip_qual</span><span class="p">,</span> <span class="n">skip_dp</span><span class="p">,</span> <span class="n">skip_af</span><span class="p">,</span> <span class="n">skip_indels</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">vcf_handle</span> <span class="ow">in</span> <span class="n">vcf_handles</span><span class="p">:</span>
        <span class="n">accepted_snps</span><span class="p">,</span> <span class="n">skip_qual</span><span class="p">,</span> <span class="n">skip_dp</span><span class="p">,</span> <span class="n">skip_af</span><span class="p">,</span> <span class="n">skip_indels</span> <span class="o">=</span> <span class="n">parse_vcf</span><span class="p">(</span>
                  <span class="n">vcf_handle</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">,</span> <span class="n">annotations</span><span class="p">,</span> <span class="n">seqs</span><span class="p">,</span>
                  <span class="n">min_qual</span><span class="p">,</span> <span class="n">min_reads</span><span class="p">,</span> <span class="n">min_freq</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span> <span class="n">num_lines</span><span class="p">,</span>
                  <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">vcf_handles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">accepted_snps</span> <span class="o">+=</span> <span class="n">accepted_snps</span>
        <span class="n">skip_qual</span> <span class="o">+=</span> <span class="n">skip_qual</span>
        <span class="n">skip_dp</span> <span class="o">+=</span> <span class="n">skip_dp</span>
        <span class="n">skip_af</span> <span class="o">+=</span> <span class="n">skip_af</span>
        <span class="n">skip_indels</span> <span class="o">+=</span> <span class="n">skip_indels</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Finished parsing, SNPs passed </span><span class="si">%d</span><span class="s2">; skipped for: qual </span><span class="si">%d</span><span class="s2">, &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;depth </span><span class="si">%d</span><span class="s2">, freq </span><span class="si">%d</span><span class="s2">, indels </span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">accepted_snps</span><span class="p">,</span> <span class="n">skip_qual</span><span class="p">,</span> <span class="n">skip_dp</span><span class="p">,</span> <span class="n">skip_af</span><span class="p">,</span> <span class="n">skip_indels</span>
        <span class="p">)</span>

    <span class="n">save_data</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">)</span>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">MGKit: Metagenomic framework</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../library.html">Library Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changes.html">Changes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  <li><a href="../../mgkit.html">mgkit</a><ul>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2013-2020, Francesco Rubino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>