

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>mgkit.workflow.snp_parser &mdash; MGKit: Metagenomic framework 0.5.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/graphviz.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> MGKit: Metagenomic framework
          

          
          </a>

          
            
            
              <div class="version">
                0.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../library.html">Library Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changes.html">Changes</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">MGKit: Metagenomic framework</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
          <li><a href="../../mgkit.html">mgkit</a> &raquo;</li>
        
      <li>mgkit.workflow.snp_parser</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for mgkit.workflow.snp_parser</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This script parses results of SNPs analysis from any tool for SNP calling [#]_</span>
<span class="sd">and integrates them into a format that can be later used for other scripts in</span>
<span class="sd">the pipeline.</span>

<span class="sd">It integrates coverage and expected number of syn/nonsyn change and taxonomy</span>
<span class="sd">from a GFF file, SNP data from a VCF file.</span>

<span class="sd">.. note::</span>

<span class="sd">    The script accept gzipped VCF files</span>

<span class="sd">.. [#] GATK pipeline was tested, but it is possible to use samtools and</span>
<span class="sd">    bcftools</span>

<span class="sd">Changes</span>
<span class="sd">*******</span>

<span class="sd">.. versionchanged:: 0.2.1</span>
<span class="sd">    added *-s* option for VCF files generated using bcftools</span>

<span class="sd">.. versionchanged:: 0.1.16</span>
<span class="sd">    reworkked internals and removed SNPDat, syn/nonsyn evaluation is internal</span>

<span class="sd">.. versionchanged:: 0.1.13</span>
<span class="sd">    reworked the internals and the classes used, including options -m and -s</span>


<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">division</span>
<span class="kn">from</span> <span class="nn">builtins</span> <span class="kn">import</span> <span class="nb">zip</span>
<span class="kn">import</span> <span class="nn">HTSeq</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">from</span> <span class="nn">.</span> <span class="kn">import</span> <span class="n">utils</span>
<span class="kn">from</span> <span class="nn">..io</span> <span class="kn">import</span> <span class="n">gff</span><span class="p">,</span> <span class="n">compressed_handle</span><span class="p">,</span> <span class="n">fasta</span>
<span class="kn">from</span> <span class="nn">..</span> <span class="kn">import</span> <span class="n">logger</span>
<span class="kn">from</span> <span class="nn">..snps.classes</span> <span class="kn">import</span> <span class="n">GeneSNP</span><span class="p">,</span> <span class="n">SNPType</span>
<span class="kn">import</span> <span class="nn">pickle</span>


<span class="n">LOG</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<div class="viewcode-block" id="set_parser"><a class="viewcode-back" href="../../../api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.set_parser">[docs]</a><span class="k">def</span> <span class="nf">set_parser</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sets command line arguments parser</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s1">&#39;SNPs analysis, requires a vcf file and SNPDat results&#39;</span><span class="p">,</span>
        <span class="n">formatter_class</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentDefaultsHelpFormatter</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-o&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--output-file&#39;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s1">&#39;snp_data.pickle&#39;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">FileType</span><span class="p">(</span><span class="s1">&#39;wb&#39;</span><span class="p">),</span>
        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Ouput file&#39;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-q&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--min-qual&#39;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store&#39;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Minimum SNP quality (Phred score)&#39;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-f&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--min-freq&#39;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store&#39;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Minimum allele frequency&#39;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-r&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--min-reads&#39;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store&#39;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Minimum number of reads to accept the SNP&#39;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-g&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--gff-file&#39;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">FileType</span><span class="p">(</span><span class="s1">&#39;rb&#39;</span><span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store&#39;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;GFF file with annotations&#39;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-p&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--vcf-file&#39;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">FileType</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store&#39;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Merged VCF file&#39;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-a&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--reference&#39;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">FileType</span><span class="p">(</span><span class="s1">&#39;rb&#39;</span><span class="p">),</span>
        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Fasta file with the GFF Reference&#39;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-m&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--samples-id&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s1">&#39;append&#39;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;the ids of the samples used in the analysis&#39;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-c&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--cov-suff&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store&#39;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s1">&#39;_cov&#39;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Per sample coverage suffix in the GFF&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-s&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--bcftools-vcf&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_true&#39;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;bcftools call was used to produce the VCF file&quot;</span>
    <span class="p">)</span>

    <span class="n">utils</span><span class="o">.</span><span class="n">add_basic_options</span><span class="p">(</span><span class="n">parser</span><span class="p">,</span> <span class="n">manual</span><span class="o">=</span><span class="vm">__doc__</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span></div>


<div class="viewcode-block" id="init_count_set"><a class="viewcode-back" href="../../../api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.init_count_set">[docs]</a><span class="k">def</span> <span class="nf">init_count_set</span><span class="p">(</span><span class="n">annotations</span><span class="p">):</span>
    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Init data structures&quot;</span><span class="p">)</span>

    <span class="n">samples</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">annotations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sample_coverage</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

    <span class="n">snp_data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
        <span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="p">{})</span> <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">samples</span>
    <span class="p">)</span>

    <span class="k">for</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">:</span>

        <span class="n">taxon_id</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">taxon_id</span>

        <span class="n">uid</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">uid</span>

        <span class="n">sample_coverage</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">sample_coverage</span>

        <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">sample_coverage</span><span class="p">:</span>
            <span class="n">snp_data</span><span class="p">[</span><span class="n">sample</span><span class="p">][</span><span class="n">uid</span><span class="p">]</span> <span class="o">=</span> <span class="n">GeneSNP</span><span class="p">(</span>
                <span class="n">uid</span><span class="o">=</span><span class="n">uid</span><span class="p">,</span>
                <span class="n">gene_id</span><span class="o">=</span><span class="n">annotation</span><span class="o">.</span><span class="n">gene_id</span><span class="p">,</span>
                <span class="n">taxon_id</span><span class="o">=</span><span class="n">taxon_id</span><span class="p">,</span>
                <span class="n">exp_syn</span><span class="o">=</span><span class="n">annotation</span><span class="o">.</span><span class="n">exp_syn</span><span class="p">,</span>
                <span class="n">exp_nonsyn</span><span class="o">=</span><span class="n">annotation</span><span class="o">.</span><span class="n">exp_nonsyn</span><span class="p">,</span>
                <span class="n">coverage</span><span class="o">=</span><span class="n">sample_coverage</span><span class="p">[</span><span class="n">sample</span><span class="p">],</span>
            <span class="p">)</span>

    <span class="k">return</span> <span class="n">snp_data</span></div>


<div class="viewcode-block" id="check_snp_in_set"><a class="viewcode-back" href="../../../api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.check_snp_in_set">[docs]</a><span class="k">def</span> <span class="nf">check_snp_in_set</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">change</span><span class="p">,</span> <span class="n">annotations</span><span class="p">,</span> <span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Used by :func:`parse_vcf` to check if a SNP</span>

<span class="sd">    :param iterable samples: list of samples that contain the SNP</span>
<span class="sd">    :param dict snp_data: dictionary from :func:`init_count_set` with per</span>
<span class="sd">        sample SNPs information</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">pos</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">annotation</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">is_syn</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">change</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
            <span class="n">snp_type</span> <span class="o">=</span> <span class="n">SNPType</span><span class="o">.</span><span class="n">syn</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">snp_type</span> <span class="o">=</span> <span class="n">SNPType</span><span class="o">.</span><span class="n">nonsyn</span>

        <span class="n">uid</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">uid</span>
        <span class="n">rel_pos</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get_relative_pos</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">:</span>
            <span class="n">snp_data</span><span class="p">[</span><span class="n">sample</span><span class="p">][</span><span class="n">uid</span><span class="p">]</span><span class="o">.</span><span class="n">add_snp</span><span class="p">(</span><span class="n">rel_pos</span><span class="p">,</span> <span class="n">change</span><span class="p">,</span> <span class="n">snp_type</span><span class="o">=</span><span class="n">snp_type</span><span class="p">)</span></div>


<div class="viewcode-block" id="parse_vcf"><a class="viewcode-back" href="../../../api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.parse_vcf">[docs]</a><span class="k">def</span> <span class="nf">parse_vcf</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">,</span> <span class="n">min_reads</span><span class="p">,</span> <span class="n">min_af</span><span class="p">,</span> <span class="n">min_qual</span><span class="p">,</span> <span class="n">annotations</span><span class="p">,</span>
              <span class="n">seqs</span><span class="p">,</span> <span class="n">options</span><span class="p">,</span> <span class="n">line_num</span><span class="o">=</span><span class="mi">100000</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse VCF file counts synonymous and non-synonymous SNPs</span>

<span class="sd">    :param file vcf_file: file handle to a VCF file</span>
<span class="sd">    :param dict snp_data: dictionary from :func:`init_count_set` with per</span>
<span class="sd">        sample SNPs information</span>
<span class="sd">    :param int min_reads: minimum number of reads to accept a SNP</span>
<span class="sd">    :param float min_af: minimum allele frequency to accept a SNP</span>
<span class="sd">    :param int min_qual: minimum quality (Phred score) to accept a SNP</span>
<span class="sd">    :param dict annotations: annotations grouped by their reference sequence</span>
<span class="sd">    :param dict seqs: reference sequences</span>
<span class="sd">    :param int line_num: the interval in number of lines at which progress</span>
<span class="sd">        will be printed</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">vcf_handle</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">VCF_Reader</span><span class="p">(</span><span class="n">compressed_handle</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">))</span>

    <span class="n">vcf_handle</span><span class="o">.</span><span class="n">parse_meta</span><span class="p">()</span>
    <span class="n">vcf_handle</span><span class="o">.</span><span class="n">make_info_dict</span><span class="p">()</span>

    <span class="c1"># total number of SNPs accepted</span>
    <span class="n">count_tot</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># number of SNPs skipped for low depth</span>
    <span class="n">skip_dp</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># number of SNPs skipped for low allele frequency</span>
    <span class="n">skip_af</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># number of SNPs skipped for low quality</span>
    <span class="n">skip_qual</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># indels</span>
    <span class="n">skip_indels</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">vcf_record</span> <span class="ow">in</span> <span class="n">vcf_handle</span><span class="p">:</span>
        <span class="c1"># the SNP is a sequence with no annotations</span>
        <span class="k">if</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="nb">float</span><span class="p">(</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">qual</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_qual</span><span class="p">:</span>
            <span class="c1"># low quality SNP</span>
            <span class="n">skip_qual</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">continue</span>

        <span class="c1"># unpack info records (needed for vcf_record.info to be a dictionary)</span>
        <span class="n">vcf_record</span><span class="o">.</span><span class="n">unpack_info</span><span class="p">(</span><span class="n">vcf_handle</span><span class="o">.</span><span class="n">infodict</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;INDEL&#39;</span><span class="p">]:</span>
            <span class="n">skip_indels</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;DP&#39;</span><span class="p">],</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">LOG</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;DP&#39;</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;DP&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">min_reads</span><span class="p">:</span>
            <span class="c1"># not enough reads (depth) for the SNP</span>
            <span class="n">skip_dp</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">continue</span>

        <span class="c1"># Samtools mpileup -&gt; bcftools call doesn&#39;t output the allele freq.</span>
        <span class="c1"># it can be calculated with AC/AN for each ALT nucleotide</span>
        <span class="c1"># checked on bfctools (roh command) manual</span>
        <span class="c1"># https://samtools.github.io/bcftools/bcftools.html</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">allele_freqs</span> <span class="o">=</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;AF&#39;</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;AC&#39;</span><span class="p">],</span> <span class="nb">list</span><span class="p">):</span>
                <span class="n">allele_freqs</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">AC</span> <span class="o">/</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;AN&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">AC</span> <span class="ow">in</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;AC&#39;</span><span class="p">]</span>
                <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">allele_freqs</span> <span class="o">=</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;AC&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;AN&#39;</span><span class="p">]</span>

        <span class="c1"># if the allele frequency is a single value, make it a list, so</span>
        <span class="c1"># the iteration below works anyway</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">allele_freqs</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="n">allele_freqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">allele_freqs</span><span class="p">]</span>

        <span class="c1"># alt is the nucleotidic change</span>
        <span class="n">iter_data</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="n">allele_freqs</span><span class="p">,</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">alt_index</span><span class="p">,</span> <span class="p">(</span><span class="n">allele_freq</span><span class="p">,</span> <span class="n">change</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">iter_data</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">allele_freq</span> <span class="o">&lt;</span> <span class="n">min_af</span><span class="p">:</span>
                <span class="c1"># the allele frequency for the SNP is too low, it&#39;ll be</span>
                <span class="c1"># skipped</span>
                <span class="n">skip_af</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">continue</span>

            <span class="c1"># the samples that contain the SNP is a string separated by &#39;-&#39;</span>
            <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">bcftools_vcf</span><span class="p">:</span>
                <span class="n">samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">sample_info</span> <span class="ow">in</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="c1"># prepare the genotype list, to make the comparison easier</span>
                    <span class="c1"># the genotype separator to &#39;/&#39; only, to use only one</span>
                    <span class="c1"># type of split</span>
                    <span class="n">sample_info_gt</span> <span class="o">=</span> <span class="n">sample_info</span><span class="p">[</span><span class="s1">&#39;GT&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">,</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
                    <span class="n">sample_info_gt</span> <span class="o">=</span> <span class="n">sample_info_gt</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">genotype</span> <span class="ow">in</span> <span class="n">sample_info_gt</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s1">&#39;.&#39;</span><span class="p">:</span>
                            <span class="k">continue</span>
                        <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="n">alt_index</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                            <span class="n">samples</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">sample_id</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">samples</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">sample</span>
                    <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;set&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
                <span class="p">]</span>
            <span class="n">check_snp_in_set</span><span class="p">(</span>
                <span class="n">samples</span><span class="p">,</span>
                <span class="n">snp_data</span><span class="p">,</span>
                <span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="o">.</span><span class="n">start</span><span class="p">,</span>
                <span class="n">change</span><span class="p">,</span>
                <span class="n">annotations</span><span class="p">[</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">],</span>
                <span class="n">seqs</span><span class="p">[</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="c1"># increase the total number of snps available</span>
            <span class="n">count_tot</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">vcf_handle</span><span class="o">.</span><span class="n">line_no</span> <span class="o">%</span> <span class="n">line_num</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Line </span><span class="si">%d</span><span class="s2">, SNPs passed </span><span class="si">%d</span><span class="s2">; skipped for: qual </span><span class="si">%d</span><span class="s2">, &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;depth </span><span class="si">%d</span><span class="s2">, freq </span><span class="si">%d</span><span class="s2">, indels </span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">vcf_handle</span><span class="o">.</span><span class="n">line_no</span><span class="p">,</span> <span class="n">count_tot</span><span class="p">,</span> <span class="n">skip_qual</span><span class="p">,</span> <span class="n">skip_dp</span><span class="p">,</span> <span class="n">skip_af</span><span class="p">,</span>
                <span class="n">skip_indels</span>
            <span class="p">)</span></div>


<div class="viewcode-block" id="save_data"><a class="viewcode-back" href="../../../api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.save_data">[docs]</a><span class="k">def</span> <span class="nf">save_data</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Pickle data structures to the disk.</span>

<span class="sd">    :param str output_file: base name for pickle files</span>
<span class="sd">    :param dict snp_data: dictionary from :func:`init_count_set` with per</span>
<span class="sd">        sample SNPs information</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">LOG</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Saving sample SNPs to </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">snp_data</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span></div>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../api/mgkit.workflow.snp_parser.html#mgkit.workflow.snp_parser.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="s2">&quot;Main function&quot;</span>
    <span class="n">options</span> <span class="o">=</span> <span class="n">set_parser</span><span class="p">()</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="c1"># configs log and set log level</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">config_log</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>

    <span class="n">seqs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">load_fasta</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">reference</span><span class="p">))</span>

    <span class="c1"># Loads them as list because it&#39;s easier to init the data structure</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gff</span><span class="o">.</span><span class="n">parse_gff</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">gff_file</span><span class="p">))</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annotations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sample_coverage</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">samples_id</span><span class="p">):</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">exit_script</span><span class="p">(</span>
            <span class="s2">&quot;Coverage information was not found for all samples&quot;</span><span class="p">,</span> <span class="mi">2</span>
        <span class="p">)</span>

    <span class="n">snp_data</span> <span class="o">=</span> <span class="n">init_count_set</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span>

    <span class="c1"># Group annotations by their reference sequence</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="n">gff</span><span class="o">.</span><span class="n">group_annotations</span><span class="p">(</span>
        <span class="n">annotations</span><span class="p">,</span>
        <span class="n">key_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">seq_id</span>
    <span class="p">)</span>

    <span class="n">parse_vcf</span><span class="p">(</span>
        <span class="n">options</span><span class="o">.</span><span class="n">vcf_file</span><span class="p">,</span>
        <span class="n">snp_data</span><span class="p">,</span>
        <span class="n">options</span><span class="o">.</span><span class="n">min_reads</span><span class="p">,</span>
        <span class="n">options</span><span class="o">.</span><span class="n">min_freq</span><span class="p">,</span>
        <span class="n">options</span><span class="o">.</span><span class="n">min_qual</span><span class="p">,</span>
        <span class="n">annotations</span><span class="p">,</span>
        <span class="n">seqs</span><span class="p">,</span>
        <span class="n">options</span>
    <span class="p">)</span>

    <span class="n">save_data</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">output_file</span><span class="p">,</span> <span class="n">snp_data</span><span class="p">)</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2013-2015, Francesco Rubino

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>