

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>mgkit.workflow.add_gff_info module &mdash; MGKit: Metagenomic framework 0.5.5 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="mgkit.workflow.blast2gff module" href="mgkit.workflow.blast2gff.html" />
    <link rel="prev" title="mgkit.workflow package" href="mgkit.workflow.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> MGKit: Metagenomic framework
          

          
          </a>

          
            
            
              <div class="version">
                0.5.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">Library Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mgkit.html">mgkit package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.align.html">mgkit.align module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.consts.html">mgkit.consts module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.html">mgkit.counts package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.func.html">mgkit.counts.func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.glm.html">mgkit.counts.glm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.scaling.html">mgkit.counts.scaling module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.html">mgkit.db package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.dbm.html">mgkit.db.dbm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.mongo.html">mgkit.db.mongo module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.html">mgkit.filter package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.common.html">mgkit.filter.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.gff.html">mgkit.filter.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.lists.html">mgkit.filter.lists module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.reads.html">mgkit.filter.reads module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.taxon.html">mgkit.filter.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.graphs.html">mgkit.graphs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.html">mgkit.io package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.blast.html">mgkit.io.blast module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fasta.html">mgkit.io.fasta module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fastq.html">mgkit.io.fastq module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.gff.html">mgkit.io.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.glimmer.html">mgkit.io.glimmer module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.snpdat.html">mgkit.io.snpdat module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.uniprot.html">mgkit.io.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.utils.html">mgkit.io.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.kegg.html">mgkit.kegg module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.logger.html">mgkit.logger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.html">mgkit.mappings package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.cazy.html">mgkit.mappings.cazy module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.eggnog.html">mgkit.mappings.eggnog module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.enzyme.html">mgkit.mappings.enzyme module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.go.html">mgkit.mappings.go module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.pandas_map.html">mgkit.mappings.pandas_map module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.taxon.html">mgkit.mappings.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.utils.html">mgkit.mappings.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.html">mgkit.net package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.embl.html">mgkit.net.embl module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.pfam.html">mgkit.net.pfam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.uniprot.html">mgkit.net.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.utils.html">mgkit.net.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.html">mgkit.plots package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.abund.html">mgkit.plots.abund module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.boxplot.html">mgkit.plots.boxplot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.colors.html">mgkit.plots.colors module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.heatmap.html">mgkit.plots.heatmap module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.unused.html">mgkit.plots.unused module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.utils.html">mgkit.plots.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.simple_cache.html">mgkit.simple_cache module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.html">mgkit.snps package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.classes.html">mgkit.snps.classes module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.conv_func.html">mgkit.snps.conv_func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.filter.html">mgkit.snps.filter module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.funcs.html">mgkit.snps.funcs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.mapper.html">mgkit.snps.mapper module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.taxon.html">mgkit.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.html">mgkit.utils package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.common.html">mgkit.utils.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.dictionary.html">mgkit.utils.dictionary module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.sequence.html">mgkit.utils.sequence module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.trans_tables.html">mgkit.utils.trans_tables module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.html">mgkit.workflow package</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">mgkit.workflow.add_gff_info module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#uniprot-command">Uniprot Command</a></li>
<li class="toctree-l3"><a class="reference internal" href="#coverage-command">Coverage Command</a></li>
<li class="toctree-l3"><a class="reference internal" href="#adding-coverage-from-samtools-depth">Adding Coverage from samtools depth</a></li>
<li class="toctree-l3"><a class="reference internal" href="#uniprot-offline-mappings">Uniprot Offline Mappings</a></li>
<li class="toctree-l3"><a class="reference internal" href="#kegg-information">Kegg Information</a></li>
<li class="toctree-l3"><a class="reference internal" href="#expected-aminoacidic-changes">Expected Aminoacidic Changes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#adding-count-data">Adding Count Data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#adding-taxonomy-from-a-table">Adding Taxonomy from a Table</a></li>
<li class="toctree-l3"><a class="reference internal" href="#adding-information-from-pfam">Adding information from Pfam</a></li>
<li class="toctree-l3"><a class="reference internal" href="#changes">Changes</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.blast2gff.html">mgkit.workflow.blast2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.extract_gff_info.html">mgkit.workflow.extract_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fasta_utils.html">mgkit.workflow.fasta_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fastq_utils.html">mgkit.workflow.fastq_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.filter_gff.html">mgkit.workflow.filter_gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.hmmer2gff.html">mgkit.workflow.hmmer2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.json2gff.html">mgkit.workflow.json2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.sampling_utils.html">mgkit.workflow.sampling_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.snp_parser.html">mgkit.workflow.snp_parser module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.taxon_utils.html">mgkit.workflow.taxon_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.utils.html">mgkit.workflow.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html">mgkit</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MGKit: Metagenomic framework</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../library.html">Library Reference</a> &raquo;</li>
        
          <li><a href="mgkit.html">mgkit package</a> &raquo;</li>
        
          <li><a href="mgkit.workflow.html">mgkit.workflow package</a> &raquo;</li>
        
      <li>mgkit.workflow.add_gff_info module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/api/mgkit.workflow.add_gff_info.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="module-mgkit.workflow.add_gff_info">
<span id="mgkit-workflow-add-gff-info-module"></span><h1>mgkit.workflow.add_gff_info module<a class="headerlink" href="#module-mgkit.workflow.add_gff_info" title="Permalink to this headline">¶</a></h1>
<p>Add more information to GFF annotations: gene mappings, coverage, taxonomy,
etc..</p>
<div class="section" id="uniprot-command">
<h2>Uniprot Command<a class="headerlink" href="#uniprot-command" title="Permalink to this headline">¶</a></h2>
<p>If the <em>gene_id</em> of an annotation is a Uniprot ID, the script queries Uniprot
for the requested information. At the moment the information that can be added
is the taxon_id, taxon_name, lineage and mapping to EC, KO, eggNOG IDs.</p>
<p>It’s also possible to add mappings to other databases using the <em>-m</em> option
with the correct identifier for the mapping, which can be found at <a class="reference external" href="http://www.uniprot.org/faq/28">this page</a>; for example if it’s we want to add the
mappings of uniprot IDs to <em>BioCyc</em>, in the <em>abbreviation</em> column of the
mappings we find that it’s identifier is <em>REACTOME_ID</em>, so we pass
<em>-m REACTOME</em> to the script (leaving <em>_ID</em> out). Mapped IDs are separated by
commas.</p>
<p>The taxonomy IDs are not overwritten if they are found in the annotations, the
<em>-f</em> is provided to force the overwriting of those values.</p>
<p>See also <a class="reference internal" href="../gff.html#gff-specs"><span class="std std-ref">MGKit GFF Specifications</span></a> for more informations about the GFF specifications
used.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>As the script needs to query Uniprot a lot, it is recommended to split
the GFF in several files, so an error in the connection doesn’t waste time.</p>
<p>However, a cache is kept to reduce the number of connections</p>
</div>
</div>
<div class="section" id="coverage-command">
<h2>Coverage Command<a class="headerlink" href="#coverage-command" title="Permalink to this headline">¶</a></h2>
<p>Adds coverage information from BAM alignment files to a GFF file, using the
function <a class="reference internal" href="mgkit.align.html#mgkit.align.add_coverage_info" title="mgkit.align.add_coverage_info"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.align.add_coverage_info()</span></code></a>, the user needs to supply for
each sample a BAM file, using the <cite>-a</cite> option, whose parameter is in the form
<cite>sample,samplealg.bam</cite>. More samples can be supplied adding more <cite>-a</cite>
arguments.</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>As an example, to add coverage for <cite>sample1</cite>, <cite>sample2</cite> the command line
is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">add</span><span class="o">-</span><span class="n">gff</span><span class="o">-</span><span class="n">info</span> <span class="n">coverage</span> <span class="o">-</span><span class="n">a</span> <span class="n">sample1</span><span class="p">,</span><span class="n">sample1</span><span class="o">.</span><span class="n">bam</span> <span class="o">-</span><span class="n">a</span> <span class="n">sample2</span><span class="p">,</span><span class="n">sample2</span><span class="o">.</span><span class="n">bam</span> \
<span class="n">inputgff</span> <span class="n">outputgff</span>
</pre></div>
</div>
</div>
<p>A total coverage for the annotation is also calculated and stored in the
<cite>cov</cite> attribute, while each sample coverage is stored into <cite>sample_cov</cite> as per
<a class="reference internal" href="../gff.html#gff-specs"><span class="std std-ref">MGKit GFF Specifications</span></a>.</p>
</div>
<div class="section" id="adding-coverage-from-samtools-depth">
<h2>Adding Coverage from samtools depth<a class="headerlink" href="#adding-coverage-from-samtools-depth" title="Permalink to this headline">¶</a></h2>
<p>The <em>cov_samtools</em> allows the use of the output of <em>samtools</em> <strong>depth</strong>
command. The command work similarly to <em>coverage</em>, accepting compressed <em>depth</em>
files as well. If only one <em>depth</em> file is passed and no sample is passed, the
attribute in the GFF will be <em>cov</em>, otherwise the attribute will be
<em>sample1_cov</em>, <em>sample2_cov</em>, etc.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>From version 0.4.2 the behaviour is different in the following ways:</p>
<ul class="simple">
<li><dl class="simple">
<dt>the GFF file is loaded in memory to get information that helps reduce</dt><dd><p>the memory size</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>the depth file is scanned and each time a sequence in the GFF file is</dt><dd><p>found, coverage data is added, this way memory is freed more often</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>the GFF file is written at the very end, so chaining multiple commands</dt><dd><p>(one for each sample) doesn’t give any advantage now</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>there’s no need to use the <cite>-aa</cite> option of <cite>samtools depth</cite>, since some</dt><dd><p>assumptions are made when scanning the fils (see next point)</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>when the depth file is scanned completely, the sequences in the GFF file</dt><dd><p>with coverage information are set coverage 0.0, instead of raising an
error</p>
</dd>
</dl>
</li>
<li><p>average coverage is only calculated if requested, with the <cite>-m</cite> option</p></li>
</ul>
</div>
<p>The GFF file is first loaded in memory to get information about the maximum
array size for each sequence, then the script proceeds for each depth file, to
scan it until a) all sequences in the GFF are found or b) until the end of the
file. If b) is the case, all sequences with not found in the depth file are set
to a coverage of 0.0. A warning with the number of sequences lacking coverage
is printed, for diagnostic purposes. If the the mean coverage is requested with
the <cite>-m</cite> option, coverage is then calculated</p>
<p>To create samtools <em>depth</em> files, you can use this command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools depth [bam_file] &gt; [depth_file]

where [bam_file] is the alignment to be used and [depth_file] where the
file will be stored (by redirection). You could create a gzipped file, as
the file can be huge (and in my experience compressing reduces by ~10x the
file size)::

$ samtools depth [bam_file] | gzip &gt; [depth_file.gz]
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>if version &lt; 0.4.2:</p>
<p>The <em>-aa</em> options must be used to pass information about all base
pairs and sequences coverage in the BAM/SAM file.</p>
<p>$ samtools depth -aa bam_file</p>
</div>
</div>
<div class="section" id="uniprot-offline-mappings">
<h2>Uniprot Offline Mappings<a class="headerlink" href="#uniprot-offline-mappings" title="Permalink to this headline">¶</a></h2>
<p>Similar to the <em>uniprot</em> command, it uses the <a class="reference external" href="ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/idmapping.dat.gz">idmapping</a>
file provided by Uniprot, which speeds up the process of adding mappings and
taxonomy IDs from Uniprot gene IDs. It’s not possible tough to add <em>EC</em>
mappings with this command, as those are not included in the file.</p>
</div>
<div class="section" id="kegg-information">
<h2>Kegg Information<a class="headerlink" href="#kegg-information" title="Permalink to this headline">¶</a></h2>
<p>The <em>kegg</em> command allows to add information to each annotation. Right now the
information that can be added is restricted to the pathway(s) (reference KO) a
KO is part of and both the KO and pathway(s) descriptions. This information is
stored in keys starting with <strong>ko_</strong>.</p>
</div>
<div class="section" id="expected-aminoacidic-changes">
<h2>Expected Aminoacidic Changes<a class="headerlink" href="#expected-aminoacidic-changes" title="Permalink to this headline">¶</a></h2>
<p>Some scripts, like <a class="reference internal" href="../scripts/snp_analysis.html#snp-parser"><span class="std std-ref">snp_parser - SNPs analysis</span></a>, require information about the expected
number of synonymous and non-synonymous changes of an annotation. This can be
done using <a class="reference internal" href="mgkit.io.gff.html#mgkit.io.gff.Annotation.add_exp_syn_count" title="mgkit.io.gff.Annotation.add_exp_syn_count"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation.add_exp_syn_count()</span></code></a> by the user of the
command <cite>exp_syn</cite> of this script. The attributes added to each annotation are
explained in the <a class="reference internal" href="../gff.html#gff-specs"><span class="std std-ref">MGKit GFF Specifications</span></a></p>
</div>
<div class="section" id="adding-count-data">
<h2>Adding Count Data<a class="headerlink" href="#adding-count-data" title="Permalink to this headline">¶</a></h2>
<p>Count data on a per-sample basis can be added with the <em>counts</em> command. The
accepted inputs are from HTSeq-count and featureCounts. The ouput produced by
featureCounts, is the one from using its <strong>-f</strong> option must be used.</p>
<p>This script accept by default a tab separated file, with a uid in the first
column and the other columns are the counts for each sample, in the same order
as they are passed to the <strong>-s</strong> option. To use the featureCounts file format,
this script <strong>-e</strong> option must be used.</p>
<p>The sample names must be provided in the same order as the columns in the input
files. If the counts are FPKMS the <em>-f</em> option can be used.</p>
</div>
<div class="section" id="adding-taxonomy-from-a-table">
<h2>Adding Taxonomy from a Table<a class="headerlink" href="#adding-taxonomy-from-a-table" title="Permalink to this headline">¶</a></h2>
<p>There are cases where it may needed or preferred to add the taxonomy from a
<em>gene_id</em> already provided in the GFF file. For such cases the <em>addtaxa</em>
command can be used. It works in a similar way to the <em>taxonomy</em> command, only
it expect three different type of inputs:</p>
<blockquote>
<div><ul class="simple">
<li><p><em>GI-Taxa</em> table from NCBI (e.g. gi_taxid_nucl.dmp, )</p></li>
<li><p>tab separated table</p></li>
<li><p>dictionary</p></li>
<li><p>HDF5</p></li>
</ul>
</div></blockquote>
<p>The first two are tab separated files, where on each line, the first column is
the <em>gene_id</em> that is found in the first column, while the second if the
<em>taxon_id</em>.</p>
<p>The third option is a serialised Python <em>dict</em>/hash table, whose keys are the
<em>gene_id</em> and the value is that gene corresponding <em>taxon_id</em>. The serialised
formats accepted are msgpack, json and pickle. The <em>msgpack</em> module must be
importable. The option to use json and msgpack allow to integrate this script
with other languages without resorting to a text file.</p>
<p>The last option is a HDF5 created using the <em>to_hdf</em> command in
<a class="reference internal" href="../scripts/taxon_utils.html#taxon-utils"><span class="std std-ref">taxon-utils - Taxonomy Utilities</span></a>. This requires <cite>pandas</cite> installed and <cite>pytables</cite> and it
provides faster lookup of IDs in the table.</p>
<p>While the default is to look for the <em>gene_id</em> attribute in the GFF annotation,
another attribute can be specified, using the <strong>-gene-attr</strong> option.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>the dictionary content is loaded after the table files and its keys and
corresponding values takes precedence over the text files.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>from September 2016 NCBI will retire the GI. In that case the same
kind of table can be built from the <em>nucl_gb.accession2taxid.gz</em> file
The format is different, but some information can be found in
<a class="reference internal" href="mgkit.io.blast.html#mgkit.io.blast.parse_accession_taxa_table" title="mgkit.io.blast.parse_accession_taxa_table"><code class="xref py py-func docutils literal notranslate"><span class="pre">mgkit.io.blast.parse_accession_taxa_table()</span></code></a></p>
</div>
</div>
<div class="section" id="adding-information-from-pfam">
<h2>Adding information from Pfam<a class="headerlink" href="#adding-information-from-pfam" title="Permalink to this headline">¶</a></h2>
<p>Adds the Pfam description for the annotation, by downloading the list from
Pfam.</p>
<p>The options allow to specify in which attribute the ID/ACCESSION is stored
(defaults to <em>gene_id</em>) and which one between ID/ACCESSION is the value of that
attribute (defaults to <em>ID</em>). if no description is found for the family, a
warning message is logged.</p>
</div>
<div class="section" id="changes">
<h2>Changes<a class="headerlink" href="#changes" title="Permalink to this headline">¶</a></h2>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.4.2: </span>added <em>-m</em> option for <em>cov_samtools</em> command, to calculate the average
coverage for an annotation (<em>cov</em> attribute). Fixed loading compressed
files</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.4: </span>removed the <em>taxonomy</em> command, since a similar result can be obtained with
<em>taxon-utils lca</em> and <em>add-gff-info addtaxa</em>. Removed <em>eggnog</em> command and
added option to verbose the logging in <em>cov_samtools</em> (now is quiet), also
changed the interface</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.3: </span>changed how <em>addtaxa</em> <em>-a</em> works, to allow the use of <em>seq_id</em> as key to
add the taxon_id</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.0: </span>added <em>cov_samtools</em> command, <em>–split</em> option to <em>exp_syn</em>, <em>-c</em> option to
<em>addtaxa</em>. <em>kegg</em> now does not skip annotations when the attribute is not
found.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.6: </span>added <em>skip-no-taxon</em> option to <em>addtaxa</em></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.5: </span>if a dictionary is supplied to <em>addtaxa</em>, the GFF is not preloaded</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.3: </span>added <em>pfam</em> command, renamed <em>gitaxa</em> to <em>addtaxa</em> and made it general</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.2: </span>added <em>eggnog</em>, <em>gitaxa</em> and <em>counts</em> command</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.1.</span></p>
</div>
<ul class="simple">
<li><p>added <em>-d</em> to <em>uniprot</em> command</p></li>
<li><p>added cache to <em>uniprot</em> command</p></li>
<li><p>added <em>kegg</em> command (cached)</p></li>
</ul>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.1.16: </span>added <em>exp_syn</em> command</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.1.15: </span><em>taxonomy</em> command <em>-b</em> option changed</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.1.13.</span></p>
</div>
<ul class="simple">
<li><p>added <em>–force-taxon-id</em> option to the <em>uniprot</em> command</p></li>
<li><p>added <em>coverage</em> command</p></li>
<li><p>added <em>taxonomy</em> command</p></li>
<li><p>added <em>unipfile</em> command</p></li>
</ul>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.12.</span></p>
</div>
<dl class="py function">
<dt id="mgkit.workflow.add_gff_info.add_uniprot_info">
<code class="sig-prename descclassname">mgkit.workflow.add_gff_info.</code><code class="sig-name descname">add_uniprot_info</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">annotations</span></em>, <em class="sig-param"><span class="n">email</span></em>, <em class="sig-param"><span class="n">force_taxon_id</span></em>, <em class="sig-param"><span class="n">taxon_id</span></em>, <em class="sig-param"><span class="n">lineage</span></em>, <em class="sig-param"><span class="n">eggnog</span></em>, <em class="sig-param"><span class="n">enzymes</span></em>, <em class="sig-param"><span class="n">kegg_orthologs</span></em>, <em class="sig-param"><span class="n">protein_names</span></em>, <em class="sig-param"><span class="n">mapping</span></em>, <em class="sig-param"><span class="n">info_cache</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/add_gff_info.html#add_uniprot_info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.add_gff_info.add_uniprot_info" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.add_gff_info.load_featurecounts_files">
<code class="sig-prename descclassname">mgkit.workflow.add_gff_info.</code><code class="sig-name descname">load_featurecounts_files</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">count_files</span></em>, <em class="sig-param"><span class="n">samples</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/add_gff_info.html#load_featurecounts_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.add_gff_info.load_featurecounts_files" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.add_gff_info.load_htseq_count_files">
<code class="sig-prename descclassname">mgkit.workflow.add_gff_info.</code><code class="sig-name descname">load_htseq_count_files</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">count_files</span></em>, <em class="sig-param"><span class="n">samples</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/add_gff_info.html#load_htseq_count_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.add_gff_info.load_htseq_count_files" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.add_gff_info.parse_hdf5_arg">
<code class="sig-prename descclassname">mgkit.workflow.add_gff_info.</code><code class="sig-name descname">parse_hdf5_arg</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ctx</span></em>, <em class="sig-param"><span class="n">param</span></em>, <em class="sig-param"><span class="n">values</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/add_gff_info.html#parse_hdf5_arg"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.add_gff_info.parse_hdf5_arg" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.add_gff_info.split_sample_alg">
<code class="sig-prename descclassname">mgkit.workflow.add_gff_info.</code><code class="sig-name descname">split_sample_alg</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ctx</span></em>, <em class="sig-param"><span class="n">param</span></em>, <em class="sig-param"><span class="n">values</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/add_gff_info.html#split_sample_alg"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.add_gff_info.split_sample_alg" title="Permalink to this definition">¶</a></dt>
<dd><p>Split sample/alignment option</p>
</dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.add_gff_info.update_annotation_coverage">
<code class="sig-prename descclassname">mgkit.workflow.add_gff_info.</code><code class="sig-name descname">update_annotation_coverage</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">annotations</span></em>, <em class="sig-param"><span class="n">depth</span></em>, <em class="sig-param"><span class="n">sample</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/add_gff_info.html#update_annotation_coverage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.add_gff_info.update_annotation_coverage" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="mgkit.workflow.blast2gff.html" class="btn btn-neutral float-right" title="mgkit.workflow.blast2gff module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="mgkit.workflow.html" class="btn btn-neutral float-left" title="mgkit.workflow package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2013-2020, Francesco Rubino

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>