

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>mgkit.snps.classes module &mdash; MGKit: Metagenomic framework 0.5.4 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="mgkit.snps.conv_func module" href="mgkit.snps.conv_func.html" />
    <link rel="prev" title="mgkit.snps package" href="mgkit.snps.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> MGKit: Metagenomic framework
          

          
          </a>

          
            
            
              <div class="version">
                0.5.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">Library Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mgkit.html">mgkit package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.align.html">mgkit.align module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.consts.html">mgkit.consts module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.html">mgkit.counts package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.func.html">mgkit.counts.func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.glm.html">mgkit.counts.glm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.scaling.html">mgkit.counts.scaling module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.html">mgkit.db package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.dbm.html">mgkit.db.dbm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.mongo.html">mgkit.db.mongo module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.html">mgkit.filter package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.common.html">mgkit.filter.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.gff.html">mgkit.filter.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.lists.html">mgkit.filter.lists module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.reads.html">mgkit.filter.reads module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.taxon.html">mgkit.filter.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.graphs.html">mgkit.graphs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.html">mgkit.io package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.blast.html">mgkit.io.blast module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fasta.html">mgkit.io.fasta module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fastq.html">mgkit.io.fastq module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.gff.html">mgkit.io.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.glimmer.html">mgkit.io.glimmer module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.snpdat.html">mgkit.io.snpdat module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.uniprot.html">mgkit.io.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.utils.html">mgkit.io.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.kegg.html">mgkit.kegg module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.logger.html">mgkit.logger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.html">mgkit.mappings package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.cazy.html">mgkit.mappings.cazy module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.eggnog.html">mgkit.mappings.eggnog module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.enzyme.html">mgkit.mappings.enzyme module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.go.html">mgkit.mappings.go module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.pandas_map.html">mgkit.mappings.pandas_map module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.taxon.html">mgkit.mappings.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.utils.html">mgkit.mappings.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.html">mgkit.net package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.embl.html">mgkit.net.embl module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.pfam.html">mgkit.net.pfam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.uniprot.html">mgkit.net.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.utils.html">mgkit.net.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.html">mgkit.plots package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.abund.html">mgkit.plots.abund module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.boxplot.html">mgkit.plots.boxplot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.colors.html">mgkit.plots.colors module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.heatmap.html">mgkit.plots.heatmap module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.unused.html">mgkit.plots.unused module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.utils.html">mgkit.plots.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.simple_cache.html">mgkit.simple_cache module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.html">mgkit.snps package</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">mgkit.snps.classes module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.conv_func.html">mgkit.snps.conv_func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.filter.html">mgkit.snps.filter module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.funcs.html">mgkit.snps.funcs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.mapper.html">mgkit.snps.mapper module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.taxon.html">mgkit.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.html">mgkit.utils package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.common.html">mgkit.utils.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.dictionary.html">mgkit.utils.dictionary module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.sequence.html">mgkit.utils.sequence module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.trans_tables.html">mgkit.utils.trans_tables module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.html">mgkit.workflow package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.add_gff_info.html">mgkit.workflow.add_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.blast2gff.html">mgkit.workflow.blast2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.extract_gff_info.html">mgkit.workflow.extract_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fasta_utils.html">mgkit.workflow.fasta_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fastq_utils.html">mgkit.workflow.fastq_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.filter_gff.html">mgkit.workflow.filter_gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.hmmer2gff.html">mgkit.workflow.hmmer2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.json2gff.html">mgkit.workflow.json2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.sampling_utils.html">mgkit.workflow.sampling_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.snp_parser.html">mgkit.workflow.snp_parser module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.taxon_utils.html">mgkit.workflow.taxon_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.utils.html">mgkit.workflow.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html">mgkit</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MGKit: Metagenomic framework</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../library.html">Library Reference</a> &raquo;</li>
        
          <li><a href="mgkit.html">mgkit package</a> &raquo;</li>
        
          <li><a href="mgkit.snps.html">mgkit.snps package</a> &raquo;</li>
        
      <li>mgkit.snps.classes module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/api/mgkit.snps.classes.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="module-mgkit.snps.classes">
<span id="mgkit-snps-classes-module"></span><h1>mgkit.snps.classes module<a class="headerlink" href="#module-mgkit.snps.classes" title="Permalink to this headline">¶</a></h1>
<p>Manage SNP data.</p>
<dl class="py class">
<dt id="mgkit.snps.classes.GeneSNP">
<em class="property">class </em><code class="sig-prename descclassname">mgkit.snps.classes.</code><code class="sig-name descname">GeneSNP</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_id</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">taxon_id</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">exp_syn</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">exp_nonsyn</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">coverage</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">snps</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">uid</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">json_data</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/classes.html#GeneSNP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.GeneSNP" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#mgkit.snps.classes.RatioMixIn" title="mgkit.snps.classes.RatioMixIn"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.snps.classes.RatioMixIn</span></code></a></p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.13.</span></p>
</div>
<p>Class defining gene and synonymous/non-synonymous SNPs.</p>
<p>It defines background synonymous/non-synonymous attributes and only has a
method right now, which calculate pN/pS ratio. The method is added through
a mixin object, so the ratio can be customised and be shared with the old
implementation.</p>
<dl class="py attribute">
<dt id="mgkit.snps.classes.GeneSNP.uid">
<code class="sig-name descname">uid</code><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.uid" title="Permalink to this definition">¶</a></dt>
<dd><p>unique id for the isoform (to be referenced in a GFF file)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="mgkit.snps.classes.GeneSNP.gene_id">
<code class="sig-name descname">gene_id</code><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.gene_id" title="Permalink to this definition">¶</a></dt>
<dd><p>gene id</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="mgkit.snps.classes.GeneSNP.taxon_id">
<code class="sig-name descname">taxon_id</code><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.taxon_id" title="Permalink to this definition">¶</a></dt>
<dd><p>gene taxon</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="mgkit.snps.classes.GeneSNP.exp_syn">
<code class="sig-name descname">exp_syn</code><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.exp_syn" title="Permalink to this definition">¶</a></dt>
<dd><p>expected synonymous changes</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="mgkit.snps.classes.GeneSNP.exp_nonsyn">
<code class="sig-name descname">exp_nonsyn</code><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.exp_nonsyn" title="Permalink to this definition">¶</a></dt>
<dd><p>expected non-synonymous changes</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="mgkit.snps.classes.GeneSNP.coverage">
<code class="sig-name descname">coverage</code><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.coverage" title="Permalink to this definition">¶</a></dt>
<dd><p>gene coverage</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="mgkit.snps.classes.GeneSNP.snps">
<code class="sig-name descname">snps</code><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.snps" title="Permalink to this definition">¶</a></dt>
<dd><p>list of SNPs associated with the gene, each element is a
tuple with the position (relative to the gene start), the second is
the nucleotidic change and the third is the aa SNP type as defined
by <a class="reference internal" href="#mgkit.snps.classes.SNPType" title="mgkit.snps.classes.SNPType"><code class="xref py py-class docutils literal notranslate"><span class="pre">SNPType</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a></p>
</dd>
</dl>
</dd></dl>

<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The main difference with the <code class="xref py py-class docutils literal notranslate"><span class="pre">GeneSyn</span></code> is that all snps are kept
and <cite>syn</cite> and <cite>nonsyn</cite> are not attributes but properties that return
the count of synonymous and non-synonymous SNPs in the <cite>snps</cite> list.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This class uses more memory than <code class="xref py py-class docutils literal notranslate"><span class="pre">GeneSyn</span></code> because it doesn’t
use __slots__, it may be changed in later versions.</p>
</div>
<dl class="py method">
<dt id="mgkit.snps.classes.GeneSNP.add">
<code class="sig-name descname">add</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/classes.html#GeneSNP.add"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.add" title="Permalink to this definition">¶</a></dt>
<dd><p>Inplace addition of another instance values. No check for them being
the same gene/taxon, it’s up to the user to check that they can be
added together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> – instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">GeneSyn</span></code> to add</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="mgkit.snps.classes.GeneSNP.add_snp">
<code class="sig-name descname">add_snp</code><span class="sig-paren">(</span><em class="sig-param">position</em>, <em class="sig-param">change</em>, <em class="sig-param">snp_type=&lt;SNPType.unknown: 0&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/classes.html#GeneSNP.add_snp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.add_snp" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a SNP to the list</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – SNP position, relative to the gene start</p></li>
<li><p><strong>change</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – nucleotidic change</p></li>
<li><p><strong>snp_type</strong> (<em>enum</em>) – one of the values defined in <a class="reference internal" href="#mgkit.snps.classes.SNPType" title="mgkit.snps.classes.SNPType"><code class="xref py py-class docutils literal notranslate"><span class="pre">SNPType</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="id0">
<code class="sig-name descname">coverage</code><em class="property"> = None</em><a class="headerlink" href="#id0" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="id1">
<code class="sig-name descname">exp_nonsyn</code><em class="property"> = None</em><a class="headerlink" href="#id1" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="id2">
<code class="sig-name descname">exp_syn</code><em class="property"> = None</em><a class="headerlink" href="#id2" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="mgkit.snps.classes.GeneSNP.from_json">
<code class="sig-name descname">from_json</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/classes.html#GeneSNP.from_json"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.from_json" title="Permalink to this definition">¶</a></dt>
<dd><p>Instantiate the instance with values from a json definition</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – json representation, as returned by
<a class="reference internal" href="#mgkit.snps.classes.GeneSNP.to_json" title="mgkit.snps.classes.GeneSNP.to_json"><code class="xref py py-meth docutils literal notranslate"><span class="pre">GeneSNP.to_json()</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="id3">
<code class="sig-name descname">gene_id</code><em class="property"> = None</em><a class="headerlink" href="#id3" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="mgkit.snps.classes.GeneSNP.nonsyn">
<em class="property">property </em><code class="sig-name descname">nonsyn</code><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.nonsyn" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the expected non-synonymous changes</p>
</dd></dl>

<dl class="py attribute">
<dt id="id4">
<code class="sig-name descname">snps</code><em class="property"> = None</em><a class="headerlink" href="#id4" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="mgkit.snps.classes.GeneSNP.syn">
<em class="property">property </em><code class="sig-name descname">syn</code><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.syn" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the expected synonymous changes</p>
</dd></dl>

<dl class="py attribute">
<dt id="id5">
<code class="sig-name descname">taxon_id</code><em class="property"> = None</em><a class="headerlink" href="#id5" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="mgkit.snps.classes.GeneSNP.to_json">
<code class="sig-name descname">to_json</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/classes.html#GeneSNP.to_json"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.GeneSNP.to_json" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a json definition of the instance</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>json representation of the instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="id6">
<code class="sig-name descname">uid</code><em class="property"> = None</em><a class="headerlink" href="#id6" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="mgkit.snps.classes.RatioMixIn">
<em class="property">class </em><code class="sig-prename descclassname">mgkit.snps.classes.</code><code class="sig-name descname">RatioMixIn</code><a class="reference internal" href="../_modules/mgkit/snps/classes.html#RatioMixIn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.RatioMixIn" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt id="mgkit.snps.classes.RatioMixIn.calc_pn">
<code class="sig-name descname">calc_pn</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/classes.html#RatioMixIn.calc_pn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.RatioMixIn.calc_pn" title="Permalink to this definition">¶</a></dt>
<dd><p>Method that returns only the pN part of the pN/pS ratio.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the pN value, unless self.nonsyn is 0, in which case <cite>numpy.nan</cite> is returned</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="mgkit.snps.classes.RatioMixIn.calc_ps">
<code class="sig-name descname">calc_ps</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/classes.html#RatioMixIn.calc_ps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.RatioMixIn.calc_ps" title="Permalink to this definition">¶</a></dt>
<dd><p>Method that returns only the pS part of the pN/pS ratio.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the pS value, unless self.syn is 0, in which case <cite>numpy.nan</cite> is returned</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="mgkit.snps.classes.RatioMixIn.calc_ratio">
<code class="sig-name descname">calc_ratio</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">haplotypes</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/classes.html#RatioMixIn.calc_ratio"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.RatioMixIn.calc_ratio" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.2: </span>split the function to handle <em>flag_value</em> in another method</p>
</div>
<p>Calculate <span class="math notranslate nohighlight">\(\frac {pN}{pS}\)</span> for the gene.</p>
<div class="math notranslate nohighlight" id="equation-pn-ps">
<span class="eqno">(1)<a class="headerlink" href="#equation-pn-ps" title="Permalink to this equation">¶</a></span>\[\frac {pN}{pS} = \frac{ ^{oN}/_{eN}}{ ^{oS}/_{eS}}\]</div>
<p>WHere:</p>
<ul class="simple">
<li><p>oN (number of non-synonymous - <strong>nonsyn</strong>)</p></li>
<li><p>eN (expected number of non-synonymous - <strong>exp_nonsyn</strong>)</p></li>
<li><p>oS (number of synonymous - <strong>syn</strong>)</p></li>
<li><p>eS (expected number of synonymous - <strong>exp_syn</strong>)</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flag_value</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – when there’s no way to calculate the ratio, the
possible cases will be flagged with a negative number. This
allows to make substitutions for these values</p></li>
<li><p><strong>haplotypes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – if true, coverage information is not used,
because the SNPs are assumed to come from an alignment that has
sequences having haplotypes</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>the <span class="math notranslate nohighlight">\(\frac {pN}{pS}\)</span> for the gene.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Because pN or pS can be 0, and the return value would be NaN,
we take in account some special cases. The default return value
in this cases is <code class="xref py py-const docutils literal notranslate"><span class="pre">numpy.nan</span></code>.</p>
</div>
<ul>
<li><p>Both synonymous and non-synonymous values are 0:</p>
<blockquote>
<div><ul class="simple">
<li><p>if both the syn and nonsyn attributes are 0 but there’s
coverage for this gene, we return a 0, as there’s no
evolution in this gene. Before, the coverage was checked by
this method against either the passed <em>min_cov</em> parameter
that was equal to <code class="xref py py-data docutils literal notranslate"><span class="pre">MIN_COV</span></code>. Now the case is for the
user to check the coverage and functions in
<a class="reference internal" href="mgkit.snps.conv_func.html#module-mgkit.snps.conv_func" title="mgkit.snps.conv_func"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.snps.conv_func</span></code></a> do that. If enough coverage was
achieved, the <em>haplotypes</em> parameter can be used to return a
0</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>All other cases return a NaN value</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="mgkit.snps.classes.RatioMixIn.calc_ratio_flag">
<code class="sig-name descname">calc_ratio_flag</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/classes.html#RatioMixIn.calc_ratio_flag"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.RatioMixIn.calc_ratio_flag" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.2.2.</span></p>
</div>
<p>Handles cases where it’s important to flag the returned value, as
explained in <code class="xref py py-meth docutils literal notranslate"><span class="pre">GeneSNP.calc_ratio()</span></code>, and when the both the number
of synonymous and non-synonymous is greater than 0, the pN/pS value is
returned.</p>
<ul>
<li><dl>
<dt>The number of non-synonymous is greater than 0 but the number of</dt><dd><blockquote>
<div><p>synonymous is 0:</p>
<blockquote>
<div><ul class="simple">
<li><p>if <strong>flag_value</strong> is <strong>True</strong>, the returned value is <strong>-1</strong></p></li>
</ul>
</div></blockquote>
</div></blockquote>
<ul>
<li><p>The number of synonymous is greater than 0 but the number of
non-synonymous is 0:</p>
<blockquote>
<div><ul class="simple">
<li><p>if <strong>flag_value</strong> is <strong>True</strong>, the returned value is <strong>-2</strong></p></li>
</ul>
</div></blockquote>
</li>
</ul>
</dd>
</dl>
</li>
</ul>
<table class="docutils align-default">
<colgroup>
<col style="width: 32%" />
<col style="width: 32%" />
<col style="width: 37%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p><span class="math notranslate nohighlight">\(oS\)</span></p></th>
<th class="head"><p><span class="math notranslate nohighlight">\(oN\)</span></p></th>
<th class="head"><p>return value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>&gt;0</p></td>
<td><p>&gt;0</p></td>
<td><p><strong>pN/pS</strong></p></td>
</tr>
<tr class="row-odd"><td><p>0</p></td>
<td><p>0</p></td>
<td><p><strong>-3</strong></p></td>
</tr>
<tr class="row-even"><td><p>&gt;0</p></td>
<td><p>0</p></td>
<td><p><strong>-1</strong></p></td>
</tr>
<tr class="row-odd"><td><p>0</p></td>
<td><p>&gt;0</p></td>
<td><p><strong>-2</strong></p></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="mgkit.snps.classes.SNPType">
<em class="property">class </em><code class="sig-prename descclassname">mgkit.snps.classes.</code><code class="sig-name descname">SNPType</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">value</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/classes.html#SNPType"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.classes.SNPType" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/enum.html#enum.Enum" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">enum.Enum</span></code></a></p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.13.</span></p>
</div>
<p>Enum that defines SNP types. Supported at the moment:</p>
<ul class="simple">
<li><p>unknown = 0</p></li>
<li><p>syn (synonymous) = 1</p></li>
<li><p>nonsyn (non-synonymous) = 2</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>No support is planned at the moment to support indel mutations</p>
</div>
<dl class="py attribute">
<dt id="mgkit.snps.classes.SNPType.nonsyn">
<code class="sig-name descname">nonsyn</code><em class="property"> = 2</em><a class="headerlink" href="#mgkit.snps.classes.SNPType.nonsyn" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="mgkit.snps.classes.SNPType.syn">
<code class="sig-name descname">syn</code><em class="property"> = 1</em><a class="headerlink" href="#mgkit.snps.classes.SNPType.syn" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="mgkit.snps.classes.SNPType.unknown">
<code class="sig-name descname">unknown</code><em class="property"> = 0</em><a class="headerlink" href="#mgkit.snps.classes.SNPType.unknown" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="mgkit.snps.conv_func.html" class="btn btn-neutral float-right" title="mgkit.snps.conv_func module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="mgkit.snps.html" class="btn btn-neutral float-left" title="mgkit.snps package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2013-2020, Francesco Rubino

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>