
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>mgkit.snps.funcs module &#8212; MGKit: Metagenomic framework 0.5.7 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="mgkit.snps.mapper module" href="mgkit.snps.mapper.html" />
    <link rel="prev" title="mgkit.snps.filter module" href="mgkit.snps.filter.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-mgkit.snps.funcs">
<span id="mgkit-snps-funcs-module"></span><h1>mgkit.snps.funcs module<a class="headerlink" href="#module-mgkit.snps.funcs" title="Permalink to this headline">¶</a></h1>
<p>Functions used in SNPs manipulation</p>
<dl class="py function">
<dt id="mgkit.snps.funcs.build_rank_matrix">
<code class="sig-prename descclassname">mgkit.snps.funcs.</code><code class="sig-name descname">build_rank_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dataframe</span></em>, <em class="sig-param"><span class="n">taxonomy</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">taxon_rank</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/funcs.html#build_rank_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.funcs.build_rank_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a rank matrix from a <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code> with the pN/pS values of a
dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code> instance with a MultiIndex
(gene-taxon)</p></li>
<li><p><strong>taxonomy</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">taxon.Taxonomy</span></code> instance with the full
taxonomy</p></li>
<li><p><strong>taxon_rank</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – taxon rank to limit the specifity of the taxa
included</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.snps.funcs.combine_sample_snps">
<code class="sig-prename descclassname">mgkit.snps.funcs.</code><code class="sig-name descname">combine_sample_snps</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">snps_data</span></em>, <em class="sig-param"><span class="n">min_num</span></em>, <em class="sig-param"><span class="n">filters</span></em>, <em class="sig-param"><span class="n">index_type</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_func</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">taxon_func</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_uid</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">flag_values</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">haplotypes</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">store_uids</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">partial_calc</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">partial_syn</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/funcs.html#combine_sample_snps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.funcs.combine_sample_snps" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.2: </span>added <em>use_uid</em> argument</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.1: </span>added <em>haplotypes</em></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.4.0: </span>added <em>store_uids</em></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.5.3: </span>added <em>partial_calc</em> and <em>partial_type</em></p>
</div>
<p>Combine a dictionary sample-&gt;gene_index-&gt;GeneSyn into a
<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>. The dictionary is first filtered with the
functions in <cite>filters</cite>, mapped to different taxa and genes using
<cite>taxon_func</cite> and <cite>gene_func</cite> respectively. The returned DataFrame is also
filtered for each row having at least a <cite>min_num</cite> of not NaN values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>snps_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a>) – dictionary with the <cite>GeneSNP</cite> instances</p></li>
<li><p><strong>min_num</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – the minimum number of not NaN values necessary in a row
to be returned</p></li>
<li><p><strong>filters</strong> (<em>iterable</em>) – iterable containing filter functions, a list can be
found in <a class="reference internal" href="mgkit.snps.filter.html#module-mgkit.snps.filter" title="mgkit.snps.filter"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.snps.filter</span></code></a></p></li>
<li><p><strong>index_type</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.10)"><em>None</em></a>) – if <cite>None</cite>, each row index for the DataFrame
will be a MultiIndex with <cite>gene</cite> and <cite>taxon</cite> as elements. If the
equals ‘gene’, the row index will be gene based and if ‘taxon’ will
be taxon based</p></li>
<li><p><strong>gene_func</strong> (<em>func</em>) – a function to map a gene_id to a gene_map. See
<a class="reference internal" href="mgkit.snps.mapper.html#mgkit.snps.mapper.map_gene_id" title="mgkit.snps.mapper.map_gene_id"><code class="xref py py-func docutils literal notranslate"><span class="pre">mapper.map_gene_id()</span></code></a> for an example</p></li>
<li><p><strong>taxon_func</strong> (<em>func</em>) – a function to map a taxon_id to a list of IDs. See
<code class="xref py py-mod docutils literal notranslate"><span class="pre">mapper.map_taxon_id_to_rank</span></code> or
<code class="xref py py-mod docutils literal notranslate"><span class="pre">mapper.map_taxon_id_to_ancestor</span></code> for examples</p></li>
<li><p><strong>use_uid</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – if True, uses the <cite>GeneSNP.uid</cite> instead of
<cite>GeneSNP.gene_id</cite></p></li>
<li><p><strong>flag_values</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – if True,
<code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.snps.classes.GeneSNP.calc_ratio_flag()</span></code> will be used,
instead of <code class="xref py py-meth docutils literal notranslate"><span class="pre">mgkit.snps.classes.GeneSNP.calc_ratio()</span></code></p></li>
<li><p><strong>haplotypes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – if <em>flag_values</em> is False, and <em>haplotypes</em> is
True, the 0/0 case will be returned as 0 instead of NaN</p></li>
<li><p><strong>store_uids</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – if True a dictionary with the uid used for each
cell (e.g. gene/taxon/sample)</p></li>
<li><p><strong>partial_calc</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – if True, only pS or pN values will be calculated,
depending on the value of <em>partial_syn</em></p></li>
<li><p><strong>partial_syn</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – if both <em>partial_calc</em> and this are True, only pS
values will be calculated. If this parameter is False, pN values
will be calculated</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> with the pN/pS values for the
input SNPs, with the columns being the samples. if <em>store_uids</em> is True
the return value is a tuple (DataFrame, dict)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.snps.funcs.flat_sample_snps">
<code class="sig-prename descclassname">mgkit.snps.funcs.</code><code class="sig-name descname">flat_sample_snps</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">snps_data</span></em>, <em class="sig-param"><span class="n">min_cov</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/funcs.html#flat_sample_snps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.funcs.flat_sample_snps" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.11.</span></p>
</div>
<p>Adds all the values of a gene across all samples into one instance of
<code class="xref py py-class docutils literal notranslate"><span class="pre">classes.GeneSNP</span></code>, giving the average gene among all samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>snps_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a>) – dictionary with the instances of
<code class="xref py py-class docutils literal notranslate"><span class="pre">classes.GeneSNP</span></code></p></li>
<li><p><strong>min_cov</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – minimum coverage required for the each instance to be
added</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the dictionary with only one key (<cite>all_samples</cite>), which can be
used with <a class="reference internal" href="#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">combine_sample_snps()</span></code></a></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.snps.funcs.group_rank_matrix">
<code class="sig-prename descclassname">mgkit.snps.funcs.</code><code class="sig-name descname">group_rank_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dataframe</span></em>, <em class="sig-param"><span class="n">gene_map</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/funcs.html#group_rank_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.funcs.group_rank_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Group a rank matrix using a mapping, in the form map_id-&gt;ko_ids.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> – instance of a rank matrix from <a class="reference internal" href="#mgkit.snps.funcs.build_rank_matrix" title="mgkit.snps.funcs.build_rank_matrix"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_rank_matrix()</span></code></a></p></li>
<li><p><strong>gene_map</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a>) – dictionary with the mapping</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.snps.funcs.order_ratios">
<code class="sig-prename descclassname">mgkit.snps.funcs.</code><code class="sig-name descname">order_ratios</code><span class="sig-paren">(</span><em class="sig-param">ratios</em>, <em class="sig-param">aggr_func=&lt;function median&gt;</em>, <em class="sig-param">reverse=False</em>, <em class="sig-param">key_filter=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/funcs.html#order_ratios"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.funcs.order_ratios" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a dictionary of id-&gt;iterable where iterable contains the values of
interest, the function uses aggr_func to sort (ascending by default) it and
return a list with the key in the sorted order.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ratios</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a>) – dictionary instance id-&gt;iterable</p></li>
<li><p><strong>aggr_func</strong> (<em>function</em>) – any function returning a value that can be used
as a key in sorting</p></li>
<li><p><strong>reverse</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – the default is ascending sorting (False), set to True
to reverse key_filter: list of keys to use for ordering, if None, every
key is used</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterable with the sort order</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.snps.funcs.significance_test">
<code class="sig-prename descclassname">mgkit.snps.funcs.</code><code class="sig-name descname">significance_test</code><span class="sig-paren">(</span><em class="sig-param">dataframe</em>, <em class="sig-param">taxon_id1</em>, <em class="sig-param">taxon_id2</em>, <em class="sig-param">test_func=&lt;function ks_2samp&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/funcs.html#significance_test"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.funcs.significance_test" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.11.</span></p>
</div>
<p>Perform a statistical test on each gene distribution in two different taxa.</p>
<p>For each gene common to the two taxa, the distribution of values in all
samples (columns) between the two specified taxa is tested.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> instance</p></li>
<li><p><strong>taxon_id1</strong> – the first taxon ID</p></li>
<li><p><strong>taxon_id2</strong> – the second taxon ID</p></li>
<li><p><strong>test_func</strong> – function used to test,
defaults to <code class="xref py py-func docutils literal notranslate"><span class="pre">scipy.stats.ks_2samp()</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>with all pvalues from the tests</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.snps.funcs.write_sign_genes_table">
<code class="sig-prename descclassname">mgkit.snps.funcs.</code><code class="sig-name descname">write_sign_genes_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">out_file</span></em>, <em class="sig-param"><span class="n">dataframe</span></em>, <em class="sig-param"><span class="n">sign_genes</span></em>, <em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">gene_names</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/funcs.html#write_sign_genes_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.funcs.write_sign_genes_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a table with the list of significant genes found in a dataframe, the
significant gene list is the result of
<code class="xref py py-func docutils literal notranslate"><span class="pre">wilcoxon_pairwise_test_dataframe()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Out_file</dt>
<dd class="field-odd"><p>the file name or file object to write the file</p>
</dd>
<dt class="field-even">Dataframe</dt>
<dd class="field-even"><p>the dataframe which was tested for significant genes</p>
</dd>
<dt class="field-odd">Sign_genes</dt>
<dd class="field-odd"><p>gene list that are significant</p>
</dd>
<dt class="field-even">Taxonomy</dt>
<dd class="field-even"><p>taxonomy object</p>
</dd>
<dt class="field-odd">Gene_names</dt>
<dd class="field-odd"><p>dictionary with the name of the the genes. Optional</p>
</dd>
</dl>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">MGKit: Metagenomic framework</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">Library Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mgkit.html">mgkit package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.align.html">mgkit.align module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.consts.html">mgkit.consts module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.html">mgkit.counts package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.func.html">mgkit.counts.func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.glm.html">mgkit.counts.glm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.scaling.html">mgkit.counts.scaling module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.html">mgkit.db package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.dbm.html">mgkit.db.dbm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.mongo.html">mgkit.db.mongo module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.html">mgkit.filter package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.common.html">mgkit.filter.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.gff.html">mgkit.filter.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.lists.html">mgkit.filter.lists module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.reads.html">mgkit.filter.reads module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.taxon.html">mgkit.filter.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.graphs.html">mgkit.graphs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.html">mgkit.io package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.blast.html">mgkit.io.blast module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fasta.html">mgkit.io.fasta module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fastq.html">mgkit.io.fastq module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.gff.html">mgkit.io.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.glimmer.html">mgkit.io.glimmer module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.snpdat.html">mgkit.io.snpdat module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.uniprot.html">mgkit.io.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.utils.html">mgkit.io.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.kegg.html">mgkit.kegg module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.logger.html">mgkit.logger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.html">mgkit.mappings package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.cazy.html">mgkit.mappings.cazy module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.eggnog.html">mgkit.mappings.eggnog module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.enzyme.html">mgkit.mappings.enzyme module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.go.html">mgkit.mappings.go module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.pandas_map.html">mgkit.mappings.pandas_map module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.taxon.html">mgkit.mappings.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.utils.html">mgkit.mappings.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.html">mgkit.net package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.embl.html">mgkit.net.embl module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.pfam.html">mgkit.net.pfam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.uniprot.html">mgkit.net.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.utils.html">mgkit.net.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.html">mgkit.plots package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.abund.html">mgkit.plots.abund module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.boxplot.html">mgkit.plots.boxplot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.colors.html">mgkit.plots.colors module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.heatmap.html">mgkit.plots.heatmap module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.unused.html">mgkit.plots.unused module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.utils.html">mgkit.plots.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.simple_cache.html">mgkit.simple_cache module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.html">mgkit.snps package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.classes.html">mgkit.snps.classes module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.conv_func.html">mgkit.snps.conv_func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.filter.html">mgkit.snps.filter module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">mgkit.snps.funcs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.mapper.html">mgkit.snps.mapper module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.taxon.html">mgkit.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.html">mgkit.utils package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.common.html">mgkit.utils.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.dictionary.html">mgkit.utils.dictionary module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.sequence.html">mgkit.utils.sequence module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.trans_tables.html">mgkit.utils.trans_tables module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.html">mgkit.workflow package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.add_gff_info.html">mgkit.workflow.add_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.alg_utils.html">mgkit.workflow.alg_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.blast2gff.html">mgkit.workflow.blast2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.count_utils.html">mgkit.workflow.count_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.dict_utils.html">mgkit.workflow.dict_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.edit_gff.html">mgkit.workflow.edit_gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.extract_gff_info.html">mgkit.workflow.extract_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fasta_utils.html">mgkit.workflow.fasta_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fastq_utils.html">mgkit.workflow.fastq_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.filter_gff.html">mgkit.workflow.filter_gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.hmmer2gff.html">mgkit.workflow.hmmer2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.json2gff.html">mgkit.workflow.json2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.pnps_gen.html">mgkit.workflow.pnps_gen module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.sampling_utils.html">mgkit.workflow.sampling_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.snp_parser.html">mgkit.workflow.snp_parser module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.taxon_utils.html">mgkit.workflow.taxon_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.utils.html">mgkit.workflow.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html">mgkit</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="../library.html">Library Reference</a><ul>
  <li><a href="mgkit.html">mgkit package</a><ul>
  <li><a href="mgkit.snps.html">mgkit.snps package</a><ul>
      <li>Previous: <a href="mgkit.snps.filter.html" title="previous chapter">mgkit.snps.filter module</a></li>
      <li>Next: <a href="mgkit.snps.mapper.html" title="next chapter">mgkit.snps.mapper module</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2013-2020, Francesco Rubino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/api/mgkit.snps.funcs.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>