

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>mgkit.align module &mdash; MGKit: Metagenomic framework 0.5.4 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="mgkit.consts module" href="mgkit.consts.html" />
    <link rel="prev" title="mgkit.workflow.utils module" href="mgkit.workflow.utils.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> MGKit: Metagenomic framework
          

          
          </a>

          
            
            
              <div class="version">
                0.5.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">Library Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mgkit.html">mgkit package</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">mgkit.align module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.consts.html">mgkit.consts module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.html">mgkit.counts package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.func.html">mgkit.counts.func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.glm.html">mgkit.counts.glm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.scaling.html">mgkit.counts.scaling module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.html">mgkit.db package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.dbm.html">mgkit.db.dbm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.mongo.html">mgkit.db.mongo module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.html">mgkit.filter package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.common.html">mgkit.filter.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.gff.html">mgkit.filter.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.lists.html">mgkit.filter.lists module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.reads.html">mgkit.filter.reads module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.taxon.html">mgkit.filter.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.graphs.html">mgkit.graphs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.html">mgkit.io package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.blast.html">mgkit.io.blast module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fasta.html">mgkit.io.fasta module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fastq.html">mgkit.io.fastq module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.gff.html">mgkit.io.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.glimmer.html">mgkit.io.glimmer module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.snpdat.html">mgkit.io.snpdat module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.uniprot.html">mgkit.io.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.utils.html">mgkit.io.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.kegg.html">mgkit.kegg module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.logger.html">mgkit.logger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.html">mgkit.mappings package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.cazy.html">mgkit.mappings.cazy module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.eggnog.html">mgkit.mappings.eggnog module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.enzyme.html">mgkit.mappings.enzyme module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.go.html">mgkit.mappings.go module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.pandas_map.html">mgkit.mappings.pandas_map module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.taxon.html">mgkit.mappings.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.utils.html">mgkit.mappings.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.html">mgkit.net package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.embl.html">mgkit.net.embl module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.pfam.html">mgkit.net.pfam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.uniprot.html">mgkit.net.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.utils.html">mgkit.net.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.html">mgkit.plots package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.abund.html">mgkit.plots.abund module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.boxplot.html">mgkit.plots.boxplot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.colors.html">mgkit.plots.colors module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.heatmap.html">mgkit.plots.heatmap module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.unused.html">mgkit.plots.unused module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.utils.html">mgkit.plots.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.simple_cache.html">mgkit.simple_cache module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.html">mgkit.snps package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.classes.html">mgkit.snps.classes module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.conv_func.html">mgkit.snps.conv_func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.filter.html">mgkit.snps.filter module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.funcs.html">mgkit.snps.funcs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.mapper.html">mgkit.snps.mapper module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.taxon.html">mgkit.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.html">mgkit.utils package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.common.html">mgkit.utils.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.dictionary.html">mgkit.utils.dictionary module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.sequence.html">mgkit.utils.sequence module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.trans_tables.html">mgkit.utils.trans_tables module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.html">mgkit.workflow package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.add_gff_info.html">mgkit.workflow.add_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.blast2gff.html">mgkit.workflow.blast2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.extract_gff_info.html">mgkit.workflow.extract_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fasta_utils.html">mgkit.workflow.fasta_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fastq_utils.html">mgkit.workflow.fastq_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.filter_gff.html">mgkit.workflow.filter_gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.hmmer2gff.html">mgkit.workflow.hmmer2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.json2gff.html">mgkit.workflow.json2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.sampling_utils.html">mgkit.workflow.sampling_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.snp_parser.html">mgkit.workflow.snp_parser module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.taxon_utils.html">mgkit.workflow.taxon_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.utils.html">mgkit.workflow.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html">mgkit</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MGKit: Metagenomic framework</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../library.html">Library Reference</a> &raquo;</li>
        
          <li><a href="mgkit.html">mgkit package</a> &raquo;</li>
        
      <li>mgkit.align module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/api/mgkit.align.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="module-mgkit.align">
<span id="mgkit-align-module"></span><h1>mgkit.align module<a class="headerlink" href="#module-mgkit.align" title="Permalink to this headline">¶</a></h1>
<p>Module dealing with BAM/SAM files</p>
<dl class="py class">
<dt id="mgkit.align.SamtoolsDepth">
<em class="property">class </em><code class="sig-prename descclassname">mgkit.align.</code><code class="sig-name descname">SamtoolsDepth</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_handle</span></em>, <em class="sig-param"><span class="n">num_seqs</span><span class="o">=</span><span class="default_value">10000</span></em>, <em class="sig-param"><span class="n">max_size</span><span class="o">=</span><span class="default_value">1000000</span></em>, <em class="sig-param"><span class="n">max_size_dict</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">calc_density</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">dtype</span><span class="o">=</span><span class="default_value">'uint32'</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/align.html#SamtoolsDepth"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.align.SamtoolsDepth" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.4.2: </span>several optimisations and changes to support a scanning approach,
instead of lookup table. No exception is raised when a sequence is not
found in the file, instead assuming that the coverage is 0</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.4.0: </span>uses pandas.SparseArray now. It should use less memory, but needs
pandas version &gt; 0.24</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
<p>This a class that helps use the results from <a class="reference internal" href="#mgkit.align.read_samtools_depth" title="mgkit.align.read_samtools_depth"><code class="xref py py-func docutils literal notranslate"><span class="pre">read_samtools_depth()</span></code></a>.</p>
<p>There are 2 modes of operations:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Request a region coverage via <a class="reference internal" href="#mgkit.align.SamtoolsDepth.region_coverage" title="mgkit.align.SamtoolsDepth.region_coverage"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SamtoolsDepth.region_coverage()</span></code></a></p></li>
<li><p>Advance the <cite>samtools depth</cite> file until a sequence coverage is read</p></li>
</ol>
</div></blockquote>
<p>The method 1) was the default in MGKit &lt; 0.4.2, when the user would request
a region coverage and this class would advance the reading until the
requested region is found. While scanning, all the sequences encountered
are kept as pandas Series with SparseArray declared, reindexed from 0 to
the <em>max_size_dict</em> values if provided, or <em>max_size</em>. The advantage is that
the it’s easier to use, but with bigger datasets will keep high memory
usage. This is the case for case 2) which involves calling directly
<a class="reference internal" href="#mgkit.align.SamtoolsDepth.advance_file" title="mgkit.align.SamtoolsDepth.advance_file"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SamtoolsDepth.advance_file()</span></code></a> returning the sequence ID read. Then the
region coverage can be requested the same way as before, but won’t involve
scanning the file.</p>
<p>The use of a SparseArray in a pandas Series allows the use of
<cite>samtools depth</cite> files that weren’t produced with <cite>samtools depth -aa</cite> and
save memory.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The amount of memory used is fairly high, especially with the increasing
number of sequences in a GFF/Depth file. It is recoommended to use
max_size_dict to 1) only creates arrays for sequences needed and 2) use
arrays of smaller size</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>In MGKit 0.4.2, the internal dictionary to keep the SparseArray(s)
is a <a class="reference external" href="https://docs.python.org/3/library/weakref.html#weakref.WeakValueDictionary" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">weakref.WeakValueDictionary</span></code></a>, to improve the release
of memory. That on Linux seems to release the array <strong>too fast</strong>.
Upgrade to MGKit version 0.4.3 if this class is needed</p>
</div>
<dl class="py method">
<dt id="mgkit.align.SamtoolsDepth.advance_file">
<code class="sig-name descname">advance_file</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/align.html#SamtoolsDepth.advance_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.align.SamtoolsDepth.advance_file" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.4.2.</span></p>
</div>
<p>Requests the scan of the file and returns the sequence found, the
coverage can then be retrivied with <a class="reference internal" href="#mgkit.align.SamtoolsDepth.region_coverage" title="mgkit.align.SamtoolsDepth.region_coverage"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SamtoolsDepth.region_coverage()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the return value is None if the file is fully read,
otherwise the the sequnece ID found is returned</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="mgkit.align.SamtoolsDepth.closed">
<code class="sig-name descname">closed</code><em class="property"> = False</em><a class="headerlink" href="#mgkit.align.SamtoolsDepth.closed" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="mgkit.align.SamtoolsDepth.data">
<code class="sig-name descname">data</code><em class="property"> = None</em><a class="headerlink" href="#mgkit.align.SamtoolsDepth.data" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="mgkit.align.SamtoolsDepth.density">
<code class="sig-name descname">density</code><em class="property"> = None</em><a class="headerlink" href="#mgkit.align.SamtoolsDepth.density" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="mgkit.align.SamtoolsDepth.drop_sequence">
<code class="sig-name descname">drop_sequence</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq_id</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/align.html#SamtoolsDepth.drop_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.align.SamtoolsDepth.drop_sequence" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.4.2.</span></p>
</div>
<p>Remove the sequence passed from the internal dictionary</p>
</dd></dl>

<dl class="py attribute">
<dt id="mgkit.align.SamtoolsDepth.file_handle">
<code class="sig-name descname">file_handle</code><em class="property"> = None</em><a class="headerlink" href="#mgkit.align.SamtoolsDepth.file_handle" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="mgkit.align.SamtoolsDepth.max_size">
<code class="sig-name descname">max_size</code><em class="property"> = None</em><a class="headerlink" href="#mgkit.align.SamtoolsDepth.max_size" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="mgkit.align.SamtoolsDepth.max_size_dict">
<code class="sig-name descname">max_size_dict</code><em class="property"> = None</em><a class="headerlink" href="#mgkit.align.SamtoolsDepth.max_size_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="mgkit.align.SamtoolsDepth.not_found">
<code class="sig-name descname">not_found</code><em class="property"> = None</em><a class="headerlink" href="#mgkit.align.SamtoolsDepth.not_found" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="mgkit.align.SamtoolsDepth.region_coverage">
<code class="sig-name descname">region_coverage</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq_id</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/align.html#SamtoolsDepth.region_coverage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.align.SamtoolsDepth.region_coverage" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.4.2: </span>now using <a class="reference internal" href="#mgkit.align.SamtoolsDepth.advance_file" title="mgkit.align.SamtoolsDepth.advance_file"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SamtoolsDepth.advance_file()</span></code></a> to scan the file</p>
</div>
<p>Returns the mean coverage of a region. The <em>start</em> and <em>end</em> parameters
are expected to be 1-based coordinates, like the correspondent
attributes in <a class="reference internal" href="mgkit.io.gff.html#mgkit.io.gff.Annotation" title="mgkit.io.gff.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.Annotation</span></code></a> or
<a class="reference internal" href="mgkit.io.gff.html#mgkit.io.gff.GenomicRange" title="mgkit.io.gff.GenomicRange"><code class="xref py py-class docutils literal notranslate"><span class="pre">mgkit.io.gff.GenomicRange</span></code></a>.</p>
<p>If the sequence for which the coverage is requested is not found, the
<em>depth</em> file is read (and cached) until it is found.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – sequence for which to return mean coverage</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – start of the region</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – end of the region</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>mean coverage of the requested region</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="mgkit.align.add_coverage_info">
<code class="sig-prename descclassname">mgkit.align.</code><code class="sig-name descname">add_coverage_info</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">annotations</span></em>, <em class="sig-param"><span class="n">bam_files</span></em>, <em class="sig-param"><span class="n">samples</span></em>, <em class="sig-param"><span class="n">attr_suff</span><span class="o">=</span><span class="default_value">'_cov'</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/align.html#add_coverage_info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.align.add_coverage_info" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.4: </span>the coverage now is returned as floats instead of int</p>
</div>
<p>Adds coverage information to annotations, using BAM files.</p>
<p>The coverage information is added for each sample as a ‘sample_cov’ and the
total coverage as as ‘cov’ attribute in the annotations.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The bam_files and sample variables must have the same order</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>annotations</strong> (<em>iterable</em>) – iterable of annotations</p></li>
<li><p><strong>bam_files</strong> (<em>iterable</em>) – iterable of <code class="xref py py-class docutils literal notranslate"><span class="pre">pysam.Samfile</span></code> instances</p></li>
<li><p><strong>sample</strong> (<em>iterable</em>) – names of the samples for the BAM files</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.align.covered_annotation_bp">
<code class="sig-prename descclassname">mgkit.align.</code><code class="sig-name descname">covered_annotation_bp</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">files</span></em>, <em class="sig-param"><span class="n">annotations</span></em>, <em class="sig-param"><span class="n">min_cov</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">progress</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/align.html#covered_annotation_bp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.align.covered_annotation_bp" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.14.</span></p>
</div>
<p>Returns the number of base pairs covered of annotations over multiple
samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>files</strong> (<em>iterable</em>) – an iterable that returns the alignment file names</p></li>
<li><p><strong>annotations</strong> (<em>iterable</em>) – an iterable that returns annotations</p></li>
<li><p><strong>min_cov</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – minumum coverage for a base to counted</p></li>
<li><p><strong>progress</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – if <em>True</em>, a progress bar is used</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a dictionary whose keys are the uid and the values the number of
bases that are covered by reads among all samples</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.align.get_region_coverage">
<code class="sig-prename descclassname">mgkit.align.</code><code class="sig-name descname">get_region_coverage</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">bam_file</span></em>, <em class="sig-param"><span class="n">seq_id</span></em>, <em class="sig-param"><span class="n">feat_from</span></em>, <em class="sig-param"><span class="n">feat_to</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/align.html#get_region_coverage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.align.get_region_coverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Return coverage for an annotation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>feat_from and feat_to are 1-based indexes</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_file</strong> (<em>Samfile</em>) – instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">pysam.Samfile</span></code></p></li>
<li><p><strong>seq_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – sequence id</p></li>
<li><p><strong>feat_from</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – start position of feature</p></li>
<li><p><strong>feat_to</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – end position of feature</p></li>
</ul>
</dd>
<dt class="field-even">Return int</dt>
<dd class="field-even"><p>coverage array for the annotation</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.align.read_samtools_depth">
<code class="sig-prename descclassname">mgkit.align.</code><code class="sig-name descname">read_samtools_depth</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_handle</span></em>, <em class="sig-param"><span class="n">num_seqs</span><span class="o">=</span><span class="default_value">10000</span></em>, <em class="sig-param"><span class="n">seq_ids</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/align.html#read_samtools_depth"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.align.read_samtools_depth" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.4.2: </span>the function returns <strong>lists</strong> instead of numpy arrays for speed (at
least in my tests it seems ~4x increase)</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.4.0: </span>now returns 3 array, instead of 2. Also added <em>seq_ids</em> to skip lines</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.4: </span><em>num_seqs</em> can be None to avoid a log message</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
<p>Reads a samtools <em>depth</em> file, returning a generator that yields the
array of each base coverage on a per-sequence base.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>There’s no need anymore to use <cite>samtools depth -aa</cite>, because the
function returns the position array and this can be used to create a
Pandas SparseArray which can be reindexed to include missing positions
(with values of 0)</p>
<p><strong>Valid for version &lt; 0.4.0</strong>:</p>
<p>The information on position is not used, to use numpy and save memory.
samtools <em>depth</em> should be called with the <cite>-aa</cite> option:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>`samtools depth -aa bamfile`
</pre></div>
</div>
<p>This options will output both base position with 0 coverage and
sequneces with no aligned reads</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_handle</strong> (<em>file</em>) – file handle of the coverage file</p></li>
<li><p><strong>num_seqs</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – number of sequence that fires a log message. If
None, no message is triggered</p></li>
<li><p><strong>seq_ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#set" title="(in Python v3.9)"><em>set</em></a>) – a hashed container like a dictionary or set with
the sequences to return</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><em>tuple</em> – the first element is the sequence identifier, the second one
is the list with the positions, the third element is the list with the
coverages</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="mgkit.consts.html" class="btn btn-neutral float-right" title="mgkit.consts module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="mgkit.workflow.utils.html" class="btn btn-neutral float-left" title="mgkit.workflow.utils module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2013-2020, Francesco Rubino

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>