
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>mgkit.workflow.taxon_utils module &#8212; MGKit: Metagenomic framework 0.5.5 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="mgkit.workflow.utils module" href="mgkit.workflow.utils.html" />
    <link rel="prev" title="mgkit.workflow.snp_parser module" href="mgkit.workflow.snp_parser.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-mgkit.workflow.taxon_utils">
<span id="mgkit-workflow-taxon-utils-module"></span><h1>mgkit.workflow.taxon_utils module<a class="headerlink" href="#module-mgkit.workflow.taxon_utils" title="Permalink to this headline">¶</a></h1>
<p>The script contains commands used to access functionality related to
taxonomy, without the need to write ad-hoc code for functionality that
can be part of a workflow. One example is access to the the last common
ancestor function contained in the <a class="reference internal" href="mgkit.taxon.html#module-mgkit.taxon" title="mgkit.taxon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">mgkit.taxon</span></code></a>.</p>
<div class="section" id="last-common-ancestor-lca-and-lca-line">
<h2>Last Common Ancestor (lca and lca_line)<a class="headerlink" href="#last-common-ancestor-lca-and-lca-line" title="Permalink to this headline">¶</a></h2>
<p>These commands expose the functionality of
<code class="xref py py-func docutils literal notranslate"><span class="pre">last_common_ancestor_multiple()</span></code>, making it accessible via the command
line. They differ in the input file format and the choice of output files.</p>
<p>the <em>lca</em> command can be used to define the last common ancestor of contigs
from the annotation in a GFF file. The command uses the <em>taxon_ids</em> from all
annotations belonging to a contig/sequence, if they have a <strong>bitscore</strong> higher
or equal to the one passed (50 by default). The default output of the command
is a tab separated file where the first column is the contig/sequence name,
the taxon_id of the last common ancestor, its scientific/common name and its
lineage.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">contig_21</span>   <span class="mi">172788</span>  <span class="n">uncultured</span> <span class="n">phototrophic</span> <span class="n">eukaryote</span>   <span class="n">cellular</span> <span class="n">organisms</span><span class="p">,</span><span class="n">environmental</span> <span class="n">samples</span>
</pre></div>
</div>
<p>If the <em>-r</em> is used, by passing the fasta file containing the nucleotide
sequences the output file is a GFF where for each an annotation for the full
contig length contains the same information of the tab separated file format.</p>
<p>The <strong>lca_line</strong> command accept as input a file where each line consist of a
list of taxon_ids. The separator for the list can be changed and it defaults to
TAB. The last common ancestor for all taxa on a line is searched. The ouput of
this command is the same as the tab separated file of the <strong>lca</strong> command, with
the difference that instead of the first column, which in this command becames
a list of all <em>taxon_ids</em> that were used to find the last common ancestor for
that line. The list of <em>taxon_ids</em> is separated by semicolon “;”.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Both also accept the <em>-n</em> option, to report the config/line and the
taxon_ids that had no common ancestors. These are treated as errors and do
not appear in the output file.</p>
</div>
<div class="section" id="krona-output">
<h3>Krona Output<a class="headerlink" href="#krona-output" title="Permalink to this headline">¶</a></h3>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
<p>The <em>lca</em> command supports the writing of a file compatible with Krona. The
output file can be used with the <em>ktImportText/ImportText.pl</em> script included
with <a class="reference external" href="https://github.com/marbl/Krona/wiki">KronaTools</a>. Specifically, the
output from <em>taxon_utils</em> will be a file with all the lineages found (tab
separated), that can be used with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ktImportText -q taxon_utils_ouput
</pre></div>
</div>
<p>Note the use of <em>-q</em> to make the script count the lineages. Sequences with no
LCA found will be marked as <em>No LCA</em> in the graph, the <em>-n</em> is not required.</p>
<p>If a fasta file is passed, the format is changed to add the number of base pairs
for each contig, showing the number of bases for each taxonomic assignment. The
option <cite>-kt</cite> is not needed in this case. In that case, to generate a Krona plot:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ktImportText taxon_utils_ouput
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Please note that the output won’t include any sequence that didn’t have a
hit with the software used. If that’s important, the <strong>-kt</strong> option can be
used to add a number of <em>Unknown</em> lines at the end, to read the total
supplied.</p>
</div>
</div>
</div>
<div class="section" id="filter-by-taxon">
<h2>Filter by Taxon<a class="headerlink" href="#filter-by-taxon" title="Permalink to this headline">¶</a></h2>
<p>The <strong>filter</strong> command of this script allows to filter a GFF file using the
<em>taxon_id</em> attribute to include only some annotations, or exclude some. The
filter is based on the <cite>mgkit.taxon.is_ancestor</cite> function, and the
<cite>mgkit.filter.taxon.filter_taxon_by_id_list</cite>. It can also filter a table (tab
separated values) when the first element is an ID and the second is a taxon_id.
An example of a table of this sort is the output of the <cite>download-ncbi-taxa.sh</cite>
and <cite>download-uniprot-taxa.sh</cite>, where each accession of a database is associated
to a taxon_id.</p>
<p>Multiple taxon_id can be passed, either for inclusion or exclusion. If both
exclusion and inclusion is used, the first check is on the inclusion and then on
the exclusion. In alternative to passing taxon_id, taxon_names can be passed,
with values such as ‘cellular organisms’ that needs to be quoted. Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ taxon-utils filter -i 2 -in archaea -en prevotella -t taxonomy.pickle in.gff out.gff
</pre></div>
</div>
<p>Which will keep only line that are from Bacteria (taxon_id=2) and exclude those
from the genus <em>Prevotella</em>. It will be also include Archaea.</p>
<p>Multiple inclusion and exclusion flags can be put:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ taxon-utils filter -i 2 -i 2172 -t taxonomy in.gff out.gff
</pre></div>
</div>
<p>In particular, the inclusion flag is tested first and then the exclusion is
tested. So a line like this one:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">printf</span> <span class="s2">&quot;TEST\t838\nTEST\t1485&quot;</span> <span class="p">|</span> taxon-utils filter -p -t taxonomy.pickle -i <span class="m">2</span> -i <span class="m">1485</span> -e <span class="m">838</span>
</pre></div>
</div>
<p>Will produce <strong>TEST 1485</strong>, because both Prevotella (838) and Clostridium (1485)
are Bacteria (2) OR Prevotella, but Prevotella must be excluded according to
the exclusion option. This line also illustrate that a tab-separated file, where
the second column contains taxon IDs, can be filtered. In particular it can be
applied to files produced by <cite>download-ncbi-taxa.sh</cite> or
<cite>download-uniprot-taxa.sh</cite> (see <a class="reference internal" href="../scripts/download-taxonomy.html#download-taxonomy"><span class="std std-ref">Download Taxonomy</span></a>).</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Annotations with no taxon_id are not included in the output of both filters</p>
</div>
</div>
<div class="section" id="convert-taxa-tables-to-hdf5">
<h2>Convert Taxa Tables to HDF5<a class="headerlink" href="#convert-taxa-tables-to-hdf5" title="Permalink to this headline">¶</a></h2>
<p>This command is used to convert the taxa tables download from Uniprot and NCBI,
using the scripts mentioned in <span class="xref std std-ref">download-data</span>,
<cite>download-uniprot-taxa.sh</cite> and <cite>download-ncbi-taxa</cite> into a HDF5 file that can
be used with the <em>addtaxa</em> command in <a class="reference internal" href="../scripts/add-gff-info.html#add-gff-info"><span class="std std-ref">add-gff-info - Add informations to GFF annotations</span></a>.</p>
<p>The advantage is a faster lookup of the IDs. The other is a smaller memory
footprint when a great number of annotations are kept in memory.</p>
</div>
<div class="section" id="extract-taxonomy">
<h2>Extract taxonomy<a class="headerlink" href="#extract-taxonomy" title="Permalink to this headline">¶</a></h2>
<p>This command allows to print the taxonomy in a file created with MGKit. The
default behaviour is to print the scientific name, taxon_id, rank and lineage
(only ranked taxa) in a tab separated file (or standard output).</p>
<div class="section" id="match-names">
<h3>Match names<a class="headerlink" href="#match-names" title="Permalink to this headline">¶</a></h3>
<p>Besides the option to change the column separator, it is possible to only
print specific taxa (case insensitive search, but need the correct name) and
print headers for the output table. A partial and fuzzy search are performed,
however these are only reported, unless option <cite>-p</cite> is passed.</p>
<p>The lineage is printed only for the taxa requested, but all children can be
included by using the <cite>-c</cite> option.</p>
</div>
</div>
<div class="section" id="changes">
<h2>Changes<a class="headerlink" href="#changes" title="Permalink to this headline">¶</a></h2>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.5.0: </span>added <em>get</em> command to taxon-utils to print the taxonomy or search in it</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.4: </span>changed interface and behaviour for <em>filter</em>, also now can filter tables;
<em>lca</em> has changed the interface and allows the output of a 2 column table</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.1: </span>added <em>to_hdf</em> command</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.1: </span>added <em>-j</em> option to <em>lca</em>, which outputs a JSON file with the LCA results</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.0: </span>added <em>-k</em> and <em>-kt</em> options for Krona output, lineage now includes the LCA
also added <em>-a</em> option to select between lineages with only ranked taxa.
Now it defaults to all components.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.6: </span>added <em>feat-type</em> option to <em>lca</em> command, added phylum output to nolca</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.2.5.</span></p>
</div>
<dl class="py function">
<dt id="mgkit.workflow.taxon_utils.get_taxon_info">
<code class="sig-prename descclassname">mgkit.workflow.taxon_utils.</code><code class="sig-name descname">get_taxon_info</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">taxon_id</span></em>, <em class="sig-param"><span class="n">only_ranked</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/taxon_utils.html#get_taxon_info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.taxon_utils.get_taxon_info" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.taxon_utils.output_taxon_line">
<code class="sig-prename descclassname">mgkit.workflow.taxon_utils.</code><code class="sig-name descname">output_taxon_line</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">taxon_id</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">'\t'</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/taxon_utils.html#output_taxon_line"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.taxon_utils.output_taxon_line" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.taxon_utils.validate_taxon_ids">
<code class="sig-prename descclassname">mgkit.workflow.taxon_utils.</code><code class="sig-name descname">validate_taxon_ids</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">taxon_ids</span></em>, <em class="sig-param"><span class="n">taxonomy</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/taxon_utils.html#validate_taxon_ids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.taxon_utils.validate_taxon_ids" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.taxon_utils.validate_taxon_names">
<code class="sig-prename descclassname">mgkit.workflow.taxon_utils.</code><code class="sig-name descname">validate_taxon_names</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">taxon_names</span></em>, <em class="sig-param"><span class="n">taxonomy</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/taxon_utils.html#validate_taxon_names"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.taxon_utils.validate_taxon_names" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.taxon_utils.write_json">
<code class="sig-prename descclassname">mgkit.workflow.taxon_utils.</code><code class="sig-name descname">write_json</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">lca_dict</span></em>, <em class="sig-param"><span class="n">seq_id</span></em>, <em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">taxon_id</span></em>, <em class="sig-param"><span class="n">only_ranked</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/taxon_utils.html#write_json"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.taxon_utils.write_json" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.taxon_utils.write_krona">
<code class="sig-prename descclassname">mgkit.workflow.taxon_utils.</code><code class="sig-name descname">write_krona</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_handle</span></em>, <em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">taxon_id</span></em>, <em class="sig-param"><span class="n">only_ranked</span></em>, <em class="sig-param"><span class="n">base_pairs</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/taxon_utils.html#write_krona"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.taxon_utils.write_krona" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.taxon_utils.write_lca_gff">
<code class="sig-prename descclassname">mgkit.workflow.taxon_utils.</code><code class="sig-name descname">write_lca_gff</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_handle</span></em>, <em class="sig-param"><span class="n">seq_id</span></em>, <em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">taxon_id</span></em>, <em class="sig-param"><span class="n">taxon_name</span></em>, <em class="sig-param"><span class="n">lineage</span></em>, <em class="sig-param"><span class="n">feat_type</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/taxon_utils.html#write_lca_gff"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.taxon_utils.write_lca_gff" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.taxon_utils.write_lca_tab">
<code class="sig-prename descclassname">mgkit.workflow.taxon_utils.</code><code class="sig-name descname">write_lca_tab</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_handle</span></em>, <em class="sig-param"><span class="n">seq_id</span></em>, <em class="sig-param"><span class="n">taxon_id</span></em>, <em class="sig-param"><span class="n">taxon_name</span></em>, <em class="sig-param"><span class="n">rank</span></em>, <em class="sig-param"><span class="n">lineage</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/taxon_utils.html#write_lca_tab"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.taxon_utils.write_lca_tab" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.taxon_utils.write_lca_tab_simple">
<code class="sig-prename descclassname">mgkit.workflow.taxon_utils.</code><code class="sig-name descname">write_lca_tab_simple</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_handle</span></em>, <em class="sig-param"><span class="n">seq_id</span></em>, <em class="sig-param"><span class="n">taxon_id</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/taxon_utils.html#write_lca_tab_simple"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.taxon_utils.write_lca_tab_simple" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.taxon_utils.write_no_lca">
<code class="sig-prename descclassname">mgkit.workflow.taxon_utils.</code><code class="sig-name descname">write_no_lca</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_handle</span></em>, <em class="sig-param"><span class="n">seq_id</span></em>, <em class="sig-param"><span class="n">taxon_ids</span></em>, <em class="sig-param"><span class="n">extra</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/taxon_utils.html#write_no_lca"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.taxon_utils.write_no_lca" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">MGKit: Metagenomic framework</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">Library Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mgkit.html">mgkit package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.align.html">mgkit.align module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.consts.html">mgkit.consts module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.html">mgkit.counts package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.func.html">mgkit.counts.func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.glm.html">mgkit.counts.glm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.scaling.html">mgkit.counts.scaling module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.html">mgkit.db package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.dbm.html">mgkit.db.dbm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.mongo.html">mgkit.db.mongo module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.html">mgkit.filter package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.common.html">mgkit.filter.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.gff.html">mgkit.filter.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.lists.html">mgkit.filter.lists module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.reads.html">mgkit.filter.reads module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.taxon.html">mgkit.filter.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.graphs.html">mgkit.graphs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.html">mgkit.io package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.blast.html">mgkit.io.blast module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fasta.html">mgkit.io.fasta module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fastq.html">mgkit.io.fastq module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.gff.html">mgkit.io.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.glimmer.html">mgkit.io.glimmer module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.snpdat.html">mgkit.io.snpdat module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.uniprot.html">mgkit.io.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.utils.html">mgkit.io.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.kegg.html">mgkit.kegg module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.logger.html">mgkit.logger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.html">mgkit.mappings package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.cazy.html">mgkit.mappings.cazy module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.eggnog.html">mgkit.mappings.eggnog module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.enzyme.html">mgkit.mappings.enzyme module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.go.html">mgkit.mappings.go module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.pandas_map.html">mgkit.mappings.pandas_map module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.taxon.html">mgkit.mappings.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.utils.html">mgkit.mappings.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.html">mgkit.net package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.embl.html">mgkit.net.embl module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.pfam.html">mgkit.net.pfam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.uniprot.html">mgkit.net.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.utils.html">mgkit.net.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.html">mgkit.plots package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.abund.html">mgkit.plots.abund module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.boxplot.html">mgkit.plots.boxplot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.colors.html">mgkit.plots.colors module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.heatmap.html">mgkit.plots.heatmap module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.unused.html">mgkit.plots.unused module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.utils.html">mgkit.plots.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.simple_cache.html">mgkit.simple_cache module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.html">mgkit.snps package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.classes.html">mgkit.snps.classes module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.conv_func.html">mgkit.snps.conv_func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.filter.html">mgkit.snps.filter module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.funcs.html">mgkit.snps.funcs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.mapper.html">mgkit.snps.mapper module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.taxon.html">mgkit.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.html">mgkit.utils package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.common.html">mgkit.utils.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.dictionary.html">mgkit.utils.dictionary module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.sequence.html">mgkit.utils.sequence module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.trans_tables.html">mgkit.utils.trans_tables module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.html">mgkit.workflow package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.add_gff_info.html">mgkit.workflow.add_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.blast2gff.html">mgkit.workflow.blast2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.extract_gff_info.html">mgkit.workflow.extract_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fasta_utils.html">mgkit.workflow.fasta_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fastq_utils.html">mgkit.workflow.fastq_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.filter_gff.html">mgkit.workflow.filter_gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.hmmer2gff.html">mgkit.workflow.hmmer2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.json2gff.html">mgkit.workflow.json2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.sampling_utils.html">mgkit.workflow.sampling_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.snp_parser.html">mgkit.workflow.snp_parser module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">mgkit.workflow.taxon_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.utils.html">mgkit.workflow.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html">mgkit</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="../library.html">Library Reference</a><ul>
  <li><a href="mgkit.html">mgkit package</a><ul>
  <li><a href="mgkit.workflow.html">mgkit.workflow package</a><ul>
      <li>Previous: <a href="mgkit.workflow.snp_parser.html" title="previous chapter">mgkit.workflow.snp_parser module</a></li>
      <li>Next: <a href="mgkit.workflow.utils.html" title="next chapter">mgkit.workflow.utils module</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2013-2020, Francesco Rubino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/api/mgkit.workflow.taxon_utils.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>