

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>mgkit.snps.conv_func module &mdash; MGKit: Metagenomic framework 0.5.4 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="mgkit.snps.filter module" href="mgkit.snps.filter.html" />
    <link rel="prev" title="mgkit.snps.classes module" href="mgkit.snps.classes.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> MGKit: Metagenomic framework
          

          
          </a>

          
            
            
              <div class="version">
                0.5.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">Library Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mgkit.html">mgkit package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.align.html">mgkit.align module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.consts.html">mgkit.consts module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.html">mgkit.counts package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.func.html">mgkit.counts.func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.glm.html">mgkit.counts.glm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.scaling.html">mgkit.counts.scaling module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.html">mgkit.db package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.dbm.html">mgkit.db.dbm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.mongo.html">mgkit.db.mongo module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.html">mgkit.filter package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.common.html">mgkit.filter.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.gff.html">mgkit.filter.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.lists.html">mgkit.filter.lists module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.reads.html">mgkit.filter.reads module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.taxon.html">mgkit.filter.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.graphs.html">mgkit.graphs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.html">mgkit.io package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.blast.html">mgkit.io.blast module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fasta.html">mgkit.io.fasta module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fastq.html">mgkit.io.fastq module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.gff.html">mgkit.io.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.glimmer.html">mgkit.io.glimmer module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.snpdat.html">mgkit.io.snpdat module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.uniprot.html">mgkit.io.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.utils.html">mgkit.io.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.kegg.html">mgkit.kegg module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.logger.html">mgkit.logger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.html">mgkit.mappings package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.cazy.html">mgkit.mappings.cazy module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.eggnog.html">mgkit.mappings.eggnog module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.enzyme.html">mgkit.mappings.enzyme module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.go.html">mgkit.mappings.go module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.pandas_map.html">mgkit.mappings.pandas_map module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.taxon.html">mgkit.mappings.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.utils.html">mgkit.mappings.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.html">mgkit.net package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.embl.html">mgkit.net.embl module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.pfam.html">mgkit.net.pfam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.uniprot.html">mgkit.net.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.utils.html">mgkit.net.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.html">mgkit.plots package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.abund.html">mgkit.plots.abund module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.boxplot.html">mgkit.plots.boxplot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.colors.html">mgkit.plots.colors module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.heatmap.html">mgkit.plots.heatmap module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.unused.html">mgkit.plots.unused module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.utils.html">mgkit.plots.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.simple_cache.html">mgkit.simple_cache module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.html">mgkit.snps package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.classes.html">mgkit.snps.classes module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">mgkit.snps.conv_func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.filter.html">mgkit.snps.filter module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.funcs.html">mgkit.snps.funcs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.mapper.html">mgkit.snps.mapper module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.taxon.html">mgkit.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.html">mgkit.utils package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.common.html">mgkit.utils.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.dictionary.html">mgkit.utils.dictionary module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.sequence.html">mgkit.utils.sequence module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.trans_tables.html">mgkit.utils.trans_tables module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.html">mgkit.workflow package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.add_gff_info.html">mgkit.workflow.add_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.blast2gff.html">mgkit.workflow.blast2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.extract_gff_info.html">mgkit.workflow.extract_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fasta_utils.html">mgkit.workflow.fasta_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fastq_utils.html">mgkit.workflow.fastq_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.filter_gff.html">mgkit.workflow.filter_gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.hmmer2gff.html">mgkit.workflow.hmmer2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.json2gff.html">mgkit.workflow.json2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.sampling_utils.html">mgkit.workflow.sampling_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.snp_parser.html">mgkit.workflow.snp_parser module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.taxon_utils.html">mgkit.workflow.taxon_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.utils.html">mgkit.workflow.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html">mgkit</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MGKit: Metagenomic framework</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../library.html">Library Reference</a> &raquo;</li>
        
          <li><a href="mgkit.html">mgkit package</a> &raquo;</li>
        
          <li><a href="mgkit.snps.html">mgkit.snps package</a> &raquo;</li>
        
      <li>mgkit.snps.conv_func module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/api/mgkit.snps.conv_func.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="module-mgkit.snps.conv_func">
<span id="mgkit-snps-conv-func-module"></span><h1>mgkit.snps.conv_func module<a class="headerlink" href="#module-mgkit.snps.conv_func" title="Permalink to this headline">¶</a></h1>
<p>Wappers to use some of the general function of the snps package
in a simpler way.</p>
<dl class="py function">
<dt id="mgkit.snps.conv_func.get_full_dataframe">
<code class="sig-prename descclassname">mgkit.snps.conv_func.</code><code class="sig-name descname">get_full_dataframe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">snp_data</span></em>, <em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">min_num</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">index_type</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">filters</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/conv_func.html#get_full_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.conv_func.get_full_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.12.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.2: </span>added <em>filters</em> argument</p>
</div>
<p>Returns a <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> with the pN/pS of the given
SNPs data.</p>
<p>Shortcut for using <a class="reference internal" href="mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">combine_sample_snps()</span></code></a>, using
filters from <a class="reference internal" href="mgkit.snps.filter.html#mgkit.snps.filter.get_default_filters" title="mgkit.snps.filter.get_default_filters"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_default_filters()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>snp_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – dictionary sample-&gt;GeneSyn of SNPs data</p></li>
<li><p><strong>taxonomy</strong> – Uniprot Taxonomy</p></li>
<li><p><strong>min_num</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – minimum number of samples in which a valid pN/pS
is found</p></li>
<li><p><strong>index_type</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – type of index to return</p></li>
<li><p><strong>filters</strong> (<em>iterable</em>) – list of filters to apply, otherwise uses the
default filters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> of pN/pS values. The index type
is None (gene-taxon)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.snps.conv_func.get_gene_map_dataframe">
<code class="sig-prename descclassname">mgkit.snps.conv_func.</code><code class="sig-name descname">get_gene_map_dataframe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">snp_data</span></em>, <em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">gene_map</span></em>, <em class="sig-param"><span class="n">min_num</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">index_type</span><span class="o">=</span><span class="default_value">'gene'</span></em>, <em class="sig-param"><span class="n">filters</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/conv_func.html#get_gene_map_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.conv_func.get_gene_map_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.11.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.2: </span>added <em>filters</em> argument</p>
</div>
<p>Returns a <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> with the pN/pS of the given
SNPs data, mapping all taxa to the gene map.</p>
<p>Shortcut for using <a class="reference internal" href="mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">combine_sample_snps()</span></code></a>, using
filters from <a class="reference internal" href="mgkit.snps.filter.html#mgkit.snps.filter.get_default_filters" title="mgkit.snps.filter.get_default_filters"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_default_filters()</span></code></a> and as
<cite>gene_func</cite> parameter <a class="reference internal" href="mgkit.snps.mapper.html#mgkit.snps.mapper.map_gene_id" title="mgkit.snps.mapper.map_gene_id"><code class="xref py py-func docutils literal notranslate"><span class="pre">map_gene_id()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>snp_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – dictionary sample-&gt;GeneSyn of SNPs data</p></li>
<li><p><strong>taxonomy</strong> – Uniprot Taxonomy</p></li>
<li><p><strong>min_num</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – minimum number of samples in which a valid pN/pS
is found</p></li>
<li><p><strong>gene_map</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – dictionary of mapping for the gene_ids in in SNPs
data</p></li>
<li><p><strong>index_type</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – type of index to return</p></li>
<li><p><strong>filters</strong> (<em>iterable</em>) – list of filters to apply, otherwise uses the
default filters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> of pN/pS values. The index type
is ‘gene’</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.snps.conv_func.get_gene_taxon_dataframe">
<code class="sig-prename descclassname">mgkit.snps.conv_func.</code><code class="sig-name descname">get_gene_taxon_dataframe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">snp_data</span></em>, <em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">gene_map</span></em>, <em class="sig-param"><span class="n">min_num</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">rank</span><span class="o">=</span><span class="default_value">'genus'</span></em>, <em class="sig-param"><span class="n">index_type</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">filters</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_uid</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/conv_func.html#get_gene_taxon_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.conv_func.get_gene_taxon_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.12.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.2: </span>added <em>filters</em> argument</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.5.1: </span>gene_map can be <em>None</em>, use_uid can be passed to the underline function</p>
</div>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>edit docstring</p>
</div>
<p>Returns a <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> with the pN/pS of the given
SNPs data, mapping all taxa to the gene map.</p>
<p>Shortcut for using <a class="reference internal" href="mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">combine_sample_snps()</span></code></a>, using
filters from <a class="reference internal" href="mgkit.snps.filter.html#mgkit.snps.filter.get_default_filters" title="mgkit.snps.filter.get_default_filters"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_default_filters()</span></code></a> and as
<cite>gene_func</cite> parameter <a class="reference internal" href="mgkit.snps.mapper.html#mgkit.snps.mapper.map_gene_id" title="mgkit.snps.mapper.map_gene_id"><code class="xref py py-func docutils literal notranslate"><span class="pre">map_gene_id()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>snp_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – dictionary sample-&gt;GeneSyn of SNPs data</p></li>
<li><p><strong>taxonomy</strong> – Uniprot Taxonomy</p></li>
<li><p><strong>min_num</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – minimum number of samples in which a valid pN/pS
is found</p></li>
<li><p><strong>gene_map</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – dictionary of mapping for the gene_ids in in SNPs
data</p></li>
<li><p><strong>index_type</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – type of index to return</p></li>
<li><p><strong>filters</strong> (<em>iterable</em>) – list of filters to apply, otherwise uses the
default filters</p></li>
<li><p><strong>use_uid</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – instead of using <em>gene_id</em>, uses <em>uid</em> as gene ID</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> of pN/pS values. The index type
is ‘gene’</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.snps.conv_func.get_rank_dataframe">
<code class="sig-prename descclassname">mgkit.snps.conv_func.</code><code class="sig-name descname">get_rank_dataframe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">snp_data</span></em>, <em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">min_num</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">rank</span><span class="o">=</span><span class="default_value">'order'</span></em>, <em class="sig-param"><span class="n">index_type</span><span class="o">=</span><span class="default_value">'taxon'</span></em>, <em class="sig-param"><span class="n">filters</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/snps/conv_func.html#get_rank_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.snps.conv_func.get_rank_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.11.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.2: </span>added <em>filters</em> argument</p>
</div>
<p>Returns a <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> with the pN/pS of the given
SNPs data, mapping all taxa to the specified rank. Higher taxa won’t
be included.</p>
<p>Shortcut for using <a class="reference internal" href="mgkit.snps.funcs.html#mgkit.snps.funcs.combine_sample_snps" title="mgkit.snps.funcs.combine_sample_snps"><code class="xref py py-func docutils literal notranslate"><span class="pre">combine_sample_snps()</span></code></a>, using
filters from <a class="reference internal" href="mgkit.snps.filter.html#mgkit.snps.filter.get_default_filters" title="mgkit.snps.filter.get_default_filters"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_default_filters()</span></code></a> and as
<cite>taxon_func</cite> parameter <a class="reference internal" href="mgkit.snps.mapper.html#mgkit.snps.mapper.map_taxon_id_to_rank" title="mgkit.snps.mapper.map_taxon_id_to_rank"><code class="xref py py-func docutils literal notranslate"><span class="pre">map_taxon_id_to_rank()</span></code></a>,
with include_higher equals to False</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>snp_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – dictionary sample-&gt;GeneSyn of SNPs data</p></li>
<li><p><strong>taxonomy</strong> – Uniprot Taxonomy</p></li>
<li><p><strong>min_num</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – minimum number of samples in which a valid pN/pS
is found</p></li>
<li><p><strong>rank</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – taxon rank to map. Valid ranks are found in
<a class="reference internal" href="mgkit.taxon.html#mgkit.taxon.TAXON_RANKS" title="mgkit.taxon.TAXON_RANKS"><code class="xref py py-data docutils literal notranslate"><span class="pre">mgkit.taxon.TAXON_RANKS</span></code></a></p></li>
<li><p><strong>index_type</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – type of index to return</p></li>
<li><p><strong>filters</strong> (<em>iterable</em>) – list of filters to apply, otherwise uses the
default filters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> of pN/pS values. The index type
is ‘taxon’</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="mgkit.snps.filter.html" class="btn btn-neutral float-right" title="mgkit.snps.filter module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="mgkit.snps.classes.html" class="btn btn-neutral float-left" title="mgkit.snps.classes module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2013-2020, Francesco Rubino

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>