
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>mgkit.workflow.pnps_gen module &#8212; MGKit: Metagenomic framework 0.5.7 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="mgkit.workflow.sampling_utils module" href="mgkit.workflow.sampling_utils.html" />
    <link rel="prev" title="mgkit.workflow.json2gff module" href="mgkit.workflow.json2gff.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-mgkit.workflow.pnps_gen">
<span id="mgkit-workflow-pnps-gen-module"></span><h1>mgkit.workflow.pnps_gen module<a class="headerlink" href="#module-mgkit.workflow.pnps_gen" title="Permalink to this headline">¶</a></h1>
<div class="section" id="calculates-pn-ps-values">
<h2>Calculates pN/pS values<a class="headerlink" href="#calculates-pn-ps-values" title="Permalink to this headline">¶</a></h2>
<p>This script calculates pN/pS using the data produced by the script <cite>vcf</cite>
command. The result table is a CSV file.</p>
</div>
<div class="section" id="parse-vcf-files">
<h2>Parse VCF Files<a class="headerlink" href="#parse-vcf-files" title="Permalink to this headline">¶</a></h2>
<p>The <cite>vcf</cite> command will parse a VCF file to produce the pickle file that is used
to calculate the pN/pS.</p>
</div>
<div class="section" id="calculate-rank-pn-ps">
<h2>Calculate Rank pN/pS<a class="headerlink" href="#calculate-rank-pn-ps" title="Permalink to this headline">¶</a></h2>
<p>The <cite>rank</cite> command of the script reads SNPs information and calculate for each
element of a specific taxonomic rank (species, genus, family, etc.) its pN/pS.
Another option is the <em>None</em> rank, which makes the script use the taxonomic ID
found in the annotations.</p>
<p>For example, choosing the rank <em>genus</em> a table will be produced, similar to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Prevotella</span><span class="p">,</span><span class="mf">0.0001</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mf">1.1</span><span class="p">,</span><span class="mf">0.4</span>
<span class="n">Methanobrevibacter</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mf">0.5</span><span class="p">,</span><span class="mf">0.6</span><span class="p">,</span><span class="mf">0.8</span>
</pre></div>
</div>
<p>A pN/pS value for each genus and sample (4 in this case) will be calculated.</p>
<p>It is important to specify the taxonomic IDs to include in tha calculations. By
default only bacteria are included. To get those values, the taxonomy can be
queried using <cite>taxon-utils get</cite>.</p>
</div>
<div class="section" id="calculate-gene-rank-pn-ps">
<h2>Calculate Gene/Rank pN/pS<a class="headerlink" href="#calculate-gene-rank-pn-ps" title="Permalink to this headline">¶</a></h2>
<p>The <cite>full</cite> command create a gene/taxon table of pN/pS, internally is a pandas
MultiIndex DataFrame, written in CSV format after script execution. The
difference with the <cite>rank</cite> is the pN/pS calculation is for each gene/taxon and
by default the <cite>gene_id</cite> from the original GFF file is used (which is stored
in the file generated by <cite>snp_parser</cite>). If other gene IDs needs to be used, a
table file can be provided, which can be passed in two column formats.</p>
<p>The default in MGKit is to use Uniprot gene IDs for the functions, but we may
want to examine the Kegg Orthologs instead. A table can be passed where the
first column in the gene_id stored in the GFF file and the second is the KO:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Q7N6F9</span>  <span class="n">K05685</span>
<span class="n">Q7N6F9</span>  <span class="n">K01242</span>
<span class="n">G7E4F2</span>  <span class="n">K05625</span>
</pre></div>
</div>
<p>The <cite>Q7N6F9</cite> gene_id is repeated because it has multiple correspondences to KOs
and this format needs to be selected using the <cite>-2</cite> option of the command.</p>
<p>The default type of table expected by the command is a table with a gene ID as
first column one or more tab separated columns with mappings. The previous
table would look like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Q7N6F9</span>  <span class="n">K05685</span>  <span class="n">K01242</span>
<span class="n">G7E4F2</span>  <span class="n">K05625</span>
</pre></div>
</div>
<p>These tables can be created from the original GFF file, assuming that mappings
to KO, EC Numbers are included, with a command line like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">edit</span><span class="o">-</span><span class="n">gff</span> <span class="n">view</span> <span class="o">-</span><span class="n">a</span> <span class="n">gene_id</span> <span class="o">-</span><span class="n">a</span> <span class="n">map_KO</span> <span class="n">final</span><span class="o">.</span><span class="n">contigs</span><span class="o">-</span><span class="n">a3</span><span class="o">.</span><span class="n">gff</span><span class="o">.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">tr</span> <span class="s1">&#39;,&#39;</span> <span class="s1">&#39;      &#39;</span>
</pre></div>
</div>
<p>Extracting the KOs (which are comma separated in a MGKit GFF file) and changing
any comma to <em>tab</em>. This table can be passed to the script and will make it
possible to calculate the pN/pS for the KOs associated to the genes. Only gene
IDs present in this file have a calculated pN/pS.</p>
<p>Normally you combine the all isoforms with the same the <em>gene_id</em> to produce a
single pN/pS, but if it’s needed, the <cite>-u</cite> option can be used to calculate a
pN/pS for each line in the GFF file.</p>
</div>
<div class="section" id="changes">
<h2>Changes<a class="headerlink" href="#changes" title="Permalink to this headline">¶</a></h2>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.5.7: </span>added <em>vcf</em> command to parse VCF files and generate data for the script</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.5.1: </span>bug fix</p>
</div>
<div class="versionchanged">
<ul>
<span class="versionmodified changed">Changed in version 0.5.1: </span><li><p>added option to include the lineage as a string</p></li>
<li><p>added option to use the <em>uids</em> from the GFF instead of <em>gene_id</em>, this
does not require the GFF file, they are embedded into the <em>.pickle</em> file</p></li>
</ul>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.5.0.</span></p>
</div>
<dl class="py function">
<dt id="mgkit.workflow.pnps_gen.check_snp_in_set">
<code class="sig-prename descclassname">mgkit.workflow.pnps_gen.</code><code class="sig-name descname">check_snp_in_set</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span></em>, <em class="sig-param"><span class="n">snp_data</span></em>, <em class="sig-param"><span class="n">pos</span></em>, <em class="sig-param"><span class="n">change</span></em>, <em class="sig-param"><span class="n">annotations</span></em>, <em class="sig-param"><span class="n">seq</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/pnps_gen.html#check_snp_in_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.pnps_gen.check_snp_in_set" title="Permalink to this definition">¶</a></dt>
<dd><p>Used by <a class="reference internal" href="#mgkit.workflow.pnps_gen.parse_vcf" title="mgkit.workflow.pnps_gen.parse_vcf"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_vcf()</span></code></a> to check if a SNP</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<em>iterable</em>) – list of samples that contain the SNP</p></li>
<li><p><strong>snp_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a>) – dictionary from <a class="reference internal" href="#mgkit.workflow.pnps_gen.init_count_set" title="mgkit.workflow.pnps_gen.init_count_set"><code class="xref py py-func docutils literal notranslate"><span class="pre">init_count_set()</span></code></a> with per
sample SNPs information</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.pnps_gen.get_lineage">
<code class="sig-prename descclassname">mgkit.workflow.pnps_gen.</code><code class="sig-name descname">get_lineage</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">taxon_id</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/pnps_gen.html#get_lineage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.pnps_gen.get_lineage" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.pnps_gen.init_count_set">
<code class="sig-prename descclassname">mgkit.workflow.pnps_gen.</code><code class="sig-name descname">init_count_set</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">annotations</span></em>, <em class="sig-param"><span class="n">seqs</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/pnps_gen.html#init_count_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.pnps_gen.init_count_set" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.pnps_gen.init_count_set_sample_files">
<code class="sig-prename descclassname">mgkit.workflow.pnps_gen.</code><code class="sig-name descname">init_count_set_sample_files</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">annotations</span></em>, <em class="sig-param"><span class="n">seqs</span></em>, <em class="sig-param"><span class="n">cov_files</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/pnps_gen.html#init_count_set_sample_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.pnps_gen.init_count_set_sample_files" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.5.7.</span></p>
</div>
</dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.pnps_gen.parse_vcf">
<code class="sig-prename descclassname">mgkit.workflow.pnps_gen.</code><code class="sig-name descname">parse_vcf</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">vcf_handle</span></em>, <em class="sig-param"><span class="n">snp_data</span></em>, <em class="sig-param"><span class="n">annotations</span></em>, <em class="sig-param"><span class="n">seqs</span></em>, <em class="sig-param"><span class="n">min_qual</span></em>, <em class="sig-param"><span class="n">min_reads</span></em>, <em class="sig-param"><span class="n">min_freq</span></em>, <em class="sig-param"><span class="n">sample_ids</span></em>, <em class="sig-param"><span class="n">num_lines</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/pnps_gen.html#parse_vcf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.pnps_gen.parse_vcf" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.pnps_gen.read_gene_map_default">
<code class="sig-prename descclassname">mgkit.workflow.pnps_gen.</code><code class="sig-name descname">read_gene_map_default</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_handle</span></em>, <em class="sig-param"><span class="n">separator</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/pnps_gen.html#read_gene_map_default"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.pnps_gen.read_gene_map_default" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.pnps_gen.read_gene_map_two_columns">
<code class="sig-prename descclassname">mgkit.workflow.pnps_gen.</code><code class="sig-name descname">read_gene_map_two_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_handle</span></em>, <em class="sig-param"><span class="n">separator</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/pnps_gen.html#read_gene_map_two_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.pnps_gen.read_gene_map_two_columns" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="mgkit.workflow.pnps_gen.save_data">
<code class="sig-prename descclassname">mgkit.workflow.pnps_gen.</code><code class="sig-name descname">save_data</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">output_file</span></em>, <em class="sig-param"><span class="n">snp_data</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/workflow/pnps_gen.html#save_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.workflow.pnps_gen.save_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle data structures to the disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – base name for pickle files</p></li>
<li><p><strong>snp_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a>) – dictionary from <a class="reference internal" href="#mgkit.workflow.pnps_gen.init_count_set" title="mgkit.workflow.pnps_gen.init_count_set"><code class="xref py py-func docutils literal notranslate"><span class="pre">init_count_set()</span></code></a> with per
sample SNPs information</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">MGKit: Metagenomic framework</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">Library Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mgkit.html">mgkit package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.align.html">mgkit.align module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.consts.html">mgkit.consts module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.html">mgkit.counts package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.func.html">mgkit.counts.func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.glm.html">mgkit.counts.glm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.scaling.html">mgkit.counts.scaling module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.html">mgkit.db package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.dbm.html">mgkit.db.dbm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.mongo.html">mgkit.db.mongo module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.html">mgkit.filter package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.common.html">mgkit.filter.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.gff.html">mgkit.filter.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.lists.html">mgkit.filter.lists module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.reads.html">mgkit.filter.reads module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.taxon.html">mgkit.filter.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.graphs.html">mgkit.graphs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.html">mgkit.io package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.blast.html">mgkit.io.blast module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fasta.html">mgkit.io.fasta module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fastq.html">mgkit.io.fastq module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.gff.html">mgkit.io.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.glimmer.html">mgkit.io.glimmer module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.snpdat.html">mgkit.io.snpdat module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.uniprot.html">mgkit.io.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.utils.html">mgkit.io.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.kegg.html">mgkit.kegg module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.logger.html">mgkit.logger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.html">mgkit.mappings package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.cazy.html">mgkit.mappings.cazy module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.eggnog.html">mgkit.mappings.eggnog module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.enzyme.html">mgkit.mappings.enzyme module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.go.html">mgkit.mappings.go module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.pandas_map.html">mgkit.mappings.pandas_map module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.taxon.html">mgkit.mappings.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.utils.html">mgkit.mappings.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.html">mgkit.net package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.embl.html">mgkit.net.embl module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.pfam.html">mgkit.net.pfam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.uniprot.html">mgkit.net.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.utils.html">mgkit.net.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.html">mgkit.plots package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.abund.html">mgkit.plots.abund module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.boxplot.html">mgkit.plots.boxplot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.colors.html">mgkit.plots.colors module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.heatmap.html">mgkit.plots.heatmap module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.unused.html">mgkit.plots.unused module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.utils.html">mgkit.plots.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.simple_cache.html">mgkit.simple_cache module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.html">mgkit.snps package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.classes.html">mgkit.snps.classes module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.conv_func.html">mgkit.snps.conv_func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.filter.html">mgkit.snps.filter module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.funcs.html">mgkit.snps.funcs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.mapper.html">mgkit.snps.mapper module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.taxon.html">mgkit.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.html">mgkit.utils package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.common.html">mgkit.utils.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.dictionary.html">mgkit.utils.dictionary module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.sequence.html">mgkit.utils.sequence module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.trans_tables.html">mgkit.utils.trans_tables module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.html">mgkit.workflow package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.add_gff_info.html">mgkit.workflow.add_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.alg_utils.html">mgkit.workflow.alg_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.blast2gff.html">mgkit.workflow.blast2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.count_utils.html">mgkit.workflow.count_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.dict_utils.html">mgkit.workflow.dict_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.edit_gff.html">mgkit.workflow.edit_gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.extract_gff_info.html">mgkit.workflow.extract_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fasta_utils.html">mgkit.workflow.fasta_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fastq_utils.html">mgkit.workflow.fastq_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.filter_gff.html">mgkit.workflow.filter_gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.hmmer2gff.html">mgkit.workflow.hmmer2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.json2gff.html">mgkit.workflow.json2gff module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">mgkit.workflow.pnps_gen module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.sampling_utils.html">mgkit.workflow.sampling_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.snp_parser.html">mgkit.workflow.snp_parser module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.taxon_utils.html">mgkit.workflow.taxon_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.utils.html">mgkit.workflow.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html">mgkit</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="../library.html">Library Reference</a><ul>
  <li><a href="mgkit.html">mgkit package</a><ul>
  <li><a href="mgkit.workflow.html">mgkit.workflow package</a><ul>
      <li>Previous: <a href="mgkit.workflow.json2gff.html" title="previous chapter">mgkit.workflow.json2gff module</a></li>
      <li>Next: <a href="mgkit.workflow.sampling_utils.html" title="next chapter">mgkit.workflow.sampling_utils module</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2013-2020, Francesco Rubino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/api/mgkit.workflow.pnps_gen.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>