
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>mgkit.utils.sequence module &#8212; MGKit: Metagenomic framework 0.5.5 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="mgkit.utils.trans_tables module" href="mgkit.utils.trans_tables.html" />
    <link rel="prev" title="mgkit.utils.dictionary module" href="mgkit.utils.dictionary.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-mgkit.utils.sequence">
<span id="mgkit-utils-sequence-module"></span><h1>mgkit.utils.sequence module<a class="headerlink" href="#module-mgkit.utils.sequence" title="Permalink to this headline">¶</a></h1>
<p>Module containing functions related to sequence data</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For those functions without a docstring, look at the same with a
underscore (‘_’) prepended.</p>
</div>
<dl class="py class">
<dt id="mgkit.utils.sequence.Alignment">
<em class="property">class </em><code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">Alignment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#Alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.Alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Simple alignment class</p>
<dl class="py method">
<dt id="mgkit.utils.sequence.Alignment.add_seq">
<code class="sig-name descname">add_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">seq</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#Alignment.add_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.Alignment.add_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a sequence to the alignment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – name of the sequence</p></li>
<li><p><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – sequence</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="mgkit.utils.sequence.Alignment.add_seqs">
<code class="sig-name descname">add_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#Alignment.add_seqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.Alignment.add_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Add sequences to the alignment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seqs</strong> (<em>iterable</em>) – iterable that returns (name, seq)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="mgkit.utils.sequence.Alignment.get_consensus">
<code class="sig-name descname">get_consensus</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nucl</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#Alignment.get_consensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.Alignment.get_consensus" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.1.16: </span>added <em>nucl</em> parameter</p>
</div>
<p>The consensus sequence is constructed by checking the nucleotide that
has the maximum number of counts for each position in the alignment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>nucl</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – specify if the alignment is nucleotidic</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>consensus sequence</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="mgkit.utils.sequence.Alignment.get_position">
<code class="sig-name descname">get_position</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pos</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#Alignment.get_position"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.Alignment.get_position" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all characters at a position</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pos</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – position to return (0-based)</p>
</dd>
<dt class="field-even">Return str</dt>
<dd class="field-even"><p>all characters occuring at the position</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="mgkit.utils.sequence.Alignment.get_seq_len">
<code class="sig-name descname">get_seq_len</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#Alignment.get_seq_len"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.Alignment.get_seq_len" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the length of the alignment</p>
</dd></dl>

<dl class="py method">
<dt id="mgkit.utils.sequence.Alignment.get_snps">
<code class="sig-name descname">get_snps</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ref_seq</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">full_size</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#Alignment.get_snps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.Alignment.get_snps" title="Permalink to this definition">¶</a></dt>
<dd><p>A SNP is called for the nucleotide that has the most counts among the
ones that differ in the each site of the alignment. If two nucleotides
have the same maximum count, one is randomly chosen.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – a reference sequence can be provided, if None, a
consensus sequence is produced for the alignment</p></li>
<li><p><strong>full_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – if True a tuple is returned for each position in
the alignment. If there is no SNP at a position the value for the
SNP is None</p></li>
</ul>
</dd>
<dt class="field-even">Return list</dt>
<dd class="field-even"><p>a list of tuples (position, SNP)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py data">
<dt id="mgkit.utils.sequence.REV_COMP">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">REV_COMP</code><em class="property"> = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}</em><a class="headerlink" href="#mgkit.utils.sequence.REV_COMP" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary containing the complement of each nucleotide sequence</p>
</dd></dl>

<dl class="py data">
<dt id="mgkit.utils.sequence.TRANS_TABLE">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">TRANS_TABLE</code><em class="property"> = {'AAA': 'K', 'AAC': 'N', 'AAG': 'K', 'AAT': 'N', 'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACT': 'T', 'AGA': 'R', 'AGC': 'S', 'AGG': 'R', 'AGT': 'S', 'ATA': 'I', 'ATC': 'I', 'ATG': 'M', 'ATT': 'I', 'CAA': 'Q', 'CAC': 'H', 'CAG': 'Q', 'CAT': 'H', 'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCT': 'P', 'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGT': 'R', 'CTA': 'L', 'CTC': 'L', 'CTG': 'L', 'CTT': 'L', 'GAA': 'E', 'GAC': 'D', 'GAG': 'E', 'GAT': 'D', 'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A', 'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGT': 'G', 'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GTT': 'V', 'TAA': '*', 'TAC': 'Y', 'TAG': '*', 'TAT': 'Y', 'TCA': 'S', 'TCC': 'S', 'TCG': 'S', 'TCT': 'S', 'TGA': '*', 'TGC': 'C', 'TGG': 'W', 'TGT': 'C', 'TTA': 'L', 'TTC': 'F', 'TTG': 'L', 'TTT': 'F'}</em><a class="headerlink" href="#mgkit.utils.sequence.TRANS_TABLE" title="Permalink to this definition">¶</a></dt>
<dd><p>Translation table - Universal genetic code</p>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence._get_kmers">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">_get_kmers</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">k</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#_get_kmers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence._get_kmers" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.2.6.</span></p>
</div>
<p>Returns a generator, with every iteration yielding a kmer of size <em>k</em></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – sequence</p></li>
<li><p><strong>k</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – kmer size</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><em>str</em> – a portion of <em>seq</em>, of size <em>k</em> with a step of <em>1</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence._sequence_signature">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">_sequence_signature</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">w_size</span></em>, <em class="sig-param"><span class="n">k_size</span><span class="o">=</span><span class="default_value">4</span></em>, <em class="sig-param"><span class="n">step</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#_sequence_signature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence._sequence_signature" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.2.6.</span></p>
</div>
<p>Returns the signature of a sequence, based on a kmer length, over a sliding
window. Each sliding window signature is placed in order into a list, with
each element being a <a class="reference external" href="https://docs.python.org/3/library/collections.html#collections.Counter" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">collections.Counter</span></code></a> instance whose keys are
the kmer found in that window.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – sequence for which to get the signature</p></li>
<li><p><strong>w_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – size of the sliding window size</p></li>
<li><p><strong>k_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – size of the kmer to use <code class="xref py py-func docutils literal notranslate"><span class="pre">get_kmers()</span></code></p></li>
<li><p><strong>step</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – step to use in <code class="xref py py-func docutils literal notranslate"><span class="pre">sliding_window()</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of <a class="reference external" href="https://docs.python.org/3/library/collections.html#collections.Counter" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">collections.Counter</span></code></a> instances, for each
window used</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence._signatures_matrix">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">_signatures_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs</span></em>, <em class="sig-param"><span class="n">w_size</span></em>, <em class="sig-param"><span class="n">k_size</span><span class="o">=</span><span class="default_value">4</span></em>, <em class="sig-param"><span class="n">step</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#_signatures_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence._signatures_matrix" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.2.6.</span></p>
</div>
<p>Return a matrix (pandas.DataFrame) where the columns are the kmer found in
all sequences <em>seqs</em> and the rows are the a MultiIndex with the first level
being the sequnce name and the second the index of the sliding window for
which a signature was computed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<em>iterable</em>) – iterable that yields a tuple, with the first element
being the sequence name and the second the sequence itself</p></li>
<li><p><strong>w_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – size of the sliding window size</p></li>
<li><p><strong>k_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – size of the kmer to use <code class="xref py py-func docutils literal notranslate"><span class="pre">get_kmers()</span></code></p></li>
<li><p><strong>step</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – step to use in <code class="xref py py-func docutils literal notranslate"><span class="pre">sliding_window()</span></code>, defaults to half of
the window size</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a DataFrame where the columns are the kmers and the
rows are the signatures of each contigs/windows.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence._sliding_window">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">_sliding_window</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">size</span></em>, <em class="sig-param"><span class="n">step</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#_sliding_window"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence._sliding_window" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.2.6.</span></p>
</div>
<p>Returns a generator, with every iteration yielding a subsequence of size
<em>size</em>, with a step of <em>step</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – sequnece</p></li>
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – size of the sliding window</p></li>
<li><p><strong>step</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – the step to use in the sliding window. If <em>None</em>,
half of the sequence length is used</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><em>str</em> – a subsequence of size <em>size</em> and step <em>step</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.calc_n50">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">calc_n50</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq_lengths</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#calc_n50"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.calc_n50" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the N50 statistics for a <code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.array</span></code> of sequence
lengths.</p>
<p>The algorithm finds in the supplied array the element (contig length) for
which the sum all contig lengths equal or greater than it is equal to half
of all assembled base pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seq_lengths</strong> (<em>array</em>) – an instance of a numpy array containing the
sequence lengths</p>
</dd>
<dt class="field-even">Return int</dt>
<dd class="field-even"><p>the N50 statistics value</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.check_snp_in_seq">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">check_snp_in_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ref_seq</span></em>, <em class="sig-param"><span class="n">pos</span></em>, <em class="sig-param"><span class="n">change</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">trans_table</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#check_snp_in_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.check_snp_in_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Check a SNP in a reference sequence if it is a synonymous or non-synonymous
change.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – reference sequence</p></li>
<li><p><strong>pos</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – SNP position - it is expected to be a 1 based index</p></li>
<li><p><strong>change</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – nucleotide change occuring at <em>pos</em></p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – the starting position for the coding region - 0 based
index</p></li>
<li><p><strong>trans_table</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – translation table used - codon-&gt;AA</p></li>
</ul>
</dd>
<dt class="field-even">Return bool</dt>
<dd class="field-even"><p>True if it is a synonymous change, False if non-synonymous</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.convert_aa_to_nuc_coord">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">convert_aa_to_nuc_coord</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">frame</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#convert_aa_to_nuc_coord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.convert_aa_to_nuc_coord" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts aa coordinates to nucleotidic ones. The coordinates must be from
‘+’ strand. For the ‘-‘ strand, use <a class="reference internal" href="#mgkit.utils.sequence.reverse_aa_coord" title="mgkit.utils.sequence.reverse_aa_coord"><code class="xref py py-func docutils literal notranslate"><span class="pre">reverse_aa_coord()</span></code></a> first.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – start of the annotation (lowest number)</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – end of the annotation (highest number)</p></li>
<li><p><strong>frame</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – frame of the AA translation (0, 1 or 2)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the first element is the converted <em>start</em> and the second
element is the converted <em>end</em></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>the coordinates are assumed to be 1-based indices</p>
</div>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.extrapolate_model">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">extrapolate_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">quals</span></em>, <em class="sig-param"><span class="n">frac</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">scale_adj</span><span class="o">=</span><span class="default_value">0.5</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#extrapolate_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.extrapolate_model" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.3.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.4.4: </span>returns now the minimum and maximum quality scores found, along with
some messages for the stages of the function</p>
</div>
<p>Extrapolate a quality model from a list of qualities. It uses internally
a LOWESS as the base, which is used to estimate the noise as a normal
distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>quals</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – list of arrays of qualities, sorted by position in the
corresponding sequence</p></li>
<li><p><strong>frac</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – fraction of the data used for the LOWESS fit (uses
statsmodels)</p></li>
<li><p><strong>scale_adj</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – value by which the scale of the normal distribution
will be multiplied. Defaults to halving the scale</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the first element is the qualities fit with a LOWESS, the second
element is the distribution, the third is the minimum quality score and
the fourth element is the maximum quality score found in the sequences</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.get_contigs_info">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">get_contigs_info</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_name</span></em>, <em class="sig-param"><span class="n">pp</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#get_contigs_info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.get_contigs_info" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.2.4: </span>file_name can be a <em>dict</em> name-&gt;seq or a list of sequences</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.2.1.</span></p>
</div>
<p>Given a file name for a fasta file with sequences, a dictionary of
name-&gt;seq, or a list of sequences, returns the following information in a
tuple, or a string if <em>pp</em> is True:</p>
<ul class="simple">
<li><p>number of sequences</p></li>
<li><p>total base pairs</p></li>
<li><p>max length</p></li>
<li><p>min length</p></li>
<li><p>average length</p></li>
<li><p>N50 statistic</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – fasta file to open</p></li>
<li><p><strong>pp</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – if True, a formatted string is returned</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the returned value depends on the value of <em>pp</em>, if True a
formatted string is returned, otherwise the tuple with all values is.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.get_seq_expected_syn_count">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">get_seq_expected_syn_count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">syn_matrix</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#get_seq_expected_syn_count"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.get_seq_expected_syn_count" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.4.4: </span>if codon contains <em>N</em> or is not a multiple of 3, it is skipped</p>
</div>
<p>Calculate the expected number of synonymous and non-synonymous changes in a
nucleotide sequence. Assumes that the sequence is already in the correct
frame and its length is a multiple of 3.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>iterable</em>) – nucleotide sequence (uppercase chars)</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – frame of the sequence</p></li>
<li><p><strong>syn_matrix</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – dictionary that contains the expected number of
changes for a codon, as returned by <a class="reference internal" href="#mgkit.utils.sequence.get_syn_matrix" title="mgkit.utils.sequence.get_syn_matrix"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_syn_matrix()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Return tuple</dt>
<dd class="field-even"><p>tuple with counts of expected counts (syn, nonsyn)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.get_seq_number_of_syn">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">get_seq_number_of_syn</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ref_seq</span></em>, <em class="sig-param"><span class="n">snps</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">trans_table</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#get_seq_number_of_syn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.get_seq_number_of_syn" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a reference sequence and a list of SNPs, calculates the number of
synonymous and non-synonymous SNP.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – reference sequence</p></li>
<li><p><strong>snps</strong> (<em>iterable</em>) – list of tuples (position, SNP) - zero based index</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – the frame used for the reference {0, 1, 2}</p></li>
<li><p><strong>trans_table</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – translation table used - codon-&gt;AA</p></li>
</ul>
</dd>
<dt class="field-even">Return tuple</dt>
<dd class="field-even"><p>synonymous and non-synonymous counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.get_syn_matrix">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">get_syn_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">trans_table</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">nuc_list</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#get_syn_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.get_syn_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary containing the expected count of synonymous and
non-synonymous changes that a codon can have if one base is allowed to
change at a time.</p>
<p>There are 9 possible changes per codon.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trans_table</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – a tranlation table, defaults to
<code class="xref py py-data docutils literal notranslate"><span class="pre">seq_utils.TRANS_TABLE</span></code></p></li>
<li><p><strong>nuc_list</strong> (<em>iterable</em>) – a list of nucleotides in which a base can change,
default to the keys of <code class="xref py py-data docutils literal notranslate"><span class="pre">seq_utils.REV_COMP</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Return dict</dt>
<dd class="field-even"><p>returns a dictionary in which for each codon a dictionary
{‘syn’: 0, ‘nonsyn’: 0} holds the number of expected changes</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.get_syn_matrix_all">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">get_syn_matrix_all</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">trans_table</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#get_syn_matrix_all"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.get_syn_matrix_all" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as <a class="reference internal" href="#mgkit.utils.sequence.get_syn_matrix" title="mgkit.utils.sequence.get_syn_matrix"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_syn_matrix()</span></code></a> but a codon can change in any of the ones
included in trans_table.</p>
<p>There are 63 possible changes per codon.</p>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.get_variant_sequence">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">get_variant_sequence</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="o">*</span><span class="n">snps</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#get_variant_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.get_variant_sequence" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.16.</span></p>
</div>
<p>Return a sequence changed in the positions requested.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – a sequence</p></li>
<li><p><strong>*snps</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a>) – each argument passed is a tuple with the first element
as a position in the sequence (1-based index) and the second
element is the character to substitute in the sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>string with the changed characters</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">get_variant_sequence</span><span class="p">(</span><span class="s1">&#39;ACTGATATATGCGCGCATCT&#39;</span><span class="p">,</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">))</span>
<span class="go">&#39;CCTGNTGTATGCGCGCATCT&#39;</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It is used for nucleotide sequences, but it is valid to use any string</p>
</div>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.make_reverse_table">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">make_reverse_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tbl</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#make_reverse_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.make_reverse_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes table to reverse complement a sequence by <a class="reference internal" href="#mgkit.utils.sequence.reverse_complement" title="mgkit.utils.sequence.reverse_complement"><code class="xref py py-func docutils literal notranslate"><span class="pre">reverse_complement()</span></code></a>.
The table used is the complement for each nucleotide, defaulting to
<a class="reference internal" href="#mgkit.utils.sequence.REV_COMP" title="mgkit.utils.sequence.REV_COMP"><code class="xref py py-data docutils literal notranslate"><span class="pre">REV_COMP</span></code></a></p>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.put_gaps_in_nuc_seq">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">put_gaps_in_nuc_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nuc_seq</span></em>, <em class="sig-param"><span class="n">aa_seq</span></em>, <em class="sig-param"><span class="n">trim</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#put_gaps_in_nuc_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.put_gaps_in_nuc_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Match the gaps in an amino-acid aligned sequence to its original nucleotide
sequence. If the nucleotide sequence is not a multiple of 3, the trim
option by default trim those bases from the output.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nuc_seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – original nucleotide sequence</p></li>
<li><p><strong>aa_seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – aligned amino-acid sequence</p></li>
<li><p><strong>trim</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – if True trim last nucleotide(s)</p></li>
</ul>
</dd>
<dt class="field-even">Return str</dt>
<dd class="field-even"><p>gapped nucleotide sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.qualities_model_constant">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">qualities_model_constant</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">length</span><span class="o">=</span><span class="default_value">150</span></em>, <em class="sig-param"><span class="n">scale</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">loc</span><span class="o">=</span><span class="default_value">35</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#qualities_model_constant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.qualities_model_constant" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.3.</span></p>
</div>
<p>Model with constant quality</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – length of the qualities</p></li>
<li><p><strong>scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – base level of the qualities</p></li>
<li><p><strong>loc</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – loc parameter of the normal distribution</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>first element is sequence qualities, the second element contains
the distribution used to randomise them</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.qualities_model_decrease">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">qualities_model_decrease</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">length</span><span class="o">=</span><span class="default_value">150</span></em>, <em class="sig-param"><span class="n">scale</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">loc</span><span class="o">=</span><span class="default_value">35</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#qualities_model_decrease"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.qualities_model_decrease" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.3.</span></p>
</div>
<p>The model is a decreasing one, from 35 and depends on the length of the
sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – length of the qualities</p></li>
<li><p><strong>scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – base level of the qualities</p></li>
<li><p><strong>loc</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – loc parameter of the normal distribution</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>first element is sequence qualities, the second element contains
the distribution used to randomise them</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.random_qualities">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">random_qualities</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">length</span><span class="o">=</span><span class="default_value">150</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_qual</span><span class="o">=</span><span class="default_value">60</span></em>, <em class="sig-param"><span class="n">min_qual</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#random_qualities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.random_qualities" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.3.</span></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – number of quality arrays to yield</p></li>
<li><p><strong>length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – length of the quality array</p></li>
<li><p><strong>model</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a>) – a tuple specifying the qualities and error distribution,
if <em>None</em> <a class="reference internal" href="#mgkit.utils.sequence.qualities_model_decrease" title="mgkit.utils.sequence.qualities_model_decrease"><code class="xref py py-func docutils literal notranslate"><span class="pre">qualities_model_decrease()</span></code></a> is used</p></li>
<li><p><strong>max_qual</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – max quality values allowed, since rarely the quality
exceed 60 in sequencing data</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><em>numpy.array</em> – numpy array of qualities, with the maximum value of 60
by default, or <em>max_qual</em>, same with <em>min_qual</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.random_sequences">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">random_sequences</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">length</span><span class="o">=</span><span class="default_value">150</span></em>, <em class="sig-param"><span class="n">p</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#random_sequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.random_sequences" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.3.</span></p>
</div>
<p>Returns an iterator of random squences, where each nucleotide probability
can be customised in the order (A, C, T, G)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – number of sequences to yield</p></li>
<li><p><strong>length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – length of each sequence</p></li>
<li><p><strong>p</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a>) – tuple with the probability of a nucleotide to occur, in the
order A, C, T, G</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><em>str</em> – string representing a nucleotidic sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.random_sequences_codon">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">random_sequences_codon</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">length</span><span class="o">=</span><span class="default_value">150</span></em>, <em class="sig-param"><span class="n">codons</span><span class="o">=</span><span class="default_value">['TTT', 'TCT', 'TAT', 'TGT', 'TTC', 'TCC', 'TAC', 'TGC', 'TTA', 'TCA', 'TAA', 'TGA', 'TTG', 'TCG', 'TAG', 'TGG', 'CTT', 'CCT', 'CAT', 'CGT', 'CTC', 'CCC', 'CAC', 'CGC', 'CTA', 'CCA', 'CAA', 'CGA', 'CTG', 'CCG', 'CAG', 'CGG', 'ATT', 'ACT', 'AAT', 'AGT', 'ATC', 'ACC', 'AAC', 'AGC', 'ATA', 'ACA', 'AAA', 'AGA', 'ATG', 'ACG', 'AAG', 'AGG', 'GTT', 'GCT', 'GAT', 'GGT', 'GTC', 'GCC', 'GAC', 'GGC', 'GTA', 'GCA', 'GAA', 'GGA', 'GTG', 'GCG', 'GAG', 'GGG']</span></em>, <em class="sig-param"><span class="n">p</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">frame</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#random_sequences_codon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.random_sequences_codon" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.3.</span></p>
</div>
<p>Returns an iterator of nucleotidic sequences, based on a defined genetic
code (passed as parameter, defaults to the universal one). The sequence if
first sampled with replacement from the codon list, with a number of codons
that covers the length chosen plus an additional one to allow a frame shift
as set by <em>frame</em></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If the probability (for each codon) are supplied, the number of
sequences required to match those probabilities within a 10% margin of
error is of at least 10.000 sequences, for 5% at leas 100.000</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – number of sequences to yield</p></li>
<li><p><strong>length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – length of the sequences</p></li>
<li><p><strong>codons</strong> (<em>iterable</em>) – codons used when generating the sequences</p></li>
<li><p><strong>p</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a>) – probability of each codon occurence, in the same order as
<em>codons</em></p></li>
<li><p><strong>frame</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – used to define a specific frame shift occuring in
the sequence (0 to 2) or a random one (if <em>None</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><em>str</em> – string representing a nucleotidic sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.reverse_aa_coord">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">reverse_aa_coord</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">seq_len</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#reverse_aa_coord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.reverse_aa_coord" title="Permalink to this definition">¶</a></dt>
<dd><p>Used to reverse amino-acid coordinates when parsing an AA annotation on
the - strand. Used when the BLAST or HMMER annotations use AA sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – start of the annotation</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – end of the annotation</p></li>
<li><p><strong>seq_len</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – aa sequence length</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>reversed (from strand - to strand +) coordinates. The first
element is the converted <em>start</em> and the second element is the
converted <em>end</em></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>start and end are 1-based indices</p></li>
</ul>
</div>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.reverse_complement">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">tbl</span><span class="o">=</span><span class="default_value">{65: 'T', 67: 'G', 71: 'C', 84: 'A'}</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement of a nucleotide sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – nucleotide sequence with uppercase characters</p></li>
<li><p><strong>tbl</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – translation table returned by <a class="reference internal" href="#mgkit.utils.sequence.make_reverse_table" title="mgkit.utils.sequence.make_reverse_table"><code class="xref py py-func docutils literal notranslate"><span class="pre">make_reverse_table()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Return str</dt>
<dd class="field-even"><p>returns the reverse complement of a nucleotide sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.reverse_complement_old">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">reverse_complement_old</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">tbl</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#reverse_complement_old"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.reverse_complement_old" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement of a nucleotide sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – nucleotide sequence with uppercase characters</p></li>
<li><p><strong>tbl</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – dictionary of complement bases, like <a class="reference internal" href="#mgkit.utils.sequence.REV_COMP" title="mgkit.utils.sequence.REV_COMP"><code class="xref py py-data docutils literal notranslate"><span class="pre">REV_COMP</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Return str</dt>
<dd class="field-even"><p>returns the reverse complement of a nucleotide sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.sequence_composition">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">sequence_composition</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence</span></em>, <em class="sig-param"><span class="n">chars</span><span class="o">=</span><span class="default_value">'A', 'T', 'C', 'G'</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#sequence_composition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.sequence_composition" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.13.</span></p>
</div>
<p>Returns the number of occurences of each unique character in the sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – sequence</p></li>
<li><p><strong>chars</strong> (<em>iterable</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – iterable of the chars to test, default to
(A, C, T, G). if None checks all unique characters in the sequencce</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><em>tuple</em> – the first element is the nucleotide and the second is the number
of occurences in <em>sequence</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.sequence_gc_content">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">sequence_gc_content</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#sequence_gc_content"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.sequence_gc_content" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.3.3: </span>in case of <cite>ZeroDivisionError</cite> returns .5</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.13.</span></p>
</div>
<p>Calculate GC content information for an annotation. The formula is:</p>
<div class="math notranslate nohighlight" id="equation-gc-content">
<span class="eqno">(1)<a class="headerlink" href="#equation-gc-content" title="Permalink to this equation">¶</a></span>\[\frac {(G + C)}{(G + C + A + T)}\]</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>GC content</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.sequence_gc_ratio">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">sequence_gc_ratio</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#sequence_gc_ratio"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.sequence_gc_ratio" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified added">New in version 0.1.13.</span></p>
</div>
<p>Calculate GC ratio information for a sequence. The formula is:</p>
<div class="math notranslate nohighlight" id="equation-gc-ratio">
<span class="eqno">(2)<a class="headerlink" href="#equation-gc-ratio" title="Permalink to this equation">¶</a></span>\[\frac {(A + T)}{(G + C)}\]</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>GC ratio, or <cite>numpy.nan</cite> if G = C = 0</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="mgkit.utils.sequence.translate_sequence">
<code class="sig-prename descclassname">mgkit.utils.sequence.</code><code class="sig-name descname">translate_sequence</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">tbl</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">reverse</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mgkit/utils/sequence.html#translate_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mgkit.utils.sequence.translate_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate a nucleotide sequence in an amino acid one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – sequence to translate, it’s expected to be all caps</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – 0-based index for the translation to start</p></li>
<li><p><strong>tbl</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – dictionary with the translation for each codon</p></li>
<li><p><strong>reverse</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – if True, <a class="reference internal" href="#mgkit.utils.sequence.reverse_complement" title="mgkit.utils.sequence.reverse_complement"><code class="xref py py-func docutils literal notranslate"><span class="pre">reverse_complement()</span></code></a> will be called and
the returned sequence translated</p></li>
</ul>
</dd>
<dt class="field-even">Return str</dt>
<dd class="field-even"><p>the translated sequence</p>
</dd>
</dl>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">MGKit: Metagenomic framework</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline/index.html">Metagenomic Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts/index.html">Scripts Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Example Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gff.html">MGKit GFF Specifications</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">Library Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mgkit.html">mgkit package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.align.html">mgkit.align module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.consts.html">mgkit.consts module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.html">mgkit.counts package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.func.html">mgkit.counts.func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.glm.html">mgkit.counts.glm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.counts.scaling.html">mgkit.counts.scaling module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.html">mgkit.db package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.dbm.html">mgkit.db.dbm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.db.mongo.html">mgkit.db.mongo module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.html">mgkit.filter package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.common.html">mgkit.filter.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.gff.html">mgkit.filter.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.lists.html">mgkit.filter.lists module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.reads.html">mgkit.filter.reads module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.filter.taxon.html">mgkit.filter.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.graphs.html">mgkit.graphs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.html">mgkit.io package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.blast.html">mgkit.io.blast module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fasta.html">mgkit.io.fasta module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.fastq.html">mgkit.io.fastq module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.gff.html">mgkit.io.gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.glimmer.html">mgkit.io.glimmer module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.snpdat.html">mgkit.io.snpdat module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.uniprot.html">mgkit.io.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.io.utils.html">mgkit.io.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.kegg.html">mgkit.kegg module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.logger.html">mgkit.logger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.html">mgkit.mappings package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.cazy.html">mgkit.mappings.cazy module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.eggnog.html">mgkit.mappings.eggnog module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.enzyme.html">mgkit.mappings.enzyme module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.go.html">mgkit.mappings.go module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.pandas_map.html">mgkit.mappings.pandas_map module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.taxon.html">mgkit.mappings.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.mappings.utils.html">mgkit.mappings.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.html">mgkit.net package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.embl.html">mgkit.net.embl module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.pfam.html">mgkit.net.pfam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.uniprot.html">mgkit.net.uniprot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.net.utils.html">mgkit.net.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.html">mgkit.plots package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.abund.html">mgkit.plots.abund module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.boxplot.html">mgkit.plots.boxplot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.colors.html">mgkit.plots.colors module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.heatmap.html">mgkit.plots.heatmap module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.unused.html">mgkit.plots.unused module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.plots.utils.html">mgkit.plots.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.simple_cache.html">mgkit.simple_cache module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.html">mgkit.snps package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.classes.html">mgkit.snps.classes module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.conv_func.html">mgkit.snps.conv_func module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.filter.html">mgkit.snps.filter module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.funcs.html">mgkit.snps.funcs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.snps.mapper.html">mgkit.snps.mapper module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.taxon.html">mgkit.taxon module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.html">mgkit.utils package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.common.html">mgkit.utils.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.dictionary.html">mgkit.utils.dictionary module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">mgkit.utils.sequence module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.utils.trans_tables.html">mgkit.utils.trans_tables module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.html">mgkit.workflow package</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.add_gff_info.html">mgkit.workflow.add_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.blast2gff.html">mgkit.workflow.blast2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.extract_gff_info.html">mgkit.workflow.extract_gff_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fasta_utils.html">mgkit.workflow.fasta_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.fastq_utils.html">mgkit.workflow.fastq_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.filter_gff.html">mgkit.workflow.filter_gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.hmmer2gff.html">mgkit.workflow.hmmer2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.json2gff.html">mgkit.workflow.json2gff module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.sampling_utils.html">mgkit.workflow.sampling_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.snp_parser.html">mgkit.workflow.snp_parser module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.taxon_utils.html">mgkit.workflow.taxon_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mgkit.workflow.utils.html">mgkit.workflow.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html">mgkit</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Changes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="../library.html">Library Reference</a><ul>
  <li><a href="mgkit.html">mgkit package</a><ul>
  <li><a href="mgkit.utils.html">mgkit.utils package</a><ul>
      <li>Previous: <a href="mgkit.utils.dictionary.html" title="previous chapter">mgkit.utils.dictionary module</a></li>
      <li>Next: <a href="mgkit.utils.trans_tables.html" title="next chapter">mgkit.utils.trans_tables module</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2013-2020, Francesco Rubino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/api/mgkit.utils.sequence.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>